
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>pydna.design &#8212; Pydna 3.0.1+2.g47ae1f65a3.dirty documentation</title>
    <link rel="stylesheet" href="../../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="../../_static/language_data.js"></script>
    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">Pydna 3.0.1+2.g47ae1f65a3.dirty documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../index.html" >Module code</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="../pydna.html" accesskey="U">pydna</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for pydna.design</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="c1"># -*- coding: utf-8 -*-</span>
<span class="c1"># Copyright 2013-2018 by Björn Johansson.  All rights reserved.</span>
<span class="c1"># This code is part of the Python-dna distribution and governed by its</span>
<span class="c1"># license.  Please see the LICENSE.txt file that should have been included</span>
<span class="c1"># as part of this package.</span>
<span class="sd">&#39;&#39;&#39;This module contain functions for primer design for various purposes. </span>

<span class="sd">- :func:primer_design for designing primers for a sequence or a matching primer for an existing primer. Returns an :class:`Amplicon` object (same as the :mod:`amplify` module returns).</span>

<span class="sd">- :func:assembly_fragments Adds tails to primers for a linear assembly through homologous recombination or Gibson assembly.</span>

<span class="sd">- :func:circular_assembly_fragments Adds tails to primers for a circular assembly through homologous recombination or Gibson assembly.</span>

<span class="sd">&#39;&#39;&#39;</span>
<span class="c1">#import warnings</span>

<span class="kn">import</span> <span class="nn">math</span>                                       <span class="k">as</span> <span class="nn">_math</span>
<span class="c1">#from operator import itemgetter                   as _itemgetter</span>
<span class="kn">import</span> <span class="nn">os</span>                                         <span class="k">as</span> <span class="nn">_os</span>
<span class="kn">import</span> <span class="nn">copy</span>                                       <span class="k">as</span> <span class="nn">_copy</span>
<span class="c1">#from Bio.Alphabet import Alphabet                 as _Alphabet</span>
<span class="c1">#from Bio.Alphabet.IUPAC import IUPACAmbiguousDNA  as _IUPACAmbiguousDNA</span>
<span class="c1">#from Bio.Seq import Seq                           as _Seq</span>
<span class="kn">from</span> <span class="nn">pydna.amplify</span> <span class="k">import</span> <span class="n">Anneal</span>                  <span class="k">as</span> <span class="n">_Anneal</span>
<span class="kn">from</span> <span class="nn">pydna.amplify</span> <span class="k">import</span> <span class="n">pcr</span>                     <span class="k">as</span> <span class="n">_pcr</span>
<span class="kn">from</span> <span class="nn">pydna.tm</span> <span class="k">import</span> <span class="n">tmbresluc</span>                    <span class="k">as</span> <span class="n">_tmbresluc</span>
<span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>           <span class="k">as</span> <span class="n">_Dseqrecord</span>
<span class="c1">#from pydna._pretty import pretty_str              as _pretty_str</span>
<span class="kn">from</span> <span class="nn">pydna.primer</span>    <span class="k">import</span> <span class="n">Primer</span>                <span class="k">as</span> <span class="n">_Primer</span>

<span class="kn">import</span> <span class="nn">logging</span>    <span class="k">as</span> <span class="nn">_logging</span>
<span class="n">_module_logger</span> <span class="o">=</span> <span class="n">_logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;pydna.&quot;</span><span class="o">+</span><span class="vm">__name__</span><span class="p">)</span>

<div class="viewcode-block" id="primer_design"><a class="viewcode-back" href="../../index.html#pydna.design.primer_design">[docs]</a><span class="k">def</span> <span class="nf">primer_design</span><span class="p">(</span> <span class="n">template</span><span class="p">,</span>
                   <span class="n">fp</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">rp</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">target_tm</span><span class="o">=</span><span class="mf">55.0</span><span class="p">,</span>
                   <span class="n">fprimerc</span><span class="o">=</span><span class="mf">1000.0</span><span class="p">,</span>  <span class="c1"># nM</span>
                   <span class="n">rprimerc</span><span class="o">=</span><span class="mf">1000.0</span><span class="p">,</span>  <span class="c1"># nM</span>
                   <span class="n">saltc</span><span class="o">=</span><span class="mf">50.0</span><span class="p">,</span>
                   <span class="n">limit</span><span class="o">=</span><span class="mi">13</span><span class="p">,</span>
                   <span class="n">formula</span> <span class="o">=</span> <span class="n">_tmbresluc</span><span class="p">):</span>

    <span class="sd">&#39;&#39;&#39;This function designs a forward primer and a reverse primer for PCR amplification </span>
<span class="sd">    of a given template sequence.</span>
<span class="sd">    </span>
<span class="sd">    The template argument is a Dseqrecord object or equivalent containing the template sequence.</span>
<span class="sd">    </span>
<span class="sd">    The optional fp and rp arguments can contain an existing primer for the sequence (either the forward or reverse primer).</span>
<span class="sd">    One or the other primers can be specified, not both (since then there is nothing to design!, use the pydna.amplify.pcr function instead).</span>
<span class="sd">    </span>
<span class="sd">    If one of the primers is given, the other primer is designed to match in terms of Tm.</span>
<span class="sd">    If both primers are designed, they will be designed to target_tm</span>
<span class="sd">    </span>
<span class="sd">    fprimerc, rprimerc and saltc are formward and reverse primer concentration (nM). Saltc is the salt concentration. </span>
<span class="sd">    These arguments might affect how Tm is calculated.</span>
<span class="sd">    </span>
<span class="sd">    formula is a function that can take at least three arguments f( str, primerc=float, saltc=float).</span>
<span class="sd">    There are several of these in the pydna.tm module.</span>
<span class="sd">    </span>
<span class="sd">    The function returns a pydna.amplicon.Amplicon class instance. This object has </span>
<span class="sd">    the object.forward_primer and object.reverse_primer properties which contain the designed primers.</span>
<span class="sd">    </span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    template : pydna.dseqrecord.Dseqrecord</span>
<span class="sd">        a Dseqrecord object. The only required argument.</span>

<span class="sd">    fp, rp : pydna.primer.Primer, optional</span>
<span class="sd">        optional pydna.primer.Primer objects containing one primer each.</span>

<span class="sd">    target_tm : float, optional</span>
<span class="sd">        target tm for the primers, set to 55°C by default.</span>

<span class="sd">    fprimerc : float, optional</span>
<span class="sd">        Concentration of forward primer in nM, set to 1000.0 nM by default.</span>

<span class="sd">    rprimerc : float, optional</span>
<span class="sd">        Concentration of reverse primer in nM, set to 1000.0 nM by default.</span>

<span class="sd">    saltc  : float, optional</span>
<span class="sd">        Salt concentration (monovalet cations) :mod:`tmbresluc` set to 50.0 mM by default</span>

<span class="sd">    formula : function</span>
<span class="sd">        formula used for tm calculation</span>
<span class="sd">        this is the name of a function.</span>
<span class="sd">        built in options are:</span>

<span class="sd">        1. :func:`pydna.amplify.tmbresluc` (default)</span>
<span class="sd">        2. :func:`pydna.amplify.basictm`</span>
<span class="sd">        3. :func:`pydna.amplify.tmstaluc98`</span>
<span class="sd">        4. :func:`pydna.amplify.tmbreslauer86`</span>

<span class="sd">        These functions are imported from the :mod:`pydna.amplify` module, but can be</span>
<span class="sd">        substituted for some other custom made function.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    result : Amplicon</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>

<span class="sd">    &gt;&gt;&gt; from pydna.dseqrecord import Dseqrecord</span>
<span class="sd">    &gt;&gt;&gt; t=Dseqrecord(&quot;atgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatg&quot;)</span>
<span class="sd">    &gt;&gt;&gt; t</span>
<span class="sd">    Dseqrecord(-64)</span>
<span class="sd">    &gt;&gt;&gt; from pydna.design import primer_design</span>
<span class="sd">    &gt;&gt;&gt; ampl = primer_design(t)</span>
<span class="sd">    &gt;&gt;&gt; ampl</span>
<span class="sd">    Amplicon(64)</span>
<span class="sd">    &gt;&gt;&gt; ampl.forward_primer</span>
<span class="sd">    f64 18-mer:5&#39;-atgactgctaacccttcc-3&#39;</span>
<span class="sd">    &gt;&gt;&gt; ampl.reverse_primer</span>
<span class="sd">    r64 19-mer:5&#39;-catcgtaagtttcgaacga-3&#39;</span>
<span class="sd">    &gt;&gt;&gt; print(ampl.figure())</span>
<span class="sd">    5atgactgctaacccttcc...tcgttcgaaacttacgatg3</span>
<span class="sd">                          ||||||||||||||||||| tm 53.8 (dbd) 60.6</span>
<span class="sd">                         3agcaagctttgaatgctac5</span>
<span class="sd">    5atgactgctaacccttcc3</span>
<span class="sd">     |||||||||||||||||| tm 54.4 (dbd) 58.4</span>
<span class="sd">    3tactgacgattgggaagg...agcaagctttgaatgctac5</span>
<span class="sd">    &gt;&gt;&gt; pf = &quot;GGATCC&quot; + ampl.forward_primer</span>
<span class="sd">    &gt;&gt;&gt; pr = &quot;GGATCC&quot; + ampl.reverse_primer  </span>
<span class="sd">    &gt;&gt;&gt; pf</span>
<span class="sd">    f64 24-mer:5&#39;-GGATCCatgactgct..tcc-3&#39;</span>
<span class="sd">    &gt;&gt;&gt; pr</span>
<span class="sd">    r64 25-mer:5&#39;-GGATCCcatcgtaag..cga-3&#39;</span>
<span class="sd">    &gt;&gt;&gt; from pydna.amplify import pcr</span>
<span class="sd">    &gt;&gt;&gt; pcr_prod = pcr(pf, pr, t)</span>
<span class="sd">    &gt;&gt;&gt; print(pcr_prod.figure())</span>
<span class="sd">          5atgactgctaacccttcc...tcgttcgaaacttacgatg3</span>
<span class="sd">                                ||||||||||||||||||| tm 53.8 (dbd) 60.6</span>
<span class="sd">                               3agcaagctttgaatgctacCCTAGG5</span>
<span class="sd">    5GGATCCatgactgctaacccttcc3</span>
<span class="sd">           |||||||||||||||||| tm 54.4 (dbd) 58.4</span>
<span class="sd">          3tactgacgattgggaagg...agcaagctttgaatgctac5</span>
<span class="sd">    &gt;&gt;&gt; print(pcr_prod.seq)</span>
<span class="sd">    GGATCCatgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatgGGATCC</span>
<span class="sd">    &gt;&gt;&gt; from pydna.primer import Primer</span>
<span class="sd">    &gt;&gt;&gt; pf = Primer(&quot;atgactgctaacccttccttggtgttg&quot;, id=&quot;myprimer&quot;)</span>
<span class="sd">    &gt;&gt;&gt; ampl = primer_design(t, fp = pf)</span>
<span class="sd">    &gt;&gt;&gt; ampl.forward_primer</span>
<span class="sd">    myprimer 27-mer:5&#39;-atgactgctaaccct..ttg-3&#39;</span>
<span class="sd">    &gt;&gt;&gt; ampl.reverse_primer</span>
<span class="sd">    r64 28-mer:5&#39;-catcgtaagtttcga..gtc-3&#39;</span>

<span class="sd">    &#39;&#39;&#39;</span>
    
    <span class="k">def</span> <span class="nf">design</span><span class="p">(</span><span class="n">target_tm</span><span class="p">,</span> <span class="n">template</span><span class="p">):</span>
        <span class="sd">&#39;&#39;&#39; returns a string &#39;&#39;&#39;</span>
        <span class="n">tmp</span><span class="o">=</span><span class="mi">0</span>
        <span class="n">length</span><span class="o">=</span><span class="n">limit</span>
        <span class="n">p</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">template</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="n">length</span><span class="p">])</span>
        <span class="k">while</span> <span class="n">tmp</span><span class="o">&lt;</span><span class="n">target_tm</span><span class="p">:</span>
            <span class="n">length</span><span class="o">+=</span><span class="mi">1</span>
            <span class="n">p</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">template</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="n">length</span><span class="p">])</span>
            <span class="n">tmp</span> <span class="o">=</span> <span class="n">formula</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>
        <span class="n">ps</span> <span class="o">=</span> <span class="n">p</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">tmps</span> <span class="o">=</span> <span class="n">formula</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">ps</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>
        <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(((</span><span class="n">p</span><span class="p">,</span>   <span class="n">tmp</span><span class="p">),(</span><span class="n">ps</span><span class="p">,</span> <span class="n">tmps</span><span class="p">)))</span>
        <span class="k">return</span> <span class="nb">min</span><span class="p">(</span> <span class="p">(</span> <span class="nb">abs</span><span class="p">(</span><span class="n">target_tm</span><span class="o">-</span><span class="n">tmp</span><span class="p">),</span> <span class="n">p</span><span class="p">),</span> <span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">target_tm</span><span class="o">-</span><span class="n">tmps</span><span class="p">),</span> <span class="n">ps</span><span class="p">)</span> <span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
    
    <span class="k">if</span> <span class="n">fp</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">rp</span><span class="p">:</span>
        <span class="n">fp</span>  <span class="o">=</span> <span class="n">_Anneal</span><span class="p">((</span><span class="n">fp</span><span class="p">,),</span> <span class="n">template</span><span class="p">)</span><span class="o">.</span><span class="n">forward_primers</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
        <span class="n">target_tm</span> <span class="o">=</span> <span class="n">formula</span><span class="p">(</span> <span class="nb">str</span><span class="p">(</span><span class="n">fp</span><span class="o">.</span><span class="n">footprint</span><span class="p">),</span> <span class="n">primerc</span><span class="o">=</span><span class="n">fprimerc</span><span class="p">,</span> <span class="n">saltc</span><span class="o">=</span><span class="n">saltc</span><span class="p">)</span>
        <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;forward primer given, design reverse primer:&quot;</span><span class="p">)</span>
        <span class="n">rp</span> <span class="o">=</span> <span class="n">_Primer</span><span class="p">(</span><span class="n">design</span><span class="p">(</span><span class="n">target_tm</span><span class="p">,</span> <span class="n">template</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()))</span>
    <span class="k">elif</span> <span class="ow">not</span> <span class="n">fp</span> <span class="ow">and</span> <span class="n">rp</span><span class="p">:</span>
        <span class="n">rp</span> <span class="o">=</span>  <span class="n">_Anneal</span><span class="p">([</span><span class="n">rp</span><span class="p">],</span> <span class="n">template</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_primers</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
        <span class="n">target_tm</span> <span class="o">=</span> <span class="n">formula</span><span class="p">(</span> <span class="nb">str</span><span class="p">(</span><span class="n">rp</span><span class="o">.</span><span class="n">footprint</span><span class="p">),</span> <span class="n">primerc</span><span class="o">=</span><span class="n">rprimerc</span><span class="p">,</span> <span class="n">saltc</span><span class="o">=</span><span class="n">saltc</span><span class="p">)</span>
        <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;reverse primer given, design forward primer:&quot;</span><span class="p">)</span>
        <span class="n">fp</span> <span class="o">=</span> <span class="n">_Primer</span><span class="p">(</span><span class="n">design</span><span class="p">(</span><span class="n">target_tm</span><span class="p">,</span> <span class="n">template</span><span class="p">))</span>
    <span class="k">elif</span> <span class="ow">not</span> <span class="n">fp</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">rp</span><span class="p">:</span>
        <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;no primer given, design forward primer:&quot;</span><span class="p">)</span>
        <span class="n">fp</span> <span class="o">=</span> <span class="n">_Primer</span><span class="p">((</span><span class="n">design</span><span class="p">(</span><span class="n">target_tm</span><span class="p">,</span> <span class="n">template</span><span class="p">)))</span>
        <span class="n">target_tm</span> <span class="o">=</span> <span class="n">formula</span><span class="p">(</span> <span class="nb">str</span><span class="p">(</span><span class="n">fp</span><span class="o">.</span><span class="n">seq</span><span class="p">),</span> <span class="n">primerc</span><span class="o">=</span><span class="n">fprimerc</span><span class="p">,</span> <span class="n">saltc</span><span class="o">=</span><span class="n">saltc</span><span class="p">)</span>
        <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;no primer given, design reverse primer:&quot;</span><span class="p">)</span>
        <span class="n">rp</span> <span class="o">=</span> <span class="n">_Primer</span><span class="p">(</span><span class="n">design</span><span class="p">(</span><span class="n">target_tm</span><span class="p">,</span> <span class="n">template</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Specify maximum one of the two primers.&quot;</span><span class="p">)</span>

    <span class="n">fp</span><span class="o">.</span><span class="n">concentration</span> <span class="o">=</span> <span class="n">fprimerc</span>
    <span class="n">rp</span><span class="o">.</span><span class="n">concentration</span> <span class="o">=</span> <span class="n">rprimerc</span>

    <span class="k">if</span> <span class="n">fp</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="c1">#&lt;unknown id&gt;</span>
        <span class="n">fp</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="s2">&quot;f</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">template</span><span class="p">))</span>
        
    <span class="k">if</span> <span class="n">rp</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="s2">&quot;id&quot;</span><span class="p">:</span>
        <span class="n">rp</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="s2">&quot;r</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">template</span><span class="p">))</span>

    <span class="k">if</span> <span class="n">fp</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span>
        <span class="n">fp</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;f</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">template</span><span class="p">))</span>
        
    <span class="k">if</span> <span class="n">rp</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span>
        <span class="n">rp</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;r</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">template</span><span class="p">))</span>

    <span class="n">fp</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="n">fp</span><span class="o">.</span><span class="n">id</span><span class="o">+</span><span class="s1">&#39; &#39;</span><span class="o">+</span><span class="n">template</span><span class="o">.</span><span class="n">accession</span>
    <span class="n">rp</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="n">rp</span><span class="o">.</span><span class="n">id</span><span class="o">+</span><span class="s1">&#39; &#39;</span><span class="o">+</span><span class="n">template</span><span class="o">.</span><span class="n">accession</span>
    
    <span class="n">ampl</span> <span class="o">=</span> <span class="n">_Anneal</span><span class="p">(</span> <span class="p">(</span><span class="n">fp</span><span class="p">,</span> <span class="n">rp</span><span class="p">),</span> <span class="n">template</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="n">limit</span><span class="p">)</span>
    
    <span class="n">prod</span> <span class="o">=</span> <span class="n">ampl</span><span class="o">.</span><span class="n">products</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ampl</span><span class="o">.</span><span class="n">products</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">1</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">warnings</span> <span class="k">as</span> <span class="nn">_warnings</span>
        <span class="kn">from</span> <span class="nn">pydna</span> <span class="k">import</span> <span class="n">_PydnaWarning</span>
        <span class="n">_warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;designed primers do not yield a unique PCR product&quot;</span><span class="p">,</span>
                       <span class="n">_PydnaWarning</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">prod</span></div>


<div class="viewcode-block" id="assembly_fragments"><a class="viewcode-back" href="../../index.html#pydna.design.assembly_fragments">[docs]</a><span class="k">def</span> <span class="nf">assembly_fragments</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">overlap</span><span class="o">=</span><span class="mi">35</span><span class="p">,</span> <span class="n">maxlink</span><span class="o">=</span><span class="mi">40</span><span class="p">):</span>    
    <span class="sd">&#39;&#39;&#39;This function return a list of :mod:`pydna.amplicon.Amplicon` objects where </span>
<span class="sd">    primers have been modified with tails so that the fragments can be fused in </span>
<span class="sd">    the order they appear in the list by for example Gibson assembly or homologous </span>
<span class="sd">    recombination.</span>
<span class="sd">    </span>
<span class="sd">    Given that we have two linear :mod:`pydna.amplicon.Amplicon` objects a and b </span>
<span class="sd">    </span>
<span class="sd">    we can modify the reverse primer of a and forward primer of b with tails to allow </span>
<span class="sd">    fusion by fusion PCR, Gibson assembly or in-vivo homologous recombination.</span>
<span class="sd">    The basic requirements for the primers for the three techniques are the same.</span>

<span class="sd">    ::</span>
<span class="sd">        </span>
<span class="sd">                                &lt;--&gt;</span>

<span class="sd">       _________ a _________           __________ b ________</span>
<span class="sd">      /                     \\         /                     \\</span>
<span class="sd">      agcctatcatcttggtctctgca         TTTATATCGCATGACTCTTCTTT</span>
<span class="sd">      |||||||||||||||||||||||         |||||||||||||||||||||||</span>
<span class="sd">                       &lt;gacgt                          &lt;AGAAA</span>
<span class="sd">      agcct&gt;                          TTTAT&gt;</span>
<span class="sd">      |||||||||||||||||||||||         |||||||||||||||||||||||</span>
<span class="sd">      tcggatagtagaaccagagacgt         AAATATAGCGTACTGAGAAGAAA</span>


<span class="sd">           agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="sd">           ||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="sd">           tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
<span class="sd">           \\___________________ c ______________________/</span>


<span class="sd">    Design tailed primers incorporating a part of the next or previous fragment to be assembled.</span>

<span class="sd">    ::</span>


<span class="sd">      agcctatcatcttggtctctgca</span>
<span class="sd">      |||||||||||||||||||||||</span>
<span class="sd">                      gagacgtAAATATA</span>

<span class="sd">      |||||||||||||||||||||||</span>
<span class="sd">      tcggatagtagaaccagagacgt</span>


<span class="sd">                             TTTATATCGCATGACTCTTCTTT</span>
<span class="sd">                             |||||||||||||||||||||||</span>

<span class="sd">                      ctctgcaTTTATAT</span>
<span class="sd">                             |||||||||||||||||||||||</span>
<span class="sd">                             AAATATAGCGTACTGAGAAGAAA</span>

<span class="sd">    PCR products with flanking sequences are formed in the PCR process.</span>

<span class="sd">    ::</span>

<span class="sd">      agcctatcatcttggtctctgcaTTTATAT</span>
<span class="sd">      ||||||||||||||||||||||||||||||</span>
<span class="sd">      tcggatagtagaaccagagacgtAAATATA</span>
<span class="sd">                      \\____________/</span>

<span class="sd">                         identical</span>
<span class="sd">                         sequences</span>
<span class="sd">                       ____________</span>
<span class="sd">                      /            \\</span>
<span class="sd">                      ctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="sd">                      ||||||||||||||||||||||||||||||</span>
<span class="sd">                      gagacgtAAATATAGCGTACTGAGAAGAAA</span>

<span class="sd">    The fragments can be fused by any of the techniques mentioned earlier to form c:</span>

<span class="sd">    ::</span>

<span class="sd">      agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="sd">      ||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="sd">      tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>


<span class="sd">    The first argument of this function is a list of sequence objects containing </span>
<span class="sd">    Amplicons and other similar objects.</span>
<span class="sd">    </span>
<span class="sd">    **At least every second sequence object needs to be an Amplicon**</span>
<span class="sd">    </span>
<span class="sd">    This rule exists because if a sequence object is that is not a PCR product</span>
<span class="sd">    is to be fused with another fragment, that other fragment needs to be an Amplicon</span>
<span class="sd">    so that the primer of the other object can be modified to include the whole stretch</span>
<span class="sd">    of sequence homology needed for the fusion. See the example below where a is a </span>
<span class="sd">    non-amplicon (a linear plasmid  vector for instance)</span>

<span class="sd">    ::</span>

<span class="sd">       _________ a _________           __________ b ________</span>
<span class="sd">      /                     \\         /                     \\</span>
<span class="sd">      agcctatcatcttggtctctgca   &lt;--&gt;  TTTATATCGCATGACTCTTCTTT</span>
<span class="sd">      |||||||||||||||||||||||         |||||||||||||||||||||||</span>
<span class="sd">      tcggatagtagaaccagagacgt                          &lt;AGAAA</span>
<span class="sd">                                      TTTAT&gt;</span>
<span class="sd">                                      |||||||||||||||||||||||</span>
<span class="sd">                                &lt;--&gt;  AAATATAGCGTACTGAGAAGAAA</span>


<span class="sd">           agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="sd">           ||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="sd">           tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
<span class="sd">           \\___________________ c ______________________/</span>


<span class="sd">    In this case only the forward primer of b is fitted with a tail with a part a:</span>

<span class="sd">    ::</span>


<span class="sd">      agcctatcatcttggtctctgca</span>
<span class="sd">      |||||||||||||||||||||||</span>
<span class="sd">      tcggatagtagaaccagagacgt</span>


<span class="sd">                             TTTATATCGCATGACTCTTCTTT</span>
<span class="sd">                             |||||||||||||||||||||||</span>
<span class="sd">                                              &lt;AGAAA</span>
<span class="sd">               tcttggtctctgcaTTTATAT</span>
<span class="sd">                             |||||||||||||||||||||||</span>
<span class="sd">                             AAATATAGCGTACTGAGAAGAAA</span>

<span class="sd">    PCR products with flanking sequences are formed in the PCR process.</span>

<span class="sd">    ::</span>

<span class="sd">      agcctatcatcttggtctctgcaTTTATAT</span>
<span class="sd">      ||||||||||||||||||||||||||||||</span>
<span class="sd">      tcggatagtagaaccagagacgtAAATATA</span>
<span class="sd">                      \\____________/</span>

<span class="sd">                         identical</span>
<span class="sd">                         sequences</span>
<span class="sd">                       ____________</span>
<span class="sd">                      /            \\</span>
<span class="sd">                      ctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="sd">                      ||||||||||||||||||||||||||||||</span>
<span class="sd">                      gagacgtAAATATAGCGTACTGAGAAGAAA</span>

<span class="sd">    The fragments can be fused by for example Gibson assembly:</span>

<span class="sd">    ::</span>

<span class="sd">      agcctatcatcttggtctctgcaTTTATAT</span>
<span class="sd">      ||||||||||||||||||||||||||||||</span>
<span class="sd">      tcggatagtagaacca</span>
<span class="sd">                      </span>
<span class="sd">                                   TCGCATGACTCTTCTTT</span>
<span class="sd">                      ||||||||||||||||||||||||||||||</span>
<span class="sd">                      gagacgtAAATATAGCGTACTGAGAAGAAA </span>
<span class="sd">                      </span>
<span class="sd">    to form c:</span>

<span class="sd">    ::</span>
<span class="sd">                </span>
<span class="sd">      agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="sd">      ||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="sd">      tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
<span class="sd">      </span>

<span class="sd">    The first argument of this function is a list of sequence objects containing </span>
<span class="sd">    Amplicons and other similar objects.</span>
<span class="sd">    </span>
<span class="sd">    The overlap argument controls how many base pairs of overlap required between </span>
<span class="sd">    adjacent sequence fragments. In the junction between Amplicons, tails with the </span>
<span class="sd">    length of about half of this value is added to the two primers</span>
<span class="sd">    closest to the junction.</span>
<span class="sd">    </span>
<span class="sd">    ::</span>
<span class="sd">        </span>
<span class="sd">            &gt;       &lt;</span>
<span class="sd">            Amplicon1</span>
<span class="sd">                     Amplicon2</span>
<span class="sd">                     &gt;       &lt;</span>
<span class="sd">                     </span>
<span class="sd">                     ⇣</span>

<span class="sd">            &gt;       &lt;-</span>
<span class="sd">            Amplicon1</span>
<span class="sd">                     Amplicon2</span>
<span class="sd">                    -&gt;       &lt;                     </span>
<span class="sd">                     </span>
<span class="sd">    In the case of an Amplicon adjacent to a Dseqrecord object, the tail will </span>
<span class="sd">    be twice as long (1*overlap) since the </span>
<span class="sd">    recombining sequence is present entirely on this primer:</span>
<span class="sd">        </span>
<span class="sd">    ::</span>
<span class="sd">        </span>
<span class="sd">            Dseqrecd1</span>
<span class="sd">                     Amplicon1</span>
<span class="sd">                     &gt;       &lt;</span>
<span class="sd">                     </span>
<span class="sd">                     ⇣</span>

<span class="sd">            Dseqrecd1</span>
<span class="sd">                     Amplicon1</span>
<span class="sd">                   --&gt;       &lt;</span>
<span class="sd">    </span>
<span class="sd">    Note that if the sequence of DNA fragments starts or stops with an Amplicon, </span>
<span class="sd">    the very first and very last prinmer will not be modified i.e. assembles are </span>
<span class="sd">    always assumed to be linear. There are simple tricks around that for circular</span>
<span class="sd">    assemblies depicted in the last two examples below.</span>
<span class="sd">    </span>
<span class="sd">    The maxlink arguments controls the cut off length for sequences that will be</span>
<span class="sd">    synhtesized by adding them to primers for the adjacent fragment(s). The </span>
<span class="sd">    argument list may contain short spacers (such as spacers between fusion proteins).</span>
<span class="sd">    </span>

<span class="sd">    ::</span>

<span class="sd">        Example 1: Linear assembly of PCR products (pydna.amplicon.Amplicon class objects) ------</span>
<span class="sd">        </span>
<span class="sd">        </span>
<span class="sd">        &gt;       &lt;         &gt;       &lt;</span>
<span class="sd">        Amplicon1         Amplicon3</span>
<span class="sd">                 Amplicon2         Amplicon4</span>
<span class="sd">                 &gt;       &lt;         &gt;       &lt;</span>
<span class="sd">        </span>
<span class="sd">                             ⇣</span>
<span class="sd">                             pydna.design.assembly_fragments</span>
<span class="sd">                             ⇣ </span>
<span class="sd">        </span>
<span class="sd">        &gt;       &lt;-       -&gt;       &lt;-                      pydna.assembly.Assembly</span>
<span class="sd">        Amplicon1         Amplicon3                         </span>
<span class="sd">                 Amplicon2         Amplicon4     ➤  Amplicon1Amplicon2Amplicon3Amplicon4</span>
<span class="sd">                -&gt;       &lt;-       -&gt;       &lt;</span>
<span class="sd">        </span>
<span class="sd">        </span>
<span class="sd">        Example 2: Linear assembly of alternating Amplicons and other fragments</span>
<span class="sd">        </span>
<span class="sd">        </span>
<span class="sd">        &gt;       &lt;         &gt;       &lt;</span>
<span class="sd">        Amplicon1         Amplicon2</span>
<span class="sd">                 Dseqrecd1         Dseqrecd2</span>
<span class="sd">                      </span>
<span class="sd">                             ⇣</span>
<span class="sd">                             pydna.design.assembly_fragments</span>
<span class="sd">                             ⇣ </span>
<span class="sd">                          </span>
<span class="sd">        &gt;       &lt;--     --&gt;       &lt;--                     pydna.assembly.Assembly</span>
<span class="sd">        Amplicon1         Amplicon2</span>
<span class="sd">                 Dseqrecd1         Dseqrecd2     ➤  Amplicon1Dseqrecd1Amplicon2Dseqrecd2</span>
<span class="sd">        </span>
<span class="sd">        </span>
<span class="sd">        Example 3: Linear assembly of alternating Amplicons and other fragments</span>
<span class="sd">        </span>
<span class="sd">        </span>
<span class="sd">        Dseqrecd1         Dseqrecd2</span>
<span class="sd">                 Amplicon1         Amplicon2</span>
<span class="sd">                 &gt;       &lt;       --&gt;       &lt;</span>
<span class="sd">        </span>
<span class="sd">                             ⇣</span>
<span class="sd">                     pydna.design.assembly_fragments</span>
<span class="sd">                             ⇣</span>
<span class="sd">                                                          pydna.assembly.Assembly</span>
<span class="sd">        Dseqrecd1         Dseqrecd2</span>
<span class="sd">                 Amplicon1         Amplicon2     ➤  Dseqrecd1Amplicon1Dseqrecd2Amplicon2</span>
<span class="sd">               --&gt;       &lt;--     --&gt;       &lt;</span>
<span class="sd">        </span>
<span class="sd">        </span>
<span class="sd">        Example 4: Circular assembly of alternating Amplicons and other fragments</span>
<span class="sd">        </span>
<span class="sd">                         -&gt;       &lt;==</span>
<span class="sd">        Dseqrecd1         Amplicon2</span>
<span class="sd">                 Amplicon1         Dseqrecd1</span>
<span class="sd">               --&gt;       &lt;-</span>
<span class="sd">                             ⇣</span>
<span class="sd">                             pydna.design.assembly_fragments</span>
<span class="sd">                             ⇣ </span>
<span class="sd">                                                           pydna.assembly.Assembly</span>
<span class="sd">                         -&gt;       &lt;==</span>
<span class="sd">        Dseqrecd1         Amplicon2                    -Dseqrecd1Amplicon1Amplicon2-  </span>
<span class="sd">                 Amplicon1                       ➤    |                             |</span>
<span class="sd">               --&gt;       &lt;-                            -----------------------------</span>
<span class="sd">        </span>
<span class="sd">        ------ Example 5: Circular assembly of Amplicons</span>
<span class="sd">        </span>
<span class="sd">        &gt;       &lt;         &gt;       &lt;</span>
<span class="sd">        Amplicon1         Amplicon3</span>
<span class="sd">                 Amplicon2         Amplicon1</span>
<span class="sd">                 &gt;       &lt;         &gt;       &lt;</span>
<span class="sd">        </span>
<span class="sd">                             ⇣</span>
<span class="sd">                             pydna.design.assembly_fragments</span>
<span class="sd">                             ⇣ </span>
<span class="sd">        </span>
<span class="sd">        &gt;       &lt;=       -&gt;       &lt;-        </span>
<span class="sd">        Amplicon1         Amplicon3                  </span>
<span class="sd">                 Amplicon2         Amplicon1</span>
<span class="sd">                -&gt;       &lt;-       +&gt;       &lt;</span>
<span class="sd">        </span>
<span class="sd">                             ⇣</span>
<span class="sd">                     make new Amplicon using the Amplicon1.template and </span>
<span class="sd">                     the last fwd primer and the first rev primer.</span>
<span class="sd">                             ⇣</span>
<span class="sd">                                                           pydna.assembly.Assembly</span>
<span class="sd">        +&gt;       &lt;=       -&gt;       &lt;-        </span>
<span class="sd">         Amplicon1         Amplicon3                  -Amplicon1Amplicon2Amplicon3-</span>
<span class="sd">                  Amplicon2                      ➤   |                             |</span>
<span class="sd">                 -&gt;       &lt;-                          -----------------------------</span>
<span class="sd">        </span>
<span class="sd">        </span>


<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    f : list of :mod:`pydna.amplicon.Amplicon` and other Dseqrecord like objects</span>
<span class="sd">        list Amplicon and Dseqrecord object for which fusion primers should be constructed.</span>

<span class="sd">    overlap : int, optional</span>
<span class="sd">        Length of required overlap between fragments.</span>

<span class="sd">    maxlink : int, optional</span>
<span class="sd">        Maximum length of spacer sequences that may be present in f. These will be included in tails for designed primers.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    seqs : list of :mod:`pydna.amplicon.Amplicon` and other Dseqrecord like objects :mod:`pydna.amplicon.Amplicon` objects</span>

<span class="sd">        ::</span>

<span class="sd">          [Amplicon1,</span>
<span class="sd">           Amplicon2, ...]</span>


<span class="sd">    Examples</span>
<span class="sd">    --------</span>

<span class="sd">    &gt;&gt;&gt; from pydna.dseqrecord import Dseqrecord    </span>
<span class="sd">    &gt;&gt;&gt; from pydna.design import primer_design</span>
<span class="sd">    &gt;&gt;&gt; a=primer_design(Dseqrecord(&quot;atgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatg&quot;))</span>
<span class="sd">    &gt;&gt;&gt; b=primer_design(Dseqrecord(&quot;ccaaacccaccaggtaccttatgtaagtacttcaagtcgccagaagacttcttggtcaagttgcc&quot;))</span>
<span class="sd">    &gt;&gt;&gt; c=primer_design(Dseqrecord(&quot;tgtactggtgctgaaccttgtatcaagttgggtgttgacgccattgccccaggtggtcgtttcgtt&quot;))</span>
<span class="sd">    &gt;&gt;&gt; from pydna.design import assembly_fragments</span>
<span class="sd">    &gt;&gt;&gt; # We would like a circular recombination, so the first sequence has to be repeated</span>
<span class="sd">    &gt;&gt;&gt; fa1,fb,fc,fa2 = assembly_fragments([a,b,c,a])</span>
<span class="sd">    &gt;&gt;&gt; # Since all fragments are Amplicons, we need to extract the rp of the 1st and fp of the last fragments.</span>
<span class="sd">    &gt;&gt;&gt; from pydna.amplify import pcr</span>
<span class="sd">    &gt;&gt;&gt; fa = pcr(fa2.forward_primer, fa1.reverse_primer, a)</span>
<span class="sd">    &gt;&gt;&gt; [fa,fb,fc]</span>
<span class="sd">    [Amplicon(100), Amplicon(101), Amplicon(102)]</span>
<span class="sd">    &gt;&gt;&gt; from pydna.assembly import Assembly</span>
<span class="sd">    &gt;&gt;&gt; assemblyobj = Assembly([fa,fb,fc])</span>
<span class="sd">    &gt;&gt;&gt; assemblyobj</span>
<span class="sd">    Assembly</span>
<span class="sd">    fragments....: 100bp 101bp 102bp</span>
<span class="sd">    limit(bp)....: 25</span>
<span class="sd">    G.nodes......: 6</span>
<span class="sd">    algorithm....: common_sub_strings</span>
<span class="sd">    &gt;&gt;&gt; assemblyobj.assemble_linear()</span>
<span class="sd">    [Contig(-231), Contig(-166), Contig(-36)]</span>
<span class="sd">    &gt;&gt;&gt; assemblyobj.assemble_circular()[0].cseguid()</span>
<span class="sd">    &#39;V3Mi8zilejgyoH833UbjJOtDMbc&#39;</span>
<span class="sd">    &gt;&gt;&gt; (a+b+c).looped().cseguid()</span>
<span class="sd">    &#39;V3Mi8zilejgyoH833UbjJOtDMbc&#39;</span>
<span class="sd">    &gt;&gt;&gt; print(assemblyobj.assemble_circular()[0].figure())</span>
<span class="sd">     -|100bp_PCR_prod|36</span>
<span class="sd">    |                 \\/</span>
<span class="sd">    |                 /\\</span>
<span class="sd">    |                 36|101bp_PCR_prod|36</span>
<span class="sd">    |                                   \\/</span>
<span class="sd">    |                                   /\\</span>
<span class="sd">    |                                   36|102bp_PCR_prod|36</span>
<span class="sd">    |                                                     \\/</span>
<span class="sd">    |                                                     /\\</span>
<span class="sd">    |                                                     36-</span>
<span class="sd">    |                                                        |</span>
<span class="sd">     --------------------------------------------------------</span>
<span class="sd">    &gt;&gt;&gt;</span>

<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># sanity check for arguments</span>
    <span class="n">nf</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">f</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">item</span><span class="p">)</span><span class="o">&gt;</span><span class="n">maxlink</span><span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="nb">hasattr</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="s2">&quot;template&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="s2">&quot;template&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">nf</span><span class="p">,</span><span class="n">nf</span><span class="p">[</span><span class="mi">1</span><span class="p">:])):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Every second fragment larger than maxlink has to be an Amplicon object&quot;</span><span class="p">)</span>
    
    <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;### assembly fragments ###&quot;</span><span class="p">)</span>
    <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;overlap     = </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">overlap</span><span class="p">)</span>
    <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;max_link    = </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">maxlink</span><span class="p">)</span>
    
    <span class="n">f</span> <span class="o">=</span> <span class="p">[</span><span class="n">_copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">f</span><span class="p">]</span>
    
    <span class="n">first_fragment_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">last_fragment_length</span>  <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
    
    <span class="k">if</span> <span class="n">first_fragment_length</span><span class="o">&lt;=</span><span class="n">maxlink</span><span class="p">:</span>
        <span class="c1"># first fragment should be removed and added to second fragment (new first fragment) forward primer</span>
        <span class="n">f</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">forward_primer</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">_data</span> <span class="o">+</span> <span class="n">f</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">forward_primer</span>
        <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;first fragment removed since len(f[0]) = </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">first_fragment_length</span><span class="p">)</span>
        <span class="n">f</span><span class="o">=</span><span class="n">f</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;first fragment stays since len(f[0]) = </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">first_fragment_length</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">last_fragment_length</span><span class="o">&lt;=</span><span class="n">maxlink</span><span class="p">:</span>
        <span class="n">f</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">reverse_primer</span> <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">_data</span> <span class="o">+</span> <span class="n">f</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">reverse_primer</span>
        <span class="n">f</span><span class="o">=</span><span class="n">f</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>  
        <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;last fragment removed since len(f[</span><span class="si">%s</span><span class="s2">]) = </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="p">),</span> <span class="n">last_fragment_length</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;last fragment stays since len(f[</span><span class="si">%s</span><span class="s2">]) = </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="p">),</span><span class="n">last_fragment_length</span><span class="p">)</span>
        
    <span class="n">empty</span> <span class="o">=</span> <span class="n">_Dseqrecord</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>    
    
    <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
    
    <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;loop through fragments in groups of three:&quot;</span><span class="p">)</span>
    
    <span class="n">tail_length</span> <span class="o">=</span> <span class="n">_math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">overlap</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span>
    
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>

        <span class="n">first</span>  <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> 
        <span class="n">secnd</span>  <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span>

        <span class="n">secnd_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">secnd</span><span class="p">)</span>
     
        <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span> <span class="s2">&quot;first = </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">first</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
        <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span> <span class="s2">&quot;secnd = </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">secnd</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
        
        <span class="k">if</span> <span class="n">secnd_len</span> <span class="o">&lt;=</span> <span class="n">maxlink</span><span class="p">:</span>  
            <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;secnd is smaller or equal to maxlink; should be added to primer(s)&quot;</span><span class="p">)</span>
            <span class="n">third</span>  <span class="o">=</span> <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">2</span><span class="p">]</span>
            <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span> <span class="s2">&quot;third = </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">third</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
            <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="s2">&quot;template&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">third</span><span class="p">,</span> <span class="s2">&quot;template&quot;</span><span class="p">):</span>
                <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;secnd is is flanked by amplicons, so half of secnd should be added each flanking primers&quot;</span><span class="p">)</span>
                
                <span class="n">first</span><span class="o">.</span><span class="n">reverse_primer</span> <span class="o">=</span> <span class="n">secnd</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="n">secnd_len</span><span class="o">//</span><span class="mi">2</span><span class="p">:]</span> <span class="o">+</span> <span class="n">first</span><span class="o">.</span><span class="n">reverse_primer</span>
                <span class="n">third</span><span class="o">.</span><span class="n">forward_primer</span> <span class="o">=</span>      <span class="n">secnd</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="n">secnd_len</span><span class="o">//</span><span class="mi">2</span><span class="p">:]</span> <span class="o">+</span> <span class="n">third</span><span class="o">.</span><span class="n">forward_primer</span>
                
                <span class="n">lnk</span> <span class="o">=</span> <span class="p">(</span><span class="n">third</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">_data</span><span class="o">+</span><span class="n">secnd</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">_data</span><span class="p">[:</span><span class="n">secnd_len</span><span class="o">//</span><span class="mi">2</span><span class="p">])[</span><span class="o">-</span><span class="n">tail_length</span><span class="p">:]</span>
                <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;1 </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">lnk</span><span class="p">)</span>
                <span class="n">first</span><span class="o">.</span><span class="n">reverse_primer</span> <span class="o">=</span> <span class="n">lnk</span> <span class="o">+</span> <span class="n">first</span><span class="o">.</span><span class="n">reverse_primer</span>
                
                <span class="n">lnk</span> <span class="o">=</span>  <span class="p">(</span><span class="n">first</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">_data</span> <span class="o">+</span> <span class="n">secnd</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">_data</span><span class="p">[:</span><span class="n">secnd_len</span><span class="o">//</span><span class="mi">2</span><span class="p">])[</span><span class="o">-</span><span class="n">tail_length</span><span class="p">:]</span>
                <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;2 </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">lnk</span><span class="p">)</span>
                <span class="n">third</span><span class="o">.</span><span class="n">forward_primer</span> <span class="o">=</span> <span class="n">lnk</span> <span class="o">+</span> <span class="n">third</span><span class="o">.</span><span class="n">forward_primer</span>                
                
            <span class="k">elif</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">first</span> <span class="p">,</span> <span class="s2">&quot;template&quot;</span><span class="p">):</span>
                <span class="n">first</span><span class="o">.</span><span class="n">reverse_primer</span> <span class="o">=</span> <span class="n">secnd</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">_data</span> <span class="o">+</span> <span class="n">first</span><span class="o">.</span><span class="n">reverse_primer</span>
                <span class="n">lnk</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">third</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="n">overlap</span><span class="p">]</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">())</span>
                <span class="n">first</span><span class="o">.</span><span class="n">reverse_primer</span> <span class="o">=</span> <span class="n">lnk</span> <span class="o">+</span> <span class="n">first</span><span class="o">.</span><span class="n">reverse_primer</span>
            <span class="k">elif</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">third</span> <span class="p">,</span> <span class="s2">&quot;template&quot;</span><span class="p">):</span>
               <span class="n">third</span><span class="o">.</span><span class="n">forward_primer</span> <span class="o">=</span> <span class="n">secnd</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">_data</span> <span class="o">+</span> <span class="n">third</span><span class="o">.</span><span class="n">forward_primer</span>
               <span class="n">lnk</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">first</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="o">-</span><span class="n">overlap</span><span class="p">:])</span>
               <span class="n">third</span><span class="o">.</span><span class="n">forward_primer</span> <span class="o">=</span> <span class="n">lnk</span> <span class="o">+</span> <span class="n">third</span><span class="o">.</span><span class="n">forward_primer</span>
            <span class="n">secnd</span><span class="o">=</span><span class="n">empty</span>
            <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">2</span><span class="p">]</span> <span class="o">=</span> <span class="n">third</span>
        <span class="k">else</span><span class="p">:</span>                    <span class="c1"># secnd is larger than maxlink</span>
            <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">first</span><span class="p">,</span> <span class="s2">&quot;template&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">secnd</span><span class="p">,</span> <span class="s2">&quot;template&quot;</span><span class="p">):</span>
                <span class="n">lnk</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">first</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="o">-</span><span class="n">tail_length</span><span class="p">:])</span>
                <span class="c1">#_module_logger.debug(&quot;4 %s&quot;, lnk)</span>
                <span class="n">secnd</span><span class="o">.</span><span class="n">forward_primer</span> <span class="o">=</span> <span class="n">lnk</span> <span class="o">+</span> <span class="n">secnd</span><span class="o">.</span><span class="n">forward_primer</span>
                <span class="n">lnk</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">secnd</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="n">tail_length</span><span class="p">]</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">())</span>
                <span class="c1">#_module_logger.debug(&quot;5 %s&quot;, lnk)</span>
                <span class="n">first</span><span class="o">.</span><span class="n">reverse_primer</span> <span class="o">=</span> <span class="n">lnk</span> <span class="o">+</span> <span class="n">first</span><span class="o">.</span><span class="n">reverse_primer</span>            
            <span class="k">elif</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">first</span> <span class="p">,</span> <span class="s2">&quot;template&quot;</span><span class="p">):</span>
                <span class="n">lnk</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">secnd</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="n">overlap</span><span class="p">]</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">())</span>
                <span class="c1">#_module_logger.debug(&quot;6 %s&quot;, lnk)</span>
                <span class="n">first</span><span class="o">.</span><span class="n">reverse_primer</span> <span class="o">=</span> <span class="n">lnk</span> <span class="o">+</span> <span class="n">first</span><span class="o">.</span><span class="n">reverse_primer</span>                
            <span class="k">elif</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">secnd</span> <span class="p">,</span> <span class="s2">&quot;template&quot;</span><span class="p">):</span>
                <span class="n">lnk</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">first</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="o">-</span><span class="n">overlap</span><span class="p">:])</span>
                <span class="c1">#_module_logger.debug(&quot;7 %s&quot;, lnk)</span>
                <span class="n">secnd</span><span class="o">.</span><span class="n">forward_primer</span> <span class="o">=</span> <span class="n">lnk</span> <span class="o">+</span> <span class="n">secnd</span><span class="o">.</span><span class="n">forward_primer</span>
        <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>   <span class="o">=</span> <span class="n">first</span>
        <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">secnd</span>

        
    <span class="n">_module_logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;loop ended&quot;</span><span class="p">)</span>
    
    <span class="n">f</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">f</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">item</span><span class="p">)]</span>
    
    <span class="k">return</span> <span class="p">[</span><span class="n">_pcr</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">forward_primer</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">reverse_primer</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">template</span><span class="p">)</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="s2">&quot;template&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="n">p</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">f</span><span class="p">]</span></div>

<div class="viewcode-block" id="circular_assembly_fragments"><a class="viewcode-back" href="../../index.html#pydna.design.circular_assembly_fragments">[docs]</a><span class="k">def</span> <span class="nf">circular_assembly_fragments</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">overlap</span><span class="o">=</span><span class="mi">35</span><span class="p">,</span> <span class="n">maxlink</span><span class="o">=</span><span class="mi">40</span><span class="p">):</span>
    
    <span class="n">fragments</span> <span class="o">=</span> <span class="n">assembly_fragments</span><span class="p">(</span> <span class="n">f</span><span class="o">+</span><span class="n">f</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">1</span><span class="p">],</span> <span class="n">overlap</span><span class="o">=</span><span class="n">overlap</span><span class="p">,</span> <span class="n">maxlink</span><span class="o">=</span><span class="n">maxlink</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span> <span class="n">fragments</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;template&quot;</span><span class="p">):</span>
        <span class="n">fragments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">_pcr</span><span class="p">(</span> <span class="p">(</span><span class="n">fragments</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">forward_primer</span><span class="p">,</span> <span class="n">fragments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">reverse_primer</span><span class="p">),</span> <span class="n">fragments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">fragments</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span></div>


<span class="k">if</span> <span class="vm">__name__</span><span class="o">==</span><span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">os</span> <span class="k">as</span> <span class="nn">_os</span>
    <span class="n">cached</span> <span class="o">=</span> <span class="n">_os</span><span class="o">.</span><span class="n">getenv</span><span class="p">(</span><span class="s2">&quot;pydna_cached_funcs&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">_os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;pydna_cached_funcs&quot;</span><span class="p">]</span><span class="o">=</span><span class="s2">&quot;&quot;</span>
    <span class="kn">import</span> <span class="nn">doctest</span>
    <span class="n">doctest</span><span class="o">.</span><span class="n">testmod</span><span class="p">(</span><span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">optionflags</span><span class="o">=</span><span class="n">doctest</span><span class="o">.</span><span class="n">ELLIPSIS</span><span class="p">)</span>
    <span class="n">_os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;pydna_cached_funcs&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">cach</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">Pydna 3.0.1+2.g47ae1f65a3.dirty documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../index.html" >Module code</a> &#187;</li>
          <li class="nav-item nav-item-2"><a href="../pydna.html" >pydna</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2013-2019, Björn Johansson.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.1.2.
    </div>
  </body>
</html>