
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Welcome to Pydna’s documentation! &#8212; Pydna 3.0.1+2.g47ae1f65a3.dirty documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">Pydna 3.0.1+2.g47ae1f65a3.dirty documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="welcome-to-pydna-s-documentation">
<h1>Welcome to Pydna’s documentation!<a class="headerlink" href="#welcome-to-pydna-s-documentation" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-pydna">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-pydna" title="Permalink to this headline">¶</a></h2>
<dl class="field-list simple">
<dt class="field-odd">copyright</dt>
<dd class="field-odd"><p>Copyright 2013 - 2019 by Björn Johansson. All rights reserved.</p>
</dd>
<dt class="field-even">license</dt>
<dd class="field-even"><p>This code is part of the pydna package, governed by the
license in LICENSE.txt that should be included as part of this package.</p>
</dd>
</dl>
<div class="section" id="pydna">
<h3>pydna<a class="headerlink" href="#pydna" title="Permalink to this headline">¶</a></h3>
<p>Pydna is a python package providing code for simulation of the creation of recombinant DNA molecules 
using <a class="reference external" href="https://en.wikipedia.org/wiki/Molecular_biology">molecular biology</a> techniques.</p>
<dl class="simple">
<dt>Provided:</dt><dd><ol class="arabic simple">
<li><p>PCR simulation</p></li>
<li><p>Assembly simulation based on shared identical sequences</p></li>
<li><p>Primer design for amplification of a given sequence</p></li>
<li><p>Automatic design of primer tails for Gibson assembly or homologous recombination.</p></li>
<li><p>Restriction digestion and cut&amp;paste cloning</p></li>
<li><p>Agarose gel simulation</p></li>
<li><p>Download sequences from Genbank</p></li>
<li><p>Parsing various sequence formats including the capacity to handle broken Genbank format</p></li>
</ol>
</dd>
</dl>
<div class="section" id="pydna-package-layout">
<h4>pydna package layout<a class="headerlink" href="#pydna-package-layout" title="Permalink to this headline">¶</a></h4>
<p>The most important modules and how to import functions or classes from them are listed below.
Class names starts with a capital letter, functions with a lowercase letter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>from pydna.module import function
from pydna.module import Class
      
Example: from pydna.gel import Gel

pydna
   ├── amplify
   │         ├── Anneal
   │         └── pcr
   ├── assembly
   │          └── Assembly
   ├── design
   │        ├── assembly_fragments
   │        └── primer_design
   ├── download
   │          └── download_text
   ├── dseqrecord
   │            └── Dseqrecord
   ├── gel
   │     └── Gel
   ├── genbank
   │         ├── genbank
   │         └── Genbank
   ├── parsers
   │         ├── parse
   │         └── parse_primers
   └── readers
             ├── read
             └── read_primers
</pre></div>
</div>
</div>
<div class="section" id="how-to-use-the-documentation">
<h4>How to use the documentation<a class="headerlink" href="#how-to-use-the-documentation" title="Permalink to this headline">¶</a></h4>
<p>Documentation is available as docstrings provided in the source code for each module.
These docstrings can be inspected by reading the source code directly. 
See further below on how to obtain the code for pydna.</p>
<p>In the python shell, use the built-in <code class="docutils literal notranslate"><span class="pre">help</span></code> function to view a function’s docstring:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna</span> <span class="k">import</span> <span class="n">readers</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">help</span><span class="p">(</span><span class="n">readers</span><span class="o">.</span><span class="n">read</span><span class="p">)</span>
<span class="gp">... </span>
</pre></div>
</div>
<p>The doctrings are also used to provide an automaticly generated reference manual available online at
<a class="reference external" href="https://pydna.readthedocs.io">read the docs</a>.</p>
<p>Docstrings can be explored using <a class="reference external" href="http://ipython.org/">IPython</a>, an advanced Python shell with
TAB-completion and introspection capabilities. To see which functions are available in <cite>pydna</cite>, 
type <cite>pydna.&lt;TAB&gt;</cite> (where <cite>&lt;TAB&gt;</cite> refers to the TAB key).
Use <cite>pydna.open_config_folder?&lt;ENTER&gt;`to view the docstring or `pydna.open_config_folder??&lt;ENTER&gt;</cite> to view the source code.</p>
<p>In the <a class="reference external" href="https://github.com/spyder-ide/spyder">Spyder IDE</a> it is possible to place the cursor immediately before 
the name of a module,class or function and press ctrl+i to bring up docstrings in a separate window in Spyder</p>
<p>Code snippets are indicated by three greater-than signs:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">=</span><span class="mi">41</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="o">+</span><span class="mi">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">42</span>
</pre></div>
</div>
</div>
<div class="section" id="pydna-source-code">
<h4>pydna source code<a class="headerlink" href="#pydna-source-code" title="Permalink to this headline">¶</a></h4>
<p>The pydna source code is <a class="reference external" href="https://github.com/BjornFJohansson/pydna">available on Github</a>.</p>
</div>
<div class="section" id="how-to-get-more-help">
<h4>How to get more help<a class="headerlink" href="#how-to-get-more-help" title="Permalink to this headline">¶</a></h4>
<p>Please join the <a class="reference external" href="https://groups.google.com/forum/#!forum/pydna">Google grooup</a> for pydna, this is the preferred 
location for help. If you find bugs in pydna itself, open an issue at the <a class="reference external" href="https://github.com/BjornFJohansson/pydna/issues">Github repository</a>.</p>
</div>
<div class="section" id="examples-of-pydna-in-use">
<h4>Examples of pydna in use<a class="headerlink" href="#examples-of-pydna-in-use" title="Permalink to this headline">¶</a></h4>
<p>See this repository for a collection of <a class="reference external" href="https://github.com/BjornFJohansson/pydna-examples">examples</a>.</p>
<dl class="function">
<dt id="pydna.get_env">
<code class="sig-prename descclassname">pydna.</code><code class="sig-name descname">get_env</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna.html#get_env"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.get_env" title="Permalink to this definition">¶</a></dt>
<dd><p>Calling this function prints a an ascii table containing all environmental
variables set by <cite>pydna</cite> and their values. Pydna related variables have names
that starts with <cite>pydna_</cite></p>
</dd></dl>

<dl class="function">
<dt id="pydna.logo">
<code class="sig-prename descclassname">pydna.</code><code class="sig-name descname">logo</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna.html#logo"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.logo" title="Permalink to this definition">¶</a></dt>
<dd><p>This function returns the ascii-art logotype of pydna.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pydna</span><span class="o">.</span><span class="n">logo</span><span class="p">())</span>
<span class="go">                 _             </span>
<span class="go">                | |            </span>
<span class="go"> ____  _   _  __| |___   __ ___</span>
<span class="go">|  _ \| | | |/ _  |  _ \(____ |</span>
<span class="go">| |_| | |_| ( (_| | | | / ___ |</span>
<span class="go">|  __/ \__  |\____|_| |_\_____|</span>
<span class="go">|_|   (____/</span>
<span class="go">&lt;BLANKLINE&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.open_cache_folder">
<code class="sig-prename descclassname">pydna.</code><code class="sig-name descname">open_cache_folder</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna.html#open_cache_folder"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.open_cache_folder" title="Permalink to this definition">¶</a></dt>
<dd><p>Calling this function opens the pydna cache folder in 
the default file manager. The location for this folder is stored
in the <em>pydna_data_dir</em> environmental variable.</p>
</dd></dl>

<dl class="function">
<dt id="pydna.open_config_folder">
<code class="sig-prename descclassname">pydna.</code><code class="sig-name descname">open_config_folder</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna.html#open_config_folder"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.open_config_folder" title="Permalink to this definition">¶</a></dt>
<dd><p>Calling this function opens the pydna configuration folder in 
the default file manager. The location for this folder is stored
in the <em>pydna_config_dir</em> environmental variable.</p>
<p>The <cite>pydna.ini</cite> file can be edited to make pydna quicker to use.
See the documentation of the :class:configparser.ConfigParser´ class.</p>
<p>Below is the content of a typical <cite>pydna.ini</cite> file on a Linux 
system.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">main</span><span class="p">]</span>
<span class="n">loglevel</span><span class="o">=</span><span class="mi">30</span>
<span class="n">email</span><span class="o">=</span><span class="n">myemail</span><span class="nd">@example</span><span class="o">.</span><span class="n">org</span>
<span class="n">data_dir</span><span class="o">=/</span><span class="n">home</span><span class="o">/</span><span class="n">bjorn</span><span class="o">/.</span><span class="n">local</span><span class="o">/</span><span class="n">share</span><span class="o">/</span><span class="n">pydna</span>
<span class="n">log_dir</span><span class="o">=/</span><span class="n">home</span><span class="o">/</span><span class="n">bjorn</span><span class="o">/.</span><span class="n">cache</span><span class="o">/</span><span class="n">pydna</span><span class="o">/</span><span class="n">log</span>
<span class="n">ape</span><span class="o">=</span><span class="n">tclsh</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">bjorn</span><span class="o">/.</span><span class="n">ApE</span><span class="o">/</span><span class="n">apeextractor</span><span class="o">/</span><span class="n">ApE</span><span class="o">.</span><span class="n">vfs</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">app</span><span class="o">-</span><span class="n">AppMain</span><span class="o">/</span><span class="n">AppMain</span><span class="o">.</span><span class="n">tcl</span>
<span class="n">cached_funcs</span><span class="o">=</span><span class="n">Genbank_nucleotide</span>
<span class="n">primers</span><span class="o">=/</span><span class="n">home</span><span class="o">/</span><span class="n">bjorn</span><span class="o">/</span><span class="n">Dropbox</span><span class="o">/</span><span class="n">wikidata</span><span class="o">/</span><span class="n">PRIMERS</span><span class="o">.</span><span class="n">txt</span>
<span class="n">enzymes</span><span class="o">=/</span><span class="n">home</span><span class="o">/</span><span class="n">bjorn</span><span class="o">/</span><span class="n">Dropbox</span><span class="o">/</span><span class="n">wikidata</span><span class="o">/</span><span class="n">RestrictionEnzymes</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>The email address is set to <a class="reference external" href="mailto:someone&#37;&#52;&#48;example&#46;com">someone<span>&#64;</span>example<span>&#46;</span>com</a> by default. If you change this 
to you own address, the <a class="reference internal" href="#pydna.genbank.genbank" title="pydna.genbank.genbank"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.genbank.genbank()</span></code></a> function can be used 
to download sequences from Genbank directly without having to explicitly add 
the email address.</p>
<p>Pydna can cache results from the following functions or methods:</p>
<ul class="simple">
<li><p><a class="reference internal" href="#pydna.genbank.Genbank.nucleotide" title="pydna.genbank.Genbank.nucleotide"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.genbank.Genbank.nucleotide()</span></code></a>   Genbank_nucleotide</p></li>
<li><p><a class="reference internal" href="#pydna.amplify.Anneal" title="pydna.amplify.Anneal"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.amplify.Anneal()</span></code></a>               amplify_Anneal</p></li>
<li><p><a class="reference internal" href="#pydna.assembly.Assembly" title="pydna.assembly.Assembly"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.assembly.Assembly()</span></code></a>            assembly_Assembly</p></li>
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.download.download_text()</span></code>       download.download_text</p></li>
<li><p><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.synced" title="pydna.dseqrecord.Dseqrecord.synced"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord.synced()</span></code></a> Dseqrecord_synced</p></li>
</ul>
<p>These can be added separated by a comma to the cached_funcs entry in <strong>pydna.ini</strong>
file or the pydna_cached_funcs environment variable.</p>
</dd></dl>

<dl class="function">
<dt id="pydna.open_current_folder">
<code class="sig-prename descclassname">pydna.</code><code class="sig-name descname">open_current_folder</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna.html#open_current_folder"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.open_current_folder" title="Permalink to this definition">¶</a></dt>
<dd><p>Calling this function opens the current working directory
in the default file manager. The location for this folder is 
given by the <a class="reference external" href="https://docs.python.org/3.6/library/os.html#os.getcwd" title="(in Python v3.6)"><code class="xref py py-func docutils literal notranslate"><span class="pre">os.getcwd()</span></code></a> function</p>
</dd></dl>

<dl class="function">
<dt id="pydna.open_log_folder">
<code class="sig-prename descclassname">pydna.</code><code class="sig-name descname">open_log_folder</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna.html#open_log_folder"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.open_log_folder" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
</div>
</div>
<div class="section" id="module-pydna.dseq">
<span id="pydna-dseq-module"></span><h2>pydna.dseq module<a class="headerlink" href="#module-pydna.dseq" title="Permalink to this headline">¶</a></h2>
<p>Provides the Dseq class for handling double stranded DNA sequences. 
Dseq is a subclass of <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Seq.Seq</span></code>. The Dseq class
is mostly useful as a part of the <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord</span></code></a> class which
can hold more meta data.</p>
<p>The Dseq class support the notion of circular and linear DNA topology.</p>
<dl class="class">
<dt id="pydna.dseq.Dseq">
<em class="property">class </em><code class="sig-prename descclassname">pydna.dseq.</code><code class="sig-name descname">Dseq</code><span class="sig-paren">(</span><em class="sig-param">watson</em>, <em class="sig-param">crick=None</em>, <em class="sig-param">ovhg=None</em>, <em class="sig-param">linear=None</em>, <em class="sig-param">circular=None</em>, <em class="sig-param">pos=0</em>, <em class="sig-param">alphabet=IUPACAmbiguousDNA()</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Seq.Seq</span></code></p>
<p>Dseq is a class designed to hold information for a double stranded
DNA fragment. Dseq also holds information describing the topology of
the DNA fragment (linear or circular).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>watson</strong><span class="classifier">str</span></dt><dd><p>a string representing the watson (sense) DNA strand.</p>
</dd>
<dt><strong>crick</strong><span class="classifier">str, optional</span></dt><dd><p>a string representing the crick (antisense) DNA strand.</p>
</dd>
<dt><strong>ovhg</strong><span class="classifier">int, optional</span></dt><dd><p>A positive or negative number to describe the stagger between the watson and crick strands.
see below for a detailed explanation.</p>
</dd>
<dt><strong>linear</strong><span class="classifier">bool, optional</span></dt><dd><p>True indicates that sequence is linear, False that it is circular.</p>
</dd>
<dt><strong>circular</strong><span class="classifier">bool, optional</span></dt><dd><p>True indicates that sequence is circular, False that it is linear.</p>
</dd>
<dt><strong>alphabet</strong><span class="classifier">Bio.Alphabet, optional</span></dt><dd><p>Bio.Alphabet.IUPAC.IUPACAmbiguousDNA from the Biopython package is set as default.</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<p>Dseq is a subclass of the Biopython Seq object. It stores two
strings representing the watson (sense) and crick(antisense) strands.
two properties called linear and circular, and a numeric value ovhg
(overhang) describing the stagger for the watson and crick strand
in the 5’ end of the fragment.</p>
<p>The most common usage is probably to create a Dseq object as a
part of a Dseqrecord object (see <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord</span></code></a>).</p>
<p>There are three ways of creating a Dseq object directly:</p>
<p>Only one argument (string):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="k">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<p>The given string will be interpreted as the watson strand of a
blunt, linear double stranded sequence object. The crick strand
is created automatically from the watson strand.</p>
<p>Two arguments (string, string):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="k">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gggaaat&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">gggaaat</span>
<span class="go">   ttt</span>
</pre></div>
</div>
<p>If both watson and crick are given, but not ovhg an attempt
will be made to find the best annealing between the strands.
There are limitations to this! For long fragments it is quite
slow. The length of the annealing sequences have to be at least
half the length of the shortest of the strands.</p>
<p>Three arguments (string, string, ovhg=int):</p>
<p>The ovhg parameter controls the stagger at the five prime end:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ovhg</span> <span class="o">=</span> <span class="o">-</span><span class="mi">2</span>

<span class="n">XXXXX</span>
  <span class="n">XXX</span>

<span class="n">ovhg</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>

<span class="n">XXXXX</span>
 <span class="n">XXXX</span>

<span class="n">ovhg</span> <span class="o">=</span> <span class="mi">0</span>

<span class="n">XXXXX</span>
<span class="n">XXXXX</span>

<span class="n">ovhg</span> <span class="o">=</span> <span class="mi">1</span>

 <span class="n">XXXX</span>
<span class="n">XXXXX</span>

<span class="n">ovhg</span> <span class="o">=</span> <span class="mi">2</span>

  <span class="n">XXX</span>
<span class="n">XXXXX</span>                          
</pre></div>
</div>
<p>Example of creating Dseq objects with different amounts of stagger:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">agt</span>
<span class="go">  attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">agt</span>
<span class="go"> attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">agt</span>
<span class="go">attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go"> agt</span>
<span class="go">attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">  agt</span>
<span class="go">attca</span>
</pre></div>
</div>
<p>If the ovhg parameter is specified a crick strand also needs to be supplied,
otherwise an exception is raised.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/dist-packages/pydna_/dsdna.py&quot;</span>, line <span class="m">169</span>, in <span class="n">__init__</span>
    <span class="k">else</span><span class="p">:</span>
<span class="gr">ValueError</span>: <span class="n">ovhg defined without crick strand!</span>
</pre></div>
</div>
<p>The default alphabet is set to Biopython IUPACAmbiguousDNA</p>
<p>The shape of the fragment is set by either:</p>
<ol class="arabic simple">
<li><p>linear   = False, True</p></li>
<li><p>circular = True, False</p></li>
</ol>
<p>Note that both ends of the DNA fragment has to be compatible to set
circular = True (or linear = False).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-4)</span>
<span class="go"> aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;ttt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-4)</span>
<span class="go">aaa</span>
<span class="go"> ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">linear</span> <span class="o">=</span> <span class="kc">False</span> <span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(o3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="kc">True</span> <span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(o3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;tttcccc&quot;</span><span class="p">,</span><span class="s2">&quot;aaacccc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-11)</span>
<span class="go">    tttcccc</span>
<span class="go">ccccaaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">ovhg</span>
<span class="go">4</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;ccccttt&quot;</span><span class="p">,</span><span class="s2">&quot;ccccaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-11)</span>
<span class="go">ccccttt</span>
<span class="go">    aaacccc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">ovhg</span>
<span class="go">-4</span>
<span class="gp">&gt;&gt;&gt; </span>  
</pre></div>
</div>
<p>Coercing to string</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<span class="go">&#39;ggggtttcccc&#39;</span>
</pre></div>
</div>
<p>A Dseq object can be longer that either the watson or crick strands.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;--</span> <span class="n">length</span> <span class="o">--&gt;</span>
<span class="n">GATCCTTT</span>
     <span class="n">AAAGCCTAG</span>

<span class="o">&lt;--</span> <span class="n">length</span> <span class="o">--&gt;</span>
      <span class="n">GATCCTTT</span>
<span class="n">AAAGCCCTA</span>
</pre></div>
</div>
<p>The slicing of a linear Dseq object works mostly as it does for a string.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s2">&quot;ggatcc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">&#39;a&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
<span class="go">&#39;at&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[::</span><span class="mi">2</span><span class="p">]</span>
<span class="go">&#39;gac&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="k">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">linear</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-1)</span>
<span class="go">a</span>
<span class="go">t</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
<span class="go">Dseq(-2)</span>
<span class="go">at</span>
<span class="go">ta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">Dseq(-0)</span>
<span class="go">&lt;BLANKLINE&gt;</span>
<span class="go">&lt;BLANKLINE&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[::</span><span class="mi">2</span><span class="p">]</span>
<span class="go">Dseq(-3)</span>
<span class="go">gac</span>
<span class="go">ctg</span>
</pre></div>
</div>
<p>The slicing of a circular Dseq object has a slightly different meaning.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s2">&quot;ggAtCc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span>
<span class="go">Dseq(o6)</span>
<span class="go">ggAtCc</span>
<span class="go">ccTaGg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">4</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-5)</span>
<span class="go">CcggA</span>
<span class="go">GgccT</span>
</pre></div>
</div>
<p>The slice [X:X] produces an empty slice for a string, while this will return
the linearized sequence starting at X:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s2">&quot;ggatcc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span>
<span class="go">Dseq(o6)</span>
<span class="go">ggatcc</span>
<span class="go">cctagg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">3</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-6)</span>
<span class="go">tccgga</span>
<span class="go">aggcct</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
<dl class="method">
<dt id="pydna.dseq.Dseq.T4">
<code class="sig-name descname">T4</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">nucleotides=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.T4"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.T4" title="Permalink to this definition">¶</a></dt>
<dd><p>Fill in five prime protruding ends and chewing back
three prime protruding ends by a DNA polymerase providing both
5’-3’ DNA polymerase activity and 3’-5’ nuclease acitivty
(such as T4 DNA polymerase). This can be done in presence of any
combination of the four A, G, C or T. Removing one or more nucleotides
can facilitate engineering of sticky ends. Default are all four nucleotides together.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>nucleotides</strong><span class="classifier">str</span></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="k">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gatcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">()</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;t&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgat</span>
<span class="go"> tagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcga</span>
<span class="go">  agctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;g&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcg</span>
<span class="go">   gctag</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.circular">
<em class="property">property </em><code class="sig-name descname">circular</code><a class="headerlink" href="#pydna.dseq.Dseq.circular" title="Permalink to this definition">¶</a></dt>
<dd><p>The circular property can not be set directly. 
Use <a class="reference internal" href="#pydna.dseq.Dseq.looped" title="pydna.dseq.Dseq.looped"><code class="xref py py-meth docutils literal notranslate"><span class="pre">looped()</span></code></a> to create a circular Dseq object</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.cut">
<code class="sig-name descname">cut</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">*enzymes</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.cut"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.cut" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of linear Dseq fragments produced in the digestion.
If there are no cuts, an empty list is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>enzymes</strong><span class="classifier">enzyme object or iterable of such objects</span></dt><dd><p>A Bio.Restriction.XXX restriction objects or iterable.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>frags</strong><span class="classifier">list</span></dt><dd><p>list of Dseq objects formed by the digestion</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="k">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;ggatccnnngaattc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Dseq(-15)</span>
<span class="go">ggatccnnngaattc</span>
<span class="go">cctaggnnncttaag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Restriction</span> <span class="k">import</span> <span class="n">BamHI</span><span class="p">,</span><span class="n">EcoRI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">))</span>
<span class="go">&lt;class &#39;tuple&#39;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">frag</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">):</span> <span class="nb">print</span><span class="p">(</span><span class="nb">repr</span><span class="p">(</span><span class="n">frag</span><span class="p">))</span>
<span class="go">Dseq(-5)</span>
<span class="go">g</span>
<span class="go">cctag</span>
<span class="go">Dseq(-14)</span>
<span class="go">gatccnnngaattc</span>
<span class="go">    gnnncttaag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">,</span> <span class="n">BamHI</span><span class="p">)</span> <span class="o">==</span>  <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">,</span> <span class="n">EcoRI</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">,</span> <span class="n">BamHI</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="o">+</span><span class="n">c</span>
<span class="go">Dseq(-15)</span>
<span class="go">ggatccnnngaattc</span>
<span class="go">cctaggnnncttaag</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.cutters">
<code class="sig-name descname">cutters</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">batch=RestrictionBatch(['AanI'</em>, <em class="sig-param">'AarI'</em>, <em class="sig-param">'AasI'</em>, <em class="sig-param">'AatII'</em>, <em class="sig-param">'AbaSI'</em>, <em class="sig-param">'AbsI'</em>, <em class="sig-param">'Acc16I'</em>, <em class="sig-param">'Acc36I'</em>, <em class="sig-param">'Acc65I'</em>, <em class="sig-param">'AccB1I'</em>, <em class="sig-param">'AccB7I'</em>, <em class="sig-param">'AccBSI'</em>, <em class="sig-param">'AccI'</em>, <em class="sig-param">'AccII'</em>, <em class="sig-param">'AccIII'</em>, <em class="sig-param">'AciI'</em>, <em class="sig-param">'AclI'</em>, <em class="sig-param">'AclWI'</em>, <em class="sig-param">'AcoI'</em>, <em class="sig-param">'AcsI'</em>, <em class="sig-param">'AcuI'</em>, <em class="sig-param">'AcvI'</em>, <em class="sig-param">'AcyI'</em>, <em class="sig-param">'AdeI'</em>, <em class="sig-param">'AfaI'</em>, <em class="sig-param">'AfeI'</em>, <em class="sig-param">'AfiI'</em>, <em class="sig-param">'AflII'</em>, <em class="sig-param">'AflIII'</em>, <em class="sig-param">'AgeI'</em>, <em class="sig-param">'AgsI'</em>, <em class="sig-param">'AhdI'</em>, <em class="sig-param">'AhlI'</em>, <em class="sig-param">'AjiI'</em>, <em class="sig-param">'AjnI'</em>, <em class="sig-param">'AjuI'</em>, <em class="sig-param">'AleI'</em>, <em class="sig-param">'AloI'</em>, <em class="sig-param">'AluBI'</em>, <em class="sig-param">'AluI'</em>, <em class="sig-param">'Alw21I'</em>, <em class="sig-param">'Alw26I'</em>, <em class="sig-param">'Alw44I'</em>, <em class="sig-param">'AlwI'</em>, <em class="sig-param">'AlwNI'</em>, <em class="sig-param">'Ama87I'</em>, <em class="sig-param">'Aor13HI'</em>, <em class="sig-param">'Aor51HI'</em>, <em class="sig-param">'AoxI'</em>, <em class="sig-param">'ApaI'</em>, <em class="sig-param">'ApaLI'</em>, <em class="sig-param">'ApeKI'</em>, <em class="sig-param">'ApoI'</em>, <em class="sig-param">'ArsI'</em>, <em class="sig-param">'AscI'</em>, <em class="sig-param">'AseI'</em>, <em class="sig-param">'AsiGI'</em>, <em class="sig-param">'AsiSI'</em>, <em class="sig-param">'Asp700I'</em>, <em class="sig-param">'Asp718I'</em>, <em class="sig-param">'AspA2I'</em>, <em class="sig-param">'AspLEI'</em>, <em class="sig-param">'AspS9I'</em>, <em class="sig-param">'AsuC2I'</em>, <em class="sig-param">'AsuHPI'</em>, <em class="sig-param">'AsuII'</em>, <em class="sig-param">'AsuNHI'</em>, <em class="sig-param">'AvaI'</em>, <em class="sig-param">'AvaII'</em>, <em class="sig-param">'AvrII'</em>, <em class="sig-param">'AxyI'</em>, <em class="sig-param">'BaeGI'</em>, <em class="sig-param">'BaeI'</em>, <em class="sig-param">'BalI'</em>, <em class="sig-param">'BamHI'</em>, <em class="sig-param">'BanI'</em>, <em class="sig-param">'BanII'</em>, <em class="sig-param">'BarI'</em>, <em class="sig-param">'BauI'</em>, <em class="sig-param">'BbrPI'</em>, <em class="sig-param">'BbsI'</em>, <em class="sig-param">'Bbv12I'</em>, <em class="sig-param">'BbvCI'</em>, <em class="sig-param">'BbvI'</em>, <em class="sig-param">'BccI'</em>, <em class="sig-param">'BceAI'</em>, <em class="sig-param">'BcgI'</em>, <em class="sig-param">'BciT130I'</em>, <em class="sig-param">'BciVI'</em>, <em class="sig-param">'BclI'</em>, <em class="sig-param">'BcnI'</em>, <em class="sig-param">'BcoDI'</em>, <em class="sig-param">'BcuI'</em>, <em class="sig-param">'BfaI'</em>, <em class="sig-param">'BfmI'</em>, <em class="sig-param">'BfoI'</em>, <em class="sig-param">'BfrI'</em>, <em class="sig-param">'BfuAI'</em>, <em class="sig-param">'BfuI'</em>, <em class="sig-param">'BglI'</em>, <em class="sig-param">'BglII'</em>, <em class="sig-param">'BisI'</em>, <em class="sig-param">'BlnI'</em>, <em class="sig-param">'BlpI'</em>, <em class="sig-param">'BlsI'</em>, <em class="sig-param">'BmcAI'</em>, <em class="sig-param">'Bme1390I'</em>, <em class="sig-param">'Bme18I'</em>, <em class="sig-param">'BmeRI'</em>, <em class="sig-param">'BmeT110I'</em>, <em class="sig-param">'BmgBI'</em>, <em class="sig-param">'BmgT120I'</em>, <em class="sig-param">'BmiI'</em>, <em class="sig-param">'BmrFI'</em>, <em class="sig-param">'BmrI'</em>, <em class="sig-param">'BmsI'</em>, <em class="sig-param">'BmtI'</em>, <em class="sig-param">'BmuI'</em>, <em class="sig-param">'BoxI'</em>, <em class="sig-param">'BpiI'</em>, <em class="sig-param">'BplI'</em>, <em class="sig-param">'BpmI'</em>, <em class="sig-param">'Bpu10I'</em>, <em class="sig-param">'Bpu1102I'</em>, <em class="sig-param">'Bpu14I'</em>, <em class="sig-param">'BpuEI'</em>, <em class="sig-param">'BpuMI'</em>, <em class="sig-param">'Bsa29I'</em>, <em class="sig-param">'BsaAI'</em>, <em class="sig-param">'BsaBI'</em>, <em class="sig-param">'BsaHI'</em>, <em class="sig-param">'BsaI'</em>, <em class="sig-param">'BsaJI'</em>, <em class="sig-param">'BsaWI'</em>, <em class="sig-param">'BsaXI'</em>, <em class="sig-param">'Bsc4I'</em>, <em class="sig-param">'Bse118I'</em>, <em class="sig-param">'Bse1I'</em>, <em class="sig-param">'Bse21I'</em>, <em class="sig-param">'Bse3DI'</em>, <em class="sig-param">'Bse8I'</em>, <em class="sig-param">'BseAI'</em>, <em class="sig-param">'BseBI'</em>, <em class="sig-param">'BseCI'</em>, <em class="sig-param">'BseDI'</em>, <em class="sig-param">'BseGI'</em>, <em class="sig-param">'BseJI'</em>, <em class="sig-param">'BseLI'</em>, <em class="sig-param">'BseMI'</em>, <em class="sig-param">'BseMII'</em>, <em class="sig-param">'BseNI'</em>, <em class="sig-param">'BsePI'</em>, <em class="sig-param">'BseRI'</em>, <em class="sig-param">'BseSI'</em>, <em class="sig-param">'BseX3I'</em>, <em class="sig-param">'BseXI'</em>, <em class="sig-param">'BseYI'</em>, <em class="sig-param">'BsgI'</em>, <em class="sig-param">'Bsh1236I'</em>, <em class="sig-param">'Bsh1285I'</em>, <em class="sig-param">'BshFI'</em>, <em class="sig-param">'BshNI'</em>, <em class="sig-param">'BshTI'</em>, <em class="sig-param">'BshVI'</em>, <em class="sig-param">'BsiEI'</em>, <em class="sig-param">'BsiHKAI'</em>, <em class="sig-param">'BsiHKCI'</em>, <em class="sig-param">'BsiSI'</em>, <em class="sig-param">'BsiWI'</em>, <em class="sig-param">'BslFI'</em>, <em class="sig-param">'BslI'</em>, <em class="sig-param">'BsmAI'</em>, <em class="sig-param">'BsmBI'</em>, <em class="sig-param">'BsmFI'</em>, <em class="sig-param">'BsmI'</em>, <em class="sig-param">'BsnI'</em>, <em class="sig-param">'Bso31I'</em>, <em class="sig-param">'BsoBI'</em>, <em class="sig-param">'Bsp119I'</em>, <em class="sig-param">'Bsp120I'</em>, <em class="sig-param">'Bsp1286I'</em>, <em class="sig-param">'Bsp13I'</em>, <em class="sig-param">'Bsp1407I'</em>, <em class="sig-param">'Bsp143I'</em>, <em class="sig-param">'Bsp1720I'</em>, <em class="sig-param">'Bsp19I'</em>, <em class="sig-param">'Bsp68I'</em>, <em class="sig-param">'BspACI'</em>, <em class="sig-param">'BspANI'</em>, <em class="sig-param">'BspCNI'</em>, <em class="sig-param">'BspDI'</em>, <em class="sig-param">'BspEI'</em>, <em class="sig-param">'BspFNI'</em>, <em class="sig-param">'BspHI'</em>, <em class="sig-param">'BspLI'</em>, <em class="sig-param">'BspMAI'</em>, <em class="sig-param">'BspMI'</em>, <em class="sig-param">'BspOI'</em>, <em class="sig-param">'BspPI'</em>, <em class="sig-param">'BspQI'</em>, <em class="sig-param">'BspT104I'</em>, <em class="sig-param">'BspT107I'</em>, <em class="sig-param">'BspTI'</em>, <em class="sig-param">'BspTNI'</em>, <em class="sig-param">'BsrBI'</em>, <em class="sig-param">'BsrDI'</em>, <em class="sig-param">'BsrFI'</em>, <em class="sig-param">'BsrGI'</em>, <em class="sig-param">'BsrI'</em>, <em class="sig-param">'BssAI'</em>, <em class="sig-param">'BssECI'</em>, <em class="sig-param">'BssHII'</em>, <em class="sig-param">'BssMI'</em>, <em class="sig-param">'BssNAI'</em>, <em class="sig-param">'BssNI'</em>, <em class="sig-param">'BssSI'</em>, <em class="sig-param">'BssT1I'</em>, <em class="sig-param">'Bst1107I'</em>, <em class="sig-param">'Bst2BI'</em>, <em class="sig-param">'Bst2UI'</em>, <em class="sig-param">'Bst4CI'</em>, <em class="sig-param">'Bst6I'</em>, <em class="sig-param">'BstACI'</em>, <em class="sig-param">'BstAFI'</em>, <em class="sig-param">'BstAPI'</em>, <em class="sig-param">'BstAUI'</em>, <em class="sig-param">'BstBAI'</em>, <em class="sig-param">'BstBI'</em>, <em class="sig-param">'BstC8I'</em>, <em class="sig-param">'BstDEI'</em>, <em class="sig-param">'BstDSI'</em>, <em class="sig-param">'BstEII'</em>, <em class="sig-param">'BstENI'</em>, <em class="sig-param">'BstF5I'</em>, <em class="sig-param">'BstFNI'</em>, <em class="sig-param">'BstH2I'</em>, <em class="sig-param">'BstHHI'</em>, <em class="sig-param">'BstKTI'</em>, <em class="sig-param">'BstMAI'</em>, <em class="sig-param">'BstMBI'</em>, <em class="sig-param">'BstMCI'</em>, <em class="sig-param">'BstMWI'</em>, <em class="sig-param">'BstNI'</em>, <em class="sig-param">'BstNSI'</em>, <em class="sig-param">'BstPAI'</em>, <em class="sig-param">'BstPI'</em>, <em class="sig-param">'BstSCI'</em>, <em class="sig-param">'BstSFI'</em>, <em class="sig-param">'BstSLI'</em>, <em class="sig-param">'BstSNI'</em>, <em class="sig-param">'BstUI'</em>, <em class="sig-param">'BstV1I'</em>, <em class="sig-param">'BstV2I'</em>, <em class="sig-param">'BstX2I'</em>, <em class="sig-param">'BstXI'</em>, <em class="sig-param">'BstYI'</em>, <em class="sig-param">'BstZ17I'</em>, <em class="sig-param">'BstZI'</em>, <em class="sig-param">'Bsu15I'</em>, <em class="sig-param">'Bsu36I'</em>, <em class="sig-param">'BsuI'</em>, <em class="sig-param">'BsuRI'</em>, <em class="sig-param">'BsuTUI'</em>, <em class="sig-param">'BtgI'</em>, <em class="sig-param">'BtgZI'</em>, <em class="sig-param">'BtrI'</em>, <em class="sig-param">'BtsCI'</em>, <em class="sig-param">'BtsI'</em>, <em class="sig-param">'BtsIMutI'</em>, <em class="sig-param">'BtuMI'</em>, <em class="sig-param">'BveI'</em>, <em class="sig-param">'Cac8I'</em>, <em class="sig-param">'CaiI'</em>, <em class="sig-param">'CciI'</em>, <em class="sig-param">'CciNI'</em>, <em class="sig-param">'CfoI'</em>, <em class="sig-param">'Cfr10I'</em>, <em class="sig-param">'Cfr13I'</em>, <em class="sig-param">'Cfr42I'</em>, <em class="sig-param">'Cfr9I'</em>, <em class="sig-param">'ClaI'</em>, <em class="sig-param">'CpoI'</em>, <em class="sig-param">'CseI'</em>, <em class="sig-param">'CsiI'</em>, <em class="sig-param">'Csp6I'</em>, <em class="sig-param">'CspAI'</em>, <em class="sig-param">'CspCI'</em>, <em class="sig-param">'CspI'</em>, <em class="sig-param">'CviAII'</em>, <em class="sig-param">'CviJI'</em>, <em class="sig-param">'CviKI_1'</em>, <em class="sig-param">'CviQI'</em>, <em class="sig-param">'DdeI'</em>, <em class="sig-param">'DinI'</em>, <em class="sig-param">'DpnI'</em>, <em class="sig-param">'DpnII'</em>, <em class="sig-param">'DraI'</em>, <em class="sig-param">'DraIII'</em>, <em class="sig-param">'DrdI'</em>, <em class="sig-param">'DriI'</em>, <em class="sig-param">'DseDI'</em>, <em class="sig-param">'EaeI'</em>, <em class="sig-param">'EagI'</em>, <em class="sig-param">'Eam1104I'</em>, <em class="sig-param">'Eam1105I'</em>, <em class="sig-param">'EarI'</em>, <em class="sig-param">'EciI'</em>, <em class="sig-param">'Ecl136II'</em>, <em class="sig-param">'EclXI'</em>, <em class="sig-param">'Eco105I'</em>, <em class="sig-param">'Eco130I'</em>, <em class="sig-param">'Eco147I'</em>, <em class="sig-param">'Eco24I'</em>, <em class="sig-param">'Eco31I'</em>, <em class="sig-param">'Eco32I'</em>, <em class="sig-param">'Eco47I'</em>, <em class="sig-param">'Eco47III'</em>, <em class="sig-param">'Eco52I'</em>, <em class="sig-param">'Eco53kI'</em>, <em class="sig-param">'Eco57I'</em>, <em class="sig-param">'Eco72I'</em>, <em class="sig-param">'Eco81I'</em>, <em class="sig-param">'Eco88I'</em>, <em class="sig-param">'Eco91I'</em>, <em class="sig-param">'EcoICRI'</em>, <em class="sig-param">'EcoNI'</em>, <em class="sig-param">'EcoO109I'</em>, <em class="sig-param">'EcoO65I'</em>, <em class="sig-param">'EcoRI'</em>, <em class="sig-param">'EcoRII'</em>, <em class="sig-param">'EcoRV'</em>, <em class="sig-param">'EcoT14I'</em>, <em class="sig-param">'EcoT22I'</em>, <em class="sig-param">'EcoT38I'</em>, <em class="sig-param">'EgeI'</em>, <em class="sig-param">'EheI'</em>, <em class="sig-param">'ErhI'</em>, <em class="sig-param">'Esp3I'</em>, <em class="sig-param">'FaeI'</em>, <em class="sig-param">'FaiI'</em>, <em class="sig-param">'FalI'</em>, <em class="sig-param">'FaqI'</em>, <em class="sig-param">'FatI'</em>, <em class="sig-param">'FauI'</em>, <em class="sig-param">'FauNDI'</em>, <em class="sig-param">'FbaI'</em>, <em class="sig-param">'FblI'</em>, <em class="sig-param">'Fnu4HI'</em>, <em class="sig-param">'FokI'</em>, <em class="sig-param">'FriOI'</em>, <em class="sig-param">'FseI'</em>, <em class="sig-param">'Fsp4HI'</em>, <em class="sig-param">'FspAI'</em>, <em class="sig-param">'FspBI'</em>, <em class="sig-param">'FspEI'</em>, <em class="sig-param">'FspI'</em>, <em class="sig-param">'GlaI'</em>, <em class="sig-param">'GluI'</em>, <em class="sig-param">'GsaI'</em>, <em class="sig-param">'GsuI'</em>, <em class="sig-param">'HaeII'</em>, <em class="sig-param">'HaeIII'</em>, <em class="sig-param">'HapII'</em>, <em class="sig-param">'HgaI'</em>, <em class="sig-param">'HhaI'</em>, <em class="sig-param">'Hin1I'</em>, <em class="sig-param">'Hin1II'</em>, <em class="sig-param">'Hin6I'</em>, <em class="sig-param">'HinP1I'</em>, <em class="sig-param">'HincII'</em>, <em class="sig-param">'HindII'</em>, <em class="sig-param">'HindIII'</em>, <em class="sig-param">'HinfI'</em>, <em class="sig-param">'HpaI'</em>, <em class="sig-param">'HpaII'</em>, <em class="sig-param">'HphI'</em>, <em class="sig-param">'Hpy166II'</em>, <em class="sig-param">'Hpy188I'</em>, <em class="sig-param">'Hpy188III'</em>, <em class="sig-param">'Hpy8I'</em>, <em class="sig-param">'Hpy99I'</em>, <em class="sig-param">'HpyAV'</em>, <em class="sig-param">'HpyCH4III'</em>, <em class="sig-param">'HpyCH4IV'</em>, <em class="sig-param">'HpyCH4V'</em>, <em class="sig-param">'HpyF10VI'</em>, <em class="sig-param">'HpyF3I'</em>, <em class="sig-param">'HpySE526I'</em>, <em class="sig-param">'Hsp92I'</em>, <em class="sig-param">'Hsp92II'</em>, <em class="sig-param">'HspAI'</em>, <em class="sig-param">'KasI'</em>, <em class="sig-param">'KflI'</em>, <em class="sig-param">'Kpn2I'</em>, <em class="sig-param">'KpnI'</em>, <em class="sig-param">'KroI'</em>, <em class="sig-param">'Ksp22I'</em>, <em class="sig-param">'KspAI'</em>, <em class="sig-param">'KspI'</em>, <em class="sig-param">'Kzo9I'</em>, <em class="sig-param">'LguI'</em>, <em class="sig-param">'LmnI'</em>, <em class="sig-param">'LpnPI'</em>, <em class="sig-param">'Lsp1109I'</em>, <em class="sig-param">'LweI'</em>, <em class="sig-param">'MabI'</em>, <em class="sig-param">'MaeI'</em>, <em class="sig-param">'MaeII'</em>, <em class="sig-param">'MaeIII'</em>, <em class="sig-param">'MalI'</em>, <em class="sig-param">'MauBI'</em>, <em class="sig-param">'MbiI'</em>, <em class="sig-param">'MboI'</em>, <em class="sig-param">'MboII'</em>, <em class="sig-param">'MfeI'</em>, <em class="sig-param">'MflI'</em>, <em class="sig-param">'MhlI'</em>, <em class="sig-param">'MlsI'</em>, <em class="sig-param">'MluCI'</em>, <em class="sig-param">'MluI'</em>, <em class="sig-param">'MluNI'</em>, <em class="sig-param">'Mly113I'</em>, <em class="sig-param">'MlyI'</em>, <em class="sig-param">'MmeI'</em>, <em class="sig-param">'MnlI'</em>, <em class="sig-param">'Mox20I'</em>, <em class="sig-param">'Mph1103I'</em>, <em class="sig-param">'MreI'</em>, <em class="sig-param">'MroI'</em>, <em class="sig-param">'MroNI'</em>, <em class="sig-param">'MroXI'</em>, <em class="sig-param">'MscI'</em>, <em class="sig-param">'MseI'</em>, <em class="sig-param">'MslI'</em>, <em class="sig-param">'Msp20I'</em>, <em class="sig-param">'MspA1I'</em>, <em class="sig-param">'MspCI'</em>, <em class="sig-param">'MspI'</em>, <em class="sig-param">'MspJI'</em>, <em class="sig-param">'MspR9I'</em>, <em class="sig-param">'MssI'</em>, <em class="sig-param">'MteI'</em>, <em class="sig-param">'MunI'</em>, <em class="sig-param">'Mva1269I'</em>, <em class="sig-param">'MvaI'</em>, <em class="sig-param">'MvnI'</em>, <em class="sig-param">'MwoI'</em>, <em class="sig-param">'NaeI'</em>, <em class="sig-param">'NarI'</em>, <em class="sig-param">'NciI'</em>, <em class="sig-param">'NcoI'</em>, <em class="sig-param">'NdeI'</em>, <em class="sig-param">'NdeII'</em>, <em class="sig-param">'NgoMIV'</em>, <em class="sig-param">'NheI'</em>, <em class="sig-param">'NlaIII'</em>, <em class="sig-param">'NlaIV'</em>, <em class="sig-param">'NmeAIII'</em>, <em class="sig-param">'NmuCI'</em>, <em class="sig-param">'NotI'</em>, <em class="sig-param">'NruI'</em>, <em class="sig-param">'NsbI'</em>, <em class="sig-param">'NsiI'</em>, <em class="sig-param">'NspI'</em>, <em class="sig-param">'NspV'</em>, <em class="sig-param">'OliI'</em>, <em class="sig-param">'PacI'</em>, <em class="sig-param">'PaeI'</em>, <em class="sig-param">'PaeR7I'</em>, <em class="sig-param">'PagI'</em>, <em class="sig-param">'PalAI'</em>, <em class="sig-param">'PasI'</em>, <em class="sig-param">'PauI'</em>, <em class="sig-param">'PceI'</em>, <em class="sig-param">'PciI'</em>, <em class="sig-param">'PciSI'</em>, <em class="sig-param">'PcsI'</em>, <em class="sig-param">'PctI'</em>, <em class="sig-param">'PdiI'</em>, <em class="sig-param">'PdmI'</em>, <em class="sig-param">'PfeI'</em>, <em class="sig-param">'Pfl23II'</em>, <em class="sig-param">'PflFI'</em>, <em class="sig-param">'PflMI'</em>, <em class="sig-param">'PfoI'</em>, <em class="sig-param">'PinAI'</em>, <em class="sig-param">'PkrI'</em>, <em class="sig-param">'Ple19I'</em>, <em class="sig-param">'PleI'</em>, <em class="sig-param">'PluTI'</em>, <em class="sig-param">'PmaCI'</em>, <em class="sig-param">'PmeI'</em>, <em class="sig-param">'PmlI'</em>, <em class="sig-param">'PpsI'</em>, <em class="sig-param">'Ppu21I'</em>, <em class="sig-param">'PpuMI'</em>, <em class="sig-param">'PscI'</em>, <em class="sig-param">'PshAI'</em>, <em class="sig-param">'PshBI'</em>, <em class="sig-param">'PsiI'</em>, <em class="sig-param">'Psp124BI'</em>, <em class="sig-param">'Psp1406I'</em>, <em class="sig-param">'Psp5II'</em>, <em class="sig-param">'Psp6I'</em>, <em class="sig-param">'PspCI'</em>, <em class="sig-param">'PspEI'</em>, <em class="sig-param">'PspFI'</em>, <em class="sig-param">'PspGI'</em>, <em class="sig-param">'PspLI'</em>, <em class="sig-param">'PspN4I'</em>, <em class="sig-param">'PspOMI'</em>, <em class="sig-param">'PspPI'</em>, <em class="sig-param">'PspPPI'</em>, <em class="sig-param">'PspXI'</em>, <em class="sig-param">'PsrI'</em>, <em class="sig-param">'PstI'</em>, <em class="sig-param">'PstNI'</em>, <em class="sig-param">'PsuI'</em>, <em class="sig-param">'PsyI'</em>, <em class="sig-param">'PteI'</em>, <em class="sig-param">'PvuI'</em>, <em class="sig-param">'PvuII'</em>, <em class="sig-param">'RgaI'</em>, <em class="sig-param">'RigI'</em>, <em class="sig-param">'RruI'</em>, <em class="sig-param">'RsaI'</em>, <em class="sig-param">'RsaNI'</em>, <em class="sig-param">'RseI'</em>, <em class="sig-param">'Rsr2I'</em>, <em class="sig-param">'RsrII'</em>, <em class="sig-param">'SacI'</em>, <em class="sig-param">'SacII'</em>, <em class="sig-param">'SalI'</em>, <em class="sig-param">'SapI'</em>, <em class="sig-param">'SaqAI'</em>, <em class="sig-param">'SatI'</em>, <em class="sig-param">'Sau3AI'</em>, <em class="sig-param">'Sau96I'</em>, <em class="sig-param">'SbfI'</em>, <em class="sig-param">'ScaI'</em>, <em class="sig-param">'SchI'</em>, <em class="sig-param">'ScrFI'</em>, <em class="sig-param">'SdaI'</em>, <em class="sig-param">'SduI'</em>, <em class="sig-param">'SetI'</em>, <em class="sig-param">'SexAI'</em>, <em class="sig-param">'SfaAI'</em>, <em class="sig-param">'SfaNI'</em>, <em class="sig-param">'SfcI'</em>, <em class="sig-param">'SfiI'</em>, <em class="sig-param">'SfoI'</em>, <em class="sig-param">'Sfr274I'</em>, <em class="sig-param">'Sfr303I'</em>, <em class="sig-param">'SfuI'</em>, <em class="sig-param">'SgeI'</em>, <em class="sig-param">'SgfI'</em>, <em class="sig-param">'SgrAI'</em>, <em class="sig-param">'SgrBI'</em>, <em class="sig-param">'SgrDI'</em>, <em class="sig-param">'SgsI'</em>, <em class="sig-param">'SinI'</em>, <em class="sig-param">'SlaI'</em>, <em class="sig-param">'SmaI'</em>, <em class="sig-param">'SmiI'</em>, <em class="sig-param">'SmiMI'</em>, <em class="sig-param">'SmlI'</em>, <em class="sig-param">'SmoI'</em>, <em class="sig-param">'SnaBI'</em>, <em class="sig-param">'SpeI'</em>, <em class="sig-param">'SphI'</em>, <em class="sig-param">'SrfI'</em>, <em class="sig-param">'Sse8387I'</em>, <em class="sig-param">'Sse9I'</em>, <em class="sig-param">'SseBI'</em>, <em class="sig-param">'SsiI'</em>, <em class="sig-param">'SspDI'</em>, <em class="sig-param">'SspI'</em>, <em class="sig-param">'SspMI'</em>, <em class="sig-param">'SstI'</em>, <em class="sig-param">'StuI'</em>, <em class="sig-param">'StyD4I'</em>, <em class="sig-param">'StyI'</em>, <em class="sig-param">'SwaI'</em>, <em class="sig-param">'TaaI'</em>, <em class="sig-param">'TaiI'</em>, <em class="sig-param">'TaqI'</em>, <em class="sig-param">'TaqII'</em>, <em class="sig-param">'TasI'</em>, <em class="sig-param">'TatI'</em>, <em class="sig-param">'TauI'</em>, <em class="sig-param">'TfiI'</em>, <em class="sig-param">'Tru1I'</em>, <em class="sig-param">'Tru9I'</em>, <em class="sig-param">'TscAI'</em>, <em class="sig-param">'TseFI'</em>, <em class="sig-param">'TseI'</em>, <em class="sig-param">'Tsp45I'</em>, <em class="sig-param">'TspDTI'</em>, <em class="sig-param">'TspGWI'</em>, <em class="sig-param">'TspMI'</em>, <em class="sig-param">'TspRI'</em>, <em class="sig-param">'Tth111I'</em>, <em class="sig-param">'Van91I'</em>, <em class="sig-param">'Vha464I'</em>, <em class="sig-param">'VneI'</em>, <em class="sig-param">'VpaK11BI'</em>, <em class="sig-param">'VspI'</em>, <em class="sig-param">'XagI'</em>, <em class="sig-param">'XapI'</em>, <em class="sig-param">'XbaI'</em>, <em class="sig-param">'XceI'</em>, <em class="sig-param">'XcmI'</em>, <em class="sig-param">'XhoI'</em>, <em class="sig-param">'XmaI'</em>, <em class="sig-param">'XmaJI'</em>, <em class="sig-param">'XmiI'</em>, <em class="sig-param">'XmnI'</em>, <em class="sig-param">'XspI'</em>, <em class="sig-param">'ZraI'</em>, <em class="sig-param">'ZrmI'</em>, <em class="sig-param">'Zsp2I'])</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.cutters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.cutters" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the enzymes in a RestrictionBatch that cut the sequence 
at least once.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.fill_in">
<code class="sig-name descname">fill_in</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">nucleotides=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.fill_in"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.fill_in" title="Permalink to this definition">¶</a></dt>
<dd><p>Fill in of five prime protruding end with a DNA polymerase
that has only DNA polymerase activity (such as exo-klenow <a class="footnote-reference brackets" href="#id2" id="id1">1</a>)
and any combination of A, G, C or T. Default are all four
nucleotides together.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>nucleotides</strong><span class="classifier">str</span></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id2"><span class="brackets"><a class="fn-backref" href="#id1">1</a></span></dt>
<dd><p><a class="reference external" href="http://en.wikipedia.org/wiki/Klenow_fragment#The_exo-_Klenow_fragment">http://en.wikipedia.org/wiki/Klenow_fragment#The_exo-_Klenow_fragment</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="k">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;caaa&quot;</span><span class="p">,</span> <span class="s2">&quot;cttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaa</span>
<span class="go"> tttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaag</span>
<span class="go">gtttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">(</span><span class="s2">&quot;g&quot;</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaag</span>
<span class="go">gtttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">(</span><span class="s2">&quot;tac&quot;</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaa</span>
<span class="go"> tttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaac&quot;</span><span class="p">,</span> <span class="s2">&quot;tttg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span>
<span class="go">Dseq(-5)</span>
<span class="go"> aaac</span>
<span class="go">gttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-5)</span>
<span class="go"> aaac</span>
<span class="go">gttt</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.find">
<code class="sig-name descname">find</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">sub</em>, <em class="sig-param">start=0</em>, <em class="sig-param">end=9223372036854775807</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.find"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.find" title="Permalink to this definition">¶</a></dt>
<dd><p>This method behaves like the python string method of the same name.</p>
<p>Returns an integer, the index of the first occurrence of substring
argument sub in the (sub)sequence given by [start:end].</p>
<p>Returns -1 if the subsequence is NOT found.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sub</strong><span class="classifier">string or Seq object</span></dt><dd><p>a string or another Seq object to look for.</p>
</dd>
<dt><strong>start</strong><span class="classifier">int, optional</span></dt><dd><p>slice start.</p>
</dd>
<dt><strong>end</strong><span class="classifier">int, optional</span></dt><dd><p>slice end.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="k">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;atcgactgacgtgtt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Dseq(-15)</span>
<span class="go">atcgactgacgtgtt</span>
<span class="go">tagctgactgcacaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;gac&quot;</span><span class="p">)</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s2">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s2">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Dseq(-7)</span>
<span class="go">agt</span>
<span class="go">  attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;taa&quot;</span><span class="p">)</span>
<span class="go">2</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.five_prime_end">
<code class="sig-name descname">five_prime_end</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.five_prime_end"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.five_prime_end" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a tuple describing the structure of the 5’ end of
the DNA fragment</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pydna.dseq.Dseq.three_prime_end" title="pydna.dseq.Dseq.three_prime_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.three_prime_end</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="k">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&#39;blunt&#39;, &#39;&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go"> aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&quot;3&#39;&quot;, &#39;t&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go">aaa</span>
<span class="go"> ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&quot;5&#39;&quot;, &#39;a&#39;)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.from_string">
<em class="property">classmethod </em><code class="sig-name descname">from_string</code><span class="sig-paren">(</span><em class="sig-param">dna: str</em>, <em class="sig-param">*args</em>, <em class="sig-param">linear=True</em>, <em class="sig-param">circular=False</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.from_string"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.from_string" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.linear">
<em class="property">property </em><code class="sig-name descname">linear</code><a class="headerlink" href="#pydna.dseq.Dseq.linear" title="Permalink to this definition">¶</a></dt>
<dd><p>The linear property can not be set directly. 
Use an empty slice [:] to create a linear object.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.looped">
<code class="sig-name descname">looped</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.looped"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.looped" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a circularized Dseq object. This can only be done if the
two ends are compatible, otherwise a TypeError is raised.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="k">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;catcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="go">Dseq(o8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;t&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgat</span>
<span class="go"> tagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;t&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="go">Dseq(o7)</span>
<span class="go">catcgat</span>
<span class="go">gtagcta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcga</span>
<span class="go">  agctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/dist-packages/pydna/dsdna.py&quot;</span>, line <span class="m">357</span>, in <span class="n">looped</span>
    <span class="k">if</span> <span class="n">type5</span> <span class="o">==</span> <span class="n">type3</span> <span class="ow">and</span> <span class="nb">str</span><span class="p">(</span><span class="n">sticky5</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">rc</span><span class="p">(</span><span class="n">sticky3</span><span class="p">)):</span>
<span class="gr">TypeError</span>: <span class="n">DNA cannot be circularized.</span>
<span class="go">5&#39; and 3&#39; sticky ends not compatible!</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.mung">
<code class="sig-name descname">mung</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.mung"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.mung" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulates treatment a nuclease with 5’-3’ and 3’-5’ single
strand specific exonuclease activity (such as mung bean nuclease <a class="footnote-reference brackets" href="#id4" id="id3">2</a>)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>    <span class="n">ggatcc</span>    <span class="o">-&gt;</span>     <span class="n">gatcc</span>
     <span class="n">ctaggg</span>          <span class="n">ctagg</span>

     <span class="n">ggatcc</span>   <span class="o">-&gt;</span>      <span class="n">ggatc</span>
    <span class="n">tcctag</span>            <span class="n">cctag</span>

<span class="o">&gt;&gt;&gt;</span> <span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="k">import</span> <span class="n">Dseq</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">b</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;caaa&quot;</span><span class="p">,</span> <span class="s2">&quot;cttt&quot;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">b</span>
<span class="n">Dseq</span><span class="p">(</span><span class="o">-</span><span class="mi">5</span><span class="p">)</span>
<span class="n">caaa</span>
 <span class="n">tttc</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">b</span><span class="o">.</span><span class="n">mung</span><span class="p">()</span>
<span class="n">Dseq</span><span class="p">(</span><span class="o">-</span><span class="mi">3</span><span class="p">)</span>
<span class="n">aaa</span>
<span class="n">ttt</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">c</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaac&quot;</span><span class="p">,</span> <span class="s2">&quot;tttg&quot;</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">c</span>
<span class="n">Dseq</span><span class="p">(</span><span class="o">-</span><span class="mi">5</span><span class="p">)</span>
 <span class="n">aaac</span>
<span class="n">gttt</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">c</span><span class="o">.</span><span class="n">mung</span><span class="p">()</span>
<span class="n">Dseq</span><span class="p">(</span><span class="o">-</span><span class="mi">3</span><span class="p">)</span>
<span class="n">aaa</span>
<span class="n">ttt</span>
</pre></div>
</div>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id4"><span class="brackets"><a class="fn-backref" href="#id3">2</a></span></dt>
<dd><p><a class="reference external" href="http://en.wikipedia.org/wiki/Mung_bean_nuclease">http://en.wikipedia.org/wiki/Mung_bean_nuclease</a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.mw">
<code class="sig-name descname">mw</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.mw"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.mw" title="Permalink to this definition">¶</a></dt>
<dd><p>This method returns the molecular weight of the DNA molecule
in g/mol. The following formula is used:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">MW</span> <span class="o">=</span> <span class="p">(</span><span class="n">A</span> <span class="n">x</span> <span class="mf">313.2</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">T</span> <span class="n">x</span> <span class="mf">304.2</span><span class="p">)</span> <span class="o">+</span> 
     <span class="p">(</span><span class="n">C</span> <span class="n">x</span> <span class="mf">289.2</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">G</span> <span class="n">x</span> <span class="mf">329.2</span><span class="p">)</span> <span class="o">+</span> 
     <span class="p">(</span><span class="n">N</span> <span class="n">x</span> <span class="mf">308.9</span><span class="p">)</span> <span class="o">+</span> <span class="mf">79.0</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.n_cutters">
<code class="sig-name descname">n_cutters</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">n=3</em>, <em class="sig-param">batch=RestrictionBatch(['AanI'</em>, <em class="sig-param">'AarI'</em>, <em class="sig-param">'AasI'</em>, <em class="sig-param">'AatII'</em>, <em class="sig-param">'AbaSI'</em>, <em class="sig-param">'AbsI'</em>, <em class="sig-param">'Acc16I'</em>, <em class="sig-param">'Acc36I'</em>, <em class="sig-param">'Acc65I'</em>, <em class="sig-param">'AccB1I'</em>, <em class="sig-param">'AccB7I'</em>, <em class="sig-param">'AccBSI'</em>, <em class="sig-param">'AccI'</em>, <em class="sig-param">'AccII'</em>, <em class="sig-param">'AccIII'</em>, <em class="sig-param">'AciI'</em>, <em class="sig-param">'AclI'</em>, <em class="sig-param">'AclWI'</em>, <em class="sig-param">'AcoI'</em>, <em class="sig-param">'AcsI'</em>, <em class="sig-param">'AcuI'</em>, <em class="sig-param">'AcvI'</em>, <em class="sig-param">'AcyI'</em>, <em class="sig-param">'AdeI'</em>, <em class="sig-param">'AfaI'</em>, <em class="sig-param">'AfeI'</em>, <em class="sig-param">'AfiI'</em>, <em class="sig-param">'AflII'</em>, <em class="sig-param">'AflIII'</em>, <em class="sig-param">'AgeI'</em>, <em class="sig-param">'AgsI'</em>, <em class="sig-param">'AhdI'</em>, <em class="sig-param">'AhlI'</em>, <em class="sig-param">'AjiI'</em>, <em class="sig-param">'AjnI'</em>, <em class="sig-param">'AjuI'</em>, <em class="sig-param">'AleI'</em>, <em class="sig-param">'AloI'</em>, <em class="sig-param">'AluBI'</em>, <em class="sig-param">'AluI'</em>, <em class="sig-param">'Alw21I'</em>, <em class="sig-param">'Alw26I'</em>, <em class="sig-param">'Alw44I'</em>, <em class="sig-param">'AlwI'</em>, <em class="sig-param">'AlwNI'</em>, <em class="sig-param">'Ama87I'</em>, <em class="sig-param">'Aor13HI'</em>, <em class="sig-param">'Aor51HI'</em>, <em class="sig-param">'AoxI'</em>, <em class="sig-param">'ApaI'</em>, <em class="sig-param">'ApaLI'</em>, <em class="sig-param">'ApeKI'</em>, <em class="sig-param">'ApoI'</em>, <em class="sig-param">'ArsI'</em>, <em class="sig-param">'AscI'</em>, <em class="sig-param">'AseI'</em>, <em class="sig-param">'AsiGI'</em>, <em class="sig-param">'AsiSI'</em>, <em class="sig-param">'Asp700I'</em>, <em class="sig-param">'Asp718I'</em>, <em class="sig-param">'AspA2I'</em>, <em class="sig-param">'AspLEI'</em>, <em class="sig-param">'AspS9I'</em>, <em class="sig-param">'AsuC2I'</em>, <em class="sig-param">'AsuHPI'</em>, <em class="sig-param">'AsuII'</em>, <em class="sig-param">'AsuNHI'</em>, <em class="sig-param">'AvaI'</em>, <em class="sig-param">'AvaII'</em>, <em class="sig-param">'AvrII'</em>, <em class="sig-param">'AxyI'</em>, <em class="sig-param">'BaeGI'</em>, <em class="sig-param">'BaeI'</em>, <em class="sig-param">'BalI'</em>, <em class="sig-param">'BamHI'</em>, <em class="sig-param">'BanI'</em>, <em class="sig-param">'BanII'</em>, <em class="sig-param">'BarI'</em>, <em class="sig-param">'BauI'</em>, <em class="sig-param">'BbrPI'</em>, <em class="sig-param">'BbsI'</em>, <em class="sig-param">'Bbv12I'</em>, <em class="sig-param">'BbvCI'</em>, <em class="sig-param">'BbvI'</em>, <em class="sig-param">'BccI'</em>, <em class="sig-param">'BceAI'</em>, <em class="sig-param">'BcgI'</em>, <em class="sig-param">'BciT130I'</em>, <em class="sig-param">'BciVI'</em>, <em class="sig-param">'BclI'</em>, <em class="sig-param">'BcnI'</em>, <em class="sig-param">'BcoDI'</em>, <em class="sig-param">'BcuI'</em>, <em class="sig-param">'BfaI'</em>, <em class="sig-param">'BfmI'</em>, <em class="sig-param">'BfoI'</em>, <em class="sig-param">'BfrI'</em>, <em class="sig-param">'BfuAI'</em>, <em class="sig-param">'BfuI'</em>, <em class="sig-param">'BglI'</em>, <em class="sig-param">'BglII'</em>, <em class="sig-param">'BisI'</em>, <em class="sig-param">'BlnI'</em>, <em class="sig-param">'BlpI'</em>, <em class="sig-param">'BlsI'</em>, <em class="sig-param">'BmcAI'</em>, <em class="sig-param">'Bme1390I'</em>, <em class="sig-param">'Bme18I'</em>, <em class="sig-param">'BmeRI'</em>, <em class="sig-param">'BmeT110I'</em>, <em class="sig-param">'BmgBI'</em>, <em class="sig-param">'BmgT120I'</em>, <em class="sig-param">'BmiI'</em>, <em class="sig-param">'BmrFI'</em>, <em class="sig-param">'BmrI'</em>, <em class="sig-param">'BmsI'</em>, <em class="sig-param">'BmtI'</em>, <em class="sig-param">'BmuI'</em>, <em class="sig-param">'BoxI'</em>, <em class="sig-param">'BpiI'</em>, <em class="sig-param">'BplI'</em>, <em class="sig-param">'BpmI'</em>, <em class="sig-param">'Bpu10I'</em>, <em class="sig-param">'Bpu1102I'</em>, <em class="sig-param">'Bpu14I'</em>, <em class="sig-param">'BpuEI'</em>, <em class="sig-param">'BpuMI'</em>, <em class="sig-param">'Bsa29I'</em>, <em class="sig-param">'BsaAI'</em>, <em class="sig-param">'BsaBI'</em>, <em class="sig-param">'BsaHI'</em>, <em class="sig-param">'BsaI'</em>, <em class="sig-param">'BsaJI'</em>, <em class="sig-param">'BsaWI'</em>, <em class="sig-param">'BsaXI'</em>, <em class="sig-param">'Bsc4I'</em>, <em class="sig-param">'Bse118I'</em>, <em class="sig-param">'Bse1I'</em>, <em class="sig-param">'Bse21I'</em>, <em class="sig-param">'Bse3DI'</em>, <em class="sig-param">'Bse8I'</em>, <em class="sig-param">'BseAI'</em>, <em class="sig-param">'BseBI'</em>, <em class="sig-param">'BseCI'</em>, <em class="sig-param">'BseDI'</em>, <em class="sig-param">'BseGI'</em>, <em class="sig-param">'BseJI'</em>, <em class="sig-param">'BseLI'</em>, <em class="sig-param">'BseMI'</em>, <em class="sig-param">'BseMII'</em>, <em class="sig-param">'BseNI'</em>, <em class="sig-param">'BsePI'</em>, <em class="sig-param">'BseRI'</em>, <em class="sig-param">'BseSI'</em>, <em class="sig-param">'BseX3I'</em>, <em class="sig-param">'BseXI'</em>, <em class="sig-param">'BseYI'</em>, <em class="sig-param">'BsgI'</em>, <em class="sig-param">'Bsh1236I'</em>, <em class="sig-param">'Bsh1285I'</em>, <em class="sig-param">'BshFI'</em>, <em class="sig-param">'BshNI'</em>, <em class="sig-param">'BshTI'</em>, <em class="sig-param">'BshVI'</em>, <em class="sig-param">'BsiEI'</em>, <em class="sig-param">'BsiHKAI'</em>, <em class="sig-param">'BsiHKCI'</em>, <em class="sig-param">'BsiSI'</em>, <em class="sig-param">'BsiWI'</em>, <em class="sig-param">'BslFI'</em>, <em class="sig-param">'BslI'</em>, <em class="sig-param">'BsmAI'</em>, <em class="sig-param">'BsmBI'</em>, <em class="sig-param">'BsmFI'</em>, <em class="sig-param">'BsmI'</em>, <em class="sig-param">'BsnI'</em>, <em class="sig-param">'Bso31I'</em>, <em class="sig-param">'BsoBI'</em>, <em class="sig-param">'Bsp119I'</em>, <em class="sig-param">'Bsp120I'</em>, <em class="sig-param">'Bsp1286I'</em>, <em class="sig-param">'Bsp13I'</em>, <em class="sig-param">'Bsp1407I'</em>, <em class="sig-param">'Bsp143I'</em>, <em class="sig-param">'Bsp1720I'</em>, <em class="sig-param">'Bsp19I'</em>, <em class="sig-param">'Bsp68I'</em>, <em class="sig-param">'BspACI'</em>, <em class="sig-param">'BspANI'</em>, <em class="sig-param">'BspCNI'</em>, <em class="sig-param">'BspDI'</em>, <em class="sig-param">'BspEI'</em>, <em class="sig-param">'BspFNI'</em>, <em class="sig-param">'BspHI'</em>, <em class="sig-param">'BspLI'</em>, <em class="sig-param">'BspMAI'</em>, <em class="sig-param">'BspMI'</em>, <em class="sig-param">'BspOI'</em>, <em class="sig-param">'BspPI'</em>, <em class="sig-param">'BspQI'</em>, <em class="sig-param">'BspT104I'</em>, <em class="sig-param">'BspT107I'</em>, <em class="sig-param">'BspTI'</em>, <em class="sig-param">'BspTNI'</em>, <em class="sig-param">'BsrBI'</em>, <em class="sig-param">'BsrDI'</em>, <em class="sig-param">'BsrFI'</em>, <em class="sig-param">'BsrGI'</em>, <em class="sig-param">'BsrI'</em>, <em class="sig-param">'BssAI'</em>, <em class="sig-param">'BssECI'</em>, <em class="sig-param">'BssHII'</em>, <em class="sig-param">'BssMI'</em>, <em class="sig-param">'BssNAI'</em>, <em class="sig-param">'BssNI'</em>, <em class="sig-param">'BssSI'</em>, <em class="sig-param">'BssT1I'</em>, <em class="sig-param">'Bst1107I'</em>, <em class="sig-param">'Bst2BI'</em>, <em class="sig-param">'Bst2UI'</em>, <em class="sig-param">'Bst4CI'</em>, <em class="sig-param">'Bst6I'</em>, <em class="sig-param">'BstACI'</em>, <em class="sig-param">'BstAFI'</em>, <em class="sig-param">'BstAPI'</em>, <em class="sig-param">'BstAUI'</em>, <em class="sig-param">'BstBAI'</em>, <em class="sig-param">'BstBI'</em>, <em class="sig-param">'BstC8I'</em>, <em class="sig-param">'BstDEI'</em>, <em class="sig-param">'BstDSI'</em>, <em class="sig-param">'BstEII'</em>, <em class="sig-param">'BstENI'</em>, <em class="sig-param">'BstF5I'</em>, <em class="sig-param">'BstFNI'</em>, <em class="sig-param">'BstH2I'</em>, <em class="sig-param">'BstHHI'</em>, <em class="sig-param">'BstKTI'</em>, <em class="sig-param">'BstMAI'</em>, <em class="sig-param">'BstMBI'</em>, <em class="sig-param">'BstMCI'</em>, <em class="sig-param">'BstMWI'</em>, <em class="sig-param">'BstNI'</em>, <em class="sig-param">'BstNSI'</em>, <em class="sig-param">'BstPAI'</em>, <em class="sig-param">'BstPI'</em>, <em class="sig-param">'BstSCI'</em>, <em class="sig-param">'BstSFI'</em>, <em class="sig-param">'BstSLI'</em>, <em class="sig-param">'BstSNI'</em>, <em class="sig-param">'BstUI'</em>, <em class="sig-param">'BstV1I'</em>, <em class="sig-param">'BstV2I'</em>, <em class="sig-param">'BstX2I'</em>, <em class="sig-param">'BstXI'</em>, <em class="sig-param">'BstYI'</em>, <em class="sig-param">'BstZ17I'</em>, <em class="sig-param">'BstZI'</em>, <em class="sig-param">'Bsu15I'</em>, <em class="sig-param">'Bsu36I'</em>, <em class="sig-param">'BsuI'</em>, <em class="sig-param">'BsuRI'</em>, <em class="sig-param">'BsuTUI'</em>, <em class="sig-param">'BtgI'</em>, <em class="sig-param">'BtgZI'</em>, <em class="sig-param">'BtrI'</em>, <em class="sig-param">'BtsCI'</em>, <em class="sig-param">'BtsI'</em>, <em class="sig-param">'BtsIMutI'</em>, <em class="sig-param">'BtuMI'</em>, <em class="sig-param">'BveI'</em>, <em class="sig-param">'Cac8I'</em>, <em class="sig-param">'CaiI'</em>, <em class="sig-param">'CciI'</em>, <em class="sig-param">'CciNI'</em>, <em class="sig-param">'CfoI'</em>, <em class="sig-param">'Cfr10I'</em>, <em class="sig-param">'Cfr13I'</em>, <em class="sig-param">'Cfr42I'</em>, <em class="sig-param">'Cfr9I'</em>, <em class="sig-param">'ClaI'</em>, <em class="sig-param">'CpoI'</em>, <em class="sig-param">'CseI'</em>, <em class="sig-param">'CsiI'</em>, <em class="sig-param">'Csp6I'</em>, <em class="sig-param">'CspAI'</em>, <em class="sig-param">'CspCI'</em>, <em class="sig-param">'CspI'</em>, <em class="sig-param">'CviAII'</em>, <em class="sig-param">'CviJI'</em>, <em class="sig-param">'CviKI_1'</em>, <em class="sig-param">'CviQI'</em>, <em class="sig-param">'DdeI'</em>, <em class="sig-param">'DinI'</em>, <em class="sig-param">'DpnI'</em>, <em class="sig-param">'DpnII'</em>, <em class="sig-param">'DraI'</em>, <em class="sig-param">'DraIII'</em>, <em class="sig-param">'DrdI'</em>, <em class="sig-param">'DriI'</em>, <em class="sig-param">'DseDI'</em>, <em class="sig-param">'EaeI'</em>, <em class="sig-param">'EagI'</em>, <em class="sig-param">'Eam1104I'</em>, <em class="sig-param">'Eam1105I'</em>, <em class="sig-param">'EarI'</em>, <em class="sig-param">'EciI'</em>, <em class="sig-param">'Ecl136II'</em>, <em class="sig-param">'EclXI'</em>, <em class="sig-param">'Eco105I'</em>, <em class="sig-param">'Eco130I'</em>, <em class="sig-param">'Eco147I'</em>, <em class="sig-param">'Eco24I'</em>, <em class="sig-param">'Eco31I'</em>, <em class="sig-param">'Eco32I'</em>, <em class="sig-param">'Eco47I'</em>, <em class="sig-param">'Eco47III'</em>, <em class="sig-param">'Eco52I'</em>, <em class="sig-param">'Eco53kI'</em>, <em class="sig-param">'Eco57I'</em>, <em class="sig-param">'Eco72I'</em>, <em class="sig-param">'Eco81I'</em>, <em class="sig-param">'Eco88I'</em>, <em class="sig-param">'Eco91I'</em>, <em class="sig-param">'EcoICRI'</em>, <em class="sig-param">'EcoNI'</em>, <em class="sig-param">'EcoO109I'</em>, <em class="sig-param">'EcoO65I'</em>, <em class="sig-param">'EcoRI'</em>, <em class="sig-param">'EcoRII'</em>, <em class="sig-param">'EcoRV'</em>, <em class="sig-param">'EcoT14I'</em>, <em class="sig-param">'EcoT22I'</em>, <em class="sig-param">'EcoT38I'</em>, <em class="sig-param">'EgeI'</em>, <em class="sig-param">'EheI'</em>, <em class="sig-param">'ErhI'</em>, <em class="sig-param">'Esp3I'</em>, <em class="sig-param">'FaeI'</em>, <em class="sig-param">'FaiI'</em>, <em class="sig-param">'FalI'</em>, <em class="sig-param">'FaqI'</em>, <em class="sig-param">'FatI'</em>, <em class="sig-param">'FauI'</em>, <em class="sig-param">'FauNDI'</em>, <em class="sig-param">'FbaI'</em>, <em class="sig-param">'FblI'</em>, <em class="sig-param">'Fnu4HI'</em>, <em class="sig-param">'FokI'</em>, <em class="sig-param">'FriOI'</em>, <em class="sig-param">'FseI'</em>, <em class="sig-param">'Fsp4HI'</em>, <em class="sig-param">'FspAI'</em>, <em class="sig-param">'FspBI'</em>, <em class="sig-param">'FspEI'</em>, <em class="sig-param">'FspI'</em>, <em class="sig-param">'GlaI'</em>, <em class="sig-param">'GluI'</em>, <em class="sig-param">'GsaI'</em>, <em class="sig-param">'GsuI'</em>, <em class="sig-param">'HaeII'</em>, <em class="sig-param">'HaeIII'</em>, <em class="sig-param">'HapII'</em>, <em class="sig-param">'HgaI'</em>, <em class="sig-param">'HhaI'</em>, <em class="sig-param">'Hin1I'</em>, <em class="sig-param">'Hin1II'</em>, <em class="sig-param">'Hin6I'</em>, <em class="sig-param">'HinP1I'</em>, <em class="sig-param">'HincII'</em>, <em class="sig-param">'HindII'</em>, <em class="sig-param">'HindIII'</em>, <em class="sig-param">'HinfI'</em>, <em class="sig-param">'HpaI'</em>, <em class="sig-param">'HpaII'</em>, <em class="sig-param">'HphI'</em>, <em class="sig-param">'Hpy166II'</em>, <em class="sig-param">'Hpy188I'</em>, <em class="sig-param">'Hpy188III'</em>, <em class="sig-param">'Hpy8I'</em>, <em class="sig-param">'Hpy99I'</em>, <em class="sig-param">'HpyAV'</em>, <em class="sig-param">'HpyCH4III'</em>, <em class="sig-param">'HpyCH4IV'</em>, <em class="sig-param">'HpyCH4V'</em>, <em class="sig-param">'HpyF10VI'</em>, <em class="sig-param">'HpyF3I'</em>, <em class="sig-param">'HpySE526I'</em>, <em class="sig-param">'Hsp92I'</em>, <em class="sig-param">'Hsp92II'</em>, <em class="sig-param">'HspAI'</em>, <em class="sig-param">'KasI'</em>, <em class="sig-param">'KflI'</em>, <em class="sig-param">'Kpn2I'</em>, <em class="sig-param">'KpnI'</em>, <em class="sig-param">'KroI'</em>, <em class="sig-param">'Ksp22I'</em>, <em class="sig-param">'KspAI'</em>, <em class="sig-param">'KspI'</em>, <em class="sig-param">'Kzo9I'</em>, <em class="sig-param">'LguI'</em>, <em class="sig-param">'LmnI'</em>, <em class="sig-param">'LpnPI'</em>, <em class="sig-param">'Lsp1109I'</em>, <em class="sig-param">'LweI'</em>, <em class="sig-param">'MabI'</em>, <em class="sig-param">'MaeI'</em>, <em class="sig-param">'MaeII'</em>, <em class="sig-param">'MaeIII'</em>, <em class="sig-param">'MalI'</em>, <em class="sig-param">'MauBI'</em>, <em class="sig-param">'MbiI'</em>, <em class="sig-param">'MboI'</em>, <em class="sig-param">'MboII'</em>, <em class="sig-param">'MfeI'</em>, <em class="sig-param">'MflI'</em>, <em class="sig-param">'MhlI'</em>, <em class="sig-param">'MlsI'</em>, <em class="sig-param">'MluCI'</em>, <em class="sig-param">'MluI'</em>, <em class="sig-param">'MluNI'</em>, <em class="sig-param">'Mly113I'</em>, <em class="sig-param">'MlyI'</em>, <em class="sig-param">'MmeI'</em>, <em class="sig-param">'MnlI'</em>, <em class="sig-param">'Mox20I'</em>, <em class="sig-param">'Mph1103I'</em>, <em class="sig-param">'MreI'</em>, <em class="sig-param">'MroI'</em>, <em class="sig-param">'MroNI'</em>, <em class="sig-param">'MroXI'</em>, <em class="sig-param">'MscI'</em>, <em class="sig-param">'MseI'</em>, <em class="sig-param">'MslI'</em>, <em class="sig-param">'Msp20I'</em>, <em class="sig-param">'MspA1I'</em>, <em class="sig-param">'MspCI'</em>, <em class="sig-param">'MspI'</em>, <em class="sig-param">'MspJI'</em>, <em class="sig-param">'MspR9I'</em>, <em class="sig-param">'MssI'</em>, <em class="sig-param">'MteI'</em>, <em class="sig-param">'MunI'</em>, <em class="sig-param">'Mva1269I'</em>, <em class="sig-param">'MvaI'</em>, <em class="sig-param">'MvnI'</em>, <em class="sig-param">'MwoI'</em>, <em class="sig-param">'NaeI'</em>, <em class="sig-param">'NarI'</em>, <em class="sig-param">'NciI'</em>, <em class="sig-param">'NcoI'</em>, <em class="sig-param">'NdeI'</em>, <em class="sig-param">'NdeII'</em>, <em class="sig-param">'NgoMIV'</em>, <em class="sig-param">'NheI'</em>, <em class="sig-param">'NlaIII'</em>, <em class="sig-param">'NlaIV'</em>, <em class="sig-param">'NmeAIII'</em>, <em class="sig-param">'NmuCI'</em>, <em class="sig-param">'NotI'</em>, <em class="sig-param">'NruI'</em>, <em class="sig-param">'NsbI'</em>, <em class="sig-param">'NsiI'</em>, <em class="sig-param">'NspI'</em>, <em class="sig-param">'NspV'</em>, <em class="sig-param">'OliI'</em>, <em class="sig-param">'PacI'</em>, <em class="sig-param">'PaeI'</em>, <em class="sig-param">'PaeR7I'</em>, <em class="sig-param">'PagI'</em>, <em class="sig-param">'PalAI'</em>, <em class="sig-param">'PasI'</em>, <em class="sig-param">'PauI'</em>, <em class="sig-param">'PceI'</em>, <em class="sig-param">'PciI'</em>, <em class="sig-param">'PciSI'</em>, <em class="sig-param">'PcsI'</em>, <em class="sig-param">'PctI'</em>, <em class="sig-param">'PdiI'</em>, <em class="sig-param">'PdmI'</em>, <em class="sig-param">'PfeI'</em>, <em class="sig-param">'Pfl23II'</em>, <em class="sig-param">'PflFI'</em>, <em class="sig-param">'PflMI'</em>, <em class="sig-param">'PfoI'</em>, <em class="sig-param">'PinAI'</em>, <em class="sig-param">'PkrI'</em>, <em class="sig-param">'Ple19I'</em>, <em class="sig-param">'PleI'</em>, <em class="sig-param">'PluTI'</em>, <em class="sig-param">'PmaCI'</em>, <em class="sig-param">'PmeI'</em>, <em class="sig-param">'PmlI'</em>, <em class="sig-param">'PpsI'</em>, <em class="sig-param">'Ppu21I'</em>, <em class="sig-param">'PpuMI'</em>, <em class="sig-param">'PscI'</em>, <em class="sig-param">'PshAI'</em>, <em class="sig-param">'PshBI'</em>, <em class="sig-param">'PsiI'</em>, <em class="sig-param">'Psp124BI'</em>, <em class="sig-param">'Psp1406I'</em>, <em class="sig-param">'Psp5II'</em>, <em class="sig-param">'Psp6I'</em>, <em class="sig-param">'PspCI'</em>, <em class="sig-param">'PspEI'</em>, <em class="sig-param">'PspFI'</em>, <em class="sig-param">'PspGI'</em>, <em class="sig-param">'PspLI'</em>, <em class="sig-param">'PspN4I'</em>, <em class="sig-param">'PspOMI'</em>, <em class="sig-param">'PspPI'</em>, <em class="sig-param">'PspPPI'</em>, <em class="sig-param">'PspXI'</em>, <em class="sig-param">'PsrI'</em>, <em class="sig-param">'PstI'</em>, <em class="sig-param">'PstNI'</em>, <em class="sig-param">'PsuI'</em>, <em class="sig-param">'PsyI'</em>, <em class="sig-param">'PteI'</em>, <em class="sig-param">'PvuI'</em>, <em class="sig-param">'PvuII'</em>, <em class="sig-param">'RgaI'</em>, <em class="sig-param">'RigI'</em>, <em class="sig-param">'RruI'</em>, <em class="sig-param">'RsaI'</em>, <em class="sig-param">'RsaNI'</em>, <em class="sig-param">'RseI'</em>, <em class="sig-param">'Rsr2I'</em>, <em class="sig-param">'RsrII'</em>, <em class="sig-param">'SacI'</em>, <em class="sig-param">'SacII'</em>, <em class="sig-param">'SalI'</em>, <em class="sig-param">'SapI'</em>, <em class="sig-param">'SaqAI'</em>, <em class="sig-param">'SatI'</em>, <em class="sig-param">'Sau3AI'</em>, <em class="sig-param">'Sau96I'</em>, <em class="sig-param">'SbfI'</em>, <em class="sig-param">'ScaI'</em>, <em class="sig-param">'SchI'</em>, <em class="sig-param">'ScrFI'</em>, <em class="sig-param">'SdaI'</em>, <em class="sig-param">'SduI'</em>, <em class="sig-param">'SetI'</em>, <em class="sig-param">'SexAI'</em>, <em class="sig-param">'SfaAI'</em>, <em class="sig-param">'SfaNI'</em>, <em class="sig-param">'SfcI'</em>, <em class="sig-param">'SfiI'</em>, <em class="sig-param">'SfoI'</em>, <em class="sig-param">'Sfr274I'</em>, <em class="sig-param">'Sfr303I'</em>, <em class="sig-param">'SfuI'</em>, <em class="sig-param">'SgeI'</em>, <em class="sig-param">'SgfI'</em>, <em class="sig-param">'SgrAI'</em>, <em class="sig-param">'SgrBI'</em>, <em class="sig-param">'SgrDI'</em>, <em class="sig-param">'SgsI'</em>, <em class="sig-param">'SinI'</em>, <em class="sig-param">'SlaI'</em>, <em class="sig-param">'SmaI'</em>, <em class="sig-param">'SmiI'</em>, <em class="sig-param">'SmiMI'</em>, <em class="sig-param">'SmlI'</em>, <em class="sig-param">'SmoI'</em>, <em class="sig-param">'SnaBI'</em>, <em class="sig-param">'SpeI'</em>, <em class="sig-param">'SphI'</em>, <em class="sig-param">'SrfI'</em>, <em class="sig-param">'Sse8387I'</em>, <em class="sig-param">'Sse9I'</em>, <em class="sig-param">'SseBI'</em>, <em class="sig-param">'SsiI'</em>, <em class="sig-param">'SspDI'</em>, <em class="sig-param">'SspI'</em>, <em class="sig-param">'SspMI'</em>, <em class="sig-param">'SstI'</em>, <em class="sig-param">'StuI'</em>, <em class="sig-param">'StyD4I'</em>, <em class="sig-param">'StyI'</em>, <em class="sig-param">'SwaI'</em>, <em class="sig-param">'TaaI'</em>, <em class="sig-param">'TaiI'</em>, <em class="sig-param">'TaqI'</em>, <em class="sig-param">'TaqII'</em>, <em class="sig-param">'TasI'</em>, <em class="sig-param">'TatI'</em>, <em class="sig-param">'TauI'</em>, <em class="sig-param">'TfiI'</em>, <em class="sig-param">'Tru1I'</em>, <em class="sig-param">'Tru9I'</em>, <em class="sig-param">'TscAI'</em>, <em class="sig-param">'TseFI'</em>, <em class="sig-param">'TseI'</em>, <em class="sig-param">'Tsp45I'</em>, <em class="sig-param">'TspDTI'</em>, <em class="sig-param">'TspGWI'</em>, <em class="sig-param">'TspMI'</em>, <em class="sig-param">'TspRI'</em>, <em class="sig-param">'Tth111I'</em>, <em class="sig-param">'Van91I'</em>, <em class="sig-param">'Vha464I'</em>, <em class="sig-param">'VneI'</em>, <em class="sig-param">'VpaK11BI'</em>, <em class="sig-param">'VspI'</em>, <em class="sig-param">'XagI'</em>, <em class="sig-param">'XapI'</em>, <em class="sig-param">'XbaI'</em>, <em class="sig-param">'XceI'</em>, <em class="sig-param">'XcmI'</em>, <em class="sig-param">'XhoI'</em>, <em class="sig-param">'XmaI'</em>, <em class="sig-param">'XmaJI'</em>, <em class="sig-param">'XmiI'</em>, <em class="sig-param">'XmnI'</em>, <em class="sig-param">'XspI'</em>, <em class="sig-param">'ZraI'</em>, <em class="sig-param">'ZrmI'</em>, <em class="sig-param">'Zsp2I'])</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.n_cutters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.n_cutters" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the enzymes in a RestrictionBatch that cut the sequence 
n times.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.no_cutters">
<code class="sig-name descname">no_cutters</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">batch=RestrictionBatch(['AanI'</em>, <em class="sig-param">'AarI'</em>, <em class="sig-param">'AasI'</em>, <em class="sig-param">'AatII'</em>, <em class="sig-param">'AbaSI'</em>, <em class="sig-param">'AbsI'</em>, <em class="sig-param">'Acc16I'</em>, <em class="sig-param">'Acc36I'</em>, <em class="sig-param">'Acc65I'</em>, <em class="sig-param">'AccB1I'</em>, <em class="sig-param">'AccB7I'</em>, <em class="sig-param">'AccBSI'</em>, <em class="sig-param">'AccI'</em>, <em class="sig-param">'AccII'</em>, <em class="sig-param">'AccIII'</em>, <em class="sig-param">'AciI'</em>, <em class="sig-param">'AclI'</em>, <em class="sig-param">'AclWI'</em>, <em class="sig-param">'AcoI'</em>, <em class="sig-param">'AcsI'</em>, <em class="sig-param">'AcuI'</em>, <em class="sig-param">'AcvI'</em>, <em class="sig-param">'AcyI'</em>, <em class="sig-param">'AdeI'</em>, <em class="sig-param">'AfaI'</em>, <em class="sig-param">'AfeI'</em>, <em class="sig-param">'AfiI'</em>, <em class="sig-param">'AflII'</em>, <em class="sig-param">'AflIII'</em>, <em class="sig-param">'AgeI'</em>, <em class="sig-param">'AgsI'</em>, <em class="sig-param">'AhdI'</em>, <em class="sig-param">'AhlI'</em>, <em class="sig-param">'AjiI'</em>, <em class="sig-param">'AjnI'</em>, <em class="sig-param">'AjuI'</em>, <em class="sig-param">'AleI'</em>, <em class="sig-param">'AloI'</em>, <em class="sig-param">'AluBI'</em>, <em class="sig-param">'AluI'</em>, <em class="sig-param">'Alw21I'</em>, <em class="sig-param">'Alw26I'</em>, <em class="sig-param">'Alw44I'</em>, <em class="sig-param">'AlwI'</em>, <em class="sig-param">'AlwNI'</em>, <em class="sig-param">'Ama87I'</em>, <em class="sig-param">'Aor13HI'</em>, <em class="sig-param">'Aor51HI'</em>, <em class="sig-param">'AoxI'</em>, <em class="sig-param">'ApaI'</em>, <em class="sig-param">'ApaLI'</em>, <em class="sig-param">'ApeKI'</em>, <em class="sig-param">'ApoI'</em>, <em class="sig-param">'ArsI'</em>, <em class="sig-param">'AscI'</em>, <em class="sig-param">'AseI'</em>, <em class="sig-param">'AsiGI'</em>, <em class="sig-param">'AsiSI'</em>, <em class="sig-param">'Asp700I'</em>, <em class="sig-param">'Asp718I'</em>, <em class="sig-param">'AspA2I'</em>, <em class="sig-param">'AspLEI'</em>, <em class="sig-param">'AspS9I'</em>, <em class="sig-param">'AsuC2I'</em>, <em class="sig-param">'AsuHPI'</em>, <em class="sig-param">'AsuII'</em>, <em class="sig-param">'AsuNHI'</em>, <em class="sig-param">'AvaI'</em>, <em class="sig-param">'AvaII'</em>, <em class="sig-param">'AvrII'</em>, <em class="sig-param">'AxyI'</em>, <em class="sig-param">'BaeGI'</em>, <em class="sig-param">'BaeI'</em>, <em class="sig-param">'BalI'</em>, <em class="sig-param">'BamHI'</em>, <em class="sig-param">'BanI'</em>, <em class="sig-param">'BanII'</em>, <em class="sig-param">'BarI'</em>, <em class="sig-param">'BauI'</em>, <em class="sig-param">'BbrPI'</em>, <em class="sig-param">'BbsI'</em>, <em class="sig-param">'Bbv12I'</em>, <em class="sig-param">'BbvCI'</em>, <em class="sig-param">'BbvI'</em>, <em class="sig-param">'BccI'</em>, <em class="sig-param">'BceAI'</em>, <em class="sig-param">'BcgI'</em>, <em class="sig-param">'BciT130I'</em>, <em class="sig-param">'BciVI'</em>, <em class="sig-param">'BclI'</em>, <em class="sig-param">'BcnI'</em>, <em class="sig-param">'BcoDI'</em>, <em class="sig-param">'BcuI'</em>, <em class="sig-param">'BfaI'</em>, <em class="sig-param">'BfmI'</em>, <em class="sig-param">'BfoI'</em>, <em class="sig-param">'BfrI'</em>, <em class="sig-param">'BfuAI'</em>, <em class="sig-param">'BfuI'</em>, <em class="sig-param">'BglI'</em>, <em class="sig-param">'BglII'</em>, <em class="sig-param">'BisI'</em>, <em class="sig-param">'BlnI'</em>, <em class="sig-param">'BlpI'</em>, <em class="sig-param">'BlsI'</em>, <em class="sig-param">'BmcAI'</em>, <em class="sig-param">'Bme1390I'</em>, <em class="sig-param">'Bme18I'</em>, <em class="sig-param">'BmeRI'</em>, <em class="sig-param">'BmeT110I'</em>, <em class="sig-param">'BmgBI'</em>, <em class="sig-param">'BmgT120I'</em>, <em class="sig-param">'BmiI'</em>, <em class="sig-param">'BmrFI'</em>, <em class="sig-param">'BmrI'</em>, <em class="sig-param">'BmsI'</em>, <em class="sig-param">'BmtI'</em>, <em class="sig-param">'BmuI'</em>, <em class="sig-param">'BoxI'</em>, <em class="sig-param">'BpiI'</em>, <em class="sig-param">'BplI'</em>, <em class="sig-param">'BpmI'</em>, <em class="sig-param">'Bpu10I'</em>, <em class="sig-param">'Bpu1102I'</em>, <em class="sig-param">'Bpu14I'</em>, <em class="sig-param">'BpuEI'</em>, <em class="sig-param">'BpuMI'</em>, <em class="sig-param">'Bsa29I'</em>, <em class="sig-param">'BsaAI'</em>, <em class="sig-param">'BsaBI'</em>, <em class="sig-param">'BsaHI'</em>, <em class="sig-param">'BsaI'</em>, <em class="sig-param">'BsaJI'</em>, <em class="sig-param">'BsaWI'</em>, <em class="sig-param">'BsaXI'</em>, <em class="sig-param">'Bsc4I'</em>, <em class="sig-param">'Bse118I'</em>, <em class="sig-param">'Bse1I'</em>, <em class="sig-param">'Bse21I'</em>, <em class="sig-param">'Bse3DI'</em>, <em class="sig-param">'Bse8I'</em>, <em class="sig-param">'BseAI'</em>, <em class="sig-param">'BseBI'</em>, <em class="sig-param">'BseCI'</em>, <em class="sig-param">'BseDI'</em>, <em class="sig-param">'BseGI'</em>, <em class="sig-param">'BseJI'</em>, <em class="sig-param">'BseLI'</em>, <em class="sig-param">'BseMI'</em>, <em class="sig-param">'BseMII'</em>, <em class="sig-param">'BseNI'</em>, <em class="sig-param">'BsePI'</em>, <em class="sig-param">'BseRI'</em>, <em class="sig-param">'BseSI'</em>, <em class="sig-param">'BseX3I'</em>, <em class="sig-param">'BseXI'</em>, <em class="sig-param">'BseYI'</em>, <em class="sig-param">'BsgI'</em>, <em class="sig-param">'Bsh1236I'</em>, <em class="sig-param">'Bsh1285I'</em>, <em class="sig-param">'BshFI'</em>, <em class="sig-param">'BshNI'</em>, <em class="sig-param">'BshTI'</em>, <em class="sig-param">'BshVI'</em>, <em class="sig-param">'BsiEI'</em>, <em class="sig-param">'BsiHKAI'</em>, <em class="sig-param">'BsiHKCI'</em>, <em class="sig-param">'BsiSI'</em>, <em class="sig-param">'BsiWI'</em>, <em class="sig-param">'BslFI'</em>, <em class="sig-param">'BslI'</em>, <em class="sig-param">'BsmAI'</em>, <em class="sig-param">'BsmBI'</em>, <em class="sig-param">'BsmFI'</em>, <em class="sig-param">'BsmI'</em>, <em class="sig-param">'BsnI'</em>, <em class="sig-param">'Bso31I'</em>, <em class="sig-param">'BsoBI'</em>, <em class="sig-param">'Bsp119I'</em>, <em class="sig-param">'Bsp120I'</em>, <em class="sig-param">'Bsp1286I'</em>, <em class="sig-param">'Bsp13I'</em>, <em class="sig-param">'Bsp1407I'</em>, <em class="sig-param">'Bsp143I'</em>, <em class="sig-param">'Bsp1720I'</em>, <em class="sig-param">'Bsp19I'</em>, <em class="sig-param">'Bsp68I'</em>, <em class="sig-param">'BspACI'</em>, <em class="sig-param">'BspANI'</em>, <em class="sig-param">'BspCNI'</em>, <em class="sig-param">'BspDI'</em>, <em class="sig-param">'BspEI'</em>, <em class="sig-param">'BspFNI'</em>, <em class="sig-param">'BspHI'</em>, <em class="sig-param">'BspLI'</em>, <em class="sig-param">'BspMAI'</em>, <em class="sig-param">'BspMI'</em>, <em class="sig-param">'BspOI'</em>, <em class="sig-param">'BspPI'</em>, <em class="sig-param">'BspQI'</em>, <em class="sig-param">'BspT104I'</em>, <em class="sig-param">'BspT107I'</em>, <em class="sig-param">'BspTI'</em>, <em class="sig-param">'BspTNI'</em>, <em class="sig-param">'BsrBI'</em>, <em class="sig-param">'BsrDI'</em>, <em class="sig-param">'BsrFI'</em>, <em class="sig-param">'BsrGI'</em>, <em class="sig-param">'BsrI'</em>, <em class="sig-param">'BssAI'</em>, <em class="sig-param">'BssECI'</em>, <em class="sig-param">'BssHII'</em>, <em class="sig-param">'BssMI'</em>, <em class="sig-param">'BssNAI'</em>, <em class="sig-param">'BssNI'</em>, <em class="sig-param">'BssSI'</em>, <em class="sig-param">'BssT1I'</em>, <em class="sig-param">'Bst1107I'</em>, <em class="sig-param">'Bst2BI'</em>, <em class="sig-param">'Bst2UI'</em>, <em class="sig-param">'Bst4CI'</em>, <em class="sig-param">'Bst6I'</em>, <em class="sig-param">'BstACI'</em>, <em class="sig-param">'BstAFI'</em>, <em class="sig-param">'BstAPI'</em>, <em class="sig-param">'BstAUI'</em>, <em class="sig-param">'BstBAI'</em>, <em class="sig-param">'BstBI'</em>, <em class="sig-param">'BstC8I'</em>, <em class="sig-param">'BstDEI'</em>, <em class="sig-param">'BstDSI'</em>, <em class="sig-param">'BstEII'</em>, <em class="sig-param">'BstENI'</em>, <em class="sig-param">'BstF5I'</em>, <em class="sig-param">'BstFNI'</em>, <em class="sig-param">'BstH2I'</em>, <em class="sig-param">'BstHHI'</em>, <em class="sig-param">'BstKTI'</em>, <em class="sig-param">'BstMAI'</em>, <em class="sig-param">'BstMBI'</em>, <em class="sig-param">'BstMCI'</em>, <em class="sig-param">'BstMWI'</em>, <em class="sig-param">'BstNI'</em>, <em class="sig-param">'BstNSI'</em>, <em class="sig-param">'BstPAI'</em>, <em class="sig-param">'BstPI'</em>, <em class="sig-param">'BstSCI'</em>, <em class="sig-param">'BstSFI'</em>, <em class="sig-param">'BstSLI'</em>, <em class="sig-param">'BstSNI'</em>, <em class="sig-param">'BstUI'</em>, <em class="sig-param">'BstV1I'</em>, <em class="sig-param">'BstV2I'</em>, <em class="sig-param">'BstX2I'</em>, <em class="sig-param">'BstXI'</em>, <em class="sig-param">'BstYI'</em>, <em class="sig-param">'BstZ17I'</em>, <em class="sig-param">'BstZI'</em>, <em class="sig-param">'Bsu15I'</em>, <em class="sig-param">'Bsu36I'</em>, <em class="sig-param">'BsuI'</em>, <em class="sig-param">'BsuRI'</em>, <em class="sig-param">'BsuTUI'</em>, <em class="sig-param">'BtgI'</em>, <em class="sig-param">'BtgZI'</em>, <em class="sig-param">'BtrI'</em>, <em class="sig-param">'BtsCI'</em>, <em class="sig-param">'BtsI'</em>, <em class="sig-param">'BtsIMutI'</em>, <em class="sig-param">'BtuMI'</em>, <em class="sig-param">'BveI'</em>, <em class="sig-param">'Cac8I'</em>, <em class="sig-param">'CaiI'</em>, <em class="sig-param">'CciI'</em>, <em class="sig-param">'CciNI'</em>, <em class="sig-param">'CfoI'</em>, <em class="sig-param">'Cfr10I'</em>, <em class="sig-param">'Cfr13I'</em>, <em class="sig-param">'Cfr42I'</em>, <em class="sig-param">'Cfr9I'</em>, <em class="sig-param">'ClaI'</em>, <em class="sig-param">'CpoI'</em>, <em class="sig-param">'CseI'</em>, <em class="sig-param">'CsiI'</em>, <em class="sig-param">'Csp6I'</em>, <em class="sig-param">'CspAI'</em>, <em class="sig-param">'CspCI'</em>, <em class="sig-param">'CspI'</em>, <em class="sig-param">'CviAII'</em>, <em class="sig-param">'CviJI'</em>, <em class="sig-param">'CviKI_1'</em>, <em class="sig-param">'CviQI'</em>, <em class="sig-param">'DdeI'</em>, <em class="sig-param">'DinI'</em>, <em class="sig-param">'DpnI'</em>, <em class="sig-param">'DpnII'</em>, <em class="sig-param">'DraI'</em>, <em class="sig-param">'DraIII'</em>, <em class="sig-param">'DrdI'</em>, <em class="sig-param">'DriI'</em>, <em class="sig-param">'DseDI'</em>, <em class="sig-param">'EaeI'</em>, <em class="sig-param">'EagI'</em>, <em class="sig-param">'Eam1104I'</em>, <em class="sig-param">'Eam1105I'</em>, <em class="sig-param">'EarI'</em>, <em class="sig-param">'EciI'</em>, <em class="sig-param">'Ecl136II'</em>, <em class="sig-param">'EclXI'</em>, <em class="sig-param">'Eco105I'</em>, <em class="sig-param">'Eco130I'</em>, <em class="sig-param">'Eco147I'</em>, <em class="sig-param">'Eco24I'</em>, <em class="sig-param">'Eco31I'</em>, <em class="sig-param">'Eco32I'</em>, <em class="sig-param">'Eco47I'</em>, <em class="sig-param">'Eco47III'</em>, <em class="sig-param">'Eco52I'</em>, <em class="sig-param">'Eco53kI'</em>, <em class="sig-param">'Eco57I'</em>, <em class="sig-param">'Eco72I'</em>, <em class="sig-param">'Eco81I'</em>, <em class="sig-param">'Eco88I'</em>, <em class="sig-param">'Eco91I'</em>, <em class="sig-param">'EcoICRI'</em>, <em class="sig-param">'EcoNI'</em>, <em class="sig-param">'EcoO109I'</em>, <em class="sig-param">'EcoO65I'</em>, <em class="sig-param">'EcoRI'</em>, <em class="sig-param">'EcoRII'</em>, <em class="sig-param">'EcoRV'</em>, <em class="sig-param">'EcoT14I'</em>, <em class="sig-param">'EcoT22I'</em>, <em class="sig-param">'EcoT38I'</em>, <em class="sig-param">'EgeI'</em>, <em class="sig-param">'EheI'</em>, <em class="sig-param">'ErhI'</em>, <em class="sig-param">'Esp3I'</em>, <em class="sig-param">'FaeI'</em>, <em class="sig-param">'FaiI'</em>, <em class="sig-param">'FalI'</em>, <em class="sig-param">'FaqI'</em>, <em class="sig-param">'FatI'</em>, <em class="sig-param">'FauI'</em>, <em class="sig-param">'FauNDI'</em>, <em class="sig-param">'FbaI'</em>, <em class="sig-param">'FblI'</em>, <em class="sig-param">'Fnu4HI'</em>, <em class="sig-param">'FokI'</em>, <em class="sig-param">'FriOI'</em>, <em class="sig-param">'FseI'</em>, <em class="sig-param">'Fsp4HI'</em>, <em class="sig-param">'FspAI'</em>, <em class="sig-param">'FspBI'</em>, <em class="sig-param">'FspEI'</em>, <em class="sig-param">'FspI'</em>, <em class="sig-param">'GlaI'</em>, <em class="sig-param">'GluI'</em>, <em class="sig-param">'GsaI'</em>, <em class="sig-param">'GsuI'</em>, <em class="sig-param">'HaeII'</em>, <em class="sig-param">'HaeIII'</em>, <em class="sig-param">'HapII'</em>, <em class="sig-param">'HgaI'</em>, <em class="sig-param">'HhaI'</em>, <em class="sig-param">'Hin1I'</em>, <em class="sig-param">'Hin1II'</em>, <em class="sig-param">'Hin6I'</em>, <em class="sig-param">'HinP1I'</em>, <em class="sig-param">'HincII'</em>, <em class="sig-param">'HindII'</em>, <em class="sig-param">'HindIII'</em>, <em class="sig-param">'HinfI'</em>, <em class="sig-param">'HpaI'</em>, <em class="sig-param">'HpaII'</em>, <em class="sig-param">'HphI'</em>, <em class="sig-param">'Hpy166II'</em>, <em class="sig-param">'Hpy188I'</em>, <em class="sig-param">'Hpy188III'</em>, <em class="sig-param">'Hpy8I'</em>, <em class="sig-param">'Hpy99I'</em>, <em class="sig-param">'HpyAV'</em>, <em class="sig-param">'HpyCH4III'</em>, <em class="sig-param">'HpyCH4IV'</em>, <em class="sig-param">'HpyCH4V'</em>, <em class="sig-param">'HpyF10VI'</em>, <em class="sig-param">'HpyF3I'</em>, <em class="sig-param">'HpySE526I'</em>, <em class="sig-param">'Hsp92I'</em>, <em class="sig-param">'Hsp92II'</em>, <em class="sig-param">'HspAI'</em>, <em class="sig-param">'KasI'</em>, <em class="sig-param">'KflI'</em>, <em class="sig-param">'Kpn2I'</em>, <em class="sig-param">'KpnI'</em>, <em class="sig-param">'KroI'</em>, <em class="sig-param">'Ksp22I'</em>, <em class="sig-param">'KspAI'</em>, <em class="sig-param">'KspI'</em>, <em class="sig-param">'Kzo9I'</em>, <em class="sig-param">'LguI'</em>, <em class="sig-param">'LmnI'</em>, <em class="sig-param">'LpnPI'</em>, <em class="sig-param">'Lsp1109I'</em>, <em class="sig-param">'LweI'</em>, <em class="sig-param">'MabI'</em>, <em class="sig-param">'MaeI'</em>, <em class="sig-param">'MaeII'</em>, <em class="sig-param">'MaeIII'</em>, <em class="sig-param">'MalI'</em>, <em class="sig-param">'MauBI'</em>, <em class="sig-param">'MbiI'</em>, <em class="sig-param">'MboI'</em>, <em class="sig-param">'MboII'</em>, <em class="sig-param">'MfeI'</em>, <em class="sig-param">'MflI'</em>, <em class="sig-param">'MhlI'</em>, <em class="sig-param">'MlsI'</em>, <em class="sig-param">'MluCI'</em>, <em class="sig-param">'MluI'</em>, <em class="sig-param">'MluNI'</em>, <em class="sig-param">'Mly113I'</em>, <em class="sig-param">'MlyI'</em>, <em class="sig-param">'MmeI'</em>, <em class="sig-param">'MnlI'</em>, <em class="sig-param">'Mox20I'</em>, <em class="sig-param">'Mph1103I'</em>, <em class="sig-param">'MreI'</em>, <em class="sig-param">'MroI'</em>, <em class="sig-param">'MroNI'</em>, <em class="sig-param">'MroXI'</em>, <em class="sig-param">'MscI'</em>, <em class="sig-param">'MseI'</em>, <em class="sig-param">'MslI'</em>, <em class="sig-param">'Msp20I'</em>, <em class="sig-param">'MspA1I'</em>, <em class="sig-param">'MspCI'</em>, <em class="sig-param">'MspI'</em>, <em class="sig-param">'MspJI'</em>, <em class="sig-param">'MspR9I'</em>, <em class="sig-param">'MssI'</em>, <em class="sig-param">'MteI'</em>, <em class="sig-param">'MunI'</em>, <em class="sig-param">'Mva1269I'</em>, <em class="sig-param">'MvaI'</em>, <em class="sig-param">'MvnI'</em>, <em class="sig-param">'MwoI'</em>, <em class="sig-param">'NaeI'</em>, <em class="sig-param">'NarI'</em>, <em class="sig-param">'NciI'</em>, <em class="sig-param">'NcoI'</em>, <em class="sig-param">'NdeI'</em>, <em class="sig-param">'NdeII'</em>, <em class="sig-param">'NgoMIV'</em>, <em class="sig-param">'NheI'</em>, <em class="sig-param">'NlaIII'</em>, <em class="sig-param">'NlaIV'</em>, <em class="sig-param">'NmeAIII'</em>, <em class="sig-param">'NmuCI'</em>, <em class="sig-param">'NotI'</em>, <em class="sig-param">'NruI'</em>, <em class="sig-param">'NsbI'</em>, <em class="sig-param">'NsiI'</em>, <em class="sig-param">'NspI'</em>, <em class="sig-param">'NspV'</em>, <em class="sig-param">'OliI'</em>, <em class="sig-param">'PacI'</em>, <em class="sig-param">'PaeI'</em>, <em class="sig-param">'PaeR7I'</em>, <em class="sig-param">'PagI'</em>, <em class="sig-param">'PalAI'</em>, <em class="sig-param">'PasI'</em>, <em class="sig-param">'PauI'</em>, <em class="sig-param">'PceI'</em>, <em class="sig-param">'PciI'</em>, <em class="sig-param">'PciSI'</em>, <em class="sig-param">'PcsI'</em>, <em class="sig-param">'PctI'</em>, <em class="sig-param">'PdiI'</em>, <em class="sig-param">'PdmI'</em>, <em class="sig-param">'PfeI'</em>, <em class="sig-param">'Pfl23II'</em>, <em class="sig-param">'PflFI'</em>, <em class="sig-param">'PflMI'</em>, <em class="sig-param">'PfoI'</em>, <em class="sig-param">'PinAI'</em>, <em class="sig-param">'PkrI'</em>, <em class="sig-param">'Ple19I'</em>, <em class="sig-param">'PleI'</em>, <em class="sig-param">'PluTI'</em>, <em class="sig-param">'PmaCI'</em>, <em class="sig-param">'PmeI'</em>, <em class="sig-param">'PmlI'</em>, <em class="sig-param">'PpsI'</em>, <em class="sig-param">'Ppu21I'</em>, <em class="sig-param">'PpuMI'</em>, <em class="sig-param">'PscI'</em>, <em class="sig-param">'PshAI'</em>, <em class="sig-param">'PshBI'</em>, <em class="sig-param">'PsiI'</em>, <em class="sig-param">'Psp124BI'</em>, <em class="sig-param">'Psp1406I'</em>, <em class="sig-param">'Psp5II'</em>, <em class="sig-param">'Psp6I'</em>, <em class="sig-param">'PspCI'</em>, <em class="sig-param">'PspEI'</em>, <em class="sig-param">'PspFI'</em>, <em class="sig-param">'PspGI'</em>, <em class="sig-param">'PspLI'</em>, <em class="sig-param">'PspN4I'</em>, <em class="sig-param">'PspOMI'</em>, <em class="sig-param">'PspPI'</em>, <em class="sig-param">'PspPPI'</em>, <em class="sig-param">'PspXI'</em>, <em class="sig-param">'PsrI'</em>, <em class="sig-param">'PstI'</em>, <em class="sig-param">'PstNI'</em>, <em class="sig-param">'PsuI'</em>, <em class="sig-param">'PsyI'</em>, <em class="sig-param">'PteI'</em>, <em class="sig-param">'PvuI'</em>, <em class="sig-param">'PvuII'</em>, <em class="sig-param">'RgaI'</em>, <em class="sig-param">'RigI'</em>, <em class="sig-param">'RruI'</em>, <em class="sig-param">'RsaI'</em>, <em class="sig-param">'RsaNI'</em>, <em class="sig-param">'RseI'</em>, <em class="sig-param">'Rsr2I'</em>, <em class="sig-param">'RsrII'</em>, <em class="sig-param">'SacI'</em>, <em class="sig-param">'SacII'</em>, <em class="sig-param">'SalI'</em>, <em class="sig-param">'SapI'</em>, <em class="sig-param">'SaqAI'</em>, <em class="sig-param">'SatI'</em>, <em class="sig-param">'Sau3AI'</em>, <em class="sig-param">'Sau96I'</em>, <em class="sig-param">'SbfI'</em>, <em class="sig-param">'ScaI'</em>, <em class="sig-param">'SchI'</em>, <em class="sig-param">'ScrFI'</em>, <em class="sig-param">'SdaI'</em>, <em class="sig-param">'SduI'</em>, <em class="sig-param">'SetI'</em>, <em class="sig-param">'SexAI'</em>, <em class="sig-param">'SfaAI'</em>, <em class="sig-param">'SfaNI'</em>, <em class="sig-param">'SfcI'</em>, <em class="sig-param">'SfiI'</em>, <em class="sig-param">'SfoI'</em>, <em class="sig-param">'Sfr274I'</em>, <em class="sig-param">'Sfr303I'</em>, <em class="sig-param">'SfuI'</em>, <em class="sig-param">'SgeI'</em>, <em class="sig-param">'SgfI'</em>, <em class="sig-param">'SgrAI'</em>, <em class="sig-param">'SgrBI'</em>, <em class="sig-param">'SgrDI'</em>, <em class="sig-param">'SgsI'</em>, <em class="sig-param">'SinI'</em>, <em class="sig-param">'SlaI'</em>, <em class="sig-param">'SmaI'</em>, <em class="sig-param">'SmiI'</em>, <em class="sig-param">'SmiMI'</em>, <em class="sig-param">'SmlI'</em>, <em class="sig-param">'SmoI'</em>, <em class="sig-param">'SnaBI'</em>, <em class="sig-param">'SpeI'</em>, <em class="sig-param">'SphI'</em>, <em class="sig-param">'SrfI'</em>, <em class="sig-param">'Sse8387I'</em>, <em class="sig-param">'Sse9I'</em>, <em class="sig-param">'SseBI'</em>, <em class="sig-param">'SsiI'</em>, <em class="sig-param">'SspDI'</em>, <em class="sig-param">'SspI'</em>, <em class="sig-param">'SspMI'</em>, <em class="sig-param">'SstI'</em>, <em class="sig-param">'StuI'</em>, <em class="sig-param">'StyD4I'</em>, <em class="sig-param">'StyI'</em>, <em class="sig-param">'SwaI'</em>, <em class="sig-param">'TaaI'</em>, <em class="sig-param">'TaiI'</em>, <em class="sig-param">'TaqI'</em>, <em class="sig-param">'TaqII'</em>, <em class="sig-param">'TasI'</em>, <em class="sig-param">'TatI'</em>, <em class="sig-param">'TauI'</em>, <em class="sig-param">'TfiI'</em>, <em class="sig-param">'Tru1I'</em>, <em class="sig-param">'Tru9I'</em>, <em class="sig-param">'TscAI'</em>, <em class="sig-param">'TseFI'</em>, <em class="sig-param">'TseI'</em>, <em class="sig-param">'Tsp45I'</em>, <em class="sig-param">'TspDTI'</em>, <em class="sig-param">'TspGWI'</em>, <em class="sig-param">'TspMI'</em>, <em class="sig-param">'TspRI'</em>, <em class="sig-param">'Tth111I'</em>, <em class="sig-param">'Van91I'</em>, <em class="sig-param">'Vha464I'</em>, <em class="sig-param">'VneI'</em>, <em class="sig-param">'VpaK11BI'</em>, <em class="sig-param">'VspI'</em>, <em class="sig-param">'XagI'</em>, <em class="sig-param">'XapI'</em>, <em class="sig-param">'XbaI'</em>, <em class="sig-param">'XceI'</em>, <em class="sig-param">'XcmI'</em>, <em class="sig-param">'XhoI'</em>, <em class="sig-param">'XmaI'</em>, <em class="sig-param">'XmaJI'</em>, <em class="sig-param">'XmiI'</em>, <em class="sig-param">'XmnI'</em>, <em class="sig-param">'XspI'</em>, <em class="sig-param">'ZraI'</em>, <em class="sig-param">'ZrmI'</em>, <em class="sig-param">'Zsp2I'])</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.no_cutters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.no_cutters" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the enzymes in a RestrictionBatch that do <strong>not</strong> 
cut the sequence.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.once_cutters">
<code class="sig-name descname">once_cutters</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">batch=RestrictionBatch(['AanI'</em>, <em class="sig-param">'AarI'</em>, <em class="sig-param">'AasI'</em>, <em class="sig-param">'AatII'</em>, <em class="sig-param">'AbaSI'</em>, <em class="sig-param">'AbsI'</em>, <em class="sig-param">'Acc16I'</em>, <em class="sig-param">'Acc36I'</em>, <em class="sig-param">'Acc65I'</em>, <em class="sig-param">'AccB1I'</em>, <em class="sig-param">'AccB7I'</em>, <em class="sig-param">'AccBSI'</em>, <em class="sig-param">'AccI'</em>, <em class="sig-param">'AccII'</em>, <em class="sig-param">'AccIII'</em>, <em class="sig-param">'AciI'</em>, <em class="sig-param">'AclI'</em>, <em class="sig-param">'AclWI'</em>, <em class="sig-param">'AcoI'</em>, <em class="sig-param">'AcsI'</em>, <em class="sig-param">'AcuI'</em>, <em class="sig-param">'AcvI'</em>, <em class="sig-param">'AcyI'</em>, <em class="sig-param">'AdeI'</em>, <em class="sig-param">'AfaI'</em>, <em class="sig-param">'AfeI'</em>, <em class="sig-param">'AfiI'</em>, <em class="sig-param">'AflII'</em>, <em class="sig-param">'AflIII'</em>, <em class="sig-param">'AgeI'</em>, <em class="sig-param">'AgsI'</em>, <em class="sig-param">'AhdI'</em>, <em class="sig-param">'AhlI'</em>, <em class="sig-param">'AjiI'</em>, <em class="sig-param">'AjnI'</em>, <em class="sig-param">'AjuI'</em>, <em class="sig-param">'AleI'</em>, <em class="sig-param">'AloI'</em>, <em class="sig-param">'AluBI'</em>, <em class="sig-param">'AluI'</em>, <em class="sig-param">'Alw21I'</em>, <em class="sig-param">'Alw26I'</em>, <em class="sig-param">'Alw44I'</em>, <em class="sig-param">'AlwI'</em>, <em class="sig-param">'AlwNI'</em>, <em class="sig-param">'Ama87I'</em>, <em class="sig-param">'Aor13HI'</em>, <em class="sig-param">'Aor51HI'</em>, <em class="sig-param">'AoxI'</em>, <em class="sig-param">'ApaI'</em>, <em class="sig-param">'ApaLI'</em>, <em class="sig-param">'ApeKI'</em>, <em class="sig-param">'ApoI'</em>, <em class="sig-param">'ArsI'</em>, <em class="sig-param">'AscI'</em>, <em class="sig-param">'AseI'</em>, <em class="sig-param">'AsiGI'</em>, <em class="sig-param">'AsiSI'</em>, <em class="sig-param">'Asp700I'</em>, <em class="sig-param">'Asp718I'</em>, <em class="sig-param">'AspA2I'</em>, <em class="sig-param">'AspLEI'</em>, <em class="sig-param">'AspS9I'</em>, <em class="sig-param">'AsuC2I'</em>, <em class="sig-param">'AsuHPI'</em>, <em class="sig-param">'AsuII'</em>, <em class="sig-param">'AsuNHI'</em>, <em class="sig-param">'AvaI'</em>, <em class="sig-param">'AvaII'</em>, <em class="sig-param">'AvrII'</em>, <em class="sig-param">'AxyI'</em>, <em class="sig-param">'BaeGI'</em>, <em class="sig-param">'BaeI'</em>, <em class="sig-param">'BalI'</em>, <em class="sig-param">'BamHI'</em>, <em class="sig-param">'BanI'</em>, <em class="sig-param">'BanII'</em>, <em class="sig-param">'BarI'</em>, <em class="sig-param">'BauI'</em>, <em class="sig-param">'BbrPI'</em>, <em class="sig-param">'BbsI'</em>, <em class="sig-param">'Bbv12I'</em>, <em class="sig-param">'BbvCI'</em>, <em class="sig-param">'BbvI'</em>, <em class="sig-param">'BccI'</em>, <em class="sig-param">'BceAI'</em>, <em class="sig-param">'BcgI'</em>, <em class="sig-param">'BciT130I'</em>, <em class="sig-param">'BciVI'</em>, <em class="sig-param">'BclI'</em>, <em class="sig-param">'BcnI'</em>, <em class="sig-param">'BcoDI'</em>, <em class="sig-param">'BcuI'</em>, <em class="sig-param">'BfaI'</em>, <em class="sig-param">'BfmI'</em>, <em class="sig-param">'BfoI'</em>, <em class="sig-param">'BfrI'</em>, <em class="sig-param">'BfuAI'</em>, <em class="sig-param">'BfuI'</em>, <em class="sig-param">'BglI'</em>, <em class="sig-param">'BglII'</em>, <em class="sig-param">'BisI'</em>, <em class="sig-param">'BlnI'</em>, <em class="sig-param">'BlpI'</em>, <em class="sig-param">'BlsI'</em>, <em class="sig-param">'BmcAI'</em>, <em class="sig-param">'Bme1390I'</em>, <em class="sig-param">'Bme18I'</em>, <em class="sig-param">'BmeRI'</em>, <em class="sig-param">'BmeT110I'</em>, <em class="sig-param">'BmgBI'</em>, <em class="sig-param">'BmgT120I'</em>, <em class="sig-param">'BmiI'</em>, <em class="sig-param">'BmrFI'</em>, <em class="sig-param">'BmrI'</em>, <em class="sig-param">'BmsI'</em>, <em class="sig-param">'BmtI'</em>, <em class="sig-param">'BmuI'</em>, <em class="sig-param">'BoxI'</em>, <em class="sig-param">'BpiI'</em>, <em class="sig-param">'BplI'</em>, <em class="sig-param">'BpmI'</em>, <em class="sig-param">'Bpu10I'</em>, <em class="sig-param">'Bpu1102I'</em>, <em class="sig-param">'Bpu14I'</em>, <em class="sig-param">'BpuEI'</em>, <em class="sig-param">'BpuMI'</em>, <em class="sig-param">'Bsa29I'</em>, <em class="sig-param">'BsaAI'</em>, <em class="sig-param">'BsaBI'</em>, <em class="sig-param">'BsaHI'</em>, <em class="sig-param">'BsaI'</em>, <em class="sig-param">'BsaJI'</em>, <em class="sig-param">'BsaWI'</em>, <em class="sig-param">'BsaXI'</em>, <em class="sig-param">'Bsc4I'</em>, <em class="sig-param">'Bse118I'</em>, <em class="sig-param">'Bse1I'</em>, <em class="sig-param">'Bse21I'</em>, <em class="sig-param">'Bse3DI'</em>, <em class="sig-param">'Bse8I'</em>, <em class="sig-param">'BseAI'</em>, <em class="sig-param">'BseBI'</em>, <em class="sig-param">'BseCI'</em>, <em class="sig-param">'BseDI'</em>, <em class="sig-param">'BseGI'</em>, <em class="sig-param">'BseJI'</em>, <em class="sig-param">'BseLI'</em>, <em class="sig-param">'BseMI'</em>, <em class="sig-param">'BseMII'</em>, <em class="sig-param">'BseNI'</em>, <em class="sig-param">'BsePI'</em>, <em class="sig-param">'BseRI'</em>, <em class="sig-param">'BseSI'</em>, <em class="sig-param">'BseX3I'</em>, <em class="sig-param">'BseXI'</em>, <em class="sig-param">'BseYI'</em>, <em class="sig-param">'BsgI'</em>, <em class="sig-param">'Bsh1236I'</em>, <em class="sig-param">'Bsh1285I'</em>, <em class="sig-param">'BshFI'</em>, <em class="sig-param">'BshNI'</em>, <em class="sig-param">'BshTI'</em>, <em class="sig-param">'BshVI'</em>, <em class="sig-param">'BsiEI'</em>, <em class="sig-param">'BsiHKAI'</em>, <em class="sig-param">'BsiHKCI'</em>, <em class="sig-param">'BsiSI'</em>, <em class="sig-param">'BsiWI'</em>, <em class="sig-param">'BslFI'</em>, <em class="sig-param">'BslI'</em>, <em class="sig-param">'BsmAI'</em>, <em class="sig-param">'BsmBI'</em>, <em class="sig-param">'BsmFI'</em>, <em class="sig-param">'BsmI'</em>, <em class="sig-param">'BsnI'</em>, <em class="sig-param">'Bso31I'</em>, <em class="sig-param">'BsoBI'</em>, <em class="sig-param">'Bsp119I'</em>, <em class="sig-param">'Bsp120I'</em>, <em class="sig-param">'Bsp1286I'</em>, <em class="sig-param">'Bsp13I'</em>, <em class="sig-param">'Bsp1407I'</em>, <em class="sig-param">'Bsp143I'</em>, <em class="sig-param">'Bsp1720I'</em>, <em class="sig-param">'Bsp19I'</em>, <em class="sig-param">'Bsp68I'</em>, <em class="sig-param">'BspACI'</em>, <em class="sig-param">'BspANI'</em>, <em class="sig-param">'BspCNI'</em>, <em class="sig-param">'BspDI'</em>, <em class="sig-param">'BspEI'</em>, <em class="sig-param">'BspFNI'</em>, <em class="sig-param">'BspHI'</em>, <em class="sig-param">'BspLI'</em>, <em class="sig-param">'BspMAI'</em>, <em class="sig-param">'BspMI'</em>, <em class="sig-param">'BspOI'</em>, <em class="sig-param">'BspPI'</em>, <em class="sig-param">'BspQI'</em>, <em class="sig-param">'BspT104I'</em>, <em class="sig-param">'BspT107I'</em>, <em class="sig-param">'BspTI'</em>, <em class="sig-param">'BspTNI'</em>, <em class="sig-param">'BsrBI'</em>, <em class="sig-param">'BsrDI'</em>, <em class="sig-param">'BsrFI'</em>, <em class="sig-param">'BsrGI'</em>, <em class="sig-param">'BsrI'</em>, <em class="sig-param">'BssAI'</em>, <em class="sig-param">'BssECI'</em>, <em class="sig-param">'BssHII'</em>, <em class="sig-param">'BssMI'</em>, <em class="sig-param">'BssNAI'</em>, <em class="sig-param">'BssNI'</em>, <em class="sig-param">'BssSI'</em>, <em class="sig-param">'BssT1I'</em>, <em class="sig-param">'Bst1107I'</em>, <em class="sig-param">'Bst2BI'</em>, <em class="sig-param">'Bst2UI'</em>, <em class="sig-param">'Bst4CI'</em>, <em class="sig-param">'Bst6I'</em>, <em class="sig-param">'BstACI'</em>, <em class="sig-param">'BstAFI'</em>, <em class="sig-param">'BstAPI'</em>, <em class="sig-param">'BstAUI'</em>, <em class="sig-param">'BstBAI'</em>, <em class="sig-param">'BstBI'</em>, <em class="sig-param">'BstC8I'</em>, <em class="sig-param">'BstDEI'</em>, <em class="sig-param">'BstDSI'</em>, <em class="sig-param">'BstEII'</em>, <em class="sig-param">'BstENI'</em>, <em class="sig-param">'BstF5I'</em>, <em class="sig-param">'BstFNI'</em>, <em class="sig-param">'BstH2I'</em>, <em class="sig-param">'BstHHI'</em>, <em class="sig-param">'BstKTI'</em>, <em class="sig-param">'BstMAI'</em>, <em class="sig-param">'BstMBI'</em>, <em class="sig-param">'BstMCI'</em>, <em class="sig-param">'BstMWI'</em>, <em class="sig-param">'BstNI'</em>, <em class="sig-param">'BstNSI'</em>, <em class="sig-param">'BstPAI'</em>, <em class="sig-param">'BstPI'</em>, <em class="sig-param">'BstSCI'</em>, <em class="sig-param">'BstSFI'</em>, <em class="sig-param">'BstSLI'</em>, <em class="sig-param">'BstSNI'</em>, <em class="sig-param">'BstUI'</em>, <em class="sig-param">'BstV1I'</em>, <em class="sig-param">'BstV2I'</em>, <em class="sig-param">'BstX2I'</em>, <em class="sig-param">'BstXI'</em>, <em class="sig-param">'BstYI'</em>, <em class="sig-param">'BstZ17I'</em>, <em class="sig-param">'BstZI'</em>, <em class="sig-param">'Bsu15I'</em>, <em class="sig-param">'Bsu36I'</em>, <em class="sig-param">'BsuI'</em>, <em class="sig-param">'BsuRI'</em>, <em class="sig-param">'BsuTUI'</em>, <em class="sig-param">'BtgI'</em>, <em class="sig-param">'BtgZI'</em>, <em class="sig-param">'BtrI'</em>, <em class="sig-param">'BtsCI'</em>, <em class="sig-param">'BtsI'</em>, <em class="sig-param">'BtsIMutI'</em>, <em class="sig-param">'BtuMI'</em>, <em class="sig-param">'BveI'</em>, <em class="sig-param">'Cac8I'</em>, <em class="sig-param">'CaiI'</em>, <em class="sig-param">'CciI'</em>, <em class="sig-param">'CciNI'</em>, <em class="sig-param">'CfoI'</em>, <em class="sig-param">'Cfr10I'</em>, <em class="sig-param">'Cfr13I'</em>, <em class="sig-param">'Cfr42I'</em>, <em class="sig-param">'Cfr9I'</em>, <em class="sig-param">'ClaI'</em>, <em class="sig-param">'CpoI'</em>, <em class="sig-param">'CseI'</em>, <em class="sig-param">'CsiI'</em>, <em class="sig-param">'Csp6I'</em>, <em class="sig-param">'CspAI'</em>, <em class="sig-param">'CspCI'</em>, <em class="sig-param">'CspI'</em>, <em class="sig-param">'CviAII'</em>, <em class="sig-param">'CviJI'</em>, <em class="sig-param">'CviKI_1'</em>, <em class="sig-param">'CviQI'</em>, <em class="sig-param">'DdeI'</em>, <em class="sig-param">'DinI'</em>, <em class="sig-param">'DpnI'</em>, <em class="sig-param">'DpnII'</em>, <em class="sig-param">'DraI'</em>, <em class="sig-param">'DraIII'</em>, <em class="sig-param">'DrdI'</em>, <em class="sig-param">'DriI'</em>, <em class="sig-param">'DseDI'</em>, <em class="sig-param">'EaeI'</em>, <em class="sig-param">'EagI'</em>, <em class="sig-param">'Eam1104I'</em>, <em class="sig-param">'Eam1105I'</em>, <em class="sig-param">'EarI'</em>, <em class="sig-param">'EciI'</em>, <em class="sig-param">'Ecl136II'</em>, <em class="sig-param">'EclXI'</em>, <em class="sig-param">'Eco105I'</em>, <em class="sig-param">'Eco130I'</em>, <em class="sig-param">'Eco147I'</em>, <em class="sig-param">'Eco24I'</em>, <em class="sig-param">'Eco31I'</em>, <em class="sig-param">'Eco32I'</em>, <em class="sig-param">'Eco47I'</em>, <em class="sig-param">'Eco47III'</em>, <em class="sig-param">'Eco52I'</em>, <em class="sig-param">'Eco53kI'</em>, <em class="sig-param">'Eco57I'</em>, <em class="sig-param">'Eco72I'</em>, <em class="sig-param">'Eco81I'</em>, <em class="sig-param">'Eco88I'</em>, <em class="sig-param">'Eco91I'</em>, <em class="sig-param">'EcoICRI'</em>, <em class="sig-param">'EcoNI'</em>, <em class="sig-param">'EcoO109I'</em>, <em class="sig-param">'EcoO65I'</em>, <em class="sig-param">'EcoRI'</em>, <em class="sig-param">'EcoRII'</em>, <em class="sig-param">'EcoRV'</em>, <em class="sig-param">'EcoT14I'</em>, <em class="sig-param">'EcoT22I'</em>, <em class="sig-param">'EcoT38I'</em>, <em class="sig-param">'EgeI'</em>, <em class="sig-param">'EheI'</em>, <em class="sig-param">'ErhI'</em>, <em class="sig-param">'Esp3I'</em>, <em class="sig-param">'FaeI'</em>, <em class="sig-param">'FaiI'</em>, <em class="sig-param">'FalI'</em>, <em class="sig-param">'FaqI'</em>, <em class="sig-param">'FatI'</em>, <em class="sig-param">'FauI'</em>, <em class="sig-param">'FauNDI'</em>, <em class="sig-param">'FbaI'</em>, <em class="sig-param">'FblI'</em>, <em class="sig-param">'Fnu4HI'</em>, <em class="sig-param">'FokI'</em>, <em class="sig-param">'FriOI'</em>, <em class="sig-param">'FseI'</em>, <em class="sig-param">'Fsp4HI'</em>, <em class="sig-param">'FspAI'</em>, <em class="sig-param">'FspBI'</em>, <em class="sig-param">'FspEI'</em>, <em class="sig-param">'FspI'</em>, <em class="sig-param">'GlaI'</em>, <em class="sig-param">'GluI'</em>, <em class="sig-param">'GsaI'</em>, <em class="sig-param">'GsuI'</em>, <em class="sig-param">'HaeII'</em>, <em class="sig-param">'HaeIII'</em>, <em class="sig-param">'HapII'</em>, <em class="sig-param">'HgaI'</em>, <em class="sig-param">'HhaI'</em>, <em class="sig-param">'Hin1I'</em>, <em class="sig-param">'Hin1II'</em>, <em class="sig-param">'Hin6I'</em>, <em class="sig-param">'HinP1I'</em>, <em class="sig-param">'HincII'</em>, <em class="sig-param">'HindII'</em>, <em class="sig-param">'HindIII'</em>, <em class="sig-param">'HinfI'</em>, <em class="sig-param">'HpaI'</em>, <em class="sig-param">'HpaII'</em>, <em class="sig-param">'HphI'</em>, <em class="sig-param">'Hpy166II'</em>, <em class="sig-param">'Hpy188I'</em>, <em class="sig-param">'Hpy188III'</em>, <em class="sig-param">'Hpy8I'</em>, <em class="sig-param">'Hpy99I'</em>, <em class="sig-param">'HpyAV'</em>, <em class="sig-param">'HpyCH4III'</em>, <em class="sig-param">'HpyCH4IV'</em>, <em class="sig-param">'HpyCH4V'</em>, <em class="sig-param">'HpyF10VI'</em>, <em class="sig-param">'HpyF3I'</em>, <em class="sig-param">'HpySE526I'</em>, <em class="sig-param">'Hsp92I'</em>, <em class="sig-param">'Hsp92II'</em>, <em class="sig-param">'HspAI'</em>, <em class="sig-param">'KasI'</em>, <em class="sig-param">'KflI'</em>, <em class="sig-param">'Kpn2I'</em>, <em class="sig-param">'KpnI'</em>, <em class="sig-param">'KroI'</em>, <em class="sig-param">'Ksp22I'</em>, <em class="sig-param">'KspAI'</em>, <em class="sig-param">'KspI'</em>, <em class="sig-param">'Kzo9I'</em>, <em class="sig-param">'LguI'</em>, <em class="sig-param">'LmnI'</em>, <em class="sig-param">'LpnPI'</em>, <em class="sig-param">'Lsp1109I'</em>, <em class="sig-param">'LweI'</em>, <em class="sig-param">'MabI'</em>, <em class="sig-param">'MaeI'</em>, <em class="sig-param">'MaeII'</em>, <em class="sig-param">'MaeIII'</em>, <em class="sig-param">'MalI'</em>, <em class="sig-param">'MauBI'</em>, <em class="sig-param">'MbiI'</em>, <em class="sig-param">'MboI'</em>, <em class="sig-param">'MboII'</em>, <em class="sig-param">'MfeI'</em>, <em class="sig-param">'MflI'</em>, <em class="sig-param">'MhlI'</em>, <em class="sig-param">'MlsI'</em>, <em class="sig-param">'MluCI'</em>, <em class="sig-param">'MluI'</em>, <em class="sig-param">'MluNI'</em>, <em class="sig-param">'Mly113I'</em>, <em class="sig-param">'MlyI'</em>, <em class="sig-param">'MmeI'</em>, <em class="sig-param">'MnlI'</em>, <em class="sig-param">'Mox20I'</em>, <em class="sig-param">'Mph1103I'</em>, <em class="sig-param">'MreI'</em>, <em class="sig-param">'MroI'</em>, <em class="sig-param">'MroNI'</em>, <em class="sig-param">'MroXI'</em>, <em class="sig-param">'MscI'</em>, <em class="sig-param">'MseI'</em>, <em class="sig-param">'MslI'</em>, <em class="sig-param">'Msp20I'</em>, <em class="sig-param">'MspA1I'</em>, <em class="sig-param">'MspCI'</em>, <em class="sig-param">'MspI'</em>, <em class="sig-param">'MspJI'</em>, <em class="sig-param">'MspR9I'</em>, <em class="sig-param">'MssI'</em>, <em class="sig-param">'MteI'</em>, <em class="sig-param">'MunI'</em>, <em class="sig-param">'Mva1269I'</em>, <em class="sig-param">'MvaI'</em>, <em class="sig-param">'MvnI'</em>, <em class="sig-param">'MwoI'</em>, <em class="sig-param">'NaeI'</em>, <em class="sig-param">'NarI'</em>, <em class="sig-param">'NciI'</em>, <em class="sig-param">'NcoI'</em>, <em class="sig-param">'NdeI'</em>, <em class="sig-param">'NdeII'</em>, <em class="sig-param">'NgoMIV'</em>, <em class="sig-param">'NheI'</em>, <em class="sig-param">'NlaIII'</em>, <em class="sig-param">'NlaIV'</em>, <em class="sig-param">'NmeAIII'</em>, <em class="sig-param">'NmuCI'</em>, <em class="sig-param">'NotI'</em>, <em class="sig-param">'NruI'</em>, <em class="sig-param">'NsbI'</em>, <em class="sig-param">'NsiI'</em>, <em class="sig-param">'NspI'</em>, <em class="sig-param">'NspV'</em>, <em class="sig-param">'OliI'</em>, <em class="sig-param">'PacI'</em>, <em class="sig-param">'PaeI'</em>, <em class="sig-param">'PaeR7I'</em>, <em class="sig-param">'PagI'</em>, <em class="sig-param">'PalAI'</em>, <em class="sig-param">'PasI'</em>, <em class="sig-param">'PauI'</em>, <em class="sig-param">'PceI'</em>, <em class="sig-param">'PciI'</em>, <em class="sig-param">'PciSI'</em>, <em class="sig-param">'PcsI'</em>, <em class="sig-param">'PctI'</em>, <em class="sig-param">'PdiI'</em>, <em class="sig-param">'PdmI'</em>, <em class="sig-param">'PfeI'</em>, <em class="sig-param">'Pfl23II'</em>, <em class="sig-param">'PflFI'</em>, <em class="sig-param">'PflMI'</em>, <em class="sig-param">'PfoI'</em>, <em class="sig-param">'PinAI'</em>, <em class="sig-param">'PkrI'</em>, <em class="sig-param">'Ple19I'</em>, <em class="sig-param">'PleI'</em>, <em class="sig-param">'PluTI'</em>, <em class="sig-param">'PmaCI'</em>, <em class="sig-param">'PmeI'</em>, <em class="sig-param">'PmlI'</em>, <em class="sig-param">'PpsI'</em>, <em class="sig-param">'Ppu21I'</em>, <em class="sig-param">'PpuMI'</em>, <em class="sig-param">'PscI'</em>, <em class="sig-param">'PshAI'</em>, <em class="sig-param">'PshBI'</em>, <em class="sig-param">'PsiI'</em>, <em class="sig-param">'Psp124BI'</em>, <em class="sig-param">'Psp1406I'</em>, <em class="sig-param">'Psp5II'</em>, <em class="sig-param">'Psp6I'</em>, <em class="sig-param">'PspCI'</em>, <em class="sig-param">'PspEI'</em>, <em class="sig-param">'PspFI'</em>, <em class="sig-param">'PspGI'</em>, <em class="sig-param">'PspLI'</em>, <em class="sig-param">'PspN4I'</em>, <em class="sig-param">'PspOMI'</em>, <em class="sig-param">'PspPI'</em>, <em class="sig-param">'PspPPI'</em>, <em class="sig-param">'PspXI'</em>, <em class="sig-param">'PsrI'</em>, <em class="sig-param">'PstI'</em>, <em class="sig-param">'PstNI'</em>, <em class="sig-param">'PsuI'</em>, <em class="sig-param">'PsyI'</em>, <em class="sig-param">'PteI'</em>, <em class="sig-param">'PvuI'</em>, <em class="sig-param">'PvuII'</em>, <em class="sig-param">'RgaI'</em>, <em class="sig-param">'RigI'</em>, <em class="sig-param">'RruI'</em>, <em class="sig-param">'RsaI'</em>, <em class="sig-param">'RsaNI'</em>, <em class="sig-param">'RseI'</em>, <em class="sig-param">'Rsr2I'</em>, <em class="sig-param">'RsrII'</em>, <em class="sig-param">'SacI'</em>, <em class="sig-param">'SacII'</em>, <em class="sig-param">'SalI'</em>, <em class="sig-param">'SapI'</em>, <em class="sig-param">'SaqAI'</em>, <em class="sig-param">'SatI'</em>, <em class="sig-param">'Sau3AI'</em>, <em class="sig-param">'Sau96I'</em>, <em class="sig-param">'SbfI'</em>, <em class="sig-param">'ScaI'</em>, <em class="sig-param">'SchI'</em>, <em class="sig-param">'ScrFI'</em>, <em class="sig-param">'SdaI'</em>, <em class="sig-param">'SduI'</em>, <em class="sig-param">'SetI'</em>, <em class="sig-param">'SexAI'</em>, <em class="sig-param">'SfaAI'</em>, <em class="sig-param">'SfaNI'</em>, <em class="sig-param">'SfcI'</em>, <em class="sig-param">'SfiI'</em>, <em class="sig-param">'SfoI'</em>, <em class="sig-param">'Sfr274I'</em>, <em class="sig-param">'Sfr303I'</em>, <em class="sig-param">'SfuI'</em>, <em class="sig-param">'SgeI'</em>, <em class="sig-param">'SgfI'</em>, <em class="sig-param">'SgrAI'</em>, <em class="sig-param">'SgrBI'</em>, <em class="sig-param">'SgrDI'</em>, <em class="sig-param">'SgsI'</em>, <em class="sig-param">'SinI'</em>, <em class="sig-param">'SlaI'</em>, <em class="sig-param">'SmaI'</em>, <em class="sig-param">'SmiI'</em>, <em class="sig-param">'SmiMI'</em>, <em class="sig-param">'SmlI'</em>, <em class="sig-param">'SmoI'</em>, <em class="sig-param">'SnaBI'</em>, <em class="sig-param">'SpeI'</em>, <em class="sig-param">'SphI'</em>, <em class="sig-param">'SrfI'</em>, <em class="sig-param">'Sse8387I'</em>, <em class="sig-param">'Sse9I'</em>, <em class="sig-param">'SseBI'</em>, <em class="sig-param">'SsiI'</em>, <em class="sig-param">'SspDI'</em>, <em class="sig-param">'SspI'</em>, <em class="sig-param">'SspMI'</em>, <em class="sig-param">'SstI'</em>, <em class="sig-param">'StuI'</em>, <em class="sig-param">'StyD4I'</em>, <em class="sig-param">'StyI'</em>, <em class="sig-param">'SwaI'</em>, <em class="sig-param">'TaaI'</em>, <em class="sig-param">'TaiI'</em>, <em class="sig-param">'TaqI'</em>, <em class="sig-param">'TaqII'</em>, <em class="sig-param">'TasI'</em>, <em class="sig-param">'TatI'</em>, <em class="sig-param">'TauI'</em>, <em class="sig-param">'TfiI'</em>, <em class="sig-param">'Tru1I'</em>, <em class="sig-param">'Tru9I'</em>, <em class="sig-param">'TscAI'</em>, <em class="sig-param">'TseFI'</em>, <em class="sig-param">'TseI'</em>, <em class="sig-param">'Tsp45I'</em>, <em class="sig-param">'TspDTI'</em>, <em class="sig-param">'TspGWI'</em>, <em class="sig-param">'TspMI'</em>, <em class="sig-param">'TspRI'</em>, <em class="sig-param">'Tth111I'</em>, <em class="sig-param">'Van91I'</em>, <em class="sig-param">'Vha464I'</em>, <em class="sig-param">'VneI'</em>, <em class="sig-param">'VpaK11BI'</em>, <em class="sig-param">'VspI'</em>, <em class="sig-param">'XagI'</em>, <em class="sig-param">'XapI'</em>, <em class="sig-param">'XbaI'</em>, <em class="sig-param">'XceI'</em>, <em class="sig-param">'XcmI'</em>, <em class="sig-param">'XhoI'</em>, <em class="sig-param">'XmaI'</em>, <em class="sig-param">'XmaJI'</em>, <em class="sig-param">'XmiI'</em>, <em class="sig-param">'XmnI'</em>, <em class="sig-param">'XspI'</em>, <em class="sig-param">'ZraI'</em>, <em class="sig-param">'ZrmI'</em>, <em class="sig-param">'Zsp2I'])</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.dseq.Dseq.once_cutters" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the enzymes in a RestrictionBatch that cut the sequence 
exactly once.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.ovhg">
<em class="property">property </em><code class="sig-name descname">ovhg</code><a class="headerlink" href="#pydna.dseq.Dseq.ovhg" title="Permalink to this definition">¶</a></dt>
<dd><p>The ovhg property. This cannot be set directly, but is a 
consequence of how the watson and crick strands anneal to 
each other</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.quick">
<em class="property">classmethod </em><code class="sig-name descname">quick</code><span class="sig-paren">(</span><em class="sig-param">watson: str</em>, <em class="sig-param">crick: str</em>, <em class="sig-param">ovhg=0</em>, <em class="sig-param">linear=True</em>, <em class="sig-param">circular=False</em>, <em class="sig-param">pos=0</em>, <em class="sig-param">alphabet=IUPACAmbiguousDNA()</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.quick"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.quick" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.rc">
<code class="sig-name descname">rc</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.dseq.Dseq.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a Dseq object where watson and crick have switched
places. This represents the same double stranded sequence.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="k">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;catcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatg</span>
<span class="go">ctagctac</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.reverse_complement">
<code class="sig-name descname">reverse_complement</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.reverse_complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a Dseq object where watson and crick have switched
places. This represents the same double stranded sequence.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="k">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;catcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatg</span>
<span class="go">ctagctac</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.seguid">
<code class="sig-name descname">seguid</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.seguid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.seguid" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the SEGUID for the Dseq. The definition 
varies with the amount of stagger between the sequences.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.shifted">
<code class="sig-name descname">shifted</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">shift</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.shifted"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.shifted" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a shifted version of a circular Dseq object.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.t4">
<code class="sig-name descname">t4</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">nucleotides=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.dseq.Dseq.t4" title="Permalink to this definition">¶</a></dt>
<dd><p>Fill in five prime protruding ends and chewing back
three prime protruding ends by a DNA polymerase providing both
5’-3’ DNA polymerase activity and 3’-5’ nuclease acitivty
(such as T4 DNA polymerase). This can be done in presence of any
combination of the four A, G, C or T. Removing one or more nucleotides
can facilitate engineering of sticky ends. Default are all four nucleotides together.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>nucleotides</strong><span class="classifier">str</span></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="k">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;gatcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">()</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;t&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgat</span>
<span class="go"> tagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;a&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcga</span>
<span class="go">  agctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s2">&quot;g&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcg</span>
<span class="go">   gctag</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.three_prime_end">
<code class="sig-name descname">three_prime_end</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.three_prime_end"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.three_prime_end" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a tuple describing the structure of the 5’ end of
the DNA fragment</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="k">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&#39;blunt&#39;, &#39;&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go"> aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&quot;3&#39;&quot;, &#39;a&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="s2">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go">aaa</span>
<span class="go"> ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&quot;5&#39;&quot;, &#39;t&#39;)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pydna.dseq.Dseq.five_prime_end" title="pydna.dseq.Dseq.five_prime_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.five_prime_end</span></code></a></dt><dd></dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.tolinear">
<code class="sig-name descname">tolinear</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.tolinear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.tolinear" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a blunt, linear copy of a circular Dseq object. This can
only be done if the Dseq object is circular, otherwise a
TypeError is raised.</p>
<p>This method is deprecated, use slicing instead. See example below.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseq</span> <span class="k">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s2">&quot;catcgatc&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(o8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="p">[:]</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.twice_cutters">
<code class="sig-name descname">twice_cutters</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">batch=RestrictionBatch(['AanI'</em>, <em class="sig-param">'AarI'</em>, <em class="sig-param">'AasI'</em>, <em class="sig-param">'AatII'</em>, <em class="sig-param">'AbaSI'</em>, <em class="sig-param">'AbsI'</em>, <em class="sig-param">'Acc16I'</em>, <em class="sig-param">'Acc36I'</em>, <em class="sig-param">'Acc65I'</em>, <em class="sig-param">'AccB1I'</em>, <em class="sig-param">'AccB7I'</em>, <em class="sig-param">'AccBSI'</em>, <em class="sig-param">'AccI'</em>, <em class="sig-param">'AccII'</em>, <em class="sig-param">'AccIII'</em>, <em class="sig-param">'AciI'</em>, <em class="sig-param">'AclI'</em>, <em class="sig-param">'AclWI'</em>, <em class="sig-param">'AcoI'</em>, <em class="sig-param">'AcsI'</em>, <em class="sig-param">'AcuI'</em>, <em class="sig-param">'AcvI'</em>, <em class="sig-param">'AcyI'</em>, <em class="sig-param">'AdeI'</em>, <em class="sig-param">'AfaI'</em>, <em class="sig-param">'AfeI'</em>, <em class="sig-param">'AfiI'</em>, <em class="sig-param">'AflII'</em>, <em class="sig-param">'AflIII'</em>, <em class="sig-param">'AgeI'</em>, <em class="sig-param">'AgsI'</em>, <em class="sig-param">'AhdI'</em>, <em class="sig-param">'AhlI'</em>, <em class="sig-param">'AjiI'</em>, <em class="sig-param">'AjnI'</em>, <em class="sig-param">'AjuI'</em>, <em class="sig-param">'AleI'</em>, <em class="sig-param">'AloI'</em>, <em class="sig-param">'AluBI'</em>, <em class="sig-param">'AluI'</em>, <em class="sig-param">'Alw21I'</em>, <em class="sig-param">'Alw26I'</em>, <em class="sig-param">'Alw44I'</em>, <em class="sig-param">'AlwI'</em>, <em class="sig-param">'AlwNI'</em>, <em class="sig-param">'Ama87I'</em>, <em class="sig-param">'Aor13HI'</em>, <em class="sig-param">'Aor51HI'</em>, <em class="sig-param">'AoxI'</em>, <em class="sig-param">'ApaI'</em>, <em class="sig-param">'ApaLI'</em>, <em class="sig-param">'ApeKI'</em>, <em class="sig-param">'ApoI'</em>, <em class="sig-param">'ArsI'</em>, <em class="sig-param">'AscI'</em>, <em class="sig-param">'AseI'</em>, <em class="sig-param">'AsiGI'</em>, <em class="sig-param">'AsiSI'</em>, <em class="sig-param">'Asp700I'</em>, <em class="sig-param">'Asp718I'</em>, <em class="sig-param">'AspA2I'</em>, <em class="sig-param">'AspLEI'</em>, <em class="sig-param">'AspS9I'</em>, <em class="sig-param">'AsuC2I'</em>, <em class="sig-param">'AsuHPI'</em>, <em class="sig-param">'AsuII'</em>, <em class="sig-param">'AsuNHI'</em>, <em class="sig-param">'AvaI'</em>, <em class="sig-param">'AvaII'</em>, <em class="sig-param">'AvrII'</em>, <em class="sig-param">'AxyI'</em>, <em class="sig-param">'BaeGI'</em>, <em class="sig-param">'BaeI'</em>, <em class="sig-param">'BalI'</em>, <em class="sig-param">'BamHI'</em>, <em class="sig-param">'BanI'</em>, <em class="sig-param">'BanII'</em>, <em class="sig-param">'BarI'</em>, <em class="sig-param">'BauI'</em>, <em class="sig-param">'BbrPI'</em>, <em class="sig-param">'BbsI'</em>, <em class="sig-param">'Bbv12I'</em>, <em class="sig-param">'BbvCI'</em>, <em class="sig-param">'BbvI'</em>, <em class="sig-param">'BccI'</em>, <em class="sig-param">'BceAI'</em>, <em class="sig-param">'BcgI'</em>, <em class="sig-param">'BciT130I'</em>, <em class="sig-param">'BciVI'</em>, <em class="sig-param">'BclI'</em>, <em class="sig-param">'BcnI'</em>, <em class="sig-param">'BcoDI'</em>, <em class="sig-param">'BcuI'</em>, <em class="sig-param">'BfaI'</em>, <em class="sig-param">'BfmI'</em>, <em class="sig-param">'BfoI'</em>, <em class="sig-param">'BfrI'</em>, <em class="sig-param">'BfuAI'</em>, <em class="sig-param">'BfuI'</em>, <em class="sig-param">'BglI'</em>, <em class="sig-param">'BglII'</em>, <em class="sig-param">'BisI'</em>, <em class="sig-param">'BlnI'</em>, <em class="sig-param">'BlpI'</em>, <em class="sig-param">'BlsI'</em>, <em class="sig-param">'BmcAI'</em>, <em class="sig-param">'Bme1390I'</em>, <em class="sig-param">'Bme18I'</em>, <em class="sig-param">'BmeRI'</em>, <em class="sig-param">'BmeT110I'</em>, <em class="sig-param">'BmgBI'</em>, <em class="sig-param">'BmgT120I'</em>, <em class="sig-param">'BmiI'</em>, <em class="sig-param">'BmrFI'</em>, <em class="sig-param">'BmrI'</em>, <em class="sig-param">'BmsI'</em>, <em class="sig-param">'BmtI'</em>, <em class="sig-param">'BmuI'</em>, <em class="sig-param">'BoxI'</em>, <em class="sig-param">'BpiI'</em>, <em class="sig-param">'BplI'</em>, <em class="sig-param">'BpmI'</em>, <em class="sig-param">'Bpu10I'</em>, <em class="sig-param">'Bpu1102I'</em>, <em class="sig-param">'Bpu14I'</em>, <em class="sig-param">'BpuEI'</em>, <em class="sig-param">'BpuMI'</em>, <em class="sig-param">'Bsa29I'</em>, <em class="sig-param">'BsaAI'</em>, <em class="sig-param">'BsaBI'</em>, <em class="sig-param">'BsaHI'</em>, <em class="sig-param">'BsaI'</em>, <em class="sig-param">'BsaJI'</em>, <em class="sig-param">'BsaWI'</em>, <em class="sig-param">'BsaXI'</em>, <em class="sig-param">'Bsc4I'</em>, <em class="sig-param">'Bse118I'</em>, <em class="sig-param">'Bse1I'</em>, <em class="sig-param">'Bse21I'</em>, <em class="sig-param">'Bse3DI'</em>, <em class="sig-param">'Bse8I'</em>, <em class="sig-param">'BseAI'</em>, <em class="sig-param">'BseBI'</em>, <em class="sig-param">'BseCI'</em>, <em class="sig-param">'BseDI'</em>, <em class="sig-param">'BseGI'</em>, <em class="sig-param">'BseJI'</em>, <em class="sig-param">'BseLI'</em>, <em class="sig-param">'BseMI'</em>, <em class="sig-param">'BseMII'</em>, <em class="sig-param">'BseNI'</em>, <em class="sig-param">'BsePI'</em>, <em class="sig-param">'BseRI'</em>, <em class="sig-param">'BseSI'</em>, <em class="sig-param">'BseX3I'</em>, <em class="sig-param">'BseXI'</em>, <em class="sig-param">'BseYI'</em>, <em class="sig-param">'BsgI'</em>, <em class="sig-param">'Bsh1236I'</em>, <em class="sig-param">'Bsh1285I'</em>, <em class="sig-param">'BshFI'</em>, <em class="sig-param">'BshNI'</em>, <em class="sig-param">'BshTI'</em>, <em class="sig-param">'BshVI'</em>, <em class="sig-param">'BsiEI'</em>, <em class="sig-param">'BsiHKAI'</em>, <em class="sig-param">'BsiHKCI'</em>, <em class="sig-param">'BsiSI'</em>, <em class="sig-param">'BsiWI'</em>, <em class="sig-param">'BslFI'</em>, <em class="sig-param">'BslI'</em>, <em class="sig-param">'BsmAI'</em>, <em class="sig-param">'BsmBI'</em>, <em class="sig-param">'BsmFI'</em>, <em class="sig-param">'BsmI'</em>, <em class="sig-param">'BsnI'</em>, <em class="sig-param">'Bso31I'</em>, <em class="sig-param">'BsoBI'</em>, <em class="sig-param">'Bsp119I'</em>, <em class="sig-param">'Bsp120I'</em>, <em class="sig-param">'Bsp1286I'</em>, <em class="sig-param">'Bsp13I'</em>, <em class="sig-param">'Bsp1407I'</em>, <em class="sig-param">'Bsp143I'</em>, <em class="sig-param">'Bsp1720I'</em>, <em class="sig-param">'Bsp19I'</em>, <em class="sig-param">'Bsp68I'</em>, <em class="sig-param">'BspACI'</em>, <em class="sig-param">'BspANI'</em>, <em class="sig-param">'BspCNI'</em>, <em class="sig-param">'BspDI'</em>, <em class="sig-param">'BspEI'</em>, <em class="sig-param">'BspFNI'</em>, <em class="sig-param">'BspHI'</em>, <em class="sig-param">'BspLI'</em>, <em class="sig-param">'BspMAI'</em>, <em class="sig-param">'BspMI'</em>, <em class="sig-param">'BspOI'</em>, <em class="sig-param">'BspPI'</em>, <em class="sig-param">'BspQI'</em>, <em class="sig-param">'BspT104I'</em>, <em class="sig-param">'BspT107I'</em>, <em class="sig-param">'BspTI'</em>, <em class="sig-param">'BspTNI'</em>, <em class="sig-param">'BsrBI'</em>, <em class="sig-param">'BsrDI'</em>, <em class="sig-param">'BsrFI'</em>, <em class="sig-param">'BsrGI'</em>, <em class="sig-param">'BsrI'</em>, <em class="sig-param">'BssAI'</em>, <em class="sig-param">'BssECI'</em>, <em class="sig-param">'BssHII'</em>, <em class="sig-param">'BssMI'</em>, <em class="sig-param">'BssNAI'</em>, <em class="sig-param">'BssNI'</em>, <em class="sig-param">'BssSI'</em>, <em class="sig-param">'BssT1I'</em>, <em class="sig-param">'Bst1107I'</em>, <em class="sig-param">'Bst2BI'</em>, <em class="sig-param">'Bst2UI'</em>, <em class="sig-param">'Bst4CI'</em>, <em class="sig-param">'Bst6I'</em>, <em class="sig-param">'BstACI'</em>, <em class="sig-param">'BstAFI'</em>, <em class="sig-param">'BstAPI'</em>, <em class="sig-param">'BstAUI'</em>, <em class="sig-param">'BstBAI'</em>, <em class="sig-param">'BstBI'</em>, <em class="sig-param">'BstC8I'</em>, <em class="sig-param">'BstDEI'</em>, <em class="sig-param">'BstDSI'</em>, <em class="sig-param">'BstEII'</em>, <em class="sig-param">'BstENI'</em>, <em class="sig-param">'BstF5I'</em>, <em class="sig-param">'BstFNI'</em>, <em class="sig-param">'BstH2I'</em>, <em class="sig-param">'BstHHI'</em>, <em class="sig-param">'BstKTI'</em>, <em class="sig-param">'BstMAI'</em>, <em class="sig-param">'BstMBI'</em>, <em class="sig-param">'BstMCI'</em>, <em class="sig-param">'BstMWI'</em>, <em class="sig-param">'BstNI'</em>, <em class="sig-param">'BstNSI'</em>, <em class="sig-param">'BstPAI'</em>, <em class="sig-param">'BstPI'</em>, <em class="sig-param">'BstSCI'</em>, <em class="sig-param">'BstSFI'</em>, <em class="sig-param">'BstSLI'</em>, <em class="sig-param">'BstSNI'</em>, <em class="sig-param">'BstUI'</em>, <em class="sig-param">'BstV1I'</em>, <em class="sig-param">'BstV2I'</em>, <em class="sig-param">'BstX2I'</em>, <em class="sig-param">'BstXI'</em>, <em class="sig-param">'BstYI'</em>, <em class="sig-param">'BstZ17I'</em>, <em class="sig-param">'BstZI'</em>, <em class="sig-param">'Bsu15I'</em>, <em class="sig-param">'Bsu36I'</em>, <em class="sig-param">'BsuI'</em>, <em class="sig-param">'BsuRI'</em>, <em class="sig-param">'BsuTUI'</em>, <em class="sig-param">'BtgI'</em>, <em class="sig-param">'BtgZI'</em>, <em class="sig-param">'BtrI'</em>, <em class="sig-param">'BtsCI'</em>, <em class="sig-param">'BtsI'</em>, <em class="sig-param">'BtsIMutI'</em>, <em class="sig-param">'BtuMI'</em>, <em class="sig-param">'BveI'</em>, <em class="sig-param">'Cac8I'</em>, <em class="sig-param">'CaiI'</em>, <em class="sig-param">'CciI'</em>, <em class="sig-param">'CciNI'</em>, <em class="sig-param">'CfoI'</em>, <em class="sig-param">'Cfr10I'</em>, <em class="sig-param">'Cfr13I'</em>, <em class="sig-param">'Cfr42I'</em>, <em class="sig-param">'Cfr9I'</em>, <em class="sig-param">'ClaI'</em>, <em class="sig-param">'CpoI'</em>, <em class="sig-param">'CseI'</em>, <em class="sig-param">'CsiI'</em>, <em class="sig-param">'Csp6I'</em>, <em class="sig-param">'CspAI'</em>, <em class="sig-param">'CspCI'</em>, <em class="sig-param">'CspI'</em>, <em class="sig-param">'CviAII'</em>, <em class="sig-param">'CviJI'</em>, <em class="sig-param">'CviKI_1'</em>, <em class="sig-param">'CviQI'</em>, <em class="sig-param">'DdeI'</em>, <em class="sig-param">'DinI'</em>, <em class="sig-param">'DpnI'</em>, <em class="sig-param">'DpnII'</em>, <em class="sig-param">'DraI'</em>, <em class="sig-param">'DraIII'</em>, <em class="sig-param">'DrdI'</em>, <em class="sig-param">'DriI'</em>, <em class="sig-param">'DseDI'</em>, <em class="sig-param">'EaeI'</em>, <em class="sig-param">'EagI'</em>, <em class="sig-param">'Eam1104I'</em>, <em class="sig-param">'Eam1105I'</em>, <em class="sig-param">'EarI'</em>, <em class="sig-param">'EciI'</em>, <em class="sig-param">'Ecl136II'</em>, <em class="sig-param">'EclXI'</em>, <em class="sig-param">'Eco105I'</em>, <em class="sig-param">'Eco130I'</em>, <em class="sig-param">'Eco147I'</em>, <em class="sig-param">'Eco24I'</em>, <em class="sig-param">'Eco31I'</em>, <em class="sig-param">'Eco32I'</em>, <em class="sig-param">'Eco47I'</em>, <em class="sig-param">'Eco47III'</em>, <em class="sig-param">'Eco52I'</em>, <em class="sig-param">'Eco53kI'</em>, <em class="sig-param">'Eco57I'</em>, <em class="sig-param">'Eco72I'</em>, <em class="sig-param">'Eco81I'</em>, <em class="sig-param">'Eco88I'</em>, <em class="sig-param">'Eco91I'</em>, <em class="sig-param">'EcoICRI'</em>, <em class="sig-param">'EcoNI'</em>, <em class="sig-param">'EcoO109I'</em>, <em class="sig-param">'EcoO65I'</em>, <em class="sig-param">'EcoRI'</em>, <em class="sig-param">'EcoRII'</em>, <em class="sig-param">'EcoRV'</em>, <em class="sig-param">'EcoT14I'</em>, <em class="sig-param">'EcoT22I'</em>, <em class="sig-param">'EcoT38I'</em>, <em class="sig-param">'EgeI'</em>, <em class="sig-param">'EheI'</em>, <em class="sig-param">'ErhI'</em>, <em class="sig-param">'Esp3I'</em>, <em class="sig-param">'FaeI'</em>, <em class="sig-param">'FaiI'</em>, <em class="sig-param">'FalI'</em>, <em class="sig-param">'FaqI'</em>, <em class="sig-param">'FatI'</em>, <em class="sig-param">'FauI'</em>, <em class="sig-param">'FauNDI'</em>, <em class="sig-param">'FbaI'</em>, <em class="sig-param">'FblI'</em>, <em class="sig-param">'Fnu4HI'</em>, <em class="sig-param">'FokI'</em>, <em class="sig-param">'FriOI'</em>, <em class="sig-param">'FseI'</em>, <em class="sig-param">'Fsp4HI'</em>, <em class="sig-param">'FspAI'</em>, <em class="sig-param">'FspBI'</em>, <em class="sig-param">'FspEI'</em>, <em class="sig-param">'FspI'</em>, <em class="sig-param">'GlaI'</em>, <em class="sig-param">'GluI'</em>, <em class="sig-param">'GsaI'</em>, <em class="sig-param">'GsuI'</em>, <em class="sig-param">'HaeII'</em>, <em class="sig-param">'HaeIII'</em>, <em class="sig-param">'HapII'</em>, <em class="sig-param">'HgaI'</em>, <em class="sig-param">'HhaI'</em>, <em class="sig-param">'Hin1I'</em>, <em class="sig-param">'Hin1II'</em>, <em class="sig-param">'Hin6I'</em>, <em class="sig-param">'HinP1I'</em>, <em class="sig-param">'HincII'</em>, <em class="sig-param">'HindII'</em>, <em class="sig-param">'HindIII'</em>, <em class="sig-param">'HinfI'</em>, <em class="sig-param">'HpaI'</em>, <em class="sig-param">'HpaII'</em>, <em class="sig-param">'HphI'</em>, <em class="sig-param">'Hpy166II'</em>, <em class="sig-param">'Hpy188I'</em>, <em class="sig-param">'Hpy188III'</em>, <em class="sig-param">'Hpy8I'</em>, <em class="sig-param">'Hpy99I'</em>, <em class="sig-param">'HpyAV'</em>, <em class="sig-param">'HpyCH4III'</em>, <em class="sig-param">'HpyCH4IV'</em>, <em class="sig-param">'HpyCH4V'</em>, <em class="sig-param">'HpyF10VI'</em>, <em class="sig-param">'HpyF3I'</em>, <em class="sig-param">'HpySE526I'</em>, <em class="sig-param">'Hsp92I'</em>, <em class="sig-param">'Hsp92II'</em>, <em class="sig-param">'HspAI'</em>, <em class="sig-param">'KasI'</em>, <em class="sig-param">'KflI'</em>, <em class="sig-param">'Kpn2I'</em>, <em class="sig-param">'KpnI'</em>, <em class="sig-param">'KroI'</em>, <em class="sig-param">'Ksp22I'</em>, <em class="sig-param">'KspAI'</em>, <em class="sig-param">'KspI'</em>, <em class="sig-param">'Kzo9I'</em>, <em class="sig-param">'LguI'</em>, <em class="sig-param">'LmnI'</em>, <em class="sig-param">'LpnPI'</em>, <em class="sig-param">'Lsp1109I'</em>, <em class="sig-param">'LweI'</em>, <em class="sig-param">'MabI'</em>, <em class="sig-param">'MaeI'</em>, <em class="sig-param">'MaeII'</em>, <em class="sig-param">'MaeIII'</em>, <em class="sig-param">'MalI'</em>, <em class="sig-param">'MauBI'</em>, <em class="sig-param">'MbiI'</em>, <em class="sig-param">'MboI'</em>, <em class="sig-param">'MboII'</em>, <em class="sig-param">'MfeI'</em>, <em class="sig-param">'MflI'</em>, <em class="sig-param">'MhlI'</em>, <em class="sig-param">'MlsI'</em>, <em class="sig-param">'MluCI'</em>, <em class="sig-param">'MluI'</em>, <em class="sig-param">'MluNI'</em>, <em class="sig-param">'Mly113I'</em>, <em class="sig-param">'MlyI'</em>, <em class="sig-param">'MmeI'</em>, <em class="sig-param">'MnlI'</em>, <em class="sig-param">'Mox20I'</em>, <em class="sig-param">'Mph1103I'</em>, <em class="sig-param">'MreI'</em>, <em class="sig-param">'MroI'</em>, <em class="sig-param">'MroNI'</em>, <em class="sig-param">'MroXI'</em>, <em class="sig-param">'MscI'</em>, <em class="sig-param">'MseI'</em>, <em class="sig-param">'MslI'</em>, <em class="sig-param">'Msp20I'</em>, <em class="sig-param">'MspA1I'</em>, <em class="sig-param">'MspCI'</em>, <em class="sig-param">'MspI'</em>, <em class="sig-param">'MspJI'</em>, <em class="sig-param">'MspR9I'</em>, <em class="sig-param">'MssI'</em>, <em class="sig-param">'MteI'</em>, <em class="sig-param">'MunI'</em>, <em class="sig-param">'Mva1269I'</em>, <em class="sig-param">'MvaI'</em>, <em class="sig-param">'MvnI'</em>, <em class="sig-param">'MwoI'</em>, <em class="sig-param">'NaeI'</em>, <em class="sig-param">'NarI'</em>, <em class="sig-param">'NciI'</em>, <em class="sig-param">'NcoI'</em>, <em class="sig-param">'NdeI'</em>, <em class="sig-param">'NdeII'</em>, <em class="sig-param">'NgoMIV'</em>, <em class="sig-param">'NheI'</em>, <em class="sig-param">'NlaIII'</em>, <em class="sig-param">'NlaIV'</em>, <em class="sig-param">'NmeAIII'</em>, <em class="sig-param">'NmuCI'</em>, <em class="sig-param">'NotI'</em>, <em class="sig-param">'NruI'</em>, <em class="sig-param">'NsbI'</em>, <em class="sig-param">'NsiI'</em>, <em class="sig-param">'NspI'</em>, <em class="sig-param">'NspV'</em>, <em class="sig-param">'OliI'</em>, <em class="sig-param">'PacI'</em>, <em class="sig-param">'PaeI'</em>, <em class="sig-param">'PaeR7I'</em>, <em class="sig-param">'PagI'</em>, <em class="sig-param">'PalAI'</em>, <em class="sig-param">'PasI'</em>, <em class="sig-param">'PauI'</em>, <em class="sig-param">'PceI'</em>, <em class="sig-param">'PciI'</em>, <em class="sig-param">'PciSI'</em>, <em class="sig-param">'PcsI'</em>, <em class="sig-param">'PctI'</em>, <em class="sig-param">'PdiI'</em>, <em class="sig-param">'PdmI'</em>, <em class="sig-param">'PfeI'</em>, <em class="sig-param">'Pfl23II'</em>, <em class="sig-param">'PflFI'</em>, <em class="sig-param">'PflMI'</em>, <em class="sig-param">'PfoI'</em>, <em class="sig-param">'PinAI'</em>, <em class="sig-param">'PkrI'</em>, <em class="sig-param">'Ple19I'</em>, <em class="sig-param">'PleI'</em>, <em class="sig-param">'PluTI'</em>, <em class="sig-param">'PmaCI'</em>, <em class="sig-param">'PmeI'</em>, <em class="sig-param">'PmlI'</em>, <em class="sig-param">'PpsI'</em>, <em class="sig-param">'Ppu21I'</em>, <em class="sig-param">'PpuMI'</em>, <em class="sig-param">'PscI'</em>, <em class="sig-param">'PshAI'</em>, <em class="sig-param">'PshBI'</em>, <em class="sig-param">'PsiI'</em>, <em class="sig-param">'Psp124BI'</em>, <em class="sig-param">'Psp1406I'</em>, <em class="sig-param">'Psp5II'</em>, <em class="sig-param">'Psp6I'</em>, <em class="sig-param">'PspCI'</em>, <em class="sig-param">'PspEI'</em>, <em class="sig-param">'PspFI'</em>, <em class="sig-param">'PspGI'</em>, <em class="sig-param">'PspLI'</em>, <em class="sig-param">'PspN4I'</em>, <em class="sig-param">'PspOMI'</em>, <em class="sig-param">'PspPI'</em>, <em class="sig-param">'PspPPI'</em>, <em class="sig-param">'PspXI'</em>, <em class="sig-param">'PsrI'</em>, <em class="sig-param">'PstI'</em>, <em class="sig-param">'PstNI'</em>, <em class="sig-param">'PsuI'</em>, <em class="sig-param">'PsyI'</em>, <em class="sig-param">'PteI'</em>, <em class="sig-param">'PvuI'</em>, <em class="sig-param">'PvuII'</em>, <em class="sig-param">'RgaI'</em>, <em class="sig-param">'RigI'</em>, <em class="sig-param">'RruI'</em>, <em class="sig-param">'RsaI'</em>, <em class="sig-param">'RsaNI'</em>, <em class="sig-param">'RseI'</em>, <em class="sig-param">'Rsr2I'</em>, <em class="sig-param">'RsrII'</em>, <em class="sig-param">'SacI'</em>, <em class="sig-param">'SacII'</em>, <em class="sig-param">'SalI'</em>, <em class="sig-param">'SapI'</em>, <em class="sig-param">'SaqAI'</em>, <em class="sig-param">'SatI'</em>, <em class="sig-param">'Sau3AI'</em>, <em class="sig-param">'Sau96I'</em>, <em class="sig-param">'SbfI'</em>, <em class="sig-param">'ScaI'</em>, <em class="sig-param">'SchI'</em>, <em class="sig-param">'ScrFI'</em>, <em class="sig-param">'SdaI'</em>, <em class="sig-param">'SduI'</em>, <em class="sig-param">'SetI'</em>, <em class="sig-param">'SexAI'</em>, <em class="sig-param">'SfaAI'</em>, <em class="sig-param">'SfaNI'</em>, <em class="sig-param">'SfcI'</em>, <em class="sig-param">'SfiI'</em>, <em class="sig-param">'SfoI'</em>, <em class="sig-param">'Sfr274I'</em>, <em class="sig-param">'Sfr303I'</em>, <em class="sig-param">'SfuI'</em>, <em class="sig-param">'SgeI'</em>, <em class="sig-param">'SgfI'</em>, <em class="sig-param">'SgrAI'</em>, <em class="sig-param">'SgrBI'</em>, <em class="sig-param">'SgrDI'</em>, <em class="sig-param">'SgsI'</em>, <em class="sig-param">'SinI'</em>, <em class="sig-param">'SlaI'</em>, <em class="sig-param">'SmaI'</em>, <em class="sig-param">'SmiI'</em>, <em class="sig-param">'SmiMI'</em>, <em class="sig-param">'SmlI'</em>, <em class="sig-param">'SmoI'</em>, <em class="sig-param">'SnaBI'</em>, <em class="sig-param">'SpeI'</em>, <em class="sig-param">'SphI'</em>, <em class="sig-param">'SrfI'</em>, <em class="sig-param">'Sse8387I'</em>, <em class="sig-param">'Sse9I'</em>, <em class="sig-param">'SseBI'</em>, <em class="sig-param">'SsiI'</em>, <em class="sig-param">'SspDI'</em>, <em class="sig-param">'SspI'</em>, <em class="sig-param">'SspMI'</em>, <em class="sig-param">'SstI'</em>, <em class="sig-param">'StuI'</em>, <em class="sig-param">'StyD4I'</em>, <em class="sig-param">'StyI'</em>, <em class="sig-param">'SwaI'</em>, <em class="sig-param">'TaaI'</em>, <em class="sig-param">'TaiI'</em>, <em class="sig-param">'TaqI'</em>, <em class="sig-param">'TaqII'</em>, <em class="sig-param">'TasI'</em>, <em class="sig-param">'TatI'</em>, <em class="sig-param">'TauI'</em>, <em class="sig-param">'TfiI'</em>, <em class="sig-param">'Tru1I'</em>, <em class="sig-param">'Tru9I'</em>, <em class="sig-param">'TscAI'</em>, <em class="sig-param">'TseFI'</em>, <em class="sig-param">'TseI'</em>, <em class="sig-param">'Tsp45I'</em>, <em class="sig-param">'TspDTI'</em>, <em class="sig-param">'TspGWI'</em>, <em class="sig-param">'TspMI'</em>, <em class="sig-param">'TspRI'</em>, <em class="sig-param">'Tth111I'</em>, <em class="sig-param">'Van91I'</em>, <em class="sig-param">'Vha464I'</em>, <em class="sig-param">'VneI'</em>, <em class="sig-param">'VpaK11BI'</em>, <em class="sig-param">'VspI'</em>, <em class="sig-param">'XagI'</em>, <em class="sig-param">'XapI'</em>, <em class="sig-param">'XbaI'</em>, <em class="sig-param">'XceI'</em>, <em class="sig-param">'XcmI'</em>, <em class="sig-param">'XhoI'</em>, <em class="sig-param">'XmaI'</em>, <em class="sig-param">'XmaJI'</em>, <em class="sig-param">'XmiI'</em>, <em class="sig-param">'XmnI'</em>, <em class="sig-param">'XspI'</em>, <em class="sig-param">'ZraI'</em>, <em class="sig-param">'ZrmI'</em>, <em class="sig-param">'Zsp2I'])</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.twice_cutters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.twice_cutters" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the enzymes in a RestrictionBatch that cut the sequence 
exactly twice.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dseq.Dseq.unique_cutters">
<code class="sig-name descname">unique_cutters</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">batch=RestrictionBatch(['AanI'</em>, <em class="sig-param">'AarI'</em>, <em class="sig-param">'AasI'</em>, <em class="sig-param">'AatII'</em>, <em class="sig-param">'AbaSI'</em>, <em class="sig-param">'AbsI'</em>, <em class="sig-param">'Acc16I'</em>, <em class="sig-param">'Acc36I'</em>, <em class="sig-param">'Acc65I'</em>, <em class="sig-param">'AccB1I'</em>, <em class="sig-param">'AccB7I'</em>, <em class="sig-param">'AccBSI'</em>, <em class="sig-param">'AccI'</em>, <em class="sig-param">'AccII'</em>, <em class="sig-param">'AccIII'</em>, <em class="sig-param">'AciI'</em>, <em class="sig-param">'AclI'</em>, <em class="sig-param">'AclWI'</em>, <em class="sig-param">'AcoI'</em>, <em class="sig-param">'AcsI'</em>, <em class="sig-param">'AcuI'</em>, <em class="sig-param">'AcvI'</em>, <em class="sig-param">'AcyI'</em>, <em class="sig-param">'AdeI'</em>, <em class="sig-param">'AfaI'</em>, <em class="sig-param">'AfeI'</em>, <em class="sig-param">'AfiI'</em>, <em class="sig-param">'AflII'</em>, <em class="sig-param">'AflIII'</em>, <em class="sig-param">'AgeI'</em>, <em class="sig-param">'AgsI'</em>, <em class="sig-param">'AhdI'</em>, <em class="sig-param">'AhlI'</em>, <em class="sig-param">'AjiI'</em>, <em class="sig-param">'AjnI'</em>, <em class="sig-param">'AjuI'</em>, <em class="sig-param">'AleI'</em>, <em class="sig-param">'AloI'</em>, <em class="sig-param">'AluBI'</em>, <em class="sig-param">'AluI'</em>, <em class="sig-param">'Alw21I'</em>, <em class="sig-param">'Alw26I'</em>, <em class="sig-param">'Alw44I'</em>, <em class="sig-param">'AlwI'</em>, <em class="sig-param">'AlwNI'</em>, <em class="sig-param">'Ama87I'</em>, <em class="sig-param">'Aor13HI'</em>, <em class="sig-param">'Aor51HI'</em>, <em class="sig-param">'AoxI'</em>, <em class="sig-param">'ApaI'</em>, <em class="sig-param">'ApaLI'</em>, <em class="sig-param">'ApeKI'</em>, <em class="sig-param">'ApoI'</em>, <em class="sig-param">'ArsI'</em>, <em class="sig-param">'AscI'</em>, <em class="sig-param">'AseI'</em>, <em class="sig-param">'AsiGI'</em>, <em class="sig-param">'AsiSI'</em>, <em class="sig-param">'Asp700I'</em>, <em class="sig-param">'Asp718I'</em>, <em class="sig-param">'AspA2I'</em>, <em class="sig-param">'AspLEI'</em>, <em class="sig-param">'AspS9I'</em>, <em class="sig-param">'AsuC2I'</em>, <em class="sig-param">'AsuHPI'</em>, <em class="sig-param">'AsuII'</em>, <em class="sig-param">'AsuNHI'</em>, <em class="sig-param">'AvaI'</em>, <em class="sig-param">'AvaII'</em>, <em class="sig-param">'AvrII'</em>, <em class="sig-param">'AxyI'</em>, <em class="sig-param">'BaeGI'</em>, <em class="sig-param">'BaeI'</em>, <em class="sig-param">'BalI'</em>, <em class="sig-param">'BamHI'</em>, <em class="sig-param">'BanI'</em>, <em class="sig-param">'BanII'</em>, <em class="sig-param">'BarI'</em>, <em class="sig-param">'BauI'</em>, <em class="sig-param">'BbrPI'</em>, <em class="sig-param">'BbsI'</em>, <em class="sig-param">'Bbv12I'</em>, <em class="sig-param">'BbvCI'</em>, <em class="sig-param">'BbvI'</em>, <em class="sig-param">'BccI'</em>, <em class="sig-param">'BceAI'</em>, <em class="sig-param">'BcgI'</em>, <em class="sig-param">'BciT130I'</em>, <em class="sig-param">'BciVI'</em>, <em class="sig-param">'BclI'</em>, <em class="sig-param">'BcnI'</em>, <em class="sig-param">'BcoDI'</em>, <em class="sig-param">'BcuI'</em>, <em class="sig-param">'BfaI'</em>, <em class="sig-param">'BfmI'</em>, <em class="sig-param">'BfoI'</em>, <em class="sig-param">'BfrI'</em>, <em class="sig-param">'BfuAI'</em>, <em class="sig-param">'BfuI'</em>, <em class="sig-param">'BglI'</em>, <em class="sig-param">'BglII'</em>, <em class="sig-param">'BisI'</em>, <em class="sig-param">'BlnI'</em>, <em class="sig-param">'BlpI'</em>, <em class="sig-param">'BlsI'</em>, <em class="sig-param">'BmcAI'</em>, <em class="sig-param">'Bme1390I'</em>, <em class="sig-param">'Bme18I'</em>, <em class="sig-param">'BmeRI'</em>, <em class="sig-param">'BmeT110I'</em>, <em class="sig-param">'BmgBI'</em>, <em class="sig-param">'BmgT120I'</em>, <em class="sig-param">'BmiI'</em>, <em class="sig-param">'BmrFI'</em>, <em class="sig-param">'BmrI'</em>, <em class="sig-param">'BmsI'</em>, <em class="sig-param">'BmtI'</em>, <em class="sig-param">'BmuI'</em>, <em class="sig-param">'BoxI'</em>, <em class="sig-param">'BpiI'</em>, <em class="sig-param">'BplI'</em>, <em class="sig-param">'BpmI'</em>, <em class="sig-param">'Bpu10I'</em>, <em class="sig-param">'Bpu1102I'</em>, <em class="sig-param">'Bpu14I'</em>, <em class="sig-param">'BpuEI'</em>, <em class="sig-param">'BpuMI'</em>, <em class="sig-param">'Bsa29I'</em>, <em class="sig-param">'BsaAI'</em>, <em class="sig-param">'BsaBI'</em>, <em class="sig-param">'BsaHI'</em>, <em class="sig-param">'BsaI'</em>, <em class="sig-param">'BsaJI'</em>, <em class="sig-param">'BsaWI'</em>, <em class="sig-param">'BsaXI'</em>, <em class="sig-param">'Bsc4I'</em>, <em class="sig-param">'Bse118I'</em>, <em class="sig-param">'Bse1I'</em>, <em class="sig-param">'Bse21I'</em>, <em class="sig-param">'Bse3DI'</em>, <em class="sig-param">'Bse8I'</em>, <em class="sig-param">'BseAI'</em>, <em class="sig-param">'BseBI'</em>, <em class="sig-param">'BseCI'</em>, <em class="sig-param">'BseDI'</em>, <em class="sig-param">'BseGI'</em>, <em class="sig-param">'BseJI'</em>, <em class="sig-param">'BseLI'</em>, <em class="sig-param">'BseMI'</em>, <em class="sig-param">'BseMII'</em>, <em class="sig-param">'BseNI'</em>, <em class="sig-param">'BsePI'</em>, <em class="sig-param">'BseRI'</em>, <em class="sig-param">'BseSI'</em>, <em class="sig-param">'BseX3I'</em>, <em class="sig-param">'BseXI'</em>, <em class="sig-param">'BseYI'</em>, <em class="sig-param">'BsgI'</em>, <em class="sig-param">'Bsh1236I'</em>, <em class="sig-param">'Bsh1285I'</em>, <em class="sig-param">'BshFI'</em>, <em class="sig-param">'BshNI'</em>, <em class="sig-param">'BshTI'</em>, <em class="sig-param">'BshVI'</em>, <em class="sig-param">'BsiEI'</em>, <em class="sig-param">'BsiHKAI'</em>, <em class="sig-param">'BsiHKCI'</em>, <em class="sig-param">'BsiSI'</em>, <em class="sig-param">'BsiWI'</em>, <em class="sig-param">'BslFI'</em>, <em class="sig-param">'BslI'</em>, <em class="sig-param">'BsmAI'</em>, <em class="sig-param">'BsmBI'</em>, <em class="sig-param">'BsmFI'</em>, <em class="sig-param">'BsmI'</em>, <em class="sig-param">'BsnI'</em>, <em class="sig-param">'Bso31I'</em>, <em class="sig-param">'BsoBI'</em>, <em class="sig-param">'Bsp119I'</em>, <em class="sig-param">'Bsp120I'</em>, <em class="sig-param">'Bsp1286I'</em>, <em class="sig-param">'Bsp13I'</em>, <em class="sig-param">'Bsp1407I'</em>, <em class="sig-param">'Bsp143I'</em>, <em class="sig-param">'Bsp1720I'</em>, <em class="sig-param">'Bsp19I'</em>, <em class="sig-param">'Bsp68I'</em>, <em class="sig-param">'BspACI'</em>, <em class="sig-param">'BspANI'</em>, <em class="sig-param">'BspCNI'</em>, <em class="sig-param">'BspDI'</em>, <em class="sig-param">'BspEI'</em>, <em class="sig-param">'BspFNI'</em>, <em class="sig-param">'BspHI'</em>, <em class="sig-param">'BspLI'</em>, <em class="sig-param">'BspMAI'</em>, <em class="sig-param">'BspMI'</em>, <em class="sig-param">'BspOI'</em>, <em class="sig-param">'BspPI'</em>, <em class="sig-param">'BspQI'</em>, <em class="sig-param">'BspT104I'</em>, <em class="sig-param">'BspT107I'</em>, <em class="sig-param">'BspTI'</em>, <em class="sig-param">'BspTNI'</em>, <em class="sig-param">'BsrBI'</em>, <em class="sig-param">'BsrDI'</em>, <em class="sig-param">'BsrFI'</em>, <em class="sig-param">'BsrGI'</em>, <em class="sig-param">'BsrI'</em>, <em class="sig-param">'BssAI'</em>, <em class="sig-param">'BssECI'</em>, <em class="sig-param">'BssHII'</em>, <em class="sig-param">'BssMI'</em>, <em class="sig-param">'BssNAI'</em>, <em class="sig-param">'BssNI'</em>, <em class="sig-param">'BssSI'</em>, <em class="sig-param">'BssT1I'</em>, <em class="sig-param">'Bst1107I'</em>, <em class="sig-param">'Bst2BI'</em>, <em class="sig-param">'Bst2UI'</em>, <em class="sig-param">'Bst4CI'</em>, <em class="sig-param">'Bst6I'</em>, <em class="sig-param">'BstACI'</em>, <em class="sig-param">'BstAFI'</em>, <em class="sig-param">'BstAPI'</em>, <em class="sig-param">'BstAUI'</em>, <em class="sig-param">'BstBAI'</em>, <em class="sig-param">'BstBI'</em>, <em class="sig-param">'BstC8I'</em>, <em class="sig-param">'BstDEI'</em>, <em class="sig-param">'BstDSI'</em>, <em class="sig-param">'BstEII'</em>, <em class="sig-param">'BstENI'</em>, <em class="sig-param">'BstF5I'</em>, <em class="sig-param">'BstFNI'</em>, <em class="sig-param">'BstH2I'</em>, <em class="sig-param">'BstHHI'</em>, <em class="sig-param">'BstKTI'</em>, <em class="sig-param">'BstMAI'</em>, <em class="sig-param">'BstMBI'</em>, <em class="sig-param">'BstMCI'</em>, <em class="sig-param">'BstMWI'</em>, <em class="sig-param">'BstNI'</em>, <em class="sig-param">'BstNSI'</em>, <em class="sig-param">'BstPAI'</em>, <em class="sig-param">'BstPI'</em>, <em class="sig-param">'BstSCI'</em>, <em class="sig-param">'BstSFI'</em>, <em class="sig-param">'BstSLI'</em>, <em class="sig-param">'BstSNI'</em>, <em class="sig-param">'BstUI'</em>, <em class="sig-param">'BstV1I'</em>, <em class="sig-param">'BstV2I'</em>, <em class="sig-param">'BstX2I'</em>, <em class="sig-param">'BstXI'</em>, <em class="sig-param">'BstYI'</em>, <em class="sig-param">'BstZ17I'</em>, <em class="sig-param">'BstZI'</em>, <em class="sig-param">'Bsu15I'</em>, <em class="sig-param">'Bsu36I'</em>, <em class="sig-param">'BsuI'</em>, <em class="sig-param">'BsuRI'</em>, <em class="sig-param">'BsuTUI'</em>, <em class="sig-param">'BtgI'</em>, <em class="sig-param">'BtgZI'</em>, <em class="sig-param">'BtrI'</em>, <em class="sig-param">'BtsCI'</em>, <em class="sig-param">'BtsI'</em>, <em class="sig-param">'BtsIMutI'</em>, <em class="sig-param">'BtuMI'</em>, <em class="sig-param">'BveI'</em>, <em class="sig-param">'Cac8I'</em>, <em class="sig-param">'CaiI'</em>, <em class="sig-param">'CciI'</em>, <em class="sig-param">'CciNI'</em>, <em class="sig-param">'CfoI'</em>, <em class="sig-param">'Cfr10I'</em>, <em class="sig-param">'Cfr13I'</em>, <em class="sig-param">'Cfr42I'</em>, <em class="sig-param">'Cfr9I'</em>, <em class="sig-param">'ClaI'</em>, <em class="sig-param">'CpoI'</em>, <em class="sig-param">'CseI'</em>, <em class="sig-param">'CsiI'</em>, <em class="sig-param">'Csp6I'</em>, <em class="sig-param">'CspAI'</em>, <em class="sig-param">'CspCI'</em>, <em class="sig-param">'CspI'</em>, <em class="sig-param">'CviAII'</em>, <em class="sig-param">'CviJI'</em>, <em class="sig-param">'CviKI_1'</em>, <em class="sig-param">'CviQI'</em>, <em class="sig-param">'DdeI'</em>, <em class="sig-param">'DinI'</em>, <em class="sig-param">'DpnI'</em>, <em class="sig-param">'DpnII'</em>, <em class="sig-param">'DraI'</em>, <em class="sig-param">'DraIII'</em>, <em class="sig-param">'DrdI'</em>, <em class="sig-param">'DriI'</em>, <em class="sig-param">'DseDI'</em>, <em class="sig-param">'EaeI'</em>, <em class="sig-param">'EagI'</em>, <em class="sig-param">'Eam1104I'</em>, <em class="sig-param">'Eam1105I'</em>, <em class="sig-param">'EarI'</em>, <em class="sig-param">'EciI'</em>, <em class="sig-param">'Ecl136II'</em>, <em class="sig-param">'EclXI'</em>, <em class="sig-param">'Eco105I'</em>, <em class="sig-param">'Eco130I'</em>, <em class="sig-param">'Eco147I'</em>, <em class="sig-param">'Eco24I'</em>, <em class="sig-param">'Eco31I'</em>, <em class="sig-param">'Eco32I'</em>, <em class="sig-param">'Eco47I'</em>, <em class="sig-param">'Eco47III'</em>, <em class="sig-param">'Eco52I'</em>, <em class="sig-param">'Eco53kI'</em>, <em class="sig-param">'Eco57I'</em>, <em class="sig-param">'Eco72I'</em>, <em class="sig-param">'Eco81I'</em>, <em class="sig-param">'Eco88I'</em>, <em class="sig-param">'Eco91I'</em>, <em class="sig-param">'EcoICRI'</em>, <em class="sig-param">'EcoNI'</em>, <em class="sig-param">'EcoO109I'</em>, <em class="sig-param">'EcoO65I'</em>, <em class="sig-param">'EcoRI'</em>, <em class="sig-param">'EcoRII'</em>, <em class="sig-param">'EcoRV'</em>, <em class="sig-param">'EcoT14I'</em>, <em class="sig-param">'EcoT22I'</em>, <em class="sig-param">'EcoT38I'</em>, <em class="sig-param">'EgeI'</em>, <em class="sig-param">'EheI'</em>, <em class="sig-param">'ErhI'</em>, <em class="sig-param">'Esp3I'</em>, <em class="sig-param">'FaeI'</em>, <em class="sig-param">'FaiI'</em>, <em class="sig-param">'FalI'</em>, <em class="sig-param">'FaqI'</em>, <em class="sig-param">'FatI'</em>, <em class="sig-param">'FauI'</em>, <em class="sig-param">'FauNDI'</em>, <em class="sig-param">'FbaI'</em>, <em class="sig-param">'FblI'</em>, <em class="sig-param">'Fnu4HI'</em>, <em class="sig-param">'FokI'</em>, <em class="sig-param">'FriOI'</em>, <em class="sig-param">'FseI'</em>, <em class="sig-param">'Fsp4HI'</em>, <em class="sig-param">'FspAI'</em>, <em class="sig-param">'FspBI'</em>, <em class="sig-param">'FspEI'</em>, <em class="sig-param">'FspI'</em>, <em class="sig-param">'GlaI'</em>, <em class="sig-param">'GluI'</em>, <em class="sig-param">'GsaI'</em>, <em class="sig-param">'GsuI'</em>, <em class="sig-param">'HaeII'</em>, <em class="sig-param">'HaeIII'</em>, <em class="sig-param">'HapII'</em>, <em class="sig-param">'HgaI'</em>, <em class="sig-param">'HhaI'</em>, <em class="sig-param">'Hin1I'</em>, <em class="sig-param">'Hin1II'</em>, <em class="sig-param">'Hin6I'</em>, <em class="sig-param">'HinP1I'</em>, <em class="sig-param">'HincII'</em>, <em class="sig-param">'HindII'</em>, <em class="sig-param">'HindIII'</em>, <em class="sig-param">'HinfI'</em>, <em class="sig-param">'HpaI'</em>, <em class="sig-param">'HpaII'</em>, <em class="sig-param">'HphI'</em>, <em class="sig-param">'Hpy166II'</em>, <em class="sig-param">'Hpy188I'</em>, <em class="sig-param">'Hpy188III'</em>, <em class="sig-param">'Hpy8I'</em>, <em class="sig-param">'Hpy99I'</em>, <em class="sig-param">'HpyAV'</em>, <em class="sig-param">'HpyCH4III'</em>, <em class="sig-param">'HpyCH4IV'</em>, <em class="sig-param">'HpyCH4V'</em>, <em class="sig-param">'HpyF10VI'</em>, <em class="sig-param">'HpyF3I'</em>, <em class="sig-param">'HpySE526I'</em>, <em class="sig-param">'Hsp92I'</em>, <em class="sig-param">'Hsp92II'</em>, <em class="sig-param">'HspAI'</em>, <em class="sig-param">'KasI'</em>, <em class="sig-param">'KflI'</em>, <em class="sig-param">'Kpn2I'</em>, <em class="sig-param">'KpnI'</em>, <em class="sig-param">'KroI'</em>, <em class="sig-param">'Ksp22I'</em>, <em class="sig-param">'KspAI'</em>, <em class="sig-param">'KspI'</em>, <em class="sig-param">'Kzo9I'</em>, <em class="sig-param">'LguI'</em>, <em class="sig-param">'LmnI'</em>, <em class="sig-param">'LpnPI'</em>, <em class="sig-param">'Lsp1109I'</em>, <em class="sig-param">'LweI'</em>, <em class="sig-param">'MabI'</em>, <em class="sig-param">'MaeI'</em>, <em class="sig-param">'MaeII'</em>, <em class="sig-param">'MaeIII'</em>, <em class="sig-param">'MalI'</em>, <em class="sig-param">'MauBI'</em>, <em class="sig-param">'MbiI'</em>, <em class="sig-param">'MboI'</em>, <em class="sig-param">'MboII'</em>, <em class="sig-param">'MfeI'</em>, <em class="sig-param">'MflI'</em>, <em class="sig-param">'MhlI'</em>, <em class="sig-param">'MlsI'</em>, <em class="sig-param">'MluCI'</em>, <em class="sig-param">'MluI'</em>, <em class="sig-param">'MluNI'</em>, <em class="sig-param">'Mly113I'</em>, <em class="sig-param">'MlyI'</em>, <em class="sig-param">'MmeI'</em>, <em class="sig-param">'MnlI'</em>, <em class="sig-param">'Mox20I'</em>, <em class="sig-param">'Mph1103I'</em>, <em class="sig-param">'MreI'</em>, <em class="sig-param">'MroI'</em>, <em class="sig-param">'MroNI'</em>, <em class="sig-param">'MroXI'</em>, <em class="sig-param">'MscI'</em>, <em class="sig-param">'MseI'</em>, <em class="sig-param">'MslI'</em>, <em class="sig-param">'Msp20I'</em>, <em class="sig-param">'MspA1I'</em>, <em class="sig-param">'MspCI'</em>, <em class="sig-param">'MspI'</em>, <em class="sig-param">'MspJI'</em>, <em class="sig-param">'MspR9I'</em>, <em class="sig-param">'MssI'</em>, <em class="sig-param">'MteI'</em>, <em class="sig-param">'MunI'</em>, <em class="sig-param">'Mva1269I'</em>, <em class="sig-param">'MvaI'</em>, <em class="sig-param">'MvnI'</em>, <em class="sig-param">'MwoI'</em>, <em class="sig-param">'NaeI'</em>, <em class="sig-param">'NarI'</em>, <em class="sig-param">'NciI'</em>, <em class="sig-param">'NcoI'</em>, <em class="sig-param">'NdeI'</em>, <em class="sig-param">'NdeII'</em>, <em class="sig-param">'NgoMIV'</em>, <em class="sig-param">'NheI'</em>, <em class="sig-param">'NlaIII'</em>, <em class="sig-param">'NlaIV'</em>, <em class="sig-param">'NmeAIII'</em>, <em class="sig-param">'NmuCI'</em>, <em class="sig-param">'NotI'</em>, <em class="sig-param">'NruI'</em>, <em class="sig-param">'NsbI'</em>, <em class="sig-param">'NsiI'</em>, <em class="sig-param">'NspI'</em>, <em class="sig-param">'NspV'</em>, <em class="sig-param">'OliI'</em>, <em class="sig-param">'PacI'</em>, <em class="sig-param">'PaeI'</em>, <em class="sig-param">'PaeR7I'</em>, <em class="sig-param">'PagI'</em>, <em class="sig-param">'PalAI'</em>, <em class="sig-param">'PasI'</em>, <em class="sig-param">'PauI'</em>, <em class="sig-param">'PceI'</em>, <em class="sig-param">'PciI'</em>, <em class="sig-param">'PciSI'</em>, <em class="sig-param">'PcsI'</em>, <em class="sig-param">'PctI'</em>, <em class="sig-param">'PdiI'</em>, <em class="sig-param">'PdmI'</em>, <em class="sig-param">'PfeI'</em>, <em class="sig-param">'Pfl23II'</em>, <em class="sig-param">'PflFI'</em>, <em class="sig-param">'PflMI'</em>, <em class="sig-param">'PfoI'</em>, <em class="sig-param">'PinAI'</em>, <em class="sig-param">'PkrI'</em>, <em class="sig-param">'Ple19I'</em>, <em class="sig-param">'PleI'</em>, <em class="sig-param">'PluTI'</em>, <em class="sig-param">'PmaCI'</em>, <em class="sig-param">'PmeI'</em>, <em class="sig-param">'PmlI'</em>, <em class="sig-param">'PpsI'</em>, <em class="sig-param">'Ppu21I'</em>, <em class="sig-param">'PpuMI'</em>, <em class="sig-param">'PscI'</em>, <em class="sig-param">'PshAI'</em>, <em class="sig-param">'PshBI'</em>, <em class="sig-param">'PsiI'</em>, <em class="sig-param">'Psp124BI'</em>, <em class="sig-param">'Psp1406I'</em>, <em class="sig-param">'Psp5II'</em>, <em class="sig-param">'Psp6I'</em>, <em class="sig-param">'PspCI'</em>, <em class="sig-param">'PspEI'</em>, <em class="sig-param">'PspFI'</em>, <em class="sig-param">'PspGI'</em>, <em class="sig-param">'PspLI'</em>, <em class="sig-param">'PspN4I'</em>, <em class="sig-param">'PspOMI'</em>, <em class="sig-param">'PspPI'</em>, <em class="sig-param">'PspPPI'</em>, <em class="sig-param">'PspXI'</em>, <em class="sig-param">'PsrI'</em>, <em class="sig-param">'PstI'</em>, <em class="sig-param">'PstNI'</em>, <em class="sig-param">'PsuI'</em>, <em class="sig-param">'PsyI'</em>, <em class="sig-param">'PteI'</em>, <em class="sig-param">'PvuI'</em>, <em class="sig-param">'PvuII'</em>, <em class="sig-param">'RgaI'</em>, <em class="sig-param">'RigI'</em>, <em class="sig-param">'RruI'</em>, <em class="sig-param">'RsaI'</em>, <em class="sig-param">'RsaNI'</em>, <em class="sig-param">'RseI'</em>, <em class="sig-param">'Rsr2I'</em>, <em class="sig-param">'RsrII'</em>, <em class="sig-param">'SacI'</em>, <em class="sig-param">'SacII'</em>, <em class="sig-param">'SalI'</em>, <em class="sig-param">'SapI'</em>, <em class="sig-param">'SaqAI'</em>, <em class="sig-param">'SatI'</em>, <em class="sig-param">'Sau3AI'</em>, <em class="sig-param">'Sau96I'</em>, <em class="sig-param">'SbfI'</em>, <em class="sig-param">'ScaI'</em>, <em class="sig-param">'SchI'</em>, <em class="sig-param">'ScrFI'</em>, <em class="sig-param">'SdaI'</em>, <em class="sig-param">'SduI'</em>, <em class="sig-param">'SetI'</em>, <em class="sig-param">'SexAI'</em>, <em class="sig-param">'SfaAI'</em>, <em class="sig-param">'SfaNI'</em>, <em class="sig-param">'SfcI'</em>, <em class="sig-param">'SfiI'</em>, <em class="sig-param">'SfoI'</em>, <em class="sig-param">'Sfr274I'</em>, <em class="sig-param">'Sfr303I'</em>, <em class="sig-param">'SfuI'</em>, <em class="sig-param">'SgeI'</em>, <em class="sig-param">'SgfI'</em>, <em class="sig-param">'SgrAI'</em>, <em class="sig-param">'SgrBI'</em>, <em class="sig-param">'SgrDI'</em>, <em class="sig-param">'SgsI'</em>, <em class="sig-param">'SinI'</em>, <em class="sig-param">'SlaI'</em>, <em class="sig-param">'SmaI'</em>, <em class="sig-param">'SmiI'</em>, <em class="sig-param">'SmiMI'</em>, <em class="sig-param">'SmlI'</em>, <em class="sig-param">'SmoI'</em>, <em class="sig-param">'SnaBI'</em>, <em class="sig-param">'SpeI'</em>, <em class="sig-param">'SphI'</em>, <em class="sig-param">'SrfI'</em>, <em class="sig-param">'Sse8387I'</em>, <em class="sig-param">'Sse9I'</em>, <em class="sig-param">'SseBI'</em>, <em class="sig-param">'SsiI'</em>, <em class="sig-param">'SspDI'</em>, <em class="sig-param">'SspI'</em>, <em class="sig-param">'SspMI'</em>, <em class="sig-param">'SstI'</em>, <em class="sig-param">'StuI'</em>, <em class="sig-param">'StyD4I'</em>, <em class="sig-param">'StyI'</em>, <em class="sig-param">'SwaI'</em>, <em class="sig-param">'TaaI'</em>, <em class="sig-param">'TaiI'</em>, <em class="sig-param">'TaqI'</em>, <em class="sig-param">'TaqII'</em>, <em class="sig-param">'TasI'</em>, <em class="sig-param">'TatI'</em>, <em class="sig-param">'TauI'</em>, <em class="sig-param">'TfiI'</em>, <em class="sig-param">'Tru1I'</em>, <em class="sig-param">'Tru9I'</em>, <em class="sig-param">'TscAI'</em>, <em class="sig-param">'TseFI'</em>, <em class="sig-param">'TseI'</em>, <em class="sig-param">'Tsp45I'</em>, <em class="sig-param">'TspDTI'</em>, <em class="sig-param">'TspGWI'</em>, <em class="sig-param">'TspMI'</em>, <em class="sig-param">'TspRI'</em>, <em class="sig-param">'Tth111I'</em>, <em class="sig-param">'Van91I'</em>, <em class="sig-param">'Vha464I'</em>, <em class="sig-param">'VneI'</em>, <em class="sig-param">'VpaK11BI'</em>, <em class="sig-param">'VspI'</em>, <em class="sig-param">'XagI'</em>, <em class="sig-param">'XapI'</em>, <em class="sig-param">'XbaI'</em>, <em class="sig-param">'XceI'</em>, <em class="sig-param">'XcmI'</em>, <em class="sig-param">'XhoI'</em>, <em class="sig-param">'XmaI'</em>, <em class="sig-param">'XmaJI'</em>, <em class="sig-param">'XmiI'</em>, <em class="sig-param">'XmnI'</em>, <em class="sig-param">'XspI'</em>, <em class="sig-param">'ZraI'</em>, <em class="sig-param">'ZrmI'</em>, <em class="sig-param">'Zsp2I'])</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseq.html#Dseq.unique_cutters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseq.Dseq.unique_cutters" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the enzymes in a RestrictionBatch that cut the sequence 
exactly once.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-pydna.dseqrecord">
<span id="pydna-dseqrecord-module"></span><h2>pydna.dseqrecord module<a class="headerlink" href="#module-pydna.dseqrecord" title="Permalink to this headline">¶</a></h2>
<p>This module provides the <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dseqrecord</span></code></a> class, for handling double stranded
DNA sequences. The Dseqrecord holds sequence information in the form of a <a class="reference internal" href="#pydna.dseq.Dseq" title="pydna.dseq.Dseq"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.dseq.Dseq</span></code></a>
object. The Dseq and Dseqrecord classes are subclasses of Biopythons
Seq and SeqRecord classes, respectively.</p>
<p>The Dseq and Dseqrecord classes support the notion of circular and linear DNA topology.</p>
<dl class="class">
<dt id="pydna.dseqrecord.Dseqrecord">
<em class="property">class </em><code class="sig-prename descclassname">pydna.dseqrecord.</code><code class="sig-name descname">Dseqrecord</code><span class="sig-paren">(</span><em class="sig-param">record</em>, <em class="sig-param">*args</em>, <em class="sig-param">linear=None</em>, <em class="sig-param">circular=None</em>, <em class="sig-param">n=5e-14</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.seqrecord.SeqRecord</span></code></a></p>
<p>Dseqrecord is a double stranded version of the Biopython SeqRecord <a class="footnote-reference brackets" href="#id6" id="id5">3</a> class.
The Dseqrecord object holds a Dseq object describing the sequence.
Additionally, Dseqrecord hold meta information about the sequence in the
from of a list of SeqFeatures, in the same way as the SeqRecord does.</p>
<p>The Dseqrecord can be initialized with a string, Seq, Dseq, SeqRecord
or another Dseqrecord. The sequence information will be stored in a
Dseq object in all cases.</p>
<p>Dseqrecord objects can be read or parsed from sequences in FASTA, EMBL or Genbank formats.
See the <a class="reference internal" href="#module-pydna.readers" title="pydna.readers"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.readers</span></code></a> and <a class="reference internal" href="#module-pydna.parsers" title="pydna.parsers"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.parsers</span></code></a> modules for further information.</p>
<p>There is a short representation associated with the Dseqrecord.
<code class="docutils literal notranslate"><span class="pre">Dseqrecord(-3)</span></code> represents a linear sequence of length 2
while <code class="docutils literal notranslate"><span class="pre">Dseqrecord(o7)</span></code>
represents a circular sequence of length 7.</p>
<p>Dseqrecord and Dseq share the same concept of length. This length can be larger
than each strand alone if they are staggered as in the example below.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;--</span> <span class="n">length</span> <span class="o">--&gt;</span>
<span class="n">GATCCTTT</span>
     <span class="n">AAAGCCTAG</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>record</strong><span class="classifier">string, Seq, SeqRecord, Dseq or other Dseqrecord object</span></dt><dd><p>This data will be used to form the seq property</p>
</dd>
<dt><strong>circular</strong><span class="classifier">bool, optional</span></dt><dd><p>True or False reflecting the shape of the DNA molecule</p>
</dd>
<dt><strong>linear</strong><span class="classifier">bool, optional</span></dt><dd><p>True or False reflecting the shape of the DNA molecule</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id6"><span class="brackets"><a class="fn-backref" href="#id5">3</a></span></dt>
<dd><p><a class="reference external" href="http://biopython.org/wiki/SeqRecord">http://biopython.org/wiki/SeqRecord</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="k">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="k">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">alphabet</span>
<span class="go">IUPACAmbiguousDNA()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">alphabet</span>
<span class="go">IUPACAmbiguousDNA()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">alphabet</span>
<span class="go">IUPACAmbiguousDNA()</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.circular">
<em class="property">property </em><code class="sig-name descname">circular</code><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.circular" title="Permalink to this definition">¶</a></dt>
<dd><p>The circular property can not be set directly. 
Use <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.looped" title="pydna.dseqrecord.Dseqrecord.looped"><code class="xref py py-meth docutils literal notranslate"><span class="pre">looped()</span></code></a> or <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.tolinear" title="pydna.dseqrecord.Dseqrecord.tolinear"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tolinear()</span></code></a></p>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.cseguid">
<code class="sig-name descname">cseguid</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.cseguid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.cseguid" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the url safe cSEGUID for the sequence.</p>
<p>Only defined for circular sequences.</p>
<p>The cSEGUID checksum uniquely identifies a circular
sequence regardless of where the origin is set.
The two Dseqrecord objects below are circular
permutations.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;agtatcgtacatg&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">cseguid</span><span class="p">()</span> <span class="c1"># cseguid is CTJbs6Fat8kLQxHj+/SC0kGEiYs</span>
<span class="go">&#39;CTJbs6Fat8kLQxHj-_SC0kGEiYs&#39;</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;gagtatcgtacat&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">cseguid</span><span class="p">()</span>
<span class="go">&#39;CTJbs6Fat8kLQxHj-_SC0kGEiYs&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.cut">
<code class="sig-name descname">cut</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">*enzymes</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.cut"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.cut" title="Permalink to this definition">¶</a></dt>
<dd><p>Digest the Dseqrecord object with one or more restriction enzymes.
returns a list of linear Dseqrecords. If there are no cuts, an empty
list is returned.</p>
<p>See also <code class="xref py py-func docutils literal notranslate"><span class="pre">Dseq.cut()</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>enzymes</strong><span class="classifier">enzyme object or iterable of such objects</span></dt><dd><p>A Bio.Restriction.XXX restriction object or iterable of such.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>Dseqrecord_frags</strong><span class="classifier">list</span></dt><dd><p>list of Dseqrecord objects formed by the digestion</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggatcc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Restriction</span> <span class="k">import</span> <span class="n">BamHI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">)</span>
<span class="go">(Dseqrecord(-5), Dseqrecord(-5))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag1</span><span class="p">,</span> <span class="n">frag2</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag1</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-5)</span>
<span class="go">g</span>
<span class="go">cctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag2</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-5)</span>
<span class="go">gatcc</span>
<span class="go">    g</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.cutters">
<code class="sig-name descname">cutters</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">batch=RestrictionBatch(['AanI'</em>, <em class="sig-param">'AarI'</em>, <em class="sig-param">'AasI'</em>, <em class="sig-param">'AatII'</em>, <em class="sig-param">'AbaSI'</em>, <em class="sig-param">'AbsI'</em>, <em class="sig-param">'Acc16I'</em>, <em class="sig-param">'Acc36I'</em>, <em class="sig-param">'Acc65I'</em>, <em class="sig-param">'AccB1I'</em>, <em class="sig-param">'AccB7I'</em>, <em class="sig-param">'AccBSI'</em>, <em class="sig-param">'AccI'</em>, <em class="sig-param">'AccII'</em>, <em class="sig-param">'AccIII'</em>, <em class="sig-param">'AciI'</em>, <em class="sig-param">'AclI'</em>, <em class="sig-param">'AclWI'</em>, <em class="sig-param">'AcoI'</em>, <em class="sig-param">'AcsI'</em>, <em class="sig-param">'AcuI'</em>, <em class="sig-param">'AcvI'</em>, <em class="sig-param">'AcyI'</em>, <em class="sig-param">'AdeI'</em>, <em class="sig-param">'AfaI'</em>, <em class="sig-param">'AfeI'</em>, <em class="sig-param">'AfiI'</em>, <em class="sig-param">'AflII'</em>, <em class="sig-param">'AflIII'</em>, <em class="sig-param">'AgeI'</em>, <em class="sig-param">'AgsI'</em>, <em class="sig-param">'AhdI'</em>, <em class="sig-param">'AhlI'</em>, <em class="sig-param">'AjiI'</em>, <em class="sig-param">'AjnI'</em>, <em class="sig-param">'AjuI'</em>, <em class="sig-param">'AleI'</em>, <em class="sig-param">'AloI'</em>, <em class="sig-param">'AluBI'</em>, <em class="sig-param">'AluI'</em>, <em class="sig-param">'Alw21I'</em>, <em class="sig-param">'Alw26I'</em>, <em class="sig-param">'Alw44I'</em>, <em class="sig-param">'AlwI'</em>, <em class="sig-param">'AlwNI'</em>, <em class="sig-param">'Ama87I'</em>, <em class="sig-param">'Aor13HI'</em>, <em class="sig-param">'Aor51HI'</em>, <em class="sig-param">'AoxI'</em>, <em class="sig-param">'ApaI'</em>, <em class="sig-param">'ApaLI'</em>, <em class="sig-param">'ApeKI'</em>, <em class="sig-param">'ApoI'</em>, <em class="sig-param">'ArsI'</em>, <em class="sig-param">'AscI'</em>, <em class="sig-param">'AseI'</em>, <em class="sig-param">'AsiGI'</em>, <em class="sig-param">'AsiSI'</em>, <em class="sig-param">'Asp700I'</em>, <em class="sig-param">'Asp718I'</em>, <em class="sig-param">'AspA2I'</em>, <em class="sig-param">'AspLEI'</em>, <em class="sig-param">'AspS9I'</em>, <em class="sig-param">'AsuC2I'</em>, <em class="sig-param">'AsuHPI'</em>, <em class="sig-param">'AsuII'</em>, <em class="sig-param">'AsuNHI'</em>, <em class="sig-param">'AvaI'</em>, <em class="sig-param">'AvaII'</em>, <em class="sig-param">'AvrII'</em>, <em class="sig-param">'AxyI'</em>, <em class="sig-param">'BaeGI'</em>, <em class="sig-param">'BaeI'</em>, <em class="sig-param">'BalI'</em>, <em class="sig-param">'BamHI'</em>, <em class="sig-param">'BanI'</em>, <em class="sig-param">'BanII'</em>, <em class="sig-param">'BarI'</em>, <em class="sig-param">'BauI'</em>, <em class="sig-param">'BbrPI'</em>, <em class="sig-param">'BbsI'</em>, <em class="sig-param">'Bbv12I'</em>, <em class="sig-param">'BbvCI'</em>, <em class="sig-param">'BbvI'</em>, <em class="sig-param">'BccI'</em>, <em class="sig-param">'BceAI'</em>, <em class="sig-param">'BcgI'</em>, <em class="sig-param">'BciT130I'</em>, <em class="sig-param">'BciVI'</em>, <em class="sig-param">'BclI'</em>, <em class="sig-param">'BcnI'</em>, <em class="sig-param">'BcoDI'</em>, <em class="sig-param">'BcuI'</em>, <em class="sig-param">'BfaI'</em>, <em class="sig-param">'BfmI'</em>, <em class="sig-param">'BfoI'</em>, <em class="sig-param">'BfrI'</em>, <em class="sig-param">'BfuAI'</em>, <em class="sig-param">'BfuI'</em>, <em class="sig-param">'BglI'</em>, <em class="sig-param">'BglII'</em>, <em class="sig-param">'BisI'</em>, <em class="sig-param">'BlnI'</em>, <em class="sig-param">'BlpI'</em>, <em class="sig-param">'BlsI'</em>, <em class="sig-param">'BmcAI'</em>, <em class="sig-param">'Bme1390I'</em>, <em class="sig-param">'Bme18I'</em>, <em class="sig-param">'BmeRI'</em>, <em class="sig-param">'BmeT110I'</em>, <em class="sig-param">'BmgBI'</em>, <em class="sig-param">'BmgT120I'</em>, <em class="sig-param">'BmiI'</em>, <em class="sig-param">'BmrFI'</em>, <em class="sig-param">'BmrI'</em>, <em class="sig-param">'BmsI'</em>, <em class="sig-param">'BmtI'</em>, <em class="sig-param">'BmuI'</em>, <em class="sig-param">'BoxI'</em>, <em class="sig-param">'BpiI'</em>, <em class="sig-param">'BplI'</em>, <em class="sig-param">'BpmI'</em>, <em class="sig-param">'Bpu10I'</em>, <em class="sig-param">'Bpu1102I'</em>, <em class="sig-param">'Bpu14I'</em>, <em class="sig-param">'BpuEI'</em>, <em class="sig-param">'BpuMI'</em>, <em class="sig-param">'Bsa29I'</em>, <em class="sig-param">'BsaAI'</em>, <em class="sig-param">'BsaBI'</em>, <em class="sig-param">'BsaHI'</em>, <em class="sig-param">'BsaI'</em>, <em class="sig-param">'BsaJI'</em>, <em class="sig-param">'BsaWI'</em>, <em class="sig-param">'BsaXI'</em>, <em class="sig-param">'Bsc4I'</em>, <em class="sig-param">'Bse118I'</em>, <em class="sig-param">'Bse1I'</em>, <em class="sig-param">'Bse21I'</em>, <em class="sig-param">'Bse3DI'</em>, <em class="sig-param">'Bse8I'</em>, <em class="sig-param">'BseAI'</em>, <em class="sig-param">'BseBI'</em>, <em class="sig-param">'BseCI'</em>, <em class="sig-param">'BseDI'</em>, <em class="sig-param">'BseGI'</em>, <em class="sig-param">'BseJI'</em>, <em class="sig-param">'BseLI'</em>, <em class="sig-param">'BseMI'</em>, <em class="sig-param">'BseMII'</em>, <em class="sig-param">'BseNI'</em>, <em class="sig-param">'BsePI'</em>, <em class="sig-param">'BseRI'</em>, <em class="sig-param">'BseSI'</em>, <em class="sig-param">'BseX3I'</em>, <em class="sig-param">'BseXI'</em>, <em class="sig-param">'BseYI'</em>, <em class="sig-param">'BsgI'</em>, <em class="sig-param">'Bsh1236I'</em>, <em class="sig-param">'Bsh1285I'</em>, <em class="sig-param">'BshFI'</em>, <em class="sig-param">'BshNI'</em>, <em class="sig-param">'BshTI'</em>, <em class="sig-param">'BshVI'</em>, <em class="sig-param">'BsiEI'</em>, <em class="sig-param">'BsiHKAI'</em>, <em class="sig-param">'BsiHKCI'</em>, <em class="sig-param">'BsiSI'</em>, <em class="sig-param">'BsiWI'</em>, <em class="sig-param">'BslFI'</em>, <em class="sig-param">'BslI'</em>, <em class="sig-param">'BsmAI'</em>, <em class="sig-param">'BsmBI'</em>, <em class="sig-param">'BsmFI'</em>, <em class="sig-param">'BsmI'</em>, <em class="sig-param">'BsnI'</em>, <em class="sig-param">'Bso31I'</em>, <em class="sig-param">'BsoBI'</em>, <em class="sig-param">'Bsp119I'</em>, <em class="sig-param">'Bsp120I'</em>, <em class="sig-param">'Bsp1286I'</em>, <em class="sig-param">'Bsp13I'</em>, <em class="sig-param">'Bsp1407I'</em>, <em class="sig-param">'Bsp143I'</em>, <em class="sig-param">'Bsp1720I'</em>, <em class="sig-param">'Bsp19I'</em>, <em class="sig-param">'Bsp68I'</em>, <em class="sig-param">'BspACI'</em>, <em class="sig-param">'BspANI'</em>, <em class="sig-param">'BspCNI'</em>, <em class="sig-param">'BspDI'</em>, <em class="sig-param">'BspEI'</em>, <em class="sig-param">'BspFNI'</em>, <em class="sig-param">'BspHI'</em>, <em class="sig-param">'BspLI'</em>, <em class="sig-param">'BspMAI'</em>, <em class="sig-param">'BspMI'</em>, <em class="sig-param">'BspOI'</em>, <em class="sig-param">'BspPI'</em>, <em class="sig-param">'BspQI'</em>, <em class="sig-param">'BspT104I'</em>, <em class="sig-param">'BspT107I'</em>, <em class="sig-param">'BspTI'</em>, <em class="sig-param">'BspTNI'</em>, <em class="sig-param">'BsrBI'</em>, <em class="sig-param">'BsrDI'</em>, <em class="sig-param">'BsrFI'</em>, <em class="sig-param">'BsrGI'</em>, <em class="sig-param">'BsrI'</em>, <em class="sig-param">'BssAI'</em>, <em class="sig-param">'BssECI'</em>, <em class="sig-param">'BssHII'</em>, <em class="sig-param">'BssMI'</em>, <em class="sig-param">'BssNAI'</em>, <em class="sig-param">'BssNI'</em>, <em class="sig-param">'BssSI'</em>, <em class="sig-param">'BssT1I'</em>, <em class="sig-param">'Bst1107I'</em>, <em class="sig-param">'Bst2BI'</em>, <em class="sig-param">'Bst2UI'</em>, <em class="sig-param">'Bst4CI'</em>, <em class="sig-param">'Bst6I'</em>, <em class="sig-param">'BstACI'</em>, <em class="sig-param">'BstAFI'</em>, <em class="sig-param">'BstAPI'</em>, <em class="sig-param">'BstAUI'</em>, <em class="sig-param">'BstBAI'</em>, <em class="sig-param">'BstBI'</em>, <em class="sig-param">'BstC8I'</em>, <em class="sig-param">'BstDEI'</em>, <em class="sig-param">'BstDSI'</em>, <em class="sig-param">'BstEII'</em>, <em class="sig-param">'BstENI'</em>, <em class="sig-param">'BstF5I'</em>, <em class="sig-param">'BstFNI'</em>, <em class="sig-param">'BstH2I'</em>, <em class="sig-param">'BstHHI'</em>, <em class="sig-param">'BstKTI'</em>, <em class="sig-param">'BstMAI'</em>, <em class="sig-param">'BstMBI'</em>, <em class="sig-param">'BstMCI'</em>, <em class="sig-param">'BstMWI'</em>, <em class="sig-param">'BstNI'</em>, <em class="sig-param">'BstNSI'</em>, <em class="sig-param">'BstPAI'</em>, <em class="sig-param">'BstPI'</em>, <em class="sig-param">'BstSCI'</em>, <em class="sig-param">'BstSFI'</em>, <em class="sig-param">'BstSLI'</em>, <em class="sig-param">'BstSNI'</em>, <em class="sig-param">'BstUI'</em>, <em class="sig-param">'BstV1I'</em>, <em class="sig-param">'BstV2I'</em>, <em class="sig-param">'BstX2I'</em>, <em class="sig-param">'BstXI'</em>, <em class="sig-param">'BstYI'</em>, <em class="sig-param">'BstZ17I'</em>, <em class="sig-param">'BstZI'</em>, <em class="sig-param">'Bsu15I'</em>, <em class="sig-param">'Bsu36I'</em>, <em class="sig-param">'BsuI'</em>, <em class="sig-param">'BsuRI'</em>, <em class="sig-param">'BsuTUI'</em>, <em class="sig-param">'BtgI'</em>, <em class="sig-param">'BtgZI'</em>, <em class="sig-param">'BtrI'</em>, <em class="sig-param">'BtsCI'</em>, <em class="sig-param">'BtsI'</em>, <em class="sig-param">'BtsIMutI'</em>, <em class="sig-param">'BtuMI'</em>, <em class="sig-param">'BveI'</em>, <em class="sig-param">'Cac8I'</em>, <em class="sig-param">'CaiI'</em>, <em class="sig-param">'CciI'</em>, <em class="sig-param">'CciNI'</em>, <em class="sig-param">'CfoI'</em>, <em class="sig-param">'Cfr10I'</em>, <em class="sig-param">'Cfr13I'</em>, <em class="sig-param">'Cfr42I'</em>, <em class="sig-param">'Cfr9I'</em>, <em class="sig-param">'ClaI'</em>, <em class="sig-param">'CpoI'</em>, <em class="sig-param">'CseI'</em>, <em class="sig-param">'CsiI'</em>, <em class="sig-param">'Csp6I'</em>, <em class="sig-param">'CspAI'</em>, <em class="sig-param">'CspCI'</em>, <em class="sig-param">'CspI'</em>, <em class="sig-param">'CviAII'</em>, <em class="sig-param">'CviJI'</em>, <em class="sig-param">'CviKI_1'</em>, <em class="sig-param">'CviQI'</em>, <em class="sig-param">'DdeI'</em>, <em class="sig-param">'DinI'</em>, <em class="sig-param">'DpnI'</em>, <em class="sig-param">'DpnII'</em>, <em class="sig-param">'DraI'</em>, <em class="sig-param">'DraIII'</em>, <em class="sig-param">'DrdI'</em>, <em class="sig-param">'DriI'</em>, <em class="sig-param">'DseDI'</em>, <em class="sig-param">'EaeI'</em>, <em class="sig-param">'EagI'</em>, <em class="sig-param">'Eam1104I'</em>, <em class="sig-param">'Eam1105I'</em>, <em class="sig-param">'EarI'</em>, <em class="sig-param">'EciI'</em>, <em class="sig-param">'Ecl136II'</em>, <em class="sig-param">'EclXI'</em>, <em class="sig-param">'Eco105I'</em>, <em class="sig-param">'Eco130I'</em>, <em class="sig-param">'Eco147I'</em>, <em class="sig-param">'Eco24I'</em>, <em class="sig-param">'Eco31I'</em>, <em class="sig-param">'Eco32I'</em>, <em class="sig-param">'Eco47I'</em>, <em class="sig-param">'Eco47III'</em>, <em class="sig-param">'Eco52I'</em>, <em class="sig-param">'Eco53kI'</em>, <em class="sig-param">'Eco57I'</em>, <em class="sig-param">'Eco72I'</em>, <em class="sig-param">'Eco81I'</em>, <em class="sig-param">'Eco88I'</em>, <em class="sig-param">'Eco91I'</em>, <em class="sig-param">'EcoICRI'</em>, <em class="sig-param">'EcoNI'</em>, <em class="sig-param">'EcoO109I'</em>, <em class="sig-param">'EcoO65I'</em>, <em class="sig-param">'EcoRI'</em>, <em class="sig-param">'EcoRII'</em>, <em class="sig-param">'EcoRV'</em>, <em class="sig-param">'EcoT14I'</em>, <em class="sig-param">'EcoT22I'</em>, <em class="sig-param">'EcoT38I'</em>, <em class="sig-param">'EgeI'</em>, <em class="sig-param">'EheI'</em>, <em class="sig-param">'ErhI'</em>, <em class="sig-param">'Esp3I'</em>, <em class="sig-param">'FaeI'</em>, <em class="sig-param">'FaiI'</em>, <em class="sig-param">'FalI'</em>, <em class="sig-param">'FaqI'</em>, <em class="sig-param">'FatI'</em>, <em class="sig-param">'FauI'</em>, <em class="sig-param">'FauNDI'</em>, <em class="sig-param">'FbaI'</em>, <em class="sig-param">'FblI'</em>, <em class="sig-param">'Fnu4HI'</em>, <em class="sig-param">'FokI'</em>, <em class="sig-param">'FriOI'</em>, <em class="sig-param">'FseI'</em>, <em class="sig-param">'Fsp4HI'</em>, <em class="sig-param">'FspAI'</em>, <em class="sig-param">'FspBI'</em>, <em class="sig-param">'FspEI'</em>, <em class="sig-param">'FspI'</em>, <em class="sig-param">'GlaI'</em>, <em class="sig-param">'GluI'</em>, <em class="sig-param">'GsaI'</em>, <em class="sig-param">'GsuI'</em>, <em class="sig-param">'HaeII'</em>, <em class="sig-param">'HaeIII'</em>, <em class="sig-param">'HapII'</em>, <em class="sig-param">'HgaI'</em>, <em class="sig-param">'HhaI'</em>, <em class="sig-param">'Hin1I'</em>, <em class="sig-param">'Hin1II'</em>, <em class="sig-param">'Hin6I'</em>, <em class="sig-param">'HinP1I'</em>, <em class="sig-param">'HincII'</em>, <em class="sig-param">'HindII'</em>, <em class="sig-param">'HindIII'</em>, <em class="sig-param">'HinfI'</em>, <em class="sig-param">'HpaI'</em>, <em class="sig-param">'HpaII'</em>, <em class="sig-param">'HphI'</em>, <em class="sig-param">'Hpy166II'</em>, <em class="sig-param">'Hpy188I'</em>, <em class="sig-param">'Hpy188III'</em>, <em class="sig-param">'Hpy8I'</em>, <em class="sig-param">'Hpy99I'</em>, <em class="sig-param">'HpyAV'</em>, <em class="sig-param">'HpyCH4III'</em>, <em class="sig-param">'HpyCH4IV'</em>, <em class="sig-param">'HpyCH4V'</em>, <em class="sig-param">'HpyF10VI'</em>, <em class="sig-param">'HpyF3I'</em>, <em class="sig-param">'HpySE526I'</em>, <em class="sig-param">'Hsp92I'</em>, <em class="sig-param">'Hsp92II'</em>, <em class="sig-param">'HspAI'</em>, <em class="sig-param">'KasI'</em>, <em class="sig-param">'KflI'</em>, <em class="sig-param">'Kpn2I'</em>, <em class="sig-param">'KpnI'</em>, <em class="sig-param">'KroI'</em>, <em class="sig-param">'Ksp22I'</em>, <em class="sig-param">'KspAI'</em>, <em class="sig-param">'KspI'</em>, <em class="sig-param">'Kzo9I'</em>, <em class="sig-param">'LguI'</em>, <em class="sig-param">'LmnI'</em>, <em class="sig-param">'LpnPI'</em>, <em class="sig-param">'Lsp1109I'</em>, <em class="sig-param">'LweI'</em>, <em class="sig-param">'MabI'</em>, <em class="sig-param">'MaeI'</em>, <em class="sig-param">'MaeII'</em>, <em class="sig-param">'MaeIII'</em>, <em class="sig-param">'MalI'</em>, <em class="sig-param">'MauBI'</em>, <em class="sig-param">'MbiI'</em>, <em class="sig-param">'MboI'</em>, <em class="sig-param">'MboII'</em>, <em class="sig-param">'MfeI'</em>, <em class="sig-param">'MflI'</em>, <em class="sig-param">'MhlI'</em>, <em class="sig-param">'MlsI'</em>, <em class="sig-param">'MluCI'</em>, <em class="sig-param">'MluI'</em>, <em class="sig-param">'MluNI'</em>, <em class="sig-param">'Mly113I'</em>, <em class="sig-param">'MlyI'</em>, <em class="sig-param">'MmeI'</em>, <em class="sig-param">'MnlI'</em>, <em class="sig-param">'Mox20I'</em>, <em class="sig-param">'Mph1103I'</em>, <em class="sig-param">'MreI'</em>, <em class="sig-param">'MroI'</em>, <em class="sig-param">'MroNI'</em>, <em class="sig-param">'MroXI'</em>, <em class="sig-param">'MscI'</em>, <em class="sig-param">'MseI'</em>, <em class="sig-param">'MslI'</em>, <em class="sig-param">'Msp20I'</em>, <em class="sig-param">'MspA1I'</em>, <em class="sig-param">'MspCI'</em>, <em class="sig-param">'MspI'</em>, <em class="sig-param">'MspJI'</em>, <em class="sig-param">'MspR9I'</em>, <em class="sig-param">'MssI'</em>, <em class="sig-param">'MteI'</em>, <em class="sig-param">'MunI'</em>, <em class="sig-param">'Mva1269I'</em>, <em class="sig-param">'MvaI'</em>, <em class="sig-param">'MvnI'</em>, <em class="sig-param">'MwoI'</em>, <em class="sig-param">'NaeI'</em>, <em class="sig-param">'NarI'</em>, <em class="sig-param">'NciI'</em>, <em class="sig-param">'NcoI'</em>, <em class="sig-param">'NdeI'</em>, <em class="sig-param">'NdeII'</em>, <em class="sig-param">'NgoMIV'</em>, <em class="sig-param">'NheI'</em>, <em class="sig-param">'NlaIII'</em>, <em class="sig-param">'NlaIV'</em>, <em class="sig-param">'NmeAIII'</em>, <em class="sig-param">'NmuCI'</em>, <em class="sig-param">'NotI'</em>, <em class="sig-param">'NruI'</em>, <em class="sig-param">'NsbI'</em>, <em class="sig-param">'NsiI'</em>, <em class="sig-param">'NspI'</em>, <em class="sig-param">'NspV'</em>, <em class="sig-param">'OliI'</em>, <em class="sig-param">'PacI'</em>, <em class="sig-param">'PaeI'</em>, <em class="sig-param">'PaeR7I'</em>, <em class="sig-param">'PagI'</em>, <em class="sig-param">'PalAI'</em>, <em class="sig-param">'PasI'</em>, <em class="sig-param">'PauI'</em>, <em class="sig-param">'PceI'</em>, <em class="sig-param">'PciI'</em>, <em class="sig-param">'PciSI'</em>, <em class="sig-param">'PcsI'</em>, <em class="sig-param">'PctI'</em>, <em class="sig-param">'PdiI'</em>, <em class="sig-param">'PdmI'</em>, <em class="sig-param">'PfeI'</em>, <em class="sig-param">'Pfl23II'</em>, <em class="sig-param">'PflFI'</em>, <em class="sig-param">'PflMI'</em>, <em class="sig-param">'PfoI'</em>, <em class="sig-param">'PinAI'</em>, <em class="sig-param">'PkrI'</em>, <em class="sig-param">'Ple19I'</em>, <em class="sig-param">'PleI'</em>, <em class="sig-param">'PluTI'</em>, <em class="sig-param">'PmaCI'</em>, <em class="sig-param">'PmeI'</em>, <em class="sig-param">'PmlI'</em>, <em class="sig-param">'PpsI'</em>, <em class="sig-param">'Ppu21I'</em>, <em class="sig-param">'PpuMI'</em>, <em class="sig-param">'PscI'</em>, <em class="sig-param">'PshAI'</em>, <em class="sig-param">'PshBI'</em>, <em class="sig-param">'PsiI'</em>, <em class="sig-param">'Psp124BI'</em>, <em class="sig-param">'Psp1406I'</em>, <em class="sig-param">'Psp5II'</em>, <em class="sig-param">'Psp6I'</em>, <em class="sig-param">'PspCI'</em>, <em class="sig-param">'PspEI'</em>, <em class="sig-param">'PspFI'</em>, <em class="sig-param">'PspGI'</em>, <em class="sig-param">'PspLI'</em>, <em class="sig-param">'PspN4I'</em>, <em class="sig-param">'PspOMI'</em>, <em class="sig-param">'PspPI'</em>, <em class="sig-param">'PspPPI'</em>, <em class="sig-param">'PspXI'</em>, <em class="sig-param">'PsrI'</em>, <em class="sig-param">'PstI'</em>, <em class="sig-param">'PstNI'</em>, <em class="sig-param">'PsuI'</em>, <em class="sig-param">'PsyI'</em>, <em class="sig-param">'PteI'</em>, <em class="sig-param">'PvuI'</em>, <em class="sig-param">'PvuII'</em>, <em class="sig-param">'RgaI'</em>, <em class="sig-param">'RigI'</em>, <em class="sig-param">'RruI'</em>, <em class="sig-param">'RsaI'</em>, <em class="sig-param">'RsaNI'</em>, <em class="sig-param">'RseI'</em>, <em class="sig-param">'Rsr2I'</em>, <em class="sig-param">'RsrII'</em>, <em class="sig-param">'SacI'</em>, <em class="sig-param">'SacII'</em>, <em class="sig-param">'SalI'</em>, <em class="sig-param">'SapI'</em>, <em class="sig-param">'SaqAI'</em>, <em class="sig-param">'SatI'</em>, <em class="sig-param">'Sau3AI'</em>, <em class="sig-param">'Sau96I'</em>, <em class="sig-param">'SbfI'</em>, <em class="sig-param">'ScaI'</em>, <em class="sig-param">'SchI'</em>, <em class="sig-param">'ScrFI'</em>, <em class="sig-param">'SdaI'</em>, <em class="sig-param">'SduI'</em>, <em class="sig-param">'SetI'</em>, <em class="sig-param">'SexAI'</em>, <em class="sig-param">'SfaAI'</em>, <em class="sig-param">'SfaNI'</em>, <em class="sig-param">'SfcI'</em>, <em class="sig-param">'SfiI'</em>, <em class="sig-param">'SfoI'</em>, <em class="sig-param">'Sfr274I'</em>, <em class="sig-param">'Sfr303I'</em>, <em class="sig-param">'SfuI'</em>, <em class="sig-param">'SgeI'</em>, <em class="sig-param">'SgfI'</em>, <em class="sig-param">'SgrAI'</em>, <em class="sig-param">'SgrBI'</em>, <em class="sig-param">'SgrDI'</em>, <em class="sig-param">'SgsI'</em>, <em class="sig-param">'SinI'</em>, <em class="sig-param">'SlaI'</em>, <em class="sig-param">'SmaI'</em>, <em class="sig-param">'SmiI'</em>, <em class="sig-param">'SmiMI'</em>, <em class="sig-param">'SmlI'</em>, <em class="sig-param">'SmoI'</em>, <em class="sig-param">'SnaBI'</em>, <em class="sig-param">'SpeI'</em>, <em class="sig-param">'SphI'</em>, <em class="sig-param">'SrfI'</em>, <em class="sig-param">'Sse8387I'</em>, <em class="sig-param">'Sse9I'</em>, <em class="sig-param">'SseBI'</em>, <em class="sig-param">'SsiI'</em>, <em class="sig-param">'SspDI'</em>, <em class="sig-param">'SspI'</em>, <em class="sig-param">'SspMI'</em>, <em class="sig-param">'SstI'</em>, <em class="sig-param">'StuI'</em>, <em class="sig-param">'StyD4I'</em>, <em class="sig-param">'StyI'</em>, <em class="sig-param">'SwaI'</em>, <em class="sig-param">'TaaI'</em>, <em class="sig-param">'TaiI'</em>, <em class="sig-param">'TaqI'</em>, <em class="sig-param">'TaqII'</em>, <em class="sig-param">'TasI'</em>, <em class="sig-param">'TatI'</em>, <em class="sig-param">'TauI'</em>, <em class="sig-param">'TfiI'</em>, <em class="sig-param">'Tru1I'</em>, <em class="sig-param">'Tru9I'</em>, <em class="sig-param">'TscAI'</em>, <em class="sig-param">'TseFI'</em>, <em class="sig-param">'TseI'</em>, <em class="sig-param">'Tsp45I'</em>, <em class="sig-param">'TspDTI'</em>, <em class="sig-param">'TspGWI'</em>, <em class="sig-param">'TspMI'</em>, <em class="sig-param">'TspRI'</em>, <em class="sig-param">'Tth111I'</em>, <em class="sig-param">'Van91I'</em>, <em class="sig-param">'Vha464I'</em>, <em class="sig-param">'VneI'</em>, <em class="sig-param">'VpaK11BI'</em>, <em class="sig-param">'VspI'</em>, <em class="sig-param">'XagI'</em>, <em class="sig-param">'XapI'</em>, <em class="sig-param">'XbaI'</em>, <em class="sig-param">'XceI'</em>, <em class="sig-param">'XcmI'</em>, <em class="sig-param">'XhoI'</em>, <em class="sig-param">'XmaI'</em>, <em class="sig-param">'XmaJI'</em>, <em class="sig-param">'XmiI'</em>, <em class="sig-param">'XmnI'</em>, <em class="sig-param">'XspI'</em>, <em class="sig-param">'ZraI'</em>, <em class="sig-param">'ZrmI'</em>, <em class="sig-param">'Zsp2I'])</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.cutters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.cutters" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.extract_feature">
<code class="sig-name descname">extract_feature</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">n</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.extract_feature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.extract_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts a feature and creates a new Dseqrecord object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>n</strong><span class="classifier">int</span></dt><dd><p>Indicates the feature to extract</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">extract_feature</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-2)</span>
<span class="go">gt</span>
<span class="go">ca</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.find">
<code class="sig-name descname">find</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">other</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.find"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.find" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.find_aa">
<code class="sig-name descname">find_aa</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">other</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.find_aa" title="Permalink to this definition">¶</a></dt>
<dd><div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgtacgatcgtatgctggttatattttag&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;MYDRMLVIF*&#39;, HasStopCodon(ExtendedIUPACProtein(), &#39;*&#39;))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;RML&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;MMM&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;LKYNQHTIVH&#39;, ExtendedIUPACProtein())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;QHT&quot;</span> <span class="ow">in</span> <span class="n">s</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;QHT&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">slc</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">find_aa</span><span class="p">(</span><span class="s2">&quot;RML&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">slc</span>
<span class="go">slice(9, 18, None)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="n">slc</span><span class="p">]</span>
<span class="go">Dseqrecord(-9)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span> <span class="o">=</span> <span class="n">s</span><span class="p">[</span><span class="n">slc</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span>
<span class="go">Dseq(-9)</span>
<span class="go">cgtatgctg</span>
<span class="go">gcatacgac</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;RML&#39;, ExtendedIUPACProtein())</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.find_aminoacids">
<code class="sig-name descname">find_aminoacids</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">other</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.find_aminoacids"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.find_aminoacids" title="Permalink to this definition">¶</a></dt>
<dd><div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgtacgatcgtatgctggttatattttag&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;MYDRMLVIF*&#39;, HasStopCodon(ExtendedIUPACProtein(), &#39;*&#39;))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;RML&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;MMM&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;LKYNQHTIVH&#39;, ExtendedIUPACProtein())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;QHT&quot;</span> <span class="ow">in</span> <span class="n">s</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;QHT&quot;</span> <span class="ow">in</span> <span class="n">s</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">slc</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">find_aa</span><span class="p">(</span><span class="s2">&quot;RML&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">slc</span>
<span class="go">slice(9, 18, None)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="n">slc</span><span class="p">]</span>
<span class="go">Dseqrecord(-9)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span> <span class="o">=</span> <span class="n">s</span><span class="p">[</span><span class="n">slc</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span>
<span class="go">Dseq(-9)</span>
<span class="go">cgtatgctg</span>
<span class="go">gcatacgac</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">code</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;RML&#39;, ExtendedIUPACProtein())</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.format">
<code class="sig-name descname">format</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">f='gb'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.format"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.format" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the sequence as a string using a format supported by Biopython
SeqIO <a class="footnote-reference brackets" href="#id8" id="id7">4</a>. Default is “gb” which is short for Genbank.</p>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id8"><span class="brackets"><a class="fn-backref" href="#id7">4</a></span></dt>
<dd><p><a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s1">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;02-FEB-2013&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;gb&quot;</span><span class="p">))</span>
<span class="go">LOCUS       name                       3 bp    DNA     linear   UNK 02-FEB-2013</span>
<span class="go">DEFINITION  description.</span>
<span class="go">ACCESSION   id</span>
<span class="go">VERSION     id</span>
<span class="go">KEYWORDS    .</span>
<span class="go">SOURCE      .</span>
<span class="go">  ORGANISM  .</span>
<span class="go">            .</span>
<span class="go">FEATURES             Location/Qualifiers</span>
<span class="go">ORIGIN</span>
<span class="go">        1 aaa</span>
<span class="go">//</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.from_SeqRecord">
<em class="property">classmethod </em><code class="sig-name descname">from_SeqRecord</code><span class="sig-paren">(</span><em class="sig-param">record: pydna.seqrecord.SeqRecord</em>, <em class="sig-param">*args</em>, <em class="sig-param">linear=True</em>, <em class="sig-param">circular=False</em>, <em class="sig-param">n=5e-14</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.from_SeqRecord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.from_SeqRecord" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.from_string">
<em class="property">classmethod </em><code class="sig-name descname">from_string</code><span class="sig-paren">(</span><em class="sig-param">record: str = ''</em>, <em class="sig-param">*args</em>, <em class="sig-param">linear=True</em>, <em class="sig-param">circular=False</em>, <em class="sig-param">n=5e-14</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.from_string"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.from_string" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.linear">
<em class="property">property </em><code class="sig-name descname">linear</code><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.linear" title="Permalink to this definition">¶</a></dt>
<dd><p>The linear property can not be set directly. 
Use <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.looped" title="pydna.dseqrecord.Dseqrecord.looped"><code class="xref py py-meth docutils literal notranslate"><span class="pre">looped()</span></code></a> or <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.tolinear" title="pydna.dseqrecord.Dseqrecord.tolinear"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tolinear()</span></code></a></p>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.linearize">
<code class="sig-name descname">linearize</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">*enzymes</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.linearize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.linearize" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is similar to <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord.cut" title="pydna.dseqrecord.Dseqrecord.cut"><code class="xref py py-func docutils literal notranslate"><span class="pre">cut()</span></code></a> but throws an exception if there
is not excactly one cut i.e. none or more than one digestion products.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.looped">
<code class="sig-name descname">looped</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.looped"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.looped" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a circular version of the Dseqrecord object. The
underlying Dseq object has to have compatible ends.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pydna.dseq.Dseq.looped" title="pydna.dseq.Dseq.looped"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.looped</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(o3)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.lseguid">
<code class="sig-name descname">lseguid</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.lseguid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.lseguid" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the url safe lSEGUID for the sequence.</p>
<p>Only defined for linear double stranded sequences.</p>
<p>The lSEGUID checksum uniquely identifies a linear
sequence independent of the direction.
The two Dseqrecord objects below are each others
reverse complements, so they do in fact refer to
the same molecule.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;agtatcgtacatg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">lseguid</span><span class="p">()</span>
<span class="go">&#39;DPshMN4KeAjMovEjGEV4Kzj18lU&#39;</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;catgtacgatact&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">lseguid</span><span class="p">()</span>
<span class="go">&#39;DPshMN4KeAjMovEjGEV4Kzj18lU&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.m">
<code class="sig-name descname">m</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.m"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.m" title="Permalink to this definition">¶</a></dt>
<dd><p>This method returns the mass of the DNA molecule in grams. This is
calculated as the product between the molecular weight of the Dseq object
and the</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.map_trace_files">
<code class="sig-name descname">map_trace_files</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">pth</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.map_trace_files"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.map_trace_files" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.n_cutters">
<code class="sig-name descname">n_cutters</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">n=3</em>, <em class="sig-param">batch=RestrictionBatch(['AanI'</em>, <em class="sig-param">'AarI'</em>, <em class="sig-param">'AasI'</em>, <em class="sig-param">'AatII'</em>, <em class="sig-param">'AbaSI'</em>, <em class="sig-param">'AbsI'</em>, <em class="sig-param">'Acc16I'</em>, <em class="sig-param">'Acc36I'</em>, <em class="sig-param">'Acc65I'</em>, <em class="sig-param">'AccB1I'</em>, <em class="sig-param">'AccB7I'</em>, <em class="sig-param">'AccBSI'</em>, <em class="sig-param">'AccI'</em>, <em class="sig-param">'AccII'</em>, <em class="sig-param">'AccIII'</em>, <em class="sig-param">'AciI'</em>, <em class="sig-param">'AclI'</em>, <em class="sig-param">'AclWI'</em>, <em class="sig-param">'AcoI'</em>, <em class="sig-param">'AcsI'</em>, <em class="sig-param">'AcuI'</em>, <em class="sig-param">'AcvI'</em>, <em class="sig-param">'AcyI'</em>, <em class="sig-param">'AdeI'</em>, <em class="sig-param">'AfaI'</em>, <em class="sig-param">'AfeI'</em>, <em class="sig-param">'AfiI'</em>, <em class="sig-param">'AflII'</em>, <em class="sig-param">'AflIII'</em>, <em class="sig-param">'AgeI'</em>, <em class="sig-param">'AgsI'</em>, <em class="sig-param">'AhdI'</em>, <em class="sig-param">'AhlI'</em>, <em class="sig-param">'AjiI'</em>, <em class="sig-param">'AjnI'</em>, <em class="sig-param">'AjuI'</em>, <em class="sig-param">'AleI'</em>, <em class="sig-param">'AloI'</em>, <em class="sig-param">'AluBI'</em>, <em class="sig-param">'AluI'</em>, <em class="sig-param">'Alw21I'</em>, <em class="sig-param">'Alw26I'</em>, <em class="sig-param">'Alw44I'</em>, <em class="sig-param">'AlwI'</em>, <em class="sig-param">'AlwNI'</em>, <em class="sig-param">'Ama87I'</em>, <em class="sig-param">'Aor13HI'</em>, <em class="sig-param">'Aor51HI'</em>, <em class="sig-param">'AoxI'</em>, <em class="sig-param">'ApaI'</em>, <em class="sig-param">'ApaLI'</em>, <em class="sig-param">'ApeKI'</em>, <em class="sig-param">'ApoI'</em>, <em class="sig-param">'ArsI'</em>, <em class="sig-param">'AscI'</em>, <em class="sig-param">'AseI'</em>, <em class="sig-param">'AsiGI'</em>, <em class="sig-param">'AsiSI'</em>, <em class="sig-param">'Asp700I'</em>, <em class="sig-param">'Asp718I'</em>, <em class="sig-param">'AspA2I'</em>, <em class="sig-param">'AspLEI'</em>, <em class="sig-param">'AspS9I'</em>, <em class="sig-param">'AsuC2I'</em>, <em class="sig-param">'AsuHPI'</em>, <em class="sig-param">'AsuII'</em>, <em class="sig-param">'AsuNHI'</em>, <em class="sig-param">'AvaI'</em>, <em class="sig-param">'AvaII'</em>, <em class="sig-param">'AvrII'</em>, <em class="sig-param">'AxyI'</em>, <em class="sig-param">'BaeGI'</em>, <em class="sig-param">'BaeI'</em>, <em class="sig-param">'BalI'</em>, <em class="sig-param">'BamHI'</em>, <em class="sig-param">'BanI'</em>, <em class="sig-param">'BanII'</em>, <em class="sig-param">'BarI'</em>, <em class="sig-param">'BauI'</em>, <em class="sig-param">'BbrPI'</em>, <em class="sig-param">'BbsI'</em>, <em class="sig-param">'Bbv12I'</em>, <em class="sig-param">'BbvCI'</em>, <em class="sig-param">'BbvI'</em>, <em class="sig-param">'BccI'</em>, <em class="sig-param">'BceAI'</em>, <em class="sig-param">'BcgI'</em>, <em class="sig-param">'BciT130I'</em>, <em class="sig-param">'BciVI'</em>, <em class="sig-param">'BclI'</em>, <em class="sig-param">'BcnI'</em>, <em class="sig-param">'BcoDI'</em>, <em class="sig-param">'BcuI'</em>, <em class="sig-param">'BfaI'</em>, <em class="sig-param">'BfmI'</em>, <em class="sig-param">'BfoI'</em>, <em class="sig-param">'BfrI'</em>, <em class="sig-param">'BfuAI'</em>, <em class="sig-param">'BfuI'</em>, <em class="sig-param">'BglI'</em>, <em class="sig-param">'BglII'</em>, <em class="sig-param">'BisI'</em>, <em class="sig-param">'BlnI'</em>, <em class="sig-param">'BlpI'</em>, <em class="sig-param">'BlsI'</em>, <em class="sig-param">'BmcAI'</em>, <em class="sig-param">'Bme1390I'</em>, <em class="sig-param">'Bme18I'</em>, <em class="sig-param">'BmeRI'</em>, <em class="sig-param">'BmeT110I'</em>, <em class="sig-param">'BmgBI'</em>, <em class="sig-param">'BmgT120I'</em>, <em class="sig-param">'BmiI'</em>, <em class="sig-param">'BmrFI'</em>, <em class="sig-param">'BmrI'</em>, <em class="sig-param">'BmsI'</em>, <em class="sig-param">'BmtI'</em>, <em class="sig-param">'BmuI'</em>, <em class="sig-param">'BoxI'</em>, <em class="sig-param">'BpiI'</em>, <em class="sig-param">'BplI'</em>, <em class="sig-param">'BpmI'</em>, <em class="sig-param">'Bpu10I'</em>, <em class="sig-param">'Bpu1102I'</em>, <em class="sig-param">'Bpu14I'</em>, <em class="sig-param">'BpuEI'</em>, <em class="sig-param">'BpuMI'</em>, <em class="sig-param">'Bsa29I'</em>, <em class="sig-param">'BsaAI'</em>, <em class="sig-param">'BsaBI'</em>, <em class="sig-param">'BsaHI'</em>, <em class="sig-param">'BsaI'</em>, <em class="sig-param">'BsaJI'</em>, <em class="sig-param">'BsaWI'</em>, <em class="sig-param">'BsaXI'</em>, <em class="sig-param">'Bsc4I'</em>, <em class="sig-param">'Bse118I'</em>, <em class="sig-param">'Bse1I'</em>, <em class="sig-param">'Bse21I'</em>, <em class="sig-param">'Bse3DI'</em>, <em class="sig-param">'Bse8I'</em>, <em class="sig-param">'BseAI'</em>, <em class="sig-param">'BseBI'</em>, <em class="sig-param">'BseCI'</em>, <em class="sig-param">'BseDI'</em>, <em class="sig-param">'BseGI'</em>, <em class="sig-param">'BseJI'</em>, <em class="sig-param">'BseLI'</em>, <em class="sig-param">'BseMI'</em>, <em class="sig-param">'BseMII'</em>, <em class="sig-param">'BseNI'</em>, <em class="sig-param">'BsePI'</em>, <em class="sig-param">'BseRI'</em>, <em class="sig-param">'BseSI'</em>, <em class="sig-param">'BseX3I'</em>, <em class="sig-param">'BseXI'</em>, <em class="sig-param">'BseYI'</em>, <em class="sig-param">'BsgI'</em>, <em class="sig-param">'Bsh1236I'</em>, <em class="sig-param">'Bsh1285I'</em>, <em class="sig-param">'BshFI'</em>, <em class="sig-param">'BshNI'</em>, <em class="sig-param">'BshTI'</em>, <em class="sig-param">'BshVI'</em>, <em class="sig-param">'BsiEI'</em>, <em class="sig-param">'BsiHKAI'</em>, <em class="sig-param">'BsiHKCI'</em>, <em class="sig-param">'BsiSI'</em>, <em class="sig-param">'BsiWI'</em>, <em class="sig-param">'BslFI'</em>, <em class="sig-param">'BslI'</em>, <em class="sig-param">'BsmAI'</em>, <em class="sig-param">'BsmBI'</em>, <em class="sig-param">'BsmFI'</em>, <em class="sig-param">'BsmI'</em>, <em class="sig-param">'BsnI'</em>, <em class="sig-param">'Bso31I'</em>, <em class="sig-param">'BsoBI'</em>, <em class="sig-param">'Bsp119I'</em>, <em class="sig-param">'Bsp120I'</em>, <em class="sig-param">'Bsp1286I'</em>, <em class="sig-param">'Bsp13I'</em>, <em class="sig-param">'Bsp1407I'</em>, <em class="sig-param">'Bsp143I'</em>, <em class="sig-param">'Bsp1720I'</em>, <em class="sig-param">'Bsp19I'</em>, <em class="sig-param">'Bsp68I'</em>, <em class="sig-param">'BspACI'</em>, <em class="sig-param">'BspANI'</em>, <em class="sig-param">'BspCNI'</em>, <em class="sig-param">'BspDI'</em>, <em class="sig-param">'BspEI'</em>, <em class="sig-param">'BspFNI'</em>, <em class="sig-param">'BspHI'</em>, <em class="sig-param">'BspLI'</em>, <em class="sig-param">'BspMAI'</em>, <em class="sig-param">'BspMI'</em>, <em class="sig-param">'BspOI'</em>, <em class="sig-param">'BspPI'</em>, <em class="sig-param">'BspQI'</em>, <em class="sig-param">'BspT104I'</em>, <em class="sig-param">'BspT107I'</em>, <em class="sig-param">'BspTI'</em>, <em class="sig-param">'BspTNI'</em>, <em class="sig-param">'BsrBI'</em>, <em class="sig-param">'BsrDI'</em>, <em class="sig-param">'BsrFI'</em>, <em class="sig-param">'BsrGI'</em>, <em class="sig-param">'BsrI'</em>, <em class="sig-param">'BssAI'</em>, <em class="sig-param">'BssECI'</em>, <em class="sig-param">'BssHII'</em>, <em class="sig-param">'BssMI'</em>, <em class="sig-param">'BssNAI'</em>, <em class="sig-param">'BssNI'</em>, <em class="sig-param">'BssSI'</em>, <em class="sig-param">'BssT1I'</em>, <em class="sig-param">'Bst1107I'</em>, <em class="sig-param">'Bst2BI'</em>, <em class="sig-param">'Bst2UI'</em>, <em class="sig-param">'Bst4CI'</em>, <em class="sig-param">'Bst6I'</em>, <em class="sig-param">'BstACI'</em>, <em class="sig-param">'BstAFI'</em>, <em class="sig-param">'BstAPI'</em>, <em class="sig-param">'BstAUI'</em>, <em class="sig-param">'BstBAI'</em>, <em class="sig-param">'BstBI'</em>, <em class="sig-param">'BstC8I'</em>, <em class="sig-param">'BstDEI'</em>, <em class="sig-param">'BstDSI'</em>, <em class="sig-param">'BstEII'</em>, <em class="sig-param">'BstENI'</em>, <em class="sig-param">'BstF5I'</em>, <em class="sig-param">'BstFNI'</em>, <em class="sig-param">'BstH2I'</em>, <em class="sig-param">'BstHHI'</em>, <em class="sig-param">'BstKTI'</em>, <em class="sig-param">'BstMAI'</em>, <em class="sig-param">'BstMBI'</em>, <em class="sig-param">'BstMCI'</em>, <em class="sig-param">'BstMWI'</em>, <em class="sig-param">'BstNI'</em>, <em class="sig-param">'BstNSI'</em>, <em class="sig-param">'BstPAI'</em>, <em class="sig-param">'BstPI'</em>, <em class="sig-param">'BstSCI'</em>, <em class="sig-param">'BstSFI'</em>, <em class="sig-param">'BstSLI'</em>, <em class="sig-param">'BstSNI'</em>, <em class="sig-param">'BstUI'</em>, <em class="sig-param">'BstV1I'</em>, <em class="sig-param">'BstV2I'</em>, <em class="sig-param">'BstX2I'</em>, <em class="sig-param">'BstXI'</em>, <em class="sig-param">'BstYI'</em>, <em class="sig-param">'BstZ17I'</em>, <em class="sig-param">'BstZI'</em>, <em class="sig-param">'Bsu15I'</em>, <em class="sig-param">'Bsu36I'</em>, <em class="sig-param">'BsuI'</em>, <em class="sig-param">'BsuRI'</em>, <em class="sig-param">'BsuTUI'</em>, <em class="sig-param">'BtgI'</em>, <em class="sig-param">'BtgZI'</em>, <em class="sig-param">'BtrI'</em>, <em class="sig-param">'BtsCI'</em>, <em class="sig-param">'BtsI'</em>, <em class="sig-param">'BtsIMutI'</em>, <em class="sig-param">'BtuMI'</em>, <em class="sig-param">'BveI'</em>, <em class="sig-param">'Cac8I'</em>, <em class="sig-param">'CaiI'</em>, <em class="sig-param">'CciI'</em>, <em class="sig-param">'CciNI'</em>, <em class="sig-param">'CfoI'</em>, <em class="sig-param">'Cfr10I'</em>, <em class="sig-param">'Cfr13I'</em>, <em class="sig-param">'Cfr42I'</em>, <em class="sig-param">'Cfr9I'</em>, <em class="sig-param">'ClaI'</em>, <em class="sig-param">'CpoI'</em>, <em class="sig-param">'CseI'</em>, <em class="sig-param">'CsiI'</em>, <em class="sig-param">'Csp6I'</em>, <em class="sig-param">'CspAI'</em>, <em class="sig-param">'CspCI'</em>, <em class="sig-param">'CspI'</em>, <em class="sig-param">'CviAII'</em>, <em class="sig-param">'CviJI'</em>, <em class="sig-param">'CviKI_1'</em>, <em class="sig-param">'CviQI'</em>, <em class="sig-param">'DdeI'</em>, <em class="sig-param">'DinI'</em>, <em class="sig-param">'DpnI'</em>, <em class="sig-param">'DpnII'</em>, <em class="sig-param">'DraI'</em>, <em class="sig-param">'DraIII'</em>, <em class="sig-param">'DrdI'</em>, <em class="sig-param">'DriI'</em>, <em class="sig-param">'DseDI'</em>, <em class="sig-param">'EaeI'</em>, <em class="sig-param">'EagI'</em>, <em class="sig-param">'Eam1104I'</em>, <em class="sig-param">'Eam1105I'</em>, <em class="sig-param">'EarI'</em>, <em class="sig-param">'EciI'</em>, <em class="sig-param">'Ecl136II'</em>, <em class="sig-param">'EclXI'</em>, <em class="sig-param">'Eco105I'</em>, <em class="sig-param">'Eco130I'</em>, <em class="sig-param">'Eco147I'</em>, <em class="sig-param">'Eco24I'</em>, <em class="sig-param">'Eco31I'</em>, <em class="sig-param">'Eco32I'</em>, <em class="sig-param">'Eco47I'</em>, <em class="sig-param">'Eco47III'</em>, <em class="sig-param">'Eco52I'</em>, <em class="sig-param">'Eco53kI'</em>, <em class="sig-param">'Eco57I'</em>, <em class="sig-param">'Eco72I'</em>, <em class="sig-param">'Eco81I'</em>, <em class="sig-param">'Eco88I'</em>, <em class="sig-param">'Eco91I'</em>, <em class="sig-param">'EcoICRI'</em>, <em class="sig-param">'EcoNI'</em>, <em class="sig-param">'EcoO109I'</em>, <em class="sig-param">'EcoO65I'</em>, <em class="sig-param">'EcoRI'</em>, <em class="sig-param">'EcoRII'</em>, <em class="sig-param">'EcoRV'</em>, <em class="sig-param">'EcoT14I'</em>, <em class="sig-param">'EcoT22I'</em>, <em class="sig-param">'EcoT38I'</em>, <em class="sig-param">'EgeI'</em>, <em class="sig-param">'EheI'</em>, <em class="sig-param">'ErhI'</em>, <em class="sig-param">'Esp3I'</em>, <em class="sig-param">'FaeI'</em>, <em class="sig-param">'FaiI'</em>, <em class="sig-param">'FalI'</em>, <em class="sig-param">'FaqI'</em>, <em class="sig-param">'FatI'</em>, <em class="sig-param">'FauI'</em>, <em class="sig-param">'FauNDI'</em>, <em class="sig-param">'FbaI'</em>, <em class="sig-param">'FblI'</em>, <em class="sig-param">'Fnu4HI'</em>, <em class="sig-param">'FokI'</em>, <em class="sig-param">'FriOI'</em>, <em class="sig-param">'FseI'</em>, <em class="sig-param">'Fsp4HI'</em>, <em class="sig-param">'FspAI'</em>, <em class="sig-param">'FspBI'</em>, <em class="sig-param">'FspEI'</em>, <em class="sig-param">'FspI'</em>, <em class="sig-param">'GlaI'</em>, <em class="sig-param">'GluI'</em>, <em class="sig-param">'GsaI'</em>, <em class="sig-param">'GsuI'</em>, <em class="sig-param">'HaeII'</em>, <em class="sig-param">'HaeIII'</em>, <em class="sig-param">'HapII'</em>, <em class="sig-param">'HgaI'</em>, <em class="sig-param">'HhaI'</em>, <em class="sig-param">'Hin1I'</em>, <em class="sig-param">'Hin1II'</em>, <em class="sig-param">'Hin6I'</em>, <em class="sig-param">'HinP1I'</em>, <em class="sig-param">'HincII'</em>, <em class="sig-param">'HindII'</em>, <em class="sig-param">'HindIII'</em>, <em class="sig-param">'HinfI'</em>, <em class="sig-param">'HpaI'</em>, <em class="sig-param">'HpaII'</em>, <em class="sig-param">'HphI'</em>, <em class="sig-param">'Hpy166II'</em>, <em class="sig-param">'Hpy188I'</em>, <em class="sig-param">'Hpy188III'</em>, <em class="sig-param">'Hpy8I'</em>, <em class="sig-param">'Hpy99I'</em>, <em class="sig-param">'HpyAV'</em>, <em class="sig-param">'HpyCH4III'</em>, <em class="sig-param">'HpyCH4IV'</em>, <em class="sig-param">'HpyCH4V'</em>, <em class="sig-param">'HpyF10VI'</em>, <em class="sig-param">'HpyF3I'</em>, <em class="sig-param">'HpySE526I'</em>, <em class="sig-param">'Hsp92I'</em>, <em class="sig-param">'Hsp92II'</em>, <em class="sig-param">'HspAI'</em>, <em class="sig-param">'KasI'</em>, <em class="sig-param">'KflI'</em>, <em class="sig-param">'Kpn2I'</em>, <em class="sig-param">'KpnI'</em>, <em class="sig-param">'KroI'</em>, <em class="sig-param">'Ksp22I'</em>, <em class="sig-param">'KspAI'</em>, <em class="sig-param">'KspI'</em>, <em class="sig-param">'Kzo9I'</em>, <em class="sig-param">'LguI'</em>, <em class="sig-param">'LmnI'</em>, <em class="sig-param">'LpnPI'</em>, <em class="sig-param">'Lsp1109I'</em>, <em class="sig-param">'LweI'</em>, <em class="sig-param">'MabI'</em>, <em class="sig-param">'MaeI'</em>, <em class="sig-param">'MaeII'</em>, <em class="sig-param">'MaeIII'</em>, <em class="sig-param">'MalI'</em>, <em class="sig-param">'MauBI'</em>, <em class="sig-param">'MbiI'</em>, <em class="sig-param">'MboI'</em>, <em class="sig-param">'MboII'</em>, <em class="sig-param">'MfeI'</em>, <em class="sig-param">'MflI'</em>, <em class="sig-param">'MhlI'</em>, <em class="sig-param">'MlsI'</em>, <em class="sig-param">'MluCI'</em>, <em class="sig-param">'MluI'</em>, <em class="sig-param">'MluNI'</em>, <em class="sig-param">'Mly113I'</em>, <em class="sig-param">'MlyI'</em>, <em class="sig-param">'MmeI'</em>, <em class="sig-param">'MnlI'</em>, <em class="sig-param">'Mox20I'</em>, <em class="sig-param">'Mph1103I'</em>, <em class="sig-param">'MreI'</em>, <em class="sig-param">'MroI'</em>, <em class="sig-param">'MroNI'</em>, <em class="sig-param">'MroXI'</em>, <em class="sig-param">'MscI'</em>, <em class="sig-param">'MseI'</em>, <em class="sig-param">'MslI'</em>, <em class="sig-param">'Msp20I'</em>, <em class="sig-param">'MspA1I'</em>, <em class="sig-param">'MspCI'</em>, <em class="sig-param">'MspI'</em>, <em class="sig-param">'MspJI'</em>, <em class="sig-param">'MspR9I'</em>, <em class="sig-param">'MssI'</em>, <em class="sig-param">'MteI'</em>, <em class="sig-param">'MunI'</em>, <em class="sig-param">'Mva1269I'</em>, <em class="sig-param">'MvaI'</em>, <em class="sig-param">'MvnI'</em>, <em class="sig-param">'MwoI'</em>, <em class="sig-param">'NaeI'</em>, <em class="sig-param">'NarI'</em>, <em class="sig-param">'NciI'</em>, <em class="sig-param">'NcoI'</em>, <em class="sig-param">'NdeI'</em>, <em class="sig-param">'NdeII'</em>, <em class="sig-param">'NgoMIV'</em>, <em class="sig-param">'NheI'</em>, <em class="sig-param">'NlaIII'</em>, <em class="sig-param">'NlaIV'</em>, <em class="sig-param">'NmeAIII'</em>, <em class="sig-param">'NmuCI'</em>, <em class="sig-param">'NotI'</em>, <em class="sig-param">'NruI'</em>, <em class="sig-param">'NsbI'</em>, <em class="sig-param">'NsiI'</em>, <em class="sig-param">'NspI'</em>, <em class="sig-param">'NspV'</em>, <em class="sig-param">'OliI'</em>, <em class="sig-param">'PacI'</em>, <em class="sig-param">'PaeI'</em>, <em class="sig-param">'PaeR7I'</em>, <em class="sig-param">'PagI'</em>, <em class="sig-param">'PalAI'</em>, <em class="sig-param">'PasI'</em>, <em class="sig-param">'PauI'</em>, <em class="sig-param">'PceI'</em>, <em class="sig-param">'PciI'</em>, <em class="sig-param">'PciSI'</em>, <em class="sig-param">'PcsI'</em>, <em class="sig-param">'PctI'</em>, <em class="sig-param">'PdiI'</em>, <em class="sig-param">'PdmI'</em>, <em class="sig-param">'PfeI'</em>, <em class="sig-param">'Pfl23II'</em>, <em class="sig-param">'PflFI'</em>, <em class="sig-param">'PflMI'</em>, <em class="sig-param">'PfoI'</em>, <em class="sig-param">'PinAI'</em>, <em class="sig-param">'PkrI'</em>, <em class="sig-param">'Ple19I'</em>, <em class="sig-param">'PleI'</em>, <em class="sig-param">'PluTI'</em>, <em class="sig-param">'PmaCI'</em>, <em class="sig-param">'PmeI'</em>, <em class="sig-param">'PmlI'</em>, <em class="sig-param">'PpsI'</em>, <em class="sig-param">'Ppu21I'</em>, <em class="sig-param">'PpuMI'</em>, <em class="sig-param">'PscI'</em>, <em class="sig-param">'PshAI'</em>, <em class="sig-param">'PshBI'</em>, <em class="sig-param">'PsiI'</em>, <em class="sig-param">'Psp124BI'</em>, <em class="sig-param">'Psp1406I'</em>, <em class="sig-param">'Psp5II'</em>, <em class="sig-param">'Psp6I'</em>, <em class="sig-param">'PspCI'</em>, <em class="sig-param">'PspEI'</em>, <em class="sig-param">'PspFI'</em>, <em class="sig-param">'PspGI'</em>, <em class="sig-param">'PspLI'</em>, <em class="sig-param">'PspN4I'</em>, <em class="sig-param">'PspOMI'</em>, <em class="sig-param">'PspPI'</em>, <em class="sig-param">'PspPPI'</em>, <em class="sig-param">'PspXI'</em>, <em class="sig-param">'PsrI'</em>, <em class="sig-param">'PstI'</em>, <em class="sig-param">'PstNI'</em>, <em class="sig-param">'PsuI'</em>, <em class="sig-param">'PsyI'</em>, <em class="sig-param">'PteI'</em>, <em class="sig-param">'PvuI'</em>, <em class="sig-param">'PvuII'</em>, <em class="sig-param">'RgaI'</em>, <em class="sig-param">'RigI'</em>, <em class="sig-param">'RruI'</em>, <em class="sig-param">'RsaI'</em>, <em class="sig-param">'RsaNI'</em>, <em class="sig-param">'RseI'</em>, <em class="sig-param">'Rsr2I'</em>, <em class="sig-param">'RsrII'</em>, <em class="sig-param">'SacI'</em>, <em class="sig-param">'SacII'</em>, <em class="sig-param">'SalI'</em>, <em class="sig-param">'SapI'</em>, <em class="sig-param">'SaqAI'</em>, <em class="sig-param">'SatI'</em>, <em class="sig-param">'Sau3AI'</em>, <em class="sig-param">'Sau96I'</em>, <em class="sig-param">'SbfI'</em>, <em class="sig-param">'ScaI'</em>, <em class="sig-param">'SchI'</em>, <em class="sig-param">'ScrFI'</em>, <em class="sig-param">'SdaI'</em>, <em class="sig-param">'SduI'</em>, <em class="sig-param">'SetI'</em>, <em class="sig-param">'SexAI'</em>, <em class="sig-param">'SfaAI'</em>, <em class="sig-param">'SfaNI'</em>, <em class="sig-param">'SfcI'</em>, <em class="sig-param">'SfiI'</em>, <em class="sig-param">'SfoI'</em>, <em class="sig-param">'Sfr274I'</em>, <em class="sig-param">'Sfr303I'</em>, <em class="sig-param">'SfuI'</em>, <em class="sig-param">'SgeI'</em>, <em class="sig-param">'SgfI'</em>, <em class="sig-param">'SgrAI'</em>, <em class="sig-param">'SgrBI'</em>, <em class="sig-param">'SgrDI'</em>, <em class="sig-param">'SgsI'</em>, <em class="sig-param">'SinI'</em>, <em class="sig-param">'SlaI'</em>, <em class="sig-param">'SmaI'</em>, <em class="sig-param">'SmiI'</em>, <em class="sig-param">'SmiMI'</em>, <em class="sig-param">'SmlI'</em>, <em class="sig-param">'SmoI'</em>, <em class="sig-param">'SnaBI'</em>, <em class="sig-param">'SpeI'</em>, <em class="sig-param">'SphI'</em>, <em class="sig-param">'SrfI'</em>, <em class="sig-param">'Sse8387I'</em>, <em class="sig-param">'Sse9I'</em>, <em class="sig-param">'SseBI'</em>, <em class="sig-param">'SsiI'</em>, <em class="sig-param">'SspDI'</em>, <em class="sig-param">'SspI'</em>, <em class="sig-param">'SspMI'</em>, <em class="sig-param">'SstI'</em>, <em class="sig-param">'StuI'</em>, <em class="sig-param">'StyD4I'</em>, <em class="sig-param">'StyI'</em>, <em class="sig-param">'SwaI'</em>, <em class="sig-param">'TaaI'</em>, <em class="sig-param">'TaiI'</em>, <em class="sig-param">'TaqI'</em>, <em class="sig-param">'TaqII'</em>, <em class="sig-param">'TasI'</em>, <em class="sig-param">'TatI'</em>, <em class="sig-param">'TauI'</em>, <em class="sig-param">'TfiI'</em>, <em class="sig-param">'Tru1I'</em>, <em class="sig-param">'Tru9I'</em>, <em class="sig-param">'TscAI'</em>, <em class="sig-param">'TseFI'</em>, <em class="sig-param">'TseI'</em>, <em class="sig-param">'Tsp45I'</em>, <em class="sig-param">'TspDTI'</em>, <em class="sig-param">'TspGWI'</em>, <em class="sig-param">'TspMI'</em>, <em class="sig-param">'TspRI'</em>, <em class="sig-param">'Tth111I'</em>, <em class="sig-param">'Van91I'</em>, <em class="sig-param">'Vha464I'</em>, <em class="sig-param">'VneI'</em>, <em class="sig-param">'VpaK11BI'</em>, <em class="sig-param">'VspI'</em>, <em class="sig-param">'XagI'</em>, <em class="sig-param">'XapI'</em>, <em class="sig-param">'XbaI'</em>, <em class="sig-param">'XceI'</em>, <em class="sig-param">'XcmI'</em>, <em class="sig-param">'XhoI'</em>, <em class="sig-param">'XmaI'</em>, <em class="sig-param">'XmaJI'</em>, <em class="sig-param">'XmiI'</em>, <em class="sig-param">'XmnI'</em>, <em class="sig-param">'XspI'</em>, <em class="sig-param">'ZraI'</em>, <em class="sig-param">'ZrmI'</em>, <em class="sig-param">'Zsp2I'])</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.n_cutters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.n_cutters" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.no_cutters">
<code class="sig-name descname">no_cutters</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">batch=RestrictionBatch(['AanI'</em>, <em class="sig-param">'AarI'</em>, <em class="sig-param">'AasI'</em>, <em class="sig-param">'AatII'</em>, <em class="sig-param">'AbaSI'</em>, <em class="sig-param">'AbsI'</em>, <em class="sig-param">'Acc16I'</em>, <em class="sig-param">'Acc36I'</em>, <em class="sig-param">'Acc65I'</em>, <em class="sig-param">'AccB1I'</em>, <em class="sig-param">'AccB7I'</em>, <em class="sig-param">'AccBSI'</em>, <em class="sig-param">'AccI'</em>, <em class="sig-param">'AccII'</em>, <em class="sig-param">'AccIII'</em>, <em class="sig-param">'AciI'</em>, <em class="sig-param">'AclI'</em>, <em class="sig-param">'AclWI'</em>, <em class="sig-param">'AcoI'</em>, <em class="sig-param">'AcsI'</em>, <em class="sig-param">'AcuI'</em>, <em class="sig-param">'AcvI'</em>, <em class="sig-param">'AcyI'</em>, <em class="sig-param">'AdeI'</em>, <em class="sig-param">'AfaI'</em>, <em class="sig-param">'AfeI'</em>, <em class="sig-param">'AfiI'</em>, <em class="sig-param">'AflII'</em>, <em class="sig-param">'AflIII'</em>, <em class="sig-param">'AgeI'</em>, <em class="sig-param">'AgsI'</em>, <em class="sig-param">'AhdI'</em>, <em class="sig-param">'AhlI'</em>, <em class="sig-param">'AjiI'</em>, <em class="sig-param">'AjnI'</em>, <em class="sig-param">'AjuI'</em>, <em class="sig-param">'AleI'</em>, <em class="sig-param">'AloI'</em>, <em class="sig-param">'AluBI'</em>, <em class="sig-param">'AluI'</em>, <em class="sig-param">'Alw21I'</em>, <em class="sig-param">'Alw26I'</em>, <em class="sig-param">'Alw44I'</em>, <em class="sig-param">'AlwI'</em>, <em class="sig-param">'AlwNI'</em>, <em class="sig-param">'Ama87I'</em>, <em class="sig-param">'Aor13HI'</em>, <em class="sig-param">'Aor51HI'</em>, <em class="sig-param">'AoxI'</em>, <em class="sig-param">'ApaI'</em>, <em class="sig-param">'ApaLI'</em>, <em class="sig-param">'ApeKI'</em>, <em class="sig-param">'ApoI'</em>, <em class="sig-param">'ArsI'</em>, <em class="sig-param">'AscI'</em>, <em class="sig-param">'AseI'</em>, <em class="sig-param">'AsiGI'</em>, <em class="sig-param">'AsiSI'</em>, <em class="sig-param">'Asp700I'</em>, <em class="sig-param">'Asp718I'</em>, <em class="sig-param">'AspA2I'</em>, <em class="sig-param">'AspLEI'</em>, <em class="sig-param">'AspS9I'</em>, <em class="sig-param">'AsuC2I'</em>, <em class="sig-param">'AsuHPI'</em>, <em class="sig-param">'AsuII'</em>, <em class="sig-param">'AsuNHI'</em>, <em class="sig-param">'AvaI'</em>, <em class="sig-param">'AvaII'</em>, <em class="sig-param">'AvrII'</em>, <em class="sig-param">'AxyI'</em>, <em class="sig-param">'BaeGI'</em>, <em class="sig-param">'BaeI'</em>, <em class="sig-param">'BalI'</em>, <em class="sig-param">'BamHI'</em>, <em class="sig-param">'BanI'</em>, <em class="sig-param">'BanII'</em>, <em class="sig-param">'BarI'</em>, <em class="sig-param">'BauI'</em>, <em class="sig-param">'BbrPI'</em>, <em class="sig-param">'BbsI'</em>, <em class="sig-param">'Bbv12I'</em>, <em class="sig-param">'BbvCI'</em>, <em class="sig-param">'BbvI'</em>, <em class="sig-param">'BccI'</em>, <em class="sig-param">'BceAI'</em>, <em class="sig-param">'BcgI'</em>, <em class="sig-param">'BciT130I'</em>, <em class="sig-param">'BciVI'</em>, <em class="sig-param">'BclI'</em>, <em class="sig-param">'BcnI'</em>, <em class="sig-param">'BcoDI'</em>, <em class="sig-param">'BcuI'</em>, <em class="sig-param">'BfaI'</em>, <em class="sig-param">'BfmI'</em>, <em class="sig-param">'BfoI'</em>, <em class="sig-param">'BfrI'</em>, <em class="sig-param">'BfuAI'</em>, <em class="sig-param">'BfuI'</em>, <em class="sig-param">'BglI'</em>, <em class="sig-param">'BglII'</em>, <em class="sig-param">'BisI'</em>, <em class="sig-param">'BlnI'</em>, <em class="sig-param">'BlpI'</em>, <em class="sig-param">'BlsI'</em>, <em class="sig-param">'BmcAI'</em>, <em class="sig-param">'Bme1390I'</em>, <em class="sig-param">'Bme18I'</em>, <em class="sig-param">'BmeRI'</em>, <em class="sig-param">'BmeT110I'</em>, <em class="sig-param">'BmgBI'</em>, <em class="sig-param">'BmgT120I'</em>, <em class="sig-param">'BmiI'</em>, <em class="sig-param">'BmrFI'</em>, <em class="sig-param">'BmrI'</em>, <em class="sig-param">'BmsI'</em>, <em class="sig-param">'BmtI'</em>, <em class="sig-param">'BmuI'</em>, <em class="sig-param">'BoxI'</em>, <em class="sig-param">'BpiI'</em>, <em class="sig-param">'BplI'</em>, <em class="sig-param">'BpmI'</em>, <em class="sig-param">'Bpu10I'</em>, <em class="sig-param">'Bpu1102I'</em>, <em class="sig-param">'Bpu14I'</em>, <em class="sig-param">'BpuEI'</em>, <em class="sig-param">'BpuMI'</em>, <em class="sig-param">'Bsa29I'</em>, <em class="sig-param">'BsaAI'</em>, <em class="sig-param">'BsaBI'</em>, <em class="sig-param">'BsaHI'</em>, <em class="sig-param">'BsaI'</em>, <em class="sig-param">'BsaJI'</em>, <em class="sig-param">'BsaWI'</em>, <em class="sig-param">'BsaXI'</em>, <em class="sig-param">'Bsc4I'</em>, <em class="sig-param">'Bse118I'</em>, <em class="sig-param">'Bse1I'</em>, <em class="sig-param">'Bse21I'</em>, <em class="sig-param">'Bse3DI'</em>, <em class="sig-param">'Bse8I'</em>, <em class="sig-param">'BseAI'</em>, <em class="sig-param">'BseBI'</em>, <em class="sig-param">'BseCI'</em>, <em class="sig-param">'BseDI'</em>, <em class="sig-param">'BseGI'</em>, <em class="sig-param">'BseJI'</em>, <em class="sig-param">'BseLI'</em>, <em class="sig-param">'BseMI'</em>, <em class="sig-param">'BseMII'</em>, <em class="sig-param">'BseNI'</em>, <em class="sig-param">'BsePI'</em>, <em class="sig-param">'BseRI'</em>, <em class="sig-param">'BseSI'</em>, <em class="sig-param">'BseX3I'</em>, <em class="sig-param">'BseXI'</em>, <em class="sig-param">'BseYI'</em>, <em class="sig-param">'BsgI'</em>, <em class="sig-param">'Bsh1236I'</em>, <em class="sig-param">'Bsh1285I'</em>, <em class="sig-param">'BshFI'</em>, <em class="sig-param">'BshNI'</em>, <em class="sig-param">'BshTI'</em>, <em class="sig-param">'BshVI'</em>, <em class="sig-param">'BsiEI'</em>, <em class="sig-param">'BsiHKAI'</em>, <em class="sig-param">'BsiHKCI'</em>, <em class="sig-param">'BsiSI'</em>, <em class="sig-param">'BsiWI'</em>, <em class="sig-param">'BslFI'</em>, <em class="sig-param">'BslI'</em>, <em class="sig-param">'BsmAI'</em>, <em class="sig-param">'BsmBI'</em>, <em class="sig-param">'BsmFI'</em>, <em class="sig-param">'BsmI'</em>, <em class="sig-param">'BsnI'</em>, <em class="sig-param">'Bso31I'</em>, <em class="sig-param">'BsoBI'</em>, <em class="sig-param">'Bsp119I'</em>, <em class="sig-param">'Bsp120I'</em>, <em class="sig-param">'Bsp1286I'</em>, <em class="sig-param">'Bsp13I'</em>, <em class="sig-param">'Bsp1407I'</em>, <em class="sig-param">'Bsp143I'</em>, <em class="sig-param">'Bsp1720I'</em>, <em class="sig-param">'Bsp19I'</em>, <em class="sig-param">'Bsp68I'</em>, <em class="sig-param">'BspACI'</em>, <em class="sig-param">'BspANI'</em>, <em class="sig-param">'BspCNI'</em>, <em class="sig-param">'BspDI'</em>, <em class="sig-param">'BspEI'</em>, <em class="sig-param">'BspFNI'</em>, <em class="sig-param">'BspHI'</em>, <em class="sig-param">'BspLI'</em>, <em class="sig-param">'BspMAI'</em>, <em class="sig-param">'BspMI'</em>, <em class="sig-param">'BspOI'</em>, <em class="sig-param">'BspPI'</em>, <em class="sig-param">'BspQI'</em>, <em class="sig-param">'BspT104I'</em>, <em class="sig-param">'BspT107I'</em>, <em class="sig-param">'BspTI'</em>, <em class="sig-param">'BspTNI'</em>, <em class="sig-param">'BsrBI'</em>, <em class="sig-param">'BsrDI'</em>, <em class="sig-param">'BsrFI'</em>, <em class="sig-param">'BsrGI'</em>, <em class="sig-param">'BsrI'</em>, <em class="sig-param">'BssAI'</em>, <em class="sig-param">'BssECI'</em>, <em class="sig-param">'BssHII'</em>, <em class="sig-param">'BssMI'</em>, <em class="sig-param">'BssNAI'</em>, <em class="sig-param">'BssNI'</em>, <em class="sig-param">'BssSI'</em>, <em class="sig-param">'BssT1I'</em>, <em class="sig-param">'Bst1107I'</em>, <em class="sig-param">'Bst2BI'</em>, <em class="sig-param">'Bst2UI'</em>, <em class="sig-param">'Bst4CI'</em>, <em class="sig-param">'Bst6I'</em>, <em class="sig-param">'BstACI'</em>, <em class="sig-param">'BstAFI'</em>, <em class="sig-param">'BstAPI'</em>, <em class="sig-param">'BstAUI'</em>, <em class="sig-param">'BstBAI'</em>, <em class="sig-param">'BstBI'</em>, <em class="sig-param">'BstC8I'</em>, <em class="sig-param">'BstDEI'</em>, <em class="sig-param">'BstDSI'</em>, <em class="sig-param">'BstEII'</em>, <em class="sig-param">'BstENI'</em>, <em class="sig-param">'BstF5I'</em>, <em class="sig-param">'BstFNI'</em>, <em class="sig-param">'BstH2I'</em>, <em class="sig-param">'BstHHI'</em>, <em class="sig-param">'BstKTI'</em>, <em class="sig-param">'BstMAI'</em>, <em class="sig-param">'BstMBI'</em>, <em class="sig-param">'BstMCI'</em>, <em class="sig-param">'BstMWI'</em>, <em class="sig-param">'BstNI'</em>, <em class="sig-param">'BstNSI'</em>, <em class="sig-param">'BstPAI'</em>, <em class="sig-param">'BstPI'</em>, <em class="sig-param">'BstSCI'</em>, <em class="sig-param">'BstSFI'</em>, <em class="sig-param">'BstSLI'</em>, <em class="sig-param">'BstSNI'</em>, <em class="sig-param">'BstUI'</em>, <em class="sig-param">'BstV1I'</em>, <em class="sig-param">'BstV2I'</em>, <em class="sig-param">'BstX2I'</em>, <em class="sig-param">'BstXI'</em>, <em class="sig-param">'BstYI'</em>, <em class="sig-param">'BstZ17I'</em>, <em class="sig-param">'BstZI'</em>, <em class="sig-param">'Bsu15I'</em>, <em class="sig-param">'Bsu36I'</em>, <em class="sig-param">'BsuI'</em>, <em class="sig-param">'BsuRI'</em>, <em class="sig-param">'BsuTUI'</em>, <em class="sig-param">'BtgI'</em>, <em class="sig-param">'BtgZI'</em>, <em class="sig-param">'BtrI'</em>, <em class="sig-param">'BtsCI'</em>, <em class="sig-param">'BtsI'</em>, <em class="sig-param">'BtsIMutI'</em>, <em class="sig-param">'BtuMI'</em>, <em class="sig-param">'BveI'</em>, <em class="sig-param">'Cac8I'</em>, <em class="sig-param">'CaiI'</em>, <em class="sig-param">'CciI'</em>, <em class="sig-param">'CciNI'</em>, <em class="sig-param">'CfoI'</em>, <em class="sig-param">'Cfr10I'</em>, <em class="sig-param">'Cfr13I'</em>, <em class="sig-param">'Cfr42I'</em>, <em class="sig-param">'Cfr9I'</em>, <em class="sig-param">'ClaI'</em>, <em class="sig-param">'CpoI'</em>, <em class="sig-param">'CseI'</em>, <em class="sig-param">'CsiI'</em>, <em class="sig-param">'Csp6I'</em>, <em class="sig-param">'CspAI'</em>, <em class="sig-param">'CspCI'</em>, <em class="sig-param">'CspI'</em>, <em class="sig-param">'CviAII'</em>, <em class="sig-param">'CviJI'</em>, <em class="sig-param">'CviKI_1'</em>, <em class="sig-param">'CviQI'</em>, <em class="sig-param">'DdeI'</em>, <em class="sig-param">'DinI'</em>, <em class="sig-param">'DpnI'</em>, <em class="sig-param">'DpnII'</em>, <em class="sig-param">'DraI'</em>, <em class="sig-param">'DraIII'</em>, <em class="sig-param">'DrdI'</em>, <em class="sig-param">'DriI'</em>, <em class="sig-param">'DseDI'</em>, <em class="sig-param">'EaeI'</em>, <em class="sig-param">'EagI'</em>, <em class="sig-param">'Eam1104I'</em>, <em class="sig-param">'Eam1105I'</em>, <em class="sig-param">'EarI'</em>, <em class="sig-param">'EciI'</em>, <em class="sig-param">'Ecl136II'</em>, <em class="sig-param">'EclXI'</em>, <em class="sig-param">'Eco105I'</em>, <em class="sig-param">'Eco130I'</em>, <em class="sig-param">'Eco147I'</em>, <em class="sig-param">'Eco24I'</em>, <em class="sig-param">'Eco31I'</em>, <em class="sig-param">'Eco32I'</em>, <em class="sig-param">'Eco47I'</em>, <em class="sig-param">'Eco47III'</em>, <em class="sig-param">'Eco52I'</em>, <em class="sig-param">'Eco53kI'</em>, <em class="sig-param">'Eco57I'</em>, <em class="sig-param">'Eco72I'</em>, <em class="sig-param">'Eco81I'</em>, <em class="sig-param">'Eco88I'</em>, <em class="sig-param">'Eco91I'</em>, <em class="sig-param">'EcoICRI'</em>, <em class="sig-param">'EcoNI'</em>, <em class="sig-param">'EcoO109I'</em>, <em class="sig-param">'EcoO65I'</em>, <em class="sig-param">'EcoRI'</em>, <em class="sig-param">'EcoRII'</em>, <em class="sig-param">'EcoRV'</em>, <em class="sig-param">'EcoT14I'</em>, <em class="sig-param">'EcoT22I'</em>, <em class="sig-param">'EcoT38I'</em>, <em class="sig-param">'EgeI'</em>, <em class="sig-param">'EheI'</em>, <em class="sig-param">'ErhI'</em>, <em class="sig-param">'Esp3I'</em>, <em class="sig-param">'FaeI'</em>, <em class="sig-param">'FaiI'</em>, <em class="sig-param">'FalI'</em>, <em class="sig-param">'FaqI'</em>, <em class="sig-param">'FatI'</em>, <em class="sig-param">'FauI'</em>, <em class="sig-param">'FauNDI'</em>, <em class="sig-param">'FbaI'</em>, <em class="sig-param">'FblI'</em>, <em class="sig-param">'Fnu4HI'</em>, <em class="sig-param">'FokI'</em>, <em class="sig-param">'FriOI'</em>, <em class="sig-param">'FseI'</em>, <em class="sig-param">'Fsp4HI'</em>, <em class="sig-param">'FspAI'</em>, <em class="sig-param">'FspBI'</em>, <em class="sig-param">'FspEI'</em>, <em class="sig-param">'FspI'</em>, <em class="sig-param">'GlaI'</em>, <em class="sig-param">'GluI'</em>, <em class="sig-param">'GsaI'</em>, <em class="sig-param">'GsuI'</em>, <em class="sig-param">'HaeII'</em>, <em class="sig-param">'HaeIII'</em>, <em class="sig-param">'HapII'</em>, <em class="sig-param">'HgaI'</em>, <em class="sig-param">'HhaI'</em>, <em class="sig-param">'Hin1I'</em>, <em class="sig-param">'Hin1II'</em>, <em class="sig-param">'Hin6I'</em>, <em class="sig-param">'HinP1I'</em>, <em class="sig-param">'HincII'</em>, <em class="sig-param">'HindII'</em>, <em class="sig-param">'HindIII'</em>, <em class="sig-param">'HinfI'</em>, <em class="sig-param">'HpaI'</em>, <em class="sig-param">'HpaII'</em>, <em class="sig-param">'HphI'</em>, <em class="sig-param">'Hpy166II'</em>, <em class="sig-param">'Hpy188I'</em>, <em class="sig-param">'Hpy188III'</em>, <em class="sig-param">'Hpy8I'</em>, <em class="sig-param">'Hpy99I'</em>, <em class="sig-param">'HpyAV'</em>, <em class="sig-param">'HpyCH4III'</em>, <em class="sig-param">'HpyCH4IV'</em>, <em class="sig-param">'HpyCH4V'</em>, <em class="sig-param">'HpyF10VI'</em>, <em class="sig-param">'HpyF3I'</em>, <em class="sig-param">'HpySE526I'</em>, <em class="sig-param">'Hsp92I'</em>, <em class="sig-param">'Hsp92II'</em>, <em class="sig-param">'HspAI'</em>, <em class="sig-param">'KasI'</em>, <em class="sig-param">'KflI'</em>, <em class="sig-param">'Kpn2I'</em>, <em class="sig-param">'KpnI'</em>, <em class="sig-param">'KroI'</em>, <em class="sig-param">'Ksp22I'</em>, <em class="sig-param">'KspAI'</em>, <em class="sig-param">'KspI'</em>, <em class="sig-param">'Kzo9I'</em>, <em class="sig-param">'LguI'</em>, <em class="sig-param">'LmnI'</em>, <em class="sig-param">'LpnPI'</em>, <em class="sig-param">'Lsp1109I'</em>, <em class="sig-param">'LweI'</em>, <em class="sig-param">'MabI'</em>, <em class="sig-param">'MaeI'</em>, <em class="sig-param">'MaeII'</em>, <em class="sig-param">'MaeIII'</em>, <em class="sig-param">'MalI'</em>, <em class="sig-param">'MauBI'</em>, <em class="sig-param">'MbiI'</em>, <em class="sig-param">'MboI'</em>, <em class="sig-param">'MboII'</em>, <em class="sig-param">'MfeI'</em>, <em class="sig-param">'MflI'</em>, <em class="sig-param">'MhlI'</em>, <em class="sig-param">'MlsI'</em>, <em class="sig-param">'MluCI'</em>, <em class="sig-param">'MluI'</em>, <em class="sig-param">'MluNI'</em>, <em class="sig-param">'Mly113I'</em>, <em class="sig-param">'MlyI'</em>, <em class="sig-param">'MmeI'</em>, <em class="sig-param">'MnlI'</em>, <em class="sig-param">'Mox20I'</em>, <em class="sig-param">'Mph1103I'</em>, <em class="sig-param">'MreI'</em>, <em class="sig-param">'MroI'</em>, <em class="sig-param">'MroNI'</em>, <em class="sig-param">'MroXI'</em>, <em class="sig-param">'MscI'</em>, <em class="sig-param">'MseI'</em>, <em class="sig-param">'MslI'</em>, <em class="sig-param">'Msp20I'</em>, <em class="sig-param">'MspA1I'</em>, <em class="sig-param">'MspCI'</em>, <em class="sig-param">'MspI'</em>, <em class="sig-param">'MspJI'</em>, <em class="sig-param">'MspR9I'</em>, <em class="sig-param">'MssI'</em>, <em class="sig-param">'MteI'</em>, <em class="sig-param">'MunI'</em>, <em class="sig-param">'Mva1269I'</em>, <em class="sig-param">'MvaI'</em>, <em class="sig-param">'MvnI'</em>, <em class="sig-param">'MwoI'</em>, <em class="sig-param">'NaeI'</em>, <em class="sig-param">'NarI'</em>, <em class="sig-param">'NciI'</em>, <em class="sig-param">'NcoI'</em>, <em class="sig-param">'NdeI'</em>, <em class="sig-param">'NdeII'</em>, <em class="sig-param">'NgoMIV'</em>, <em class="sig-param">'NheI'</em>, <em class="sig-param">'NlaIII'</em>, <em class="sig-param">'NlaIV'</em>, <em class="sig-param">'NmeAIII'</em>, <em class="sig-param">'NmuCI'</em>, <em class="sig-param">'NotI'</em>, <em class="sig-param">'NruI'</em>, <em class="sig-param">'NsbI'</em>, <em class="sig-param">'NsiI'</em>, <em class="sig-param">'NspI'</em>, <em class="sig-param">'NspV'</em>, <em class="sig-param">'OliI'</em>, <em class="sig-param">'PacI'</em>, <em class="sig-param">'PaeI'</em>, <em class="sig-param">'PaeR7I'</em>, <em class="sig-param">'PagI'</em>, <em class="sig-param">'PalAI'</em>, <em class="sig-param">'PasI'</em>, <em class="sig-param">'PauI'</em>, <em class="sig-param">'PceI'</em>, <em class="sig-param">'PciI'</em>, <em class="sig-param">'PciSI'</em>, <em class="sig-param">'PcsI'</em>, <em class="sig-param">'PctI'</em>, <em class="sig-param">'PdiI'</em>, <em class="sig-param">'PdmI'</em>, <em class="sig-param">'PfeI'</em>, <em class="sig-param">'Pfl23II'</em>, <em class="sig-param">'PflFI'</em>, <em class="sig-param">'PflMI'</em>, <em class="sig-param">'PfoI'</em>, <em class="sig-param">'PinAI'</em>, <em class="sig-param">'PkrI'</em>, <em class="sig-param">'Ple19I'</em>, <em class="sig-param">'PleI'</em>, <em class="sig-param">'PluTI'</em>, <em class="sig-param">'PmaCI'</em>, <em class="sig-param">'PmeI'</em>, <em class="sig-param">'PmlI'</em>, <em class="sig-param">'PpsI'</em>, <em class="sig-param">'Ppu21I'</em>, <em class="sig-param">'PpuMI'</em>, <em class="sig-param">'PscI'</em>, <em class="sig-param">'PshAI'</em>, <em class="sig-param">'PshBI'</em>, <em class="sig-param">'PsiI'</em>, <em class="sig-param">'Psp124BI'</em>, <em class="sig-param">'Psp1406I'</em>, <em class="sig-param">'Psp5II'</em>, <em class="sig-param">'Psp6I'</em>, <em class="sig-param">'PspCI'</em>, <em class="sig-param">'PspEI'</em>, <em class="sig-param">'PspFI'</em>, <em class="sig-param">'PspGI'</em>, <em class="sig-param">'PspLI'</em>, <em class="sig-param">'PspN4I'</em>, <em class="sig-param">'PspOMI'</em>, <em class="sig-param">'PspPI'</em>, <em class="sig-param">'PspPPI'</em>, <em class="sig-param">'PspXI'</em>, <em class="sig-param">'PsrI'</em>, <em class="sig-param">'PstI'</em>, <em class="sig-param">'PstNI'</em>, <em class="sig-param">'PsuI'</em>, <em class="sig-param">'PsyI'</em>, <em class="sig-param">'PteI'</em>, <em class="sig-param">'PvuI'</em>, <em class="sig-param">'PvuII'</em>, <em class="sig-param">'RgaI'</em>, <em class="sig-param">'RigI'</em>, <em class="sig-param">'RruI'</em>, <em class="sig-param">'RsaI'</em>, <em class="sig-param">'RsaNI'</em>, <em class="sig-param">'RseI'</em>, <em class="sig-param">'Rsr2I'</em>, <em class="sig-param">'RsrII'</em>, <em class="sig-param">'SacI'</em>, <em class="sig-param">'SacII'</em>, <em class="sig-param">'SalI'</em>, <em class="sig-param">'SapI'</em>, <em class="sig-param">'SaqAI'</em>, <em class="sig-param">'SatI'</em>, <em class="sig-param">'Sau3AI'</em>, <em class="sig-param">'Sau96I'</em>, <em class="sig-param">'SbfI'</em>, <em class="sig-param">'ScaI'</em>, <em class="sig-param">'SchI'</em>, <em class="sig-param">'ScrFI'</em>, <em class="sig-param">'SdaI'</em>, <em class="sig-param">'SduI'</em>, <em class="sig-param">'SetI'</em>, <em class="sig-param">'SexAI'</em>, <em class="sig-param">'SfaAI'</em>, <em class="sig-param">'SfaNI'</em>, <em class="sig-param">'SfcI'</em>, <em class="sig-param">'SfiI'</em>, <em class="sig-param">'SfoI'</em>, <em class="sig-param">'Sfr274I'</em>, <em class="sig-param">'Sfr303I'</em>, <em class="sig-param">'SfuI'</em>, <em class="sig-param">'SgeI'</em>, <em class="sig-param">'SgfI'</em>, <em class="sig-param">'SgrAI'</em>, <em class="sig-param">'SgrBI'</em>, <em class="sig-param">'SgrDI'</em>, <em class="sig-param">'SgsI'</em>, <em class="sig-param">'SinI'</em>, <em class="sig-param">'SlaI'</em>, <em class="sig-param">'SmaI'</em>, <em class="sig-param">'SmiI'</em>, <em class="sig-param">'SmiMI'</em>, <em class="sig-param">'SmlI'</em>, <em class="sig-param">'SmoI'</em>, <em class="sig-param">'SnaBI'</em>, <em class="sig-param">'SpeI'</em>, <em class="sig-param">'SphI'</em>, <em class="sig-param">'SrfI'</em>, <em class="sig-param">'Sse8387I'</em>, <em class="sig-param">'Sse9I'</em>, <em class="sig-param">'SseBI'</em>, <em class="sig-param">'SsiI'</em>, <em class="sig-param">'SspDI'</em>, <em class="sig-param">'SspI'</em>, <em class="sig-param">'SspMI'</em>, <em class="sig-param">'SstI'</em>, <em class="sig-param">'StuI'</em>, <em class="sig-param">'StyD4I'</em>, <em class="sig-param">'StyI'</em>, <em class="sig-param">'SwaI'</em>, <em class="sig-param">'TaaI'</em>, <em class="sig-param">'TaiI'</em>, <em class="sig-param">'TaqI'</em>, <em class="sig-param">'TaqII'</em>, <em class="sig-param">'TasI'</em>, <em class="sig-param">'TatI'</em>, <em class="sig-param">'TauI'</em>, <em class="sig-param">'TfiI'</em>, <em class="sig-param">'Tru1I'</em>, <em class="sig-param">'Tru9I'</em>, <em class="sig-param">'TscAI'</em>, <em class="sig-param">'TseFI'</em>, <em class="sig-param">'TseI'</em>, <em class="sig-param">'Tsp45I'</em>, <em class="sig-param">'TspDTI'</em>, <em class="sig-param">'TspGWI'</em>, <em class="sig-param">'TspMI'</em>, <em class="sig-param">'TspRI'</em>, <em class="sig-param">'Tth111I'</em>, <em class="sig-param">'Van91I'</em>, <em class="sig-param">'Vha464I'</em>, <em class="sig-param">'VneI'</em>, <em class="sig-param">'VpaK11BI'</em>, <em class="sig-param">'VspI'</em>, <em class="sig-param">'XagI'</em>, <em class="sig-param">'XapI'</em>, <em class="sig-param">'XbaI'</em>, <em class="sig-param">'XceI'</em>, <em class="sig-param">'XcmI'</em>, <em class="sig-param">'XhoI'</em>, <em class="sig-param">'XmaI'</em>, <em class="sig-param">'XmaJI'</em>, <em class="sig-param">'XmiI'</em>, <em class="sig-param">'XmnI'</em>, <em class="sig-param">'XspI'</em>, <em class="sig-param">'ZraI'</em>, <em class="sig-param">'ZrmI'</em>, <em class="sig-param">'Zsp2I'])</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.no_cutters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.no_cutters" title="Permalink to this definition">¶</a></dt>
<dd><p>See</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.number_of_cuts">
<code class="sig-name descname">number_of_cuts</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">*enzymes</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.number_of_cuts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.number_of_cuts" title="Permalink to this definition">¶</a></dt>
<dd><p>This method returns the number of cuts by digestion with the Restriction enzymes contained in 
the iterable.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.once_cutters">
<code class="sig-name descname">once_cutters</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">batch=RestrictionBatch(['AanI'</em>, <em class="sig-param">'AarI'</em>, <em class="sig-param">'AasI'</em>, <em class="sig-param">'AatII'</em>, <em class="sig-param">'AbaSI'</em>, <em class="sig-param">'AbsI'</em>, <em class="sig-param">'Acc16I'</em>, <em class="sig-param">'Acc36I'</em>, <em class="sig-param">'Acc65I'</em>, <em class="sig-param">'AccB1I'</em>, <em class="sig-param">'AccB7I'</em>, <em class="sig-param">'AccBSI'</em>, <em class="sig-param">'AccI'</em>, <em class="sig-param">'AccII'</em>, <em class="sig-param">'AccIII'</em>, <em class="sig-param">'AciI'</em>, <em class="sig-param">'AclI'</em>, <em class="sig-param">'AclWI'</em>, <em class="sig-param">'AcoI'</em>, <em class="sig-param">'AcsI'</em>, <em class="sig-param">'AcuI'</em>, <em class="sig-param">'AcvI'</em>, <em class="sig-param">'AcyI'</em>, <em class="sig-param">'AdeI'</em>, <em class="sig-param">'AfaI'</em>, <em class="sig-param">'AfeI'</em>, <em class="sig-param">'AfiI'</em>, <em class="sig-param">'AflII'</em>, <em class="sig-param">'AflIII'</em>, <em class="sig-param">'AgeI'</em>, <em class="sig-param">'AgsI'</em>, <em class="sig-param">'AhdI'</em>, <em class="sig-param">'AhlI'</em>, <em class="sig-param">'AjiI'</em>, <em class="sig-param">'AjnI'</em>, <em class="sig-param">'AjuI'</em>, <em class="sig-param">'AleI'</em>, <em class="sig-param">'AloI'</em>, <em class="sig-param">'AluBI'</em>, <em class="sig-param">'AluI'</em>, <em class="sig-param">'Alw21I'</em>, <em class="sig-param">'Alw26I'</em>, <em class="sig-param">'Alw44I'</em>, <em class="sig-param">'AlwI'</em>, <em class="sig-param">'AlwNI'</em>, <em class="sig-param">'Ama87I'</em>, <em class="sig-param">'Aor13HI'</em>, <em class="sig-param">'Aor51HI'</em>, <em class="sig-param">'AoxI'</em>, <em class="sig-param">'ApaI'</em>, <em class="sig-param">'ApaLI'</em>, <em class="sig-param">'ApeKI'</em>, <em class="sig-param">'ApoI'</em>, <em class="sig-param">'ArsI'</em>, <em class="sig-param">'AscI'</em>, <em class="sig-param">'AseI'</em>, <em class="sig-param">'AsiGI'</em>, <em class="sig-param">'AsiSI'</em>, <em class="sig-param">'Asp700I'</em>, <em class="sig-param">'Asp718I'</em>, <em class="sig-param">'AspA2I'</em>, <em class="sig-param">'AspLEI'</em>, <em class="sig-param">'AspS9I'</em>, <em class="sig-param">'AsuC2I'</em>, <em class="sig-param">'AsuHPI'</em>, <em class="sig-param">'AsuII'</em>, <em class="sig-param">'AsuNHI'</em>, <em class="sig-param">'AvaI'</em>, <em class="sig-param">'AvaII'</em>, <em class="sig-param">'AvrII'</em>, <em class="sig-param">'AxyI'</em>, <em class="sig-param">'BaeGI'</em>, <em class="sig-param">'BaeI'</em>, <em class="sig-param">'BalI'</em>, <em class="sig-param">'BamHI'</em>, <em class="sig-param">'BanI'</em>, <em class="sig-param">'BanII'</em>, <em class="sig-param">'BarI'</em>, <em class="sig-param">'BauI'</em>, <em class="sig-param">'BbrPI'</em>, <em class="sig-param">'BbsI'</em>, <em class="sig-param">'Bbv12I'</em>, <em class="sig-param">'BbvCI'</em>, <em class="sig-param">'BbvI'</em>, <em class="sig-param">'BccI'</em>, <em class="sig-param">'BceAI'</em>, <em class="sig-param">'BcgI'</em>, <em class="sig-param">'BciT130I'</em>, <em class="sig-param">'BciVI'</em>, <em class="sig-param">'BclI'</em>, <em class="sig-param">'BcnI'</em>, <em class="sig-param">'BcoDI'</em>, <em class="sig-param">'BcuI'</em>, <em class="sig-param">'BfaI'</em>, <em class="sig-param">'BfmI'</em>, <em class="sig-param">'BfoI'</em>, <em class="sig-param">'BfrI'</em>, <em class="sig-param">'BfuAI'</em>, <em class="sig-param">'BfuI'</em>, <em class="sig-param">'BglI'</em>, <em class="sig-param">'BglII'</em>, <em class="sig-param">'BisI'</em>, <em class="sig-param">'BlnI'</em>, <em class="sig-param">'BlpI'</em>, <em class="sig-param">'BlsI'</em>, <em class="sig-param">'BmcAI'</em>, <em class="sig-param">'Bme1390I'</em>, <em class="sig-param">'Bme18I'</em>, <em class="sig-param">'BmeRI'</em>, <em class="sig-param">'BmeT110I'</em>, <em class="sig-param">'BmgBI'</em>, <em class="sig-param">'BmgT120I'</em>, <em class="sig-param">'BmiI'</em>, <em class="sig-param">'BmrFI'</em>, <em class="sig-param">'BmrI'</em>, <em class="sig-param">'BmsI'</em>, <em class="sig-param">'BmtI'</em>, <em class="sig-param">'BmuI'</em>, <em class="sig-param">'BoxI'</em>, <em class="sig-param">'BpiI'</em>, <em class="sig-param">'BplI'</em>, <em class="sig-param">'BpmI'</em>, <em class="sig-param">'Bpu10I'</em>, <em class="sig-param">'Bpu1102I'</em>, <em class="sig-param">'Bpu14I'</em>, <em class="sig-param">'BpuEI'</em>, <em class="sig-param">'BpuMI'</em>, <em class="sig-param">'Bsa29I'</em>, <em class="sig-param">'BsaAI'</em>, <em class="sig-param">'BsaBI'</em>, <em class="sig-param">'BsaHI'</em>, <em class="sig-param">'BsaI'</em>, <em class="sig-param">'BsaJI'</em>, <em class="sig-param">'BsaWI'</em>, <em class="sig-param">'BsaXI'</em>, <em class="sig-param">'Bsc4I'</em>, <em class="sig-param">'Bse118I'</em>, <em class="sig-param">'Bse1I'</em>, <em class="sig-param">'Bse21I'</em>, <em class="sig-param">'Bse3DI'</em>, <em class="sig-param">'Bse8I'</em>, <em class="sig-param">'BseAI'</em>, <em class="sig-param">'BseBI'</em>, <em class="sig-param">'BseCI'</em>, <em class="sig-param">'BseDI'</em>, <em class="sig-param">'BseGI'</em>, <em class="sig-param">'BseJI'</em>, <em class="sig-param">'BseLI'</em>, <em class="sig-param">'BseMI'</em>, <em class="sig-param">'BseMII'</em>, <em class="sig-param">'BseNI'</em>, <em class="sig-param">'BsePI'</em>, <em class="sig-param">'BseRI'</em>, <em class="sig-param">'BseSI'</em>, <em class="sig-param">'BseX3I'</em>, <em class="sig-param">'BseXI'</em>, <em class="sig-param">'BseYI'</em>, <em class="sig-param">'BsgI'</em>, <em class="sig-param">'Bsh1236I'</em>, <em class="sig-param">'Bsh1285I'</em>, <em class="sig-param">'BshFI'</em>, <em class="sig-param">'BshNI'</em>, <em class="sig-param">'BshTI'</em>, <em class="sig-param">'BshVI'</em>, <em class="sig-param">'BsiEI'</em>, <em class="sig-param">'BsiHKAI'</em>, <em class="sig-param">'BsiHKCI'</em>, <em class="sig-param">'BsiSI'</em>, <em class="sig-param">'BsiWI'</em>, <em class="sig-param">'BslFI'</em>, <em class="sig-param">'BslI'</em>, <em class="sig-param">'BsmAI'</em>, <em class="sig-param">'BsmBI'</em>, <em class="sig-param">'BsmFI'</em>, <em class="sig-param">'BsmI'</em>, <em class="sig-param">'BsnI'</em>, <em class="sig-param">'Bso31I'</em>, <em class="sig-param">'BsoBI'</em>, <em class="sig-param">'Bsp119I'</em>, <em class="sig-param">'Bsp120I'</em>, <em class="sig-param">'Bsp1286I'</em>, <em class="sig-param">'Bsp13I'</em>, <em class="sig-param">'Bsp1407I'</em>, <em class="sig-param">'Bsp143I'</em>, <em class="sig-param">'Bsp1720I'</em>, <em class="sig-param">'Bsp19I'</em>, <em class="sig-param">'Bsp68I'</em>, <em class="sig-param">'BspACI'</em>, <em class="sig-param">'BspANI'</em>, <em class="sig-param">'BspCNI'</em>, <em class="sig-param">'BspDI'</em>, <em class="sig-param">'BspEI'</em>, <em class="sig-param">'BspFNI'</em>, <em class="sig-param">'BspHI'</em>, <em class="sig-param">'BspLI'</em>, <em class="sig-param">'BspMAI'</em>, <em class="sig-param">'BspMI'</em>, <em class="sig-param">'BspOI'</em>, <em class="sig-param">'BspPI'</em>, <em class="sig-param">'BspQI'</em>, <em class="sig-param">'BspT104I'</em>, <em class="sig-param">'BspT107I'</em>, <em class="sig-param">'BspTI'</em>, <em class="sig-param">'BspTNI'</em>, <em class="sig-param">'BsrBI'</em>, <em class="sig-param">'BsrDI'</em>, <em class="sig-param">'BsrFI'</em>, <em class="sig-param">'BsrGI'</em>, <em class="sig-param">'BsrI'</em>, <em class="sig-param">'BssAI'</em>, <em class="sig-param">'BssECI'</em>, <em class="sig-param">'BssHII'</em>, <em class="sig-param">'BssMI'</em>, <em class="sig-param">'BssNAI'</em>, <em class="sig-param">'BssNI'</em>, <em class="sig-param">'BssSI'</em>, <em class="sig-param">'BssT1I'</em>, <em class="sig-param">'Bst1107I'</em>, <em class="sig-param">'Bst2BI'</em>, <em class="sig-param">'Bst2UI'</em>, <em class="sig-param">'Bst4CI'</em>, <em class="sig-param">'Bst6I'</em>, <em class="sig-param">'BstACI'</em>, <em class="sig-param">'BstAFI'</em>, <em class="sig-param">'BstAPI'</em>, <em class="sig-param">'BstAUI'</em>, <em class="sig-param">'BstBAI'</em>, <em class="sig-param">'BstBI'</em>, <em class="sig-param">'BstC8I'</em>, <em class="sig-param">'BstDEI'</em>, <em class="sig-param">'BstDSI'</em>, <em class="sig-param">'BstEII'</em>, <em class="sig-param">'BstENI'</em>, <em class="sig-param">'BstF5I'</em>, <em class="sig-param">'BstFNI'</em>, <em class="sig-param">'BstH2I'</em>, <em class="sig-param">'BstHHI'</em>, <em class="sig-param">'BstKTI'</em>, <em class="sig-param">'BstMAI'</em>, <em class="sig-param">'BstMBI'</em>, <em class="sig-param">'BstMCI'</em>, <em class="sig-param">'BstMWI'</em>, <em class="sig-param">'BstNI'</em>, <em class="sig-param">'BstNSI'</em>, <em class="sig-param">'BstPAI'</em>, <em class="sig-param">'BstPI'</em>, <em class="sig-param">'BstSCI'</em>, <em class="sig-param">'BstSFI'</em>, <em class="sig-param">'BstSLI'</em>, <em class="sig-param">'BstSNI'</em>, <em class="sig-param">'BstUI'</em>, <em class="sig-param">'BstV1I'</em>, <em class="sig-param">'BstV2I'</em>, <em class="sig-param">'BstX2I'</em>, <em class="sig-param">'BstXI'</em>, <em class="sig-param">'BstYI'</em>, <em class="sig-param">'BstZ17I'</em>, <em class="sig-param">'BstZI'</em>, <em class="sig-param">'Bsu15I'</em>, <em class="sig-param">'Bsu36I'</em>, <em class="sig-param">'BsuI'</em>, <em class="sig-param">'BsuRI'</em>, <em class="sig-param">'BsuTUI'</em>, <em class="sig-param">'BtgI'</em>, <em class="sig-param">'BtgZI'</em>, <em class="sig-param">'BtrI'</em>, <em class="sig-param">'BtsCI'</em>, <em class="sig-param">'BtsI'</em>, <em class="sig-param">'BtsIMutI'</em>, <em class="sig-param">'BtuMI'</em>, <em class="sig-param">'BveI'</em>, <em class="sig-param">'Cac8I'</em>, <em class="sig-param">'CaiI'</em>, <em class="sig-param">'CciI'</em>, <em class="sig-param">'CciNI'</em>, <em class="sig-param">'CfoI'</em>, <em class="sig-param">'Cfr10I'</em>, <em class="sig-param">'Cfr13I'</em>, <em class="sig-param">'Cfr42I'</em>, <em class="sig-param">'Cfr9I'</em>, <em class="sig-param">'ClaI'</em>, <em class="sig-param">'CpoI'</em>, <em class="sig-param">'CseI'</em>, <em class="sig-param">'CsiI'</em>, <em class="sig-param">'Csp6I'</em>, <em class="sig-param">'CspAI'</em>, <em class="sig-param">'CspCI'</em>, <em class="sig-param">'CspI'</em>, <em class="sig-param">'CviAII'</em>, <em class="sig-param">'CviJI'</em>, <em class="sig-param">'CviKI_1'</em>, <em class="sig-param">'CviQI'</em>, <em class="sig-param">'DdeI'</em>, <em class="sig-param">'DinI'</em>, <em class="sig-param">'DpnI'</em>, <em class="sig-param">'DpnII'</em>, <em class="sig-param">'DraI'</em>, <em class="sig-param">'DraIII'</em>, <em class="sig-param">'DrdI'</em>, <em class="sig-param">'DriI'</em>, <em class="sig-param">'DseDI'</em>, <em class="sig-param">'EaeI'</em>, <em class="sig-param">'EagI'</em>, <em class="sig-param">'Eam1104I'</em>, <em class="sig-param">'Eam1105I'</em>, <em class="sig-param">'EarI'</em>, <em class="sig-param">'EciI'</em>, <em class="sig-param">'Ecl136II'</em>, <em class="sig-param">'EclXI'</em>, <em class="sig-param">'Eco105I'</em>, <em class="sig-param">'Eco130I'</em>, <em class="sig-param">'Eco147I'</em>, <em class="sig-param">'Eco24I'</em>, <em class="sig-param">'Eco31I'</em>, <em class="sig-param">'Eco32I'</em>, <em class="sig-param">'Eco47I'</em>, <em class="sig-param">'Eco47III'</em>, <em class="sig-param">'Eco52I'</em>, <em class="sig-param">'Eco53kI'</em>, <em class="sig-param">'Eco57I'</em>, <em class="sig-param">'Eco72I'</em>, <em class="sig-param">'Eco81I'</em>, <em class="sig-param">'Eco88I'</em>, <em class="sig-param">'Eco91I'</em>, <em class="sig-param">'EcoICRI'</em>, <em class="sig-param">'EcoNI'</em>, <em class="sig-param">'EcoO109I'</em>, <em class="sig-param">'EcoO65I'</em>, <em class="sig-param">'EcoRI'</em>, <em class="sig-param">'EcoRII'</em>, <em class="sig-param">'EcoRV'</em>, <em class="sig-param">'EcoT14I'</em>, <em class="sig-param">'EcoT22I'</em>, <em class="sig-param">'EcoT38I'</em>, <em class="sig-param">'EgeI'</em>, <em class="sig-param">'EheI'</em>, <em class="sig-param">'ErhI'</em>, <em class="sig-param">'Esp3I'</em>, <em class="sig-param">'FaeI'</em>, <em class="sig-param">'FaiI'</em>, <em class="sig-param">'FalI'</em>, <em class="sig-param">'FaqI'</em>, <em class="sig-param">'FatI'</em>, <em class="sig-param">'FauI'</em>, <em class="sig-param">'FauNDI'</em>, <em class="sig-param">'FbaI'</em>, <em class="sig-param">'FblI'</em>, <em class="sig-param">'Fnu4HI'</em>, <em class="sig-param">'FokI'</em>, <em class="sig-param">'FriOI'</em>, <em class="sig-param">'FseI'</em>, <em class="sig-param">'Fsp4HI'</em>, <em class="sig-param">'FspAI'</em>, <em class="sig-param">'FspBI'</em>, <em class="sig-param">'FspEI'</em>, <em class="sig-param">'FspI'</em>, <em class="sig-param">'GlaI'</em>, <em class="sig-param">'GluI'</em>, <em class="sig-param">'GsaI'</em>, <em class="sig-param">'GsuI'</em>, <em class="sig-param">'HaeII'</em>, <em class="sig-param">'HaeIII'</em>, <em class="sig-param">'HapII'</em>, <em class="sig-param">'HgaI'</em>, <em class="sig-param">'HhaI'</em>, <em class="sig-param">'Hin1I'</em>, <em class="sig-param">'Hin1II'</em>, <em class="sig-param">'Hin6I'</em>, <em class="sig-param">'HinP1I'</em>, <em class="sig-param">'HincII'</em>, <em class="sig-param">'HindII'</em>, <em class="sig-param">'HindIII'</em>, <em class="sig-param">'HinfI'</em>, <em class="sig-param">'HpaI'</em>, <em class="sig-param">'HpaII'</em>, <em class="sig-param">'HphI'</em>, <em class="sig-param">'Hpy166II'</em>, <em class="sig-param">'Hpy188I'</em>, <em class="sig-param">'Hpy188III'</em>, <em class="sig-param">'Hpy8I'</em>, <em class="sig-param">'Hpy99I'</em>, <em class="sig-param">'HpyAV'</em>, <em class="sig-param">'HpyCH4III'</em>, <em class="sig-param">'HpyCH4IV'</em>, <em class="sig-param">'HpyCH4V'</em>, <em class="sig-param">'HpyF10VI'</em>, <em class="sig-param">'HpyF3I'</em>, <em class="sig-param">'HpySE526I'</em>, <em class="sig-param">'Hsp92I'</em>, <em class="sig-param">'Hsp92II'</em>, <em class="sig-param">'HspAI'</em>, <em class="sig-param">'KasI'</em>, <em class="sig-param">'KflI'</em>, <em class="sig-param">'Kpn2I'</em>, <em class="sig-param">'KpnI'</em>, <em class="sig-param">'KroI'</em>, <em class="sig-param">'Ksp22I'</em>, <em class="sig-param">'KspAI'</em>, <em class="sig-param">'KspI'</em>, <em class="sig-param">'Kzo9I'</em>, <em class="sig-param">'LguI'</em>, <em class="sig-param">'LmnI'</em>, <em class="sig-param">'LpnPI'</em>, <em class="sig-param">'Lsp1109I'</em>, <em class="sig-param">'LweI'</em>, <em class="sig-param">'MabI'</em>, <em class="sig-param">'MaeI'</em>, <em class="sig-param">'MaeII'</em>, <em class="sig-param">'MaeIII'</em>, <em class="sig-param">'MalI'</em>, <em class="sig-param">'MauBI'</em>, <em class="sig-param">'MbiI'</em>, <em class="sig-param">'MboI'</em>, <em class="sig-param">'MboII'</em>, <em class="sig-param">'MfeI'</em>, <em class="sig-param">'MflI'</em>, <em class="sig-param">'MhlI'</em>, <em class="sig-param">'MlsI'</em>, <em class="sig-param">'MluCI'</em>, <em class="sig-param">'MluI'</em>, <em class="sig-param">'MluNI'</em>, <em class="sig-param">'Mly113I'</em>, <em class="sig-param">'MlyI'</em>, <em class="sig-param">'MmeI'</em>, <em class="sig-param">'MnlI'</em>, <em class="sig-param">'Mox20I'</em>, <em class="sig-param">'Mph1103I'</em>, <em class="sig-param">'MreI'</em>, <em class="sig-param">'MroI'</em>, <em class="sig-param">'MroNI'</em>, <em class="sig-param">'MroXI'</em>, <em class="sig-param">'MscI'</em>, <em class="sig-param">'MseI'</em>, <em class="sig-param">'MslI'</em>, <em class="sig-param">'Msp20I'</em>, <em class="sig-param">'MspA1I'</em>, <em class="sig-param">'MspCI'</em>, <em class="sig-param">'MspI'</em>, <em class="sig-param">'MspJI'</em>, <em class="sig-param">'MspR9I'</em>, <em class="sig-param">'MssI'</em>, <em class="sig-param">'MteI'</em>, <em class="sig-param">'MunI'</em>, <em class="sig-param">'Mva1269I'</em>, <em class="sig-param">'MvaI'</em>, <em class="sig-param">'MvnI'</em>, <em class="sig-param">'MwoI'</em>, <em class="sig-param">'NaeI'</em>, <em class="sig-param">'NarI'</em>, <em class="sig-param">'NciI'</em>, <em class="sig-param">'NcoI'</em>, <em class="sig-param">'NdeI'</em>, <em class="sig-param">'NdeII'</em>, <em class="sig-param">'NgoMIV'</em>, <em class="sig-param">'NheI'</em>, <em class="sig-param">'NlaIII'</em>, <em class="sig-param">'NlaIV'</em>, <em class="sig-param">'NmeAIII'</em>, <em class="sig-param">'NmuCI'</em>, <em class="sig-param">'NotI'</em>, <em class="sig-param">'NruI'</em>, <em class="sig-param">'NsbI'</em>, <em class="sig-param">'NsiI'</em>, <em class="sig-param">'NspI'</em>, <em class="sig-param">'NspV'</em>, <em class="sig-param">'OliI'</em>, <em class="sig-param">'PacI'</em>, <em class="sig-param">'PaeI'</em>, <em class="sig-param">'PaeR7I'</em>, <em class="sig-param">'PagI'</em>, <em class="sig-param">'PalAI'</em>, <em class="sig-param">'PasI'</em>, <em class="sig-param">'PauI'</em>, <em class="sig-param">'PceI'</em>, <em class="sig-param">'PciI'</em>, <em class="sig-param">'PciSI'</em>, <em class="sig-param">'PcsI'</em>, <em class="sig-param">'PctI'</em>, <em class="sig-param">'PdiI'</em>, <em class="sig-param">'PdmI'</em>, <em class="sig-param">'PfeI'</em>, <em class="sig-param">'Pfl23II'</em>, <em class="sig-param">'PflFI'</em>, <em class="sig-param">'PflMI'</em>, <em class="sig-param">'PfoI'</em>, <em class="sig-param">'PinAI'</em>, <em class="sig-param">'PkrI'</em>, <em class="sig-param">'Ple19I'</em>, <em class="sig-param">'PleI'</em>, <em class="sig-param">'PluTI'</em>, <em class="sig-param">'PmaCI'</em>, <em class="sig-param">'PmeI'</em>, <em class="sig-param">'PmlI'</em>, <em class="sig-param">'PpsI'</em>, <em class="sig-param">'Ppu21I'</em>, <em class="sig-param">'PpuMI'</em>, <em class="sig-param">'PscI'</em>, <em class="sig-param">'PshAI'</em>, <em class="sig-param">'PshBI'</em>, <em class="sig-param">'PsiI'</em>, <em class="sig-param">'Psp124BI'</em>, <em class="sig-param">'Psp1406I'</em>, <em class="sig-param">'Psp5II'</em>, <em class="sig-param">'Psp6I'</em>, <em class="sig-param">'PspCI'</em>, <em class="sig-param">'PspEI'</em>, <em class="sig-param">'PspFI'</em>, <em class="sig-param">'PspGI'</em>, <em class="sig-param">'PspLI'</em>, <em class="sig-param">'PspN4I'</em>, <em class="sig-param">'PspOMI'</em>, <em class="sig-param">'PspPI'</em>, <em class="sig-param">'PspPPI'</em>, <em class="sig-param">'PspXI'</em>, <em class="sig-param">'PsrI'</em>, <em class="sig-param">'PstI'</em>, <em class="sig-param">'PstNI'</em>, <em class="sig-param">'PsuI'</em>, <em class="sig-param">'PsyI'</em>, <em class="sig-param">'PteI'</em>, <em class="sig-param">'PvuI'</em>, <em class="sig-param">'PvuII'</em>, <em class="sig-param">'RgaI'</em>, <em class="sig-param">'RigI'</em>, <em class="sig-param">'RruI'</em>, <em class="sig-param">'RsaI'</em>, <em class="sig-param">'RsaNI'</em>, <em class="sig-param">'RseI'</em>, <em class="sig-param">'Rsr2I'</em>, <em class="sig-param">'RsrII'</em>, <em class="sig-param">'SacI'</em>, <em class="sig-param">'SacII'</em>, <em class="sig-param">'SalI'</em>, <em class="sig-param">'SapI'</em>, <em class="sig-param">'SaqAI'</em>, <em class="sig-param">'SatI'</em>, <em class="sig-param">'Sau3AI'</em>, <em class="sig-param">'Sau96I'</em>, <em class="sig-param">'SbfI'</em>, <em class="sig-param">'ScaI'</em>, <em class="sig-param">'SchI'</em>, <em class="sig-param">'ScrFI'</em>, <em class="sig-param">'SdaI'</em>, <em class="sig-param">'SduI'</em>, <em class="sig-param">'SetI'</em>, <em class="sig-param">'SexAI'</em>, <em class="sig-param">'SfaAI'</em>, <em class="sig-param">'SfaNI'</em>, <em class="sig-param">'SfcI'</em>, <em class="sig-param">'SfiI'</em>, <em class="sig-param">'SfoI'</em>, <em class="sig-param">'Sfr274I'</em>, <em class="sig-param">'Sfr303I'</em>, <em class="sig-param">'SfuI'</em>, <em class="sig-param">'SgeI'</em>, <em class="sig-param">'SgfI'</em>, <em class="sig-param">'SgrAI'</em>, <em class="sig-param">'SgrBI'</em>, <em class="sig-param">'SgrDI'</em>, <em class="sig-param">'SgsI'</em>, <em class="sig-param">'SinI'</em>, <em class="sig-param">'SlaI'</em>, <em class="sig-param">'SmaI'</em>, <em class="sig-param">'SmiI'</em>, <em class="sig-param">'SmiMI'</em>, <em class="sig-param">'SmlI'</em>, <em class="sig-param">'SmoI'</em>, <em class="sig-param">'SnaBI'</em>, <em class="sig-param">'SpeI'</em>, <em class="sig-param">'SphI'</em>, <em class="sig-param">'SrfI'</em>, <em class="sig-param">'Sse8387I'</em>, <em class="sig-param">'Sse9I'</em>, <em class="sig-param">'SseBI'</em>, <em class="sig-param">'SsiI'</em>, <em class="sig-param">'SspDI'</em>, <em class="sig-param">'SspI'</em>, <em class="sig-param">'SspMI'</em>, <em class="sig-param">'SstI'</em>, <em class="sig-param">'StuI'</em>, <em class="sig-param">'StyD4I'</em>, <em class="sig-param">'StyI'</em>, <em class="sig-param">'SwaI'</em>, <em class="sig-param">'TaaI'</em>, <em class="sig-param">'TaiI'</em>, <em class="sig-param">'TaqI'</em>, <em class="sig-param">'TaqII'</em>, <em class="sig-param">'TasI'</em>, <em class="sig-param">'TatI'</em>, <em class="sig-param">'TauI'</em>, <em class="sig-param">'TfiI'</em>, <em class="sig-param">'Tru1I'</em>, <em class="sig-param">'Tru9I'</em>, <em class="sig-param">'TscAI'</em>, <em class="sig-param">'TseFI'</em>, <em class="sig-param">'TseI'</em>, <em class="sig-param">'Tsp45I'</em>, <em class="sig-param">'TspDTI'</em>, <em class="sig-param">'TspGWI'</em>, <em class="sig-param">'TspMI'</em>, <em class="sig-param">'TspRI'</em>, <em class="sig-param">'Tth111I'</em>, <em class="sig-param">'Van91I'</em>, <em class="sig-param">'Vha464I'</em>, <em class="sig-param">'VneI'</em>, <em class="sig-param">'VpaK11BI'</em>, <em class="sig-param">'VspI'</em>, <em class="sig-param">'XagI'</em>, <em class="sig-param">'XapI'</em>, <em class="sig-param">'XbaI'</em>, <em class="sig-param">'XceI'</em>, <em class="sig-param">'XcmI'</em>, <em class="sig-param">'XhoI'</em>, <em class="sig-param">'XmaI'</em>, <em class="sig-param">'XmaJI'</em>, <em class="sig-param">'XmiI'</em>, <em class="sig-param">'XmnI'</em>, <em class="sig-param">'XspI'</em>, <em class="sig-param">'ZraI'</em>, <em class="sig-param">'ZrmI'</em>, <em class="sig-param">'Zsp2I'])</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.once_cutters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.once_cutters" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.rc">
<code class="sig-name descname">rc</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reverse complement.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.reverse_complement">
<code class="sig-name descname">reverse_complement</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.reverse_complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reverse complement.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.shifted">
<code class="sig-name descname">shifted</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">shift</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.shifted"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.shifted" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a circular Dseqrecord with a new origin &lt;shift&gt;.
This only works on circular Dseqrecords. If we consider the following
circular sequence:</p>
<div class="line-block">
<div class="line"><code class="docutils literal notranslate"><span class="pre">GAAAT</span>&#160;&#160; <span class="pre">&lt;--</span> <span class="pre">watson</span> <span class="pre">strand</span></code></div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">CTTTA</span>&#160;&#160; <span class="pre">&lt;--</span> <span class="pre">crick</span> <span class="pre">strand</span></code></div>
</div>
<p>The T and the G on the watson strand are linked together as well
as the A and the C of the of the crick strand.</p>
<p>if <code class="docutils literal notranslate"><span class="pre">shift</span></code> is 1, this indicates a new origin at position 1:</p>
<div class="line-block">
<div class="line-block">
<div class="line">new origin at the | symbol:</div>
<div class="line"><br /></div>
</div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">G|AAAT</span></code></div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">C|TTTA</span></code></div>
</div>
<p>new sequence:</p>
<div class="line-block">
<div class="line"><code class="docutils literal notranslate"><span class="pre">AAATG</span></code></div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">TTTAC</span></code></div>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaat&quot;</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(o4)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">aaat</span>
<span class="go">ttta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">shifted</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(o4)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">aata</span>
<span class="go">ttat</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.stamp">
<code class="sig-name descname">stamp</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.stamp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.stamp" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a SEGUID or cSEGUID checksum and a datestring to the description property.
This will show in the genbank format.</p>
<p>For linear sequences:</p>
<p><code class="docutils literal notranslate"><span class="pre">SEGUID_&lt;seguid&gt;_&lt;datestring&gt;</span></code></p>
<p>For circular sequences:</p>
<p><code class="docutils literal notranslate"><span class="pre">cSEGUID_&lt;seguid&gt;_&lt;datestring&gt;</span></code></p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord.verify_stamp</span></code></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">stamp</span><span class="p">()</span>
<span class="go">&#39;SEGUID_YG7G6b2Kj_KtFOX63j8mRHHoIlE...&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">description</span>
<span class="go">&#39;SEGUID_YG7G6b2Kj_KtFOX63j8mRHHoIlE...&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.synced">
<code class="sig-name descname">synced</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.synced" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.tolinear">
<code class="sig-name descname">tolinear</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.tolinear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.tolinear" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a linear, blunt copy of a circular Dseqrecord object. The
underlying Dseq object has to be circular.</p>
<p>This method is deprecated, use slicing instead. See example below.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(o3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="p">[:]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-3)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.twice_cutters">
<code class="sig-name descname">twice_cutters</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">batch=RestrictionBatch(['AanI'</em>, <em class="sig-param">'AarI'</em>, <em class="sig-param">'AasI'</em>, <em class="sig-param">'AatII'</em>, <em class="sig-param">'AbaSI'</em>, <em class="sig-param">'AbsI'</em>, <em class="sig-param">'Acc16I'</em>, <em class="sig-param">'Acc36I'</em>, <em class="sig-param">'Acc65I'</em>, <em class="sig-param">'AccB1I'</em>, <em class="sig-param">'AccB7I'</em>, <em class="sig-param">'AccBSI'</em>, <em class="sig-param">'AccI'</em>, <em class="sig-param">'AccII'</em>, <em class="sig-param">'AccIII'</em>, <em class="sig-param">'AciI'</em>, <em class="sig-param">'AclI'</em>, <em class="sig-param">'AclWI'</em>, <em class="sig-param">'AcoI'</em>, <em class="sig-param">'AcsI'</em>, <em class="sig-param">'AcuI'</em>, <em class="sig-param">'AcvI'</em>, <em class="sig-param">'AcyI'</em>, <em class="sig-param">'AdeI'</em>, <em class="sig-param">'AfaI'</em>, <em class="sig-param">'AfeI'</em>, <em class="sig-param">'AfiI'</em>, <em class="sig-param">'AflII'</em>, <em class="sig-param">'AflIII'</em>, <em class="sig-param">'AgeI'</em>, <em class="sig-param">'AgsI'</em>, <em class="sig-param">'AhdI'</em>, <em class="sig-param">'AhlI'</em>, <em class="sig-param">'AjiI'</em>, <em class="sig-param">'AjnI'</em>, <em class="sig-param">'AjuI'</em>, <em class="sig-param">'AleI'</em>, <em class="sig-param">'AloI'</em>, <em class="sig-param">'AluBI'</em>, <em class="sig-param">'AluI'</em>, <em class="sig-param">'Alw21I'</em>, <em class="sig-param">'Alw26I'</em>, <em class="sig-param">'Alw44I'</em>, <em class="sig-param">'AlwI'</em>, <em class="sig-param">'AlwNI'</em>, <em class="sig-param">'Ama87I'</em>, <em class="sig-param">'Aor13HI'</em>, <em class="sig-param">'Aor51HI'</em>, <em class="sig-param">'AoxI'</em>, <em class="sig-param">'ApaI'</em>, <em class="sig-param">'ApaLI'</em>, <em class="sig-param">'ApeKI'</em>, <em class="sig-param">'ApoI'</em>, <em class="sig-param">'ArsI'</em>, <em class="sig-param">'AscI'</em>, <em class="sig-param">'AseI'</em>, <em class="sig-param">'AsiGI'</em>, <em class="sig-param">'AsiSI'</em>, <em class="sig-param">'Asp700I'</em>, <em class="sig-param">'Asp718I'</em>, <em class="sig-param">'AspA2I'</em>, <em class="sig-param">'AspLEI'</em>, <em class="sig-param">'AspS9I'</em>, <em class="sig-param">'AsuC2I'</em>, <em class="sig-param">'AsuHPI'</em>, <em class="sig-param">'AsuII'</em>, <em class="sig-param">'AsuNHI'</em>, <em class="sig-param">'AvaI'</em>, <em class="sig-param">'AvaII'</em>, <em class="sig-param">'AvrII'</em>, <em class="sig-param">'AxyI'</em>, <em class="sig-param">'BaeGI'</em>, <em class="sig-param">'BaeI'</em>, <em class="sig-param">'BalI'</em>, <em class="sig-param">'BamHI'</em>, <em class="sig-param">'BanI'</em>, <em class="sig-param">'BanII'</em>, <em class="sig-param">'BarI'</em>, <em class="sig-param">'BauI'</em>, <em class="sig-param">'BbrPI'</em>, <em class="sig-param">'BbsI'</em>, <em class="sig-param">'Bbv12I'</em>, <em class="sig-param">'BbvCI'</em>, <em class="sig-param">'BbvI'</em>, <em class="sig-param">'BccI'</em>, <em class="sig-param">'BceAI'</em>, <em class="sig-param">'BcgI'</em>, <em class="sig-param">'BciT130I'</em>, <em class="sig-param">'BciVI'</em>, <em class="sig-param">'BclI'</em>, <em class="sig-param">'BcnI'</em>, <em class="sig-param">'BcoDI'</em>, <em class="sig-param">'BcuI'</em>, <em class="sig-param">'BfaI'</em>, <em class="sig-param">'BfmI'</em>, <em class="sig-param">'BfoI'</em>, <em class="sig-param">'BfrI'</em>, <em class="sig-param">'BfuAI'</em>, <em class="sig-param">'BfuI'</em>, <em class="sig-param">'BglI'</em>, <em class="sig-param">'BglII'</em>, <em class="sig-param">'BisI'</em>, <em class="sig-param">'BlnI'</em>, <em class="sig-param">'BlpI'</em>, <em class="sig-param">'BlsI'</em>, <em class="sig-param">'BmcAI'</em>, <em class="sig-param">'Bme1390I'</em>, <em class="sig-param">'Bme18I'</em>, <em class="sig-param">'BmeRI'</em>, <em class="sig-param">'BmeT110I'</em>, <em class="sig-param">'BmgBI'</em>, <em class="sig-param">'BmgT120I'</em>, <em class="sig-param">'BmiI'</em>, <em class="sig-param">'BmrFI'</em>, <em class="sig-param">'BmrI'</em>, <em class="sig-param">'BmsI'</em>, <em class="sig-param">'BmtI'</em>, <em class="sig-param">'BmuI'</em>, <em class="sig-param">'BoxI'</em>, <em class="sig-param">'BpiI'</em>, <em class="sig-param">'BplI'</em>, <em class="sig-param">'BpmI'</em>, <em class="sig-param">'Bpu10I'</em>, <em class="sig-param">'Bpu1102I'</em>, <em class="sig-param">'Bpu14I'</em>, <em class="sig-param">'BpuEI'</em>, <em class="sig-param">'BpuMI'</em>, <em class="sig-param">'Bsa29I'</em>, <em class="sig-param">'BsaAI'</em>, <em class="sig-param">'BsaBI'</em>, <em class="sig-param">'BsaHI'</em>, <em class="sig-param">'BsaI'</em>, <em class="sig-param">'BsaJI'</em>, <em class="sig-param">'BsaWI'</em>, <em class="sig-param">'BsaXI'</em>, <em class="sig-param">'Bsc4I'</em>, <em class="sig-param">'Bse118I'</em>, <em class="sig-param">'Bse1I'</em>, <em class="sig-param">'Bse21I'</em>, <em class="sig-param">'Bse3DI'</em>, <em class="sig-param">'Bse8I'</em>, <em class="sig-param">'BseAI'</em>, <em class="sig-param">'BseBI'</em>, <em class="sig-param">'BseCI'</em>, <em class="sig-param">'BseDI'</em>, <em class="sig-param">'BseGI'</em>, <em class="sig-param">'BseJI'</em>, <em class="sig-param">'BseLI'</em>, <em class="sig-param">'BseMI'</em>, <em class="sig-param">'BseMII'</em>, <em class="sig-param">'BseNI'</em>, <em class="sig-param">'BsePI'</em>, <em class="sig-param">'BseRI'</em>, <em class="sig-param">'BseSI'</em>, <em class="sig-param">'BseX3I'</em>, <em class="sig-param">'BseXI'</em>, <em class="sig-param">'BseYI'</em>, <em class="sig-param">'BsgI'</em>, <em class="sig-param">'Bsh1236I'</em>, <em class="sig-param">'Bsh1285I'</em>, <em class="sig-param">'BshFI'</em>, <em class="sig-param">'BshNI'</em>, <em class="sig-param">'BshTI'</em>, <em class="sig-param">'BshVI'</em>, <em class="sig-param">'BsiEI'</em>, <em class="sig-param">'BsiHKAI'</em>, <em class="sig-param">'BsiHKCI'</em>, <em class="sig-param">'BsiSI'</em>, <em class="sig-param">'BsiWI'</em>, <em class="sig-param">'BslFI'</em>, <em class="sig-param">'BslI'</em>, <em class="sig-param">'BsmAI'</em>, <em class="sig-param">'BsmBI'</em>, <em class="sig-param">'BsmFI'</em>, <em class="sig-param">'BsmI'</em>, <em class="sig-param">'BsnI'</em>, <em class="sig-param">'Bso31I'</em>, <em class="sig-param">'BsoBI'</em>, <em class="sig-param">'Bsp119I'</em>, <em class="sig-param">'Bsp120I'</em>, <em class="sig-param">'Bsp1286I'</em>, <em class="sig-param">'Bsp13I'</em>, <em class="sig-param">'Bsp1407I'</em>, <em class="sig-param">'Bsp143I'</em>, <em class="sig-param">'Bsp1720I'</em>, <em class="sig-param">'Bsp19I'</em>, <em class="sig-param">'Bsp68I'</em>, <em class="sig-param">'BspACI'</em>, <em class="sig-param">'BspANI'</em>, <em class="sig-param">'BspCNI'</em>, <em class="sig-param">'BspDI'</em>, <em class="sig-param">'BspEI'</em>, <em class="sig-param">'BspFNI'</em>, <em class="sig-param">'BspHI'</em>, <em class="sig-param">'BspLI'</em>, <em class="sig-param">'BspMAI'</em>, <em class="sig-param">'BspMI'</em>, <em class="sig-param">'BspOI'</em>, <em class="sig-param">'BspPI'</em>, <em class="sig-param">'BspQI'</em>, <em class="sig-param">'BspT104I'</em>, <em class="sig-param">'BspT107I'</em>, <em class="sig-param">'BspTI'</em>, <em class="sig-param">'BspTNI'</em>, <em class="sig-param">'BsrBI'</em>, <em class="sig-param">'BsrDI'</em>, <em class="sig-param">'BsrFI'</em>, <em class="sig-param">'BsrGI'</em>, <em class="sig-param">'BsrI'</em>, <em class="sig-param">'BssAI'</em>, <em class="sig-param">'BssECI'</em>, <em class="sig-param">'BssHII'</em>, <em class="sig-param">'BssMI'</em>, <em class="sig-param">'BssNAI'</em>, <em class="sig-param">'BssNI'</em>, <em class="sig-param">'BssSI'</em>, <em class="sig-param">'BssT1I'</em>, <em class="sig-param">'Bst1107I'</em>, <em class="sig-param">'Bst2BI'</em>, <em class="sig-param">'Bst2UI'</em>, <em class="sig-param">'Bst4CI'</em>, <em class="sig-param">'Bst6I'</em>, <em class="sig-param">'BstACI'</em>, <em class="sig-param">'BstAFI'</em>, <em class="sig-param">'BstAPI'</em>, <em class="sig-param">'BstAUI'</em>, <em class="sig-param">'BstBAI'</em>, <em class="sig-param">'BstBI'</em>, <em class="sig-param">'BstC8I'</em>, <em class="sig-param">'BstDEI'</em>, <em class="sig-param">'BstDSI'</em>, <em class="sig-param">'BstEII'</em>, <em class="sig-param">'BstENI'</em>, <em class="sig-param">'BstF5I'</em>, <em class="sig-param">'BstFNI'</em>, <em class="sig-param">'BstH2I'</em>, <em class="sig-param">'BstHHI'</em>, <em class="sig-param">'BstKTI'</em>, <em class="sig-param">'BstMAI'</em>, <em class="sig-param">'BstMBI'</em>, <em class="sig-param">'BstMCI'</em>, <em class="sig-param">'BstMWI'</em>, <em class="sig-param">'BstNI'</em>, <em class="sig-param">'BstNSI'</em>, <em class="sig-param">'BstPAI'</em>, <em class="sig-param">'BstPI'</em>, <em class="sig-param">'BstSCI'</em>, <em class="sig-param">'BstSFI'</em>, <em class="sig-param">'BstSLI'</em>, <em class="sig-param">'BstSNI'</em>, <em class="sig-param">'BstUI'</em>, <em class="sig-param">'BstV1I'</em>, <em class="sig-param">'BstV2I'</em>, <em class="sig-param">'BstX2I'</em>, <em class="sig-param">'BstXI'</em>, <em class="sig-param">'BstYI'</em>, <em class="sig-param">'BstZ17I'</em>, <em class="sig-param">'BstZI'</em>, <em class="sig-param">'Bsu15I'</em>, <em class="sig-param">'Bsu36I'</em>, <em class="sig-param">'BsuI'</em>, <em class="sig-param">'BsuRI'</em>, <em class="sig-param">'BsuTUI'</em>, <em class="sig-param">'BtgI'</em>, <em class="sig-param">'BtgZI'</em>, <em class="sig-param">'BtrI'</em>, <em class="sig-param">'BtsCI'</em>, <em class="sig-param">'BtsI'</em>, <em class="sig-param">'BtsIMutI'</em>, <em class="sig-param">'BtuMI'</em>, <em class="sig-param">'BveI'</em>, <em class="sig-param">'Cac8I'</em>, <em class="sig-param">'CaiI'</em>, <em class="sig-param">'CciI'</em>, <em class="sig-param">'CciNI'</em>, <em class="sig-param">'CfoI'</em>, <em class="sig-param">'Cfr10I'</em>, <em class="sig-param">'Cfr13I'</em>, <em class="sig-param">'Cfr42I'</em>, <em class="sig-param">'Cfr9I'</em>, <em class="sig-param">'ClaI'</em>, <em class="sig-param">'CpoI'</em>, <em class="sig-param">'CseI'</em>, <em class="sig-param">'CsiI'</em>, <em class="sig-param">'Csp6I'</em>, <em class="sig-param">'CspAI'</em>, <em class="sig-param">'CspCI'</em>, <em class="sig-param">'CspI'</em>, <em class="sig-param">'CviAII'</em>, <em class="sig-param">'CviJI'</em>, <em class="sig-param">'CviKI_1'</em>, <em class="sig-param">'CviQI'</em>, <em class="sig-param">'DdeI'</em>, <em class="sig-param">'DinI'</em>, <em class="sig-param">'DpnI'</em>, <em class="sig-param">'DpnII'</em>, <em class="sig-param">'DraI'</em>, <em class="sig-param">'DraIII'</em>, <em class="sig-param">'DrdI'</em>, <em class="sig-param">'DriI'</em>, <em class="sig-param">'DseDI'</em>, <em class="sig-param">'EaeI'</em>, <em class="sig-param">'EagI'</em>, <em class="sig-param">'Eam1104I'</em>, <em class="sig-param">'Eam1105I'</em>, <em class="sig-param">'EarI'</em>, <em class="sig-param">'EciI'</em>, <em class="sig-param">'Ecl136II'</em>, <em class="sig-param">'EclXI'</em>, <em class="sig-param">'Eco105I'</em>, <em class="sig-param">'Eco130I'</em>, <em class="sig-param">'Eco147I'</em>, <em class="sig-param">'Eco24I'</em>, <em class="sig-param">'Eco31I'</em>, <em class="sig-param">'Eco32I'</em>, <em class="sig-param">'Eco47I'</em>, <em class="sig-param">'Eco47III'</em>, <em class="sig-param">'Eco52I'</em>, <em class="sig-param">'Eco53kI'</em>, <em class="sig-param">'Eco57I'</em>, <em class="sig-param">'Eco72I'</em>, <em class="sig-param">'Eco81I'</em>, <em class="sig-param">'Eco88I'</em>, <em class="sig-param">'Eco91I'</em>, <em class="sig-param">'EcoICRI'</em>, <em class="sig-param">'EcoNI'</em>, <em class="sig-param">'EcoO109I'</em>, <em class="sig-param">'EcoO65I'</em>, <em class="sig-param">'EcoRI'</em>, <em class="sig-param">'EcoRII'</em>, <em class="sig-param">'EcoRV'</em>, <em class="sig-param">'EcoT14I'</em>, <em class="sig-param">'EcoT22I'</em>, <em class="sig-param">'EcoT38I'</em>, <em class="sig-param">'EgeI'</em>, <em class="sig-param">'EheI'</em>, <em class="sig-param">'ErhI'</em>, <em class="sig-param">'Esp3I'</em>, <em class="sig-param">'FaeI'</em>, <em class="sig-param">'FaiI'</em>, <em class="sig-param">'FalI'</em>, <em class="sig-param">'FaqI'</em>, <em class="sig-param">'FatI'</em>, <em class="sig-param">'FauI'</em>, <em class="sig-param">'FauNDI'</em>, <em class="sig-param">'FbaI'</em>, <em class="sig-param">'FblI'</em>, <em class="sig-param">'Fnu4HI'</em>, <em class="sig-param">'FokI'</em>, <em class="sig-param">'FriOI'</em>, <em class="sig-param">'FseI'</em>, <em class="sig-param">'Fsp4HI'</em>, <em class="sig-param">'FspAI'</em>, <em class="sig-param">'FspBI'</em>, <em class="sig-param">'FspEI'</em>, <em class="sig-param">'FspI'</em>, <em class="sig-param">'GlaI'</em>, <em class="sig-param">'GluI'</em>, <em class="sig-param">'GsaI'</em>, <em class="sig-param">'GsuI'</em>, <em class="sig-param">'HaeII'</em>, <em class="sig-param">'HaeIII'</em>, <em class="sig-param">'HapII'</em>, <em class="sig-param">'HgaI'</em>, <em class="sig-param">'HhaI'</em>, <em class="sig-param">'Hin1I'</em>, <em class="sig-param">'Hin1II'</em>, <em class="sig-param">'Hin6I'</em>, <em class="sig-param">'HinP1I'</em>, <em class="sig-param">'HincII'</em>, <em class="sig-param">'HindII'</em>, <em class="sig-param">'HindIII'</em>, <em class="sig-param">'HinfI'</em>, <em class="sig-param">'HpaI'</em>, <em class="sig-param">'HpaII'</em>, <em class="sig-param">'HphI'</em>, <em class="sig-param">'Hpy166II'</em>, <em class="sig-param">'Hpy188I'</em>, <em class="sig-param">'Hpy188III'</em>, <em class="sig-param">'Hpy8I'</em>, <em class="sig-param">'Hpy99I'</em>, <em class="sig-param">'HpyAV'</em>, <em class="sig-param">'HpyCH4III'</em>, <em class="sig-param">'HpyCH4IV'</em>, <em class="sig-param">'HpyCH4V'</em>, <em class="sig-param">'HpyF10VI'</em>, <em class="sig-param">'HpyF3I'</em>, <em class="sig-param">'HpySE526I'</em>, <em class="sig-param">'Hsp92I'</em>, <em class="sig-param">'Hsp92II'</em>, <em class="sig-param">'HspAI'</em>, <em class="sig-param">'KasI'</em>, <em class="sig-param">'KflI'</em>, <em class="sig-param">'Kpn2I'</em>, <em class="sig-param">'KpnI'</em>, <em class="sig-param">'KroI'</em>, <em class="sig-param">'Ksp22I'</em>, <em class="sig-param">'KspAI'</em>, <em class="sig-param">'KspI'</em>, <em class="sig-param">'Kzo9I'</em>, <em class="sig-param">'LguI'</em>, <em class="sig-param">'LmnI'</em>, <em class="sig-param">'LpnPI'</em>, <em class="sig-param">'Lsp1109I'</em>, <em class="sig-param">'LweI'</em>, <em class="sig-param">'MabI'</em>, <em class="sig-param">'MaeI'</em>, <em class="sig-param">'MaeII'</em>, <em class="sig-param">'MaeIII'</em>, <em class="sig-param">'MalI'</em>, <em class="sig-param">'MauBI'</em>, <em class="sig-param">'MbiI'</em>, <em class="sig-param">'MboI'</em>, <em class="sig-param">'MboII'</em>, <em class="sig-param">'MfeI'</em>, <em class="sig-param">'MflI'</em>, <em class="sig-param">'MhlI'</em>, <em class="sig-param">'MlsI'</em>, <em class="sig-param">'MluCI'</em>, <em class="sig-param">'MluI'</em>, <em class="sig-param">'MluNI'</em>, <em class="sig-param">'Mly113I'</em>, <em class="sig-param">'MlyI'</em>, <em class="sig-param">'MmeI'</em>, <em class="sig-param">'MnlI'</em>, <em class="sig-param">'Mox20I'</em>, <em class="sig-param">'Mph1103I'</em>, <em class="sig-param">'MreI'</em>, <em class="sig-param">'MroI'</em>, <em class="sig-param">'MroNI'</em>, <em class="sig-param">'MroXI'</em>, <em class="sig-param">'MscI'</em>, <em class="sig-param">'MseI'</em>, <em class="sig-param">'MslI'</em>, <em class="sig-param">'Msp20I'</em>, <em class="sig-param">'MspA1I'</em>, <em class="sig-param">'MspCI'</em>, <em class="sig-param">'MspI'</em>, <em class="sig-param">'MspJI'</em>, <em class="sig-param">'MspR9I'</em>, <em class="sig-param">'MssI'</em>, <em class="sig-param">'MteI'</em>, <em class="sig-param">'MunI'</em>, <em class="sig-param">'Mva1269I'</em>, <em class="sig-param">'MvaI'</em>, <em class="sig-param">'MvnI'</em>, <em class="sig-param">'MwoI'</em>, <em class="sig-param">'NaeI'</em>, <em class="sig-param">'NarI'</em>, <em class="sig-param">'NciI'</em>, <em class="sig-param">'NcoI'</em>, <em class="sig-param">'NdeI'</em>, <em class="sig-param">'NdeII'</em>, <em class="sig-param">'NgoMIV'</em>, <em class="sig-param">'NheI'</em>, <em class="sig-param">'NlaIII'</em>, <em class="sig-param">'NlaIV'</em>, <em class="sig-param">'NmeAIII'</em>, <em class="sig-param">'NmuCI'</em>, <em class="sig-param">'NotI'</em>, <em class="sig-param">'NruI'</em>, <em class="sig-param">'NsbI'</em>, <em class="sig-param">'NsiI'</em>, <em class="sig-param">'NspI'</em>, <em class="sig-param">'NspV'</em>, <em class="sig-param">'OliI'</em>, <em class="sig-param">'PacI'</em>, <em class="sig-param">'PaeI'</em>, <em class="sig-param">'PaeR7I'</em>, <em class="sig-param">'PagI'</em>, <em class="sig-param">'PalAI'</em>, <em class="sig-param">'PasI'</em>, <em class="sig-param">'PauI'</em>, <em class="sig-param">'PceI'</em>, <em class="sig-param">'PciI'</em>, <em class="sig-param">'PciSI'</em>, <em class="sig-param">'PcsI'</em>, <em class="sig-param">'PctI'</em>, <em class="sig-param">'PdiI'</em>, <em class="sig-param">'PdmI'</em>, <em class="sig-param">'PfeI'</em>, <em class="sig-param">'Pfl23II'</em>, <em class="sig-param">'PflFI'</em>, <em class="sig-param">'PflMI'</em>, <em class="sig-param">'PfoI'</em>, <em class="sig-param">'PinAI'</em>, <em class="sig-param">'PkrI'</em>, <em class="sig-param">'Ple19I'</em>, <em class="sig-param">'PleI'</em>, <em class="sig-param">'PluTI'</em>, <em class="sig-param">'PmaCI'</em>, <em class="sig-param">'PmeI'</em>, <em class="sig-param">'PmlI'</em>, <em class="sig-param">'PpsI'</em>, <em class="sig-param">'Ppu21I'</em>, <em class="sig-param">'PpuMI'</em>, <em class="sig-param">'PscI'</em>, <em class="sig-param">'PshAI'</em>, <em class="sig-param">'PshBI'</em>, <em class="sig-param">'PsiI'</em>, <em class="sig-param">'Psp124BI'</em>, <em class="sig-param">'Psp1406I'</em>, <em class="sig-param">'Psp5II'</em>, <em class="sig-param">'Psp6I'</em>, <em class="sig-param">'PspCI'</em>, <em class="sig-param">'PspEI'</em>, <em class="sig-param">'PspFI'</em>, <em class="sig-param">'PspGI'</em>, <em class="sig-param">'PspLI'</em>, <em class="sig-param">'PspN4I'</em>, <em class="sig-param">'PspOMI'</em>, <em class="sig-param">'PspPI'</em>, <em class="sig-param">'PspPPI'</em>, <em class="sig-param">'PspXI'</em>, <em class="sig-param">'PsrI'</em>, <em class="sig-param">'PstI'</em>, <em class="sig-param">'PstNI'</em>, <em class="sig-param">'PsuI'</em>, <em class="sig-param">'PsyI'</em>, <em class="sig-param">'PteI'</em>, <em class="sig-param">'PvuI'</em>, <em class="sig-param">'PvuII'</em>, <em class="sig-param">'RgaI'</em>, <em class="sig-param">'RigI'</em>, <em class="sig-param">'RruI'</em>, <em class="sig-param">'RsaI'</em>, <em class="sig-param">'RsaNI'</em>, <em class="sig-param">'RseI'</em>, <em class="sig-param">'Rsr2I'</em>, <em class="sig-param">'RsrII'</em>, <em class="sig-param">'SacI'</em>, <em class="sig-param">'SacII'</em>, <em class="sig-param">'SalI'</em>, <em class="sig-param">'SapI'</em>, <em class="sig-param">'SaqAI'</em>, <em class="sig-param">'SatI'</em>, <em class="sig-param">'Sau3AI'</em>, <em class="sig-param">'Sau96I'</em>, <em class="sig-param">'SbfI'</em>, <em class="sig-param">'ScaI'</em>, <em class="sig-param">'SchI'</em>, <em class="sig-param">'ScrFI'</em>, <em class="sig-param">'SdaI'</em>, <em class="sig-param">'SduI'</em>, <em class="sig-param">'SetI'</em>, <em class="sig-param">'SexAI'</em>, <em class="sig-param">'SfaAI'</em>, <em class="sig-param">'SfaNI'</em>, <em class="sig-param">'SfcI'</em>, <em class="sig-param">'SfiI'</em>, <em class="sig-param">'SfoI'</em>, <em class="sig-param">'Sfr274I'</em>, <em class="sig-param">'Sfr303I'</em>, <em class="sig-param">'SfuI'</em>, <em class="sig-param">'SgeI'</em>, <em class="sig-param">'SgfI'</em>, <em class="sig-param">'SgrAI'</em>, <em class="sig-param">'SgrBI'</em>, <em class="sig-param">'SgrDI'</em>, <em class="sig-param">'SgsI'</em>, <em class="sig-param">'SinI'</em>, <em class="sig-param">'SlaI'</em>, <em class="sig-param">'SmaI'</em>, <em class="sig-param">'SmiI'</em>, <em class="sig-param">'SmiMI'</em>, <em class="sig-param">'SmlI'</em>, <em class="sig-param">'SmoI'</em>, <em class="sig-param">'SnaBI'</em>, <em class="sig-param">'SpeI'</em>, <em class="sig-param">'SphI'</em>, <em class="sig-param">'SrfI'</em>, <em class="sig-param">'Sse8387I'</em>, <em class="sig-param">'Sse9I'</em>, <em class="sig-param">'SseBI'</em>, <em class="sig-param">'SsiI'</em>, <em class="sig-param">'SspDI'</em>, <em class="sig-param">'SspI'</em>, <em class="sig-param">'SspMI'</em>, <em class="sig-param">'SstI'</em>, <em class="sig-param">'StuI'</em>, <em class="sig-param">'StyD4I'</em>, <em class="sig-param">'StyI'</em>, <em class="sig-param">'SwaI'</em>, <em class="sig-param">'TaaI'</em>, <em class="sig-param">'TaiI'</em>, <em class="sig-param">'TaqI'</em>, <em class="sig-param">'TaqII'</em>, <em class="sig-param">'TasI'</em>, <em class="sig-param">'TatI'</em>, <em class="sig-param">'TauI'</em>, <em class="sig-param">'TfiI'</em>, <em class="sig-param">'Tru1I'</em>, <em class="sig-param">'Tru9I'</em>, <em class="sig-param">'TscAI'</em>, <em class="sig-param">'TseFI'</em>, <em class="sig-param">'TseI'</em>, <em class="sig-param">'Tsp45I'</em>, <em class="sig-param">'TspDTI'</em>, <em class="sig-param">'TspGWI'</em>, <em class="sig-param">'TspMI'</em>, <em class="sig-param">'TspRI'</em>, <em class="sig-param">'Tth111I'</em>, <em class="sig-param">'Van91I'</em>, <em class="sig-param">'Vha464I'</em>, <em class="sig-param">'VneI'</em>, <em class="sig-param">'VpaK11BI'</em>, <em class="sig-param">'VspI'</em>, <em class="sig-param">'XagI'</em>, <em class="sig-param">'XapI'</em>, <em class="sig-param">'XbaI'</em>, <em class="sig-param">'XceI'</em>, <em class="sig-param">'XcmI'</em>, <em class="sig-param">'XhoI'</em>, <em class="sig-param">'XmaI'</em>, <em class="sig-param">'XmaJI'</em>, <em class="sig-param">'XmiI'</em>, <em class="sig-param">'XmnI'</em>, <em class="sig-param">'XspI'</em>, <em class="sig-param">'ZraI'</em>, <em class="sig-param">'ZrmI'</em>, <em class="sig-param">'Zsp2I'])</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.twice_cutters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.twice_cutters" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.unique_cutters">
<code class="sig-name descname">unique_cutters</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">batch=RestrictionBatch(['AanI'</em>, <em class="sig-param">'AarI'</em>, <em class="sig-param">'AasI'</em>, <em class="sig-param">'AatII'</em>, <em class="sig-param">'AbaSI'</em>, <em class="sig-param">'AbsI'</em>, <em class="sig-param">'Acc16I'</em>, <em class="sig-param">'Acc36I'</em>, <em class="sig-param">'Acc65I'</em>, <em class="sig-param">'AccB1I'</em>, <em class="sig-param">'AccB7I'</em>, <em class="sig-param">'AccBSI'</em>, <em class="sig-param">'AccI'</em>, <em class="sig-param">'AccII'</em>, <em class="sig-param">'AccIII'</em>, <em class="sig-param">'AciI'</em>, <em class="sig-param">'AclI'</em>, <em class="sig-param">'AclWI'</em>, <em class="sig-param">'AcoI'</em>, <em class="sig-param">'AcsI'</em>, <em class="sig-param">'AcuI'</em>, <em class="sig-param">'AcvI'</em>, <em class="sig-param">'AcyI'</em>, <em class="sig-param">'AdeI'</em>, <em class="sig-param">'AfaI'</em>, <em class="sig-param">'AfeI'</em>, <em class="sig-param">'AfiI'</em>, <em class="sig-param">'AflII'</em>, <em class="sig-param">'AflIII'</em>, <em class="sig-param">'AgeI'</em>, <em class="sig-param">'AgsI'</em>, <em class="sig-param">'AhdI'</em>, <em class="sig-param">'AhlI'</em>, <em class="sig-param">'AjiI'</em>, <em class="sig-param">'AjnI'</em>, <em class="sig-param">'AjuI'</em>, <em class="sig-param">'AleI'</em>, <em class="sig-param">'AloI'</em>, <em class="sig-param">'AluBI'</em>, <em class="sig-param">'AluI'</em>, <em class="sig-param">'Alw21I'</em>, <em class="sig-param">'Alw26I'</em>, <em class="sig-param">'Alw44I'</em>, <em class="sig-param">'AlwI'</em>, <em class="sig-param">'AlwNI'</em>, <em class="sig-param">'Ama87I'</em>, <em class="sig-param">'Aor13HI'</em>, <em class="sig-param">'Aor51HI'</em>, <em class="sig-param">'AoxI'</em>, <em class="sig-param">'ApaI'</em>, <em class="sig-param">'ApaLI'</em>, <em class="sig-param">'ApeKI'</em>, <em class="sig-param">'ApoI'</em>, <em class="sig-param">'ArsI'</em>, <em class="sig-param">'AscI'</em>, <em class="sig-param">'AseI'</em>, <em class="sig-param">'AsiGI'</em>, <em class="sig-param">'AsiSI'</em>, <em class="sig-param">'Asp700I'</em>, <em class="sig-param">'Asp718I'</em>, <em class="sig-param">'AspA2I'</em>, <em class="sig-param">'AspLEI'</em>, <em class="sig-param">'AspS9I'</em>, <em class="sig-param">'AsuC2I'</em>, <em class="sig-param">'AsuHPI'</em>, <em class="sig-param">'AsuII'</em>, <em class="sig-param">'AsuNHI'</em>, <em class="sig-param">'AvaI'</em>, <em class="sig-param">'AvaII'</em>, <em class="sig-param">'AvrII'</em>, <em class="sig-param">'AxyI'</em>, <em class="sig-param">'BaeGI'</em>, <em class="sig-param">'BaeI'</em>, <em class="sig-param">'BalI'</em>, <em class="sig-param">'BamHI'</em>, <em class="sig-param">'BanI'</em>, <em class="sig-param">'BanII'</em>, <em class="sig-param">'BarI'</em>, <em class="sig-param">'BauI'</em>, <em class="sig-param">'BbrPI'</em>, <em class="sig-param">'BbsI'</em>, <em class="sig-param">'Bbv12I'</em>, <em class="sig-param">'BbvCI'</em>, <em class="sig-param">'BbvI'</em>, <em class="sig-param">'BccI'</em>, <em class="sig-param">'BceAI'</em>, <em class="sig-param">'BcgI'</em>, <em class="sig-param">'BciT130I'</em>, <em class="sig-param">'BciVI'</em>, <em class="sig-param">'BclI'</em>, <em class="sig-param">'BcnI'</em>, <em class="sig-param">'BcoDI'</em>, <em class="sig-param">'BcuI'</em>, <em class="sig-param">'BfaI'</em>, <em class="sig-param">'BfmI'</em>, <em class="sig-param">'BfoI'</em>, <em class="sig-param">'BfrI'</em>, <em class="sig-param">'BfuAI'</em>, <em class="sig-param">'BfuI'</em>, <em class="sig-param">'BglI'</em>, <em class="sig-param">'BglII'</em>, <em class="sig-param">'BisI'</em>, <em class="sig-param">'BlnI'</em>, <em class="sig-param">'BlpI'</em>, <em class="sig-param">'BlsI'</em>, <em class="sig-param">'BmcAI'</em>, <em class="sig-param">'Bme1390I'</em>, <em class="sig-param">'Bme18I'</em>, <em class="sig-param">'BmeRI'</em>, <em class="sig-param">'BmeT110I'</em>, <em class="sig-param">'BmgBI'</em>, <em class="sig-param">'BmgT120I'</em>, <em class="sig-param">'BmiI'</em>, <em class="sig-param">'BmrFI'</em>, <em class="sig-param">'BmrI'</em>, <em class="sig-param">'BmsI'</em>, <em class="sig-param">'BmtI'</em>, <em class="sig-param">'BmuI'</em>, <em class="sig-param">'BoxI'</em>, <em class="sig-param">'BpiI'</em>, <em class="sig-param">'BplI'</em>, <em class="sig-param">'BpmI'</em>, <em class="sig-param">'Bpu10I'</em>, <em class="sig-param">'Bpu1102I'</em>, <em class="sig-param">'Bpu14I'</em>, <em class="sig-param">'BpuEI'</em>, <em class="sig-param">'BpuMI'</em>, <em class="sig-param">'Bsa29I'</em>, <em class="sig-param">'BsaAI'</em>, <em class="sig-param">'BsaBI'</em>, <em class="sig-param">'BsaHI'</em>, <em class="sig-param">'BsaI'</em>, <em class="sig-param">'BsaJI'</em>, <em class="sig-param">'BsaWI'</em>, <em class="sig-param">'BsaXI'</em>, <em class="sig-param">'Bsc4I'</em>, <em class="sig-param">'Bse118I'</em>, <em class="sig-param">'Bse1I'</em>, <em class="sig-param">'Bse21I'</em>, <em class="sig-param">'Bse3DI'</em>, <em class="sig-param">'Bse8I'</em>, <em class="sig-param">'BseAI'</em>, <em class="sig-param">'BseBI'</em>, <em class="sig-param">'BseCI'</em>, <em class="sig-param">'BseDI'</em>, <em class="sig-param">'BseGI'</em>, <em class="sig-param">'BseJI'</em>, <em class="sig-param">'BseLI'</em>, <em class="sig-param">'BseMI'</em>, <em class="sig-param">'BseMII'</em>, <em class="sig-param">'BseNI'</em>, <em class="sig-param">'BsePI'</em>, <em class="sig-param">'BseRI'</em>, <em class="sig-param">'BseSI'</em>, <em class="sig-param">'BseX3I'</em>, <em class="sig-param">'BseXI'</em>, <em class="sig-param">'BseYI'</em>, <em class="sig-param">'BsgI'</em>, <em class="sig-param">'Bsh1236I'</em>, <em class="sig-param">'Bsh1285I'</em>, <em class="sig-param">'BshFI'</em>, <em class="sig-param">'BshNI'</em>, <em class="sig-param">'BshTI'</em>, <em class="sig-param">'BshVI'</em>, <em class="sig-param">'BsiEI'</em>, <em class="sig-param">'BsiHKAI'</em>, <em class="sig-param">'BsiHKCI'</em>, <em class="sig-param">'BsiSI'</em>, <em class="sig-param">'BsiWI'</em>, <em class="sig-param">'BslFI'</em>, <em class="sig-param">'BslI'</em>, <em class="sig-param">'BsmAI'</em>, <em class="sig-param">'BsmBI'</em>, <em class="sig-param">'BsmFI'</em>, <em class="sig-param">'BsmI'</em>, <em class="sig-param">'BsnI'</em>, <em class="sig-param">'Bso31I'</em>, <em class="sig-param">'BsoBI'</em>, <em class="sig-param">'Bsp119I'</em>, <em class="sig-param">'Bsp120I'</em>, <em class="sig-param">'Bsp1286I'</em>, <em class="sig-param">'Bsp13I'</em>, <em class="sig-param">'Bsp1407I'</em>, <em class="sig-param">'Bsp143I'</em>, <em class="sig-param">'Bsp1720I'</em>, <em class="sig-param">'Bsp19I'</em>, <em class="sig-param">'Bsp68I'</em>, <em class="sig-param">'BspACI'</em>, <em class="sig-param">'BspANI'</em>, <em class="sig-param">'BspCNI'</em>, <em class="sig-param">'BspDI'</em>, <em class="sig-param">'BspEI'</em>, <em class="sig-param">'BspFNI'</em>, <em class="sig-param">'BspHI'</em>, <em class="sig-param">'BspLI'</em>, <em class="sig-param">'BspMAI'</em>, <em class="sig-param">'BspMI'</em>, <em class="sig-param">'BspOI'</em>, <em class="sig-param">'BspPI'</em>, <em class="sig-param">'BspQI'</em>, <em class="sig-param">'BspT104I'</em>, <em class="sig-param">'BspT107I'</em>, <em class="sig-param">'BspTI'</em>, <em class="sig-param">'BspTNI'</em>, <em class="sig-param">'BsrBI'</em>, <em class="sig-param">'BsrDI'</em>, <em class="sig-param">'BsrFI'</em>, <em class="sig-param">'BsrGI'</em>, <em class="sig-param">'BsrI'</em>, <em class="sig-param">'BssAI'</em>, <em class="sig-param">'BssECI'</em>, <em class="sig-param">'BssHII'</em>, <em class="sig-param">'BssMI'</em>, <em class="sig-param">'BssNAI'</em>, <em class="sig-param">'BssNI'</em>, <em class="sig-param">'BssSI'</em>, <em class="sig-param">'BssT1I'</em>, <em class="sig-param">'Bst1107I'</em>, <em class="sig-param">'Bst2BI'</em>, <em class="sig-param">'Bst2UI'</em>, <em class="sig-param">'Bst4CI'</em>, <em class="sig-param">'Bst6I'</em>, <em class="sig-param">'BstACI'</em>, <em class="sig-param">'BstAFI'</em>, <em class="sig-param">'BstAPI'</em>, <em class="sig-param">'BstAUI'</em>, <em class="sig-param">'BstBAI'</em>, <em class="sig-param">'BstBI'</em>, <em class="sig-param">'BstC8I'</em>, <em class="sig-param">'BstDEI'</em>, <em class="sig-param">'BstDSI'</em>, <em class="sig-param">'BstEII'</em>, <em class="sig-param">'BstENI'</em>, <em class="sig-param">'BstF5I'</em>, <em class="sig-param">'BstFNI'</em>, <em class="sig-param">'BstH2I'</em>, <em class="sig-param">'BstHHI'</em>, <em class="sig-param">'BstKTI'</em>, <em class="sig-param">'BstMAI'</em>, <em class="sig-param">'BstMBI'</em>, <em class="sig-param">'BstMCI'</em>, <em class="sig-param">'BstMWI'</em>, <em class="sig-param">'BstNI'</em>, <em class="sig-param">'BstNSI'</em>, <em class="sig-param">'BstPAI'</em>, <em class="sig-param">'BstPI'</em>, <em class="sig-param">'BstSCI'</em>, <em class="sig-param">'BstSFI'</em>, <em class="sig-param">'BstSLI'</em>, <em class="sig-param">'BstSNI'</em>, <em class="sig-param">'BstUI'</em>, <em class="sig-param">'BstV1I'</em>, <em class="sig-param">'BstV2I'</em>, <em class="sig-param">'BstX2I'</em>, <em class="sig-param">'BstXI'</em>, <em class="sig-param">'BstYI'</em>, <em class="sig-param">'BstZ17I'</em>, <em class="sig-param">'BstZI'</em>, <em class="sig-param">'Bsu15I'</em>, <em class="sig-param">'Bsu36I'</em>, <em class="sig-param">'BsuI'</em>, <em class="sig-param">'BsuRI'</em>, <em class="sig-param">'BsuTUI'</em>, <em class="sig-param">'BtgI'</em>, <em class="sig-param">'BtgZI'</em>, <em class="sig-param">'BtrI'</em>, <em class="sig-param">'BtsCI'</em>, <em class="sig-param">'BtsI'</em>, <em class="sig-param">'BtsIMutI'</em>, <em class="sig-param">'BtuMI'</em>, <em class="sig-param">'BveI'</em>, <em class="sig-param">'Cac8I'</em>, <em class="sig-param">'CaiI'</em>, <em class="sig-param">'CciI'</em>, <em class="sig-param">'CciNI'</em>, <em class="sig-param">'CfoI'</em>, <em class="sig-param">'Cfr10I'</em>, <em class="sig-param">'Cfr13I'</em>, <em class="sig-param">'Cfr42I'</em>, <em class="sig-param">'Cfr9I'</em>, <em class="sig-param">'ClaI'</em>, <em class="sig-param">'CpoI'</em>, <em class="sig-param">'CseI'</em>, <em class="sig-param">'CsiI'</em>, <em class="sig-param">'Csp6I'</em>, <em class="sig-param">'CspAI'</em>, <em class="sig-param">'CspCI'</em>, <em class="sig-param">'CspI'</em>, <em class="sig-param">'CviAII'</em>, <em class="sig-param">'CviJI'</em>, <em class="sig-param">'CviKI_1'</em>, <em class="sig-param">'CviQI'</em>, <em class="sig-param">'DdeI'</em>, <em class="sig-param">'DinI'</em>, <em class="sig-param">'DpnI'</em>, <em class="sig-param">'DpnII'</em>, <em class="sig-param">'DraI'</em>, <em class="sig-param">'DraIII'</em>, <em class="sig-param">'DrdI'</em>, <em class="sig-param">'DriI'</em>, <em class="sig-param">'DseDI'</em>, <em class="sig-param">'EaeI'</em>, <em class="sig-param">'EagI'</em>, <em class="sig-param">'Eam1104I'</em>, <em class="sig-param">'Eam1105I'</em>, <em class="sig-param">'EarI'</em>, <em class="sig-param">'EciI'</em>, <em class="sig-param">'Ecl136II'</em>, <em class="sig-param">'EclXI'</em>, <em class="sig-param">'Eco105I'</em>, <em class="sig-param">'Eco130I'</em>, <em class="sig-param">'Eco147I'</em>, <em class="sig-param">'Eco24I'</em>, <em class="sig-param">'Eco31I'</em>, <em class="sig-param">'Eco32I'</em>, <em class="sig-param">'Eco47I'</em>, <em class="sig-param">'Eco47III'</em>, <em class="sig-param">'Eco52I'</em>, <em class="sig-param">'Eco53kI'</em>, <em class="sig-param">'Eco57I'</em>, <em class="sig-param">'Eco72I'</em>, <em class="sig-param">'Eco81I'</em>, <em class="sig-param">'Eco88I'</em>, <em class="sig-param">'Eco91I'</em>, <em class="sig-param">'EcoICRI'</em>, <em class="sig-param">'EcoNI'</em>, <em class="sig-param">'EcoO109I'</em>, <em class="sig-param">'EcoO65I'</em>, <em class="sig-param">'EcoRI'</em>, <em class="sig-param">'EcoRII'</em>, <em class="sig-param">'EcoRV'</em>, <em class="sig-param">'EcoT14I'</em>, <em class="sig-param">'EcoT22I'</em>, <em class="sig-param">'EcoT38I'</em>, <em class="sig-param">'EgeI'</em>, <em class="sig-param">'EheI'</em>, <em class="sig-param">'ErhI'</em>, <em class="sig-param">'Esp3I'</em>, <em class="sig-param">'FaeI'</em>, <em class="sig-param">'FaiI'</em>, <em class="sig-param">'FalI'</em>, <em class="sig-param">'FaqI'</em>, <em class="sig-param">'FatI'</em>, <em class="sig-param">'FauI'</em>, <em class="sig-param">'FauNDI'</em>, <em class="sig-param">'FbaI'</em>, <em class="sig-param">'FblI'</em>, <em class="sig-param">'Fnu4HI'</em>, <em class="sig-param">'FokI'</em>, <em class="sig-param">'FriOI'</em>, <em class="sig-param">'FseI'</em>, <em class="sig-param">'Fsp4HI'</em>, <em class="sig-param">'FspAI'</em>, <em class="sig-param">'FspBI'</em>, <em class="sig-param">'FspEI'</em>, <em class="sig-param">'FspI'</em>, <em class="sig-param">'GlaI'</em>, <em class="sig-param">'GluI'</em>, <em class="sig-param">'GsaI'</em>, <em class="sig-param">'GsuI'</em>, <em class="sig-param">'HaeII'</em>, <em class="sig-param">'HaeIII'</em>, <em class="sig-param">'HapII'</em>, <em class="sig-param">'HgaI'</em>, <em class="sig-param">'HhaI'</em>, <em class="sig-param">'Hin1I'</em>, <em class="sig-param">'Hin1II'</em>, <em class="sig-param">'Hin6I'</em>, <em class="sig-param">'HinP1I'</em>, <em class="sig-param">'HincII'</em>, <em class="sig-param">'HindII'</em>, <em class="sig-param">'HindIII'</em>, <em class="sig-param">'HinfI'</em>, <em class="sig-param">'HpaI'</em>, <em class="sig-param">'HpaII'</em>, <em class="sig-param">'HphI'</em>, <em class="sig-param">'Hpy166II'</em>, <em class="sig-param">'Hpy188I'</em>, <em class="sig-param">'Hpy188III'</em>, <em class="sig-param">'Hpy8I'</em>, <em class="sig-param">'Hpy99I'</em>, <em class="sig-param">'HpyAV'</em>, <em class="sig-param">'HpyCH4III'</em>, <em class="sig-param">'HpyCH4IV'</em>, <em class="sig-param">'HpyCH4V'</em>, <em class="sig-param">'HpyF10VI'</em>, <em class="sig-param">'HpyF3I'</em>, <em class="sig-param">'HpySE526I'</em>, <em class="sig-param">'Hsp92I'</em>, <em class="sig-param">'Hsp92II'</em>, <em class="sig-param">'HspAI'</em>, <em class="sig-param">'KasI'</em>, <em class="sig-param">'KflI'</em>, <em class="sig-param">'Kpn2I'</em>, <em class="sig-param">'KpnI'</em>, <em class="sig-param">'KroI'</em>, <em class="sig-param">'Ksp22I'</em>, <em class="sig-param">'KspAI'</em>, <em class="sig-param">'KspI'</em>, <em class="sig-param">'Kzo9I'</em>, <em class="sig-param">'LguI'</em>, <em class="sig-param">'LmnI'</em>, <em class="sig-param">'LpnPI'</em>, <em class="sig-param">'Lsp1109I'</em>, <em class="sig-param">'LweI'</em>, <em class="sig-param">'MabI'</em>, <em class="sig-param">'MaeI'</em>, <em class="sig-param">'MaeII'</em>, <em class="sig-param">'MaeIII'</em>, <em class="sig-param">'MalI'</em>, <em class="sig-param">'MauBI'</em>, <em class="sig-param">'MbiI'</em>, <em class="sig-param">'MboI'</em>, <em class="sig-param">'MboII'</em>, <em class="sig-param">'MfeI'</em>, <em class="sig-param">'MflI'</em>, <em class="sig-param">'MhlI'</em>, <em class="sig-param">'MlsI'</em>, <em class="sig-param">'MluCI'</em>, <em class="sig-param">'MluI'</em>, <em class="sig-param">'MluNI'</em>, <em class="sig-param">'Mly113I'</em>, <em class="sig-param">'MlyI'</em>, <em class="sig-param">'MmeI'</em>, <em class="sig-param">'MnlI'</em>, <em class="sig-param">'Mox20I'</em>, <em class="sig-param">'Mph1103I'</em>, <em class="sig-param">'MreI'</em>, <em class="sig-param">'MroI'</em>, <em class="sig-param">'MroNI'</em>, <em class="sig-param">'MroXI'</em>, <em class="sig-param">'MscI'</em>, <em class="sig-param">'MseI'</em>, <em class="sig-param">'MslI'</em>, <em class="sig-param">'Msp20I'</em>, <em class="sig-param">'MspA1I'</em>, <em class="sig-param">'MspCI'</em>, <em class="sig-param">'MspI'</em>, <em class="sig-param">'MspJI'</em>, <em class="sig-param">'MspR9I'</em>, <em class="sig-param">'MssI'</em>, <em class="sig-param">'MteI'</em>, <em class="sig-param">'MunI'</em>, <em class="sig-param">'Mva1269I'</em>, <em class="sig-param">'MvaI'</em>, <em class="sig-param">'MvnI'</em>, <em class="sig-param">'MwoI'</em>, <em class="sig-param">'NaeI'</em>, <em class="sig-param">'NarI'</em>, <em class="sig-param">'NciI'</em>, <em class="sig-param">'NcoI'</em>, <em class="sig-param">'NdeI'</em>, <em class="sig-param">'NdeII'</em>, <em class="sig-param">'NgoMIV'</em>, <em class="sig-param">'NheI'</em>, <em class="sig-param">'NlaIII'</em>, <em class="sig-param">'NlaIV'</em>, <em class="sig-param">'NmeAIII'</em>, <em class="sig-param">'NmuCI'</em>, <em class="sig-param">'NotI'</em>, <em class="sig-param">'NruI'</em>, <em class="sig-param">'NsbI'</em>, <em class="sig-param">'NsiI'</em>, <em class="sig-param">'NspI'</em>, <em class="sig-param">'NspV'</em>, <em class="sig-param">'OliI'</em>, <em class="sig-param">'PacI'</em>, <em class="sig-param">'PaeI'</em>, <em class="sig-param">'PaeR7I'</em>, <em class="sig-param">'PagI'</em>, <em class="sig-param">'PalAI'</em>, <em class="sig-param">'PasI'</em>, <em class="sig-param">'PauI'</em>, <em class="sig-param">'PceI'</em>, <em class="sig-param">'PciI'</em>, <em class="sig-param">'PciSI'</em>, <em class="sig-param">'PcsI'</em>, <em class="sig-param">'PctI'</em>, <em class="sig-param">'PdiI'</em>, <em class="sig-param">'PdmI'</em>, <em class="sig-param">'PfeI'</em>, <em class="sig-param">'Pfl23II'</em>, <em class="sig-param">'PflFI'</em>, <em class="sig-param">'PflMI'</em>, <em class="sig-param">'PfoI'</em>, <em class="sig-param">'PinAI'</em>, <em class="sig-param">'PkrI'</em>, <em class="sig-param">'Ple19I'</em>, <em class="sig-param">'PleI'</em>, <em class="sig-param">'PluTI'</em>, <em class="sig-param">'PmaCI'</em>, <em class="sig-param">'PmeI'</em>, <em class="sig-param">'PmlI'</em>, <em class="sig-param">'PpsI'</em>, <em class="sig-param">'Ppu21I'</em>, <em class="sig-param">'PpuMI'</em>, <em class="sig-param">'PscI'</em>, <em class="sig-param">'PshAI'</em>, <em class="sig-param">'PshBI'</em>, <em class="sig-param">'PsiI'</em>, <em class="sig-param">'Psp124BI'</em>, <em class="sig-param">'Psp1406I'</em>, <em class="sig-param">'Psp5II'</em>, <em class="sig-param">'Psp6I'</em>, <em class="sig-param">'PspCI'</em>, <em class="sig-param">'PspEI'</em>, <em class="sig-param">'PspFI'</em>, <em class="sig-param">'PspGI'</em>, <em class="sig-param">'PspLI'</em>, <em class="sig-param">'PspN4I'</em>, <em class="sig-param">'PspOMI'</em>, <em class="sig-param">'PspPI'</em>, <em class="sig-param">'PspPPI'</em>, <em class="sig-param">'PspXI'</em>, <em class="sig-param">'PsrI'</em>, <em class="sig-param">'PstI'</em>, <em class="sig-param">'PstNI'</em>, <em class="sig-param">'PsuI'</em>, <em class="sig-param">'PsyI'</em>, <em class="sig-param">'PteI'</em>, <em class="sig-param">'PvuI'</em>, <em class="sig-param">'PvuII'</em>, <em class="sig-param">'RgaI'</em>, <em class="sig-param">'RigI'</em>, <em class="sig-param">'RruI'</em>, <em class="sig-param">'RsaI'</em>, <em class="sig-param">'RsaNI'</em>, <em class="sig-param">'RseI'</em>, <em class="sig-param">'Rsr2I'</em>, <em class="sig-param">'RsrII'</em>, <em class="sig-param">'SacI'</em>, <em class="sig-param">'SacII'</em>, <em class="sig-param">'SalI'</em>, <em class="sig-param">'SapI'</em>, <em class="sig-param">'SaqAI'</em>, <em class="sig-param">'SatI'</em>, <em class="sig-param">'Sau3AI'</em>, <em class="sig-param">'Sau96I'</em>, <em class="sig-param">'SbfI'</em>, <em class="sig-param">'ScaI'</em>, <em class="sig-param">'SchI'</em>, <em class="sig-param">'ScrFI'</em>, <em class="sig-param">'SdaI'</em>, <em class="sig-param">'SduI'</em>, <em class="sig-param">'SetI'</em>, <em class="sig-param">'SexAI'</em>, <em class="sig-param">'SfaAI'</em>, <em class="sig-param">'SfaNI'</em>, <em class="sig-param">'SfcI'</em>, <em class="sig-param">'SfiI'</em>, <em class="sig-param">'SfoI'</em>, <em class="sig-param">'Sfr274I'</em>, <em class="sig-param">'Sfr303I'</em>, <em class="sig-param">'SfuI'</em>, <em class="sig-param">'SgeI'</em>, <em class="sig-param">'SgfI'</em>, <em class="sig-param">'SgrAI'</em>, <em class="sig-param">'SgrBI'</em>, <em class="sig-param">'SgrDI'</em>, <em class="sig-param">'SgsI'</em>, <em class="sig-param">'SinI'</em>, <em class="sig-param">'SlaI'</em>, <em class="sig-param">'SmaI'</em>, <em class="sig-param">'SmiI'</em>, <em class="sig-param">'SmiMI'</em>, <em class="sig-param">'SmlI'</em>, <em class="sig-param">'SmoI'</em>, <em class="sig-param">'SnaBI'</em>, <em class="sig-param">'SpeI'</em>, <em class="sig-param">'SphI'</em>, <em class="sig-param">'SrfI'</em>, <em class="sig-param">'Sse8387I'</em>, <em class="sig-param">'Sse9I'</em>, <em class="sig-param">'SseBI'</em>, <em class="sig-param">'SsiI'</em>, <em class="sig-param">'SspDI'</em>, <em class="sig-param">'SspI'</em>, <em class="sig-param">'SspMI'</em>, <em class="sig-param">'SstI'</em>, <em class="sig-param">'StuI'</em>, <em class="sig-param">'StyD4I'</em>, <em class="sig-param">'StyI'</em>, <em class="sig-param">'SwaI'</em>, <em class="sig-param">'TaaI'</em>, <em class="sig-param">'TaiI'</em>, <em class="sig-param">'TaqI'</em>, <em class="sig-param">'TaqII'</em>, <em class="sig-param">'TasI'</em>, <em class="sig-param">'TatI'</em>, <em class="sig-param">'TauI'</em>, <em class="sig-param">'TfiI'</em>, <em class="sig-param">'Tru1I'</em>, <em class="sig-param">'Tru9I'</em>, <em class="sig-param">'TscAI'</em>, <em class="sig-param">'TseFI'</em>, <em class="sig-param">'TseI'</em>, <em class="sig-param">'Tsp45I'</em>, <em class="sig-param">'TspDTI'</em>, <em class="sig-param">'TspGWI'</em>, <em class="sig-param">'TspMI'</em>, <em class="sig-param">'TspRI'</em>, <em class="sig-param">'Tth111I'</em>, <em class="sig-param">'Van91I'</em>, <em class="sig-param">'Vha464I'</em>, <em class="sig-param">'VneI'</em>, <em class="sig-param">'VpaK11BI'</em>, <em class="sig-param">'VspI'</em>, <em class="sig-param">'XagI'</em>, <em class="sig-param">'XapI'</em>, <em class="sig-param">'XbaI'</em>, <em class="sig-param">'XceI'</em>, <em class="sig-param">'XcmI'</em>, <em class="sig-param">'XhoI'</em>, <em class="sig-param">'XmaI'</em>, <em class="sig-param">'XmaJI'</em>, <em class="sig-param">'XmiI'</em>, <em class="sig-param">'XmnI'</em>, <em class="sig-param">'XspI'</em>, <em class="sig-param">'ZraI'</em>, <em class="sig-param">'ZrmI'</em>, <em class="sig-param">'Zsp2I'])</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.unique_cutters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.unique_cutters" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.dseqrecord.Dseqrecord.write">
<code class="sig-name descname">write</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">filename=None</em>, <em class="sig-param">f='gb'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/dseqrecord.html#Dseqrecord.write"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dseqrecord.Dseqrecord.write" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes the Dseqrecord to a file using the format f, which must
be a format supported by Biopython SeqIO for writing <a class="footnote-reference brackets" href="#id10" id="id9">5</a>. Default
is “gb” which is short for Genbank. Note that Biopython SeqIO reads
more formats than it writes.</p>
<p>Filename is the path to the file where the sequece is to be
written. The filename is optional, if it is not given, the
description property (string) is used together with the format.</p>
<p>If obj is the Dseqrecord object, the default file name will be:</p>
<p><code class="docutils literal notranslate"><span class="pre">&lt;obj.locus&gt;.&lt;f&gt;</span></code></p>
<p>Where &lt;f&gt; is “gb” by default. If the filename already exists and
AND the sequence it contains is different, a new file name will be
used so that the old file is not lost:</p>
<p><code class="docutils literal notranslate"><span class="pre">&lt;obj.locus&gt;_NEW.&lt;f&gt;</span></code></p>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id10"><span class="brackets"><a class="fn-backref" href="#id9">5</a></span></dt>
<dd><p><a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-pydna.amplicon">
<span id="pydna-amplicon-module"></span><h2>pydna.amplicon module<a class="headerlink" href="#module-pydna.amplicon" title="Permalink to this headline">¶</a></h2>
<p>This module provides the <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-class docutils literal notranslate"><span class="pre">Amplicon</span></code></a> class for PCR simulation. This class is not meant to be use directly
but is used by the <code class="xref py py-mod docutils literal notranslate"><span class="pre">amplify</span></code> module</p>
<dl class="class">
<dt id="pydna.amplicon.Amplicon">
<em class="property">class </em><code class="sig-prename descclassname">pydna.amplicon.</code><code class="sig-name descname">Amplicon</code><span class="sig-paren">(</span><em class="sig-param">record</em>, <em class="sig-param">*args</em>, <em class="sig-param">template=None</em>, <em class="sig-param">forward_primer=None</em>, <em class="sig-param">reverse_primer=None</em>, <em class="sig-param">saltc=50.0</em>, <em class="sig-param">fprimerc=1000.0</em>, <em class="sig-param">rprimerc=1000.0</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplicon.Amplicon" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord</span></code></a></p>
<p>The Amplicon class holds information about a PCR reaction involving two
primers and one template. This class is used by the Anneal class and is not
meant to be instantiated directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>forward_primer</strong><span class="classifier">SeqRecord(Biopython)</span></dt><dd><p>SeqRecord object holding the forward (sense) primer</p>
</dd>
<dt><strong>reverse_primer</strong><span class="classifier">SeqRecord(Biopython)</span></dt><dd><p>SeqRecord object holding the reverse (antisense) primer</p>
</dd>
<dt><strong>template</strong><span class="classifier">Dseqrecord</span></dt><dd><p>Dseqrecord object holding the template (circular or linear)</p>
</dd>
<dt><strong>saltc</strong><span class="classifier">float, optional</span></dt><dd><p>saltc = monovalent cations (mM) (Na,K..)
default value is 50mM
This is used for Tm calculations.</p>
</dd>
<dt><strong>forward_primer_concentration</strong><span class="classifier">float, optional</span></dt><dd><p>primer concentration (nM)
default set to 1000nM = 1µM
This is used for Tm calculations.</p>
</dd>
<dt><strong>rc</strong><span class="classifier">float, optional</span></dt><dd><p>primer concentration (nM)
default set to 1000nM = 1µM
This is used for Tm calculations.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="method">
<dt id="pydna.amplicon.Amplicon.dbd_program">
<code class="sig-name descname">dbd_program</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon.dbd_program"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.dbd_program" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a string containing a text representation of a proposed
PCR program using a polymerase with a DNA binding domain such as Pfu-Sso7d.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Pfu-Sso7d (rate 15s/kb)             |{size}bp
Three-step|          30 cycles   |      |Tm formula: Pydna tmbresluc
98.0°C    |98.0°C                |      |SaltC 50mM
__________|_____          72.0°C |72.0°C|Primer1C   1µM
00min30s  |10s  \ 61.0°C ________|______|Primer2C   1µM
          |      \______/ 0min 0s|10min |
          |        10s           |      |4-12°C
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.amplicon.Amplicon.figure">
<code class="sig-name descname">figure</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon.figure"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.figure" title="Permalink to this definition">¶</a></dt>
<dd><p>This method returns a simple figure of the two primers binding to a part
of the template.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">5</span><span class="n">gctactacacacgtactgactg3</span>
 <span class="o">||||||||||||||||||||||</span> <span class="n">tm</span> <span class="mf">52.6</span> <span class="p">(</span><span class="n">dbd</span><span class="p">)</span> <span class="mf">58.3</span>
<span class="mi">5</span><span class="n">gctactacacacgtactgactg</span><span class="o">...</span><span class="n">caagatagagtcagtaaccaca3</span>
<span class="mi">3</span><span class="n">cgatgatgtgtgcatgactgac</span><span class="o">...</span><span class="n">gttctatctcagtcattggtgt5</span>
                          <span class="o">||||||||||||||||||||||</span> <span class="n">tm</span> <span class="mf">49.1</span> <span class="p">(</span><span class="n">dbd</span><span class="p">)</span> <span class="mf">57.7</span>
                         <span class="mi">3</span><span class="n">gttctatctcagtcattggtgt5</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>figure:string</dt><dd><p>A string containing a text representation of the primers
annealing on the template (see example above).</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>tm in the figure above is the melting temperature (tm) for each primer calculated according to
SantaLucia 1998 <a class="footnote-reference brackets" href="#id14" id="id11">6</a>.</p>
<p>dbd is the tm calculation for enzymes with dsDNA binding domains like Pfu-Sso7d <a class="footnote-reference brackets" href="#id15" id="id12">7</a>. See <a class="footnote-reference brackets" href="#id16" id="id13">8</a>
for more information.</p>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id14"><span class="brackets"><a class="fn-backref" href="#id11">6</a></span></dt>
<dd><ol class="upperalpha simple" start="10">
<li><p>SantaLucia, “A Unified View of Polymer, Dumbbell, and Oligonucleotide DNA Nearest-neighbor Thermodynamics,” Proceedings of the National Academy of Sciences 95, no. 4 (1998): 1460.</p></li>
</ol>
</dd>
<dt class="label" id="id15"><span class="brackets"><a class="fn-backref" href="#id12">7</a></span></dt>
<dd><ol class="upperalpha simple" start="13">
<li><p>Nørholm, “A Mutant Pfu DNA Polymerprimerase Designed for Advanced Uracil-excision DNA Engineering,” BMC Biotechnology 10, no. 1 (2010): 21, doi:10.1186/1472-6750-10-21.</p></li>
</ol>
</dd>
<dt class="label" id="id16"><span class="brackets"><a class="fn-backref" href="#id13">8</a></span></dt>
<dd><p><a class="reference external" href="http://www.thermoscientificbio.com/webtools/tmc/">http://www.thermoscientificbio.com/webtools/tmc/</a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="pydna.amplicon.Amplicon.from_SeqRecord">
<em class="property">classmethod </em><code class="sig-name descname">from_SeqRecord</code><span class="sig-paren">(</span><em class="sig-param">record</em>, <em class="sig-param">*args</em>, <em class="sig-param">path=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon.from_SeqRecord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.from_SeqRecord" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.amplicon.Amplicon.pfu_sso7d_program">
<code class="sig-name descname">pfu_sso7d_program</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon.pfu_sso7d_program"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.pfu_sso7d_program" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.amplicon.Amplicon.program">
<code class="sig-name descname">program</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon.program"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.program" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a string containing a text representation of a proposed
PCR program using Taq or similar polymerase.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Taq (rate 30 nt/s)
Three-step|         30 cycles     |      |Tm formula: Biopython Tm_NN
94.0°C    |94.0°C                 |      |SaltC 50mM
__________|_____          72.0°C  |72.0°C|
04min00s  |30s  \         ________|______|
          |      \ 46.0°C/ 0min 1s|10min |
          |       \_____/         |      |
          |         30s           |      |4-8°C
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.amplicon.Amplicon.rc">
<code class="sig-name descname">rc</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.amplicon.Amplicon.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reverse complement.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.amplicon.Amplicon.reverse_complement">
<code class="sig-name descname">reverse_complement</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon.reverse_complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reverse complement.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.amplicon.Amplicon.taq_program">
<code class="sig-name descname">taq_program</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplicon.html#Amplicon.taq_program"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplicon.Amplicon.taq_program" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="module-pydna.amplify">
<span id="pydna-amplify-module"></span><h2>pydna.amplify module<a class="headerlink" href="#module-pydna.amplify" title="Permalink to this headline">¶</a></h2>
<p>This module provide the <a class="reference internal" href="#pydna.amplify.Anneal" title="pydna.amplify.Anneal"><code class="xref py py-class docutils literal notranslate"><span class="pre">Anneal</span></code></a> class and the <a class="reference internal" href="#pydna.amplify.pcr" title="pydna.amplify.pcr"><code class="xref py py-func docutils literal notranslate"><span class="pre">pcr()</span></code></a> function for PCR simulation. 
The pcr function is simpler to use, but expects only one PCR product. The Anneal class should be used 
if more flexibility is required.</p>
<p>Primers with 5’ tails as well as inverse PCR on circular templates are handled correctly.</p>
<dl class="class">
<dt id="pydna.amplify.Anneal">
<em class="property">class </em><code class="sig-prename descclassname">pydna.amplify.</code><code class="sig-name descname">Anneal</code><span class="sig-paren">(</span><em class="sig-param">primers</em>, <em class="sig-param">template</em>, <em class="sig-param">limit=13</em>, <em class="sig-param">primerc=1000.0</em>, <em class="sig-param">saltc=50</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplify.html#Anneal"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Anneal" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.6/library/functions.html#object" title="(in Python v3.6)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>The Anneal class has the following important attributes:</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>forward_primers</strong><span class="classifier">list</span></dt><dd><p>Description of <cite>forward_primers</cite>.</p>
</dd>
<dt><strong>reverse_primers</strong><span class="classifier">list</span></dt><dd><p>Description of <cite>reverse_primers</cite>.</p>
</dd>
<dt><strong>template</strong><span class="classifier">Dseqrecord</span></dt><dd><p>A copy of the template argument. Primers annealing sites has been added as features that can be visualized in a seqence editor such as ApE.</p>
</dd>
<dt><strong>primerc</strong><span class="classifier">float</span></dt><dd><p>Primer concentration, affects some algorithms for primer Tm calculation. Both primers are expected to be present at the same concentration.</p>
</dd>
<dt><strong>saltc</strong><span class="classifier">int</span></dt><dd><p>Salt concentration, affects some algorithms for primer Tm calculation.</p>
</dd>
<dt><strong>limit</strong><span class="classifier">int, optional</span></dt><dd><p>The limit of PCR primer annealing, default is 13 bp.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="method">
<dt id="pydna.amplify.Anneal.products">
<em class="property">property </em><code class="sig-name descname">products</code><a class="headerlink" href="#pydna.amplify.Anneal.products" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.amplify.Anneal.report">
<code class="sig-name descname">report</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.amplify.Anneal.report" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a short report describing if or where primer
anneal on the template.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="pydna.amplify.pcr">
<code class="sig-prename descclassname">pydna.amplify.</code><code class="sig-name descname">pcr</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/amplify.html#pcr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.pcr" title="Permalink to this definition">¶</a></dt>
<dd><p>pcr is a convenience function for the Anneal class to simplify its usage,
especially from the command line. If more than one or no PCR product is
formed, a ValueError is raised.</p>
<p>args is any iterable of Dseqrecords or an iterable of iterables of Dseqrecords.
args will be greedily flattened.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>args</strong><span class="classifier">iterable containing sequence objects</span></dt><dd><p>Several arguments are also accepted.</p>
</dd>
<dt><strong>limit</strong><span class="classifier">int = 13, optional</span></dt><dd><p>limit length of the annealing part of the primers.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>product</strong><span class="classifier">Amplicon</span></dt><dd><p>An <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> object representing the PCR product.
The direction of the PCR product will be the same as for the template sequence.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>sequences in args could be of type:</p>
<ul class="simple">
<li><p>string</p></li>
<li><p>Seq</p></li>
<li><p>SeqRecord (or subclass)</p></li>
<li><p>Dseqrecord (or sublcass)</p></li>
</ul>
<p>The last sequence will be assumed to be the template while
all preceeding sequences will be assumed to be primers.</p>
<p>This is a powerful function, use with care!</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.readers</span> <span class="k">import</span> <span class="n">read</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.amplify</span> <span class="k">import</span> <span class="n">pcr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">template</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tacactcaccgtctatcattatctactatcgactgtatcatctgatagcac&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="k">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p1</span> <span class="o">=</span> <span class="n">read</span><span class="p">(</span><span class="s2">&quot;&gt;p1</span><span class="se">\n</span><span class="s2">tacactcaccgtctatcattatc&quot;</span><span class="p">,</span> <span class="n">ds</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span> <span class="o">=</span> <span class="n">read</span><span class="p">(</span><span class="s2">&quot;&gt;p2</span><span class="se">\n</span><span class="s2">cgactgtatcatctgatagcac&quot;</span><span class="p">,</span> <span class="n">ds</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">,</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Amplicon(51)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr</span><span class="p">([</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">],</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Amplicon(51)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr</span><span class="p">((</span><span class="n">p1</span><span class="p">,</span><span class="n">p2</span><span class="p">,),</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Amplicon(51)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="module-pydna.assembly">
<span id="pydna-assembly-module"></span><h2>pydna.assembly module<a class="headerlink" href="#module-pydna.assembly" title="Permalink to this headline">¶</a></h2>
<p>This module provides functions for assembly of sequences by homologous 
recombination and other related techniques. Given a list of sequences 
(Dseqrecords), all sequences are analyzed for shared homology longer than the 
set limit.</p>
<p>A graph is constructed where each overlapping region form a node and
sequences separating the overlapping regions form edges.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>            <span class="o">--</span> <span class="n">A</span> <span class="o">--</span>
<span class="n">catgatctacgtatcgtgt</span>     <span class="o">--</span> <span class="n">B</span> <span class="o">--</span>
            <span class="n">atcgtgtactgtcatattc</span>
                        <span class="n">catattcaaagttct</span>

 
 
<span class="o">--</span><span class="n">x</span><span class="o">--&gt;</span> <span class="n">A</span> <span class="o">--</span><span class="n">y</span><span class="o">--&gt;</span> <span class="n">B</span> <span class="o">--</span><span class="n">z</span><span class="o">--&gt;</span>   <span class="p">(</span><span class="n">Graph</span><span class="p">)</span>
 
<span class="n">Nodes</span><span class="p">:</span>

<span class="n">A</span> <span class="p">:</span> <span class="n">atcgtgt</span>
<span class="n">B</span> <span class="p">:</span> <span class="n">catattc</span>

<span class="n">Edges</span><span class="p">:</span>

<span class="n">x</span> <span class="p">:</span> <span class="n">catgatctacgt</span>
<span class="n">y</span> <span class="p">:</span> <span class="n">actgt</span>
<span class="n">z</span> <span class="p">:</span> <span class="n">aaagttct</span>
</pre></div>
</div>
<p>The NetworkX package is used to trace linear and circular paths through the
graph.</p>
<dl class="class">
<dt id="pydna.assembly.Assembly">
<em class="property">class </em><code class="sig-prename descclassname">pydna.assembly.</code><code class="sig-name descname">Assembly</code><span class="sig-paren">(</span><em class="sig-param">frags=None</em>, <em class="sig-param">limit=25</em>, <em class="sig-param">algorithm=&lt;function common_sub_strings&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/assembly.html#Assembly"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Assembly" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.6/library/functions.html#object" title="(in Python v3.6)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Assembly of a list of linear DNA fragments into linear or circular constructs.
The Assembly is meant to replace the Assembly method as it is easier to use.
Accepts a list of Dseqrecords (source fragments) to initiate an Assembly object.
Several methods are available for analysis of overlapping sequences, graph construction
and assembly.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>fragments</strong><span class="classifier">list</span></dt><dd><p>a list of Dseqrecord objects.</p>
</dd>
<dt><strong>limit</strong><span class="classifier">int, optional</span></dt><dd><p>The shortest shared homology to be considered</p>
</dd>
<dt><strong>algorithm</strong><span class="classifier">function, optional</span></dt><dd><p>The algorithm used to determine the shared sequences.</p>
</dd>
<dt><strong>max_nodes</strong><span class="classifier">int</span></dt><dd><blockquote>
<div><p>The maximum number of nodes in the graph. This can be tweaked to manage</p>
</div></blockquote>
<p>sequences with a high number of shared sub sequences.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.assembly</span> <span class="k">import</span> <span class="n">Assembly</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;acgatgctatactgCCCCCtgtgctgtgctcta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tgtgctgtgctctaTTTTTtattctggctgtatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tattctggctgtatcGGGGGtacgatgctatactg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">Assembly</span><span class="p">((</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span><span class="p">),</span> <span class="n">limit</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">Assembly</span>
<span class="go">fragments....: 33bp 34bp 35bp</span>
<span class="go">limit(bp)....: 14</span>
<span class="go">G.nodes......: 6</span>
<span class="go">algorithm....: common_sub_strings</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()</span>
<span class="go">[Contig(o59), Contig(o59)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">watson</span>
<span class="go">&#39;acgatgctatactgCCCCCtgtgctgtgctctaTTTTTtattctggctgtatcGGGGGt&#39;</span>
</pre></div>
</div>
<dl class="method">
<dt id="pydna.assembly.Assembly.assemble_circular">
<code class="sig-name descname">assemble_circular</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/assembly.html#Assembly.assemble_circular"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Assembly.assemble_circular" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.assembly.Assembly.assemble_linear">
<code class="sig-name descname">assemble_linear</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">start=None</em>, <em class="sig-param">end=None</em>, <em class="sig-param">max_nodes=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/assembly.html#Assembly.assemble_linear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Assembly.assemble_linear" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="module-pydna.common_sub_strings">
<span id="pydna-common-sub-strings-module"></span><h2>pydna.common_sub_strings module<a class="headerlink" href="#module-pydna.common_sub_strings" title="Permalink to this headline">¶</a></h2>
<p>This module is based on the Py-rstr-max package that was written by Romain Brixtel (rbrixtel_at_gmail_dot_com)
(<a class="reference external" href="https://brixtel.users.greyc.fr/">https://brixtel.users.greyc.fr/</a>) and is available from <a class="reference external" href="https://code.google.com/p/py-rstr-max/">https://code.google.com/p/py-rstr-max/</a>
the original code was covered by an MIT licence.</p>
<dl class="class">
<dt id="pydna.common_sub_strings.Rstr_max">
<em class="property">class </em><code class="sig-prename descclassname">pydna.common_sub_strings.</code><code class="sig-name descname">Rstr_max</code><a class="reference internal" href="_modules/pydna/common_sub_strings.html#Rstr_max"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.common_sub_strings.Rstr_max" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.6/library/functions.html#object" title="(in Python v3.6)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<dl class="method">
<dt id="pydna.common_sub_strings.Rstr_max.add_str">
<code class="sig-name descname">add_str</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">str_unicode</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/common_sub_strings.html#Rstr_max.add_str"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.common_sub_strings.Rstr_max.add_str" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.common_sub_strings.Rstr_max.go">
<code class="sig-name descname">go</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/common_sub_strings.html#Rstr_max.go"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.common_sub_strings.Rstr_max.go" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.common_sub_strings.Rstr_max.removeMany">
<code class="sig-name descname">removeMany</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">stack</em>, <em class="sig-param">results</em>, <em class="sig-param">m</em>, <em class="sig-param">idxEnd</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/common_sub_strings.html#Rstr_max.removeMany"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.common_sub_strings.Rstr_max.removeMany" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.common_sub_strings.Rstr_max.step1_sort_suffix">
<code class="sig-name descname">step1_sort_suffix</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/common_sub_strings.html#Rstr_max.step1_sort_suffix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.common_sub_strings.Rstr_max.step1_sort_suffix" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.common_sub_strings.Rstr_max.step2_lcp">
<code class="sig-name descname">step2_lcp</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/common_sub_strings.html#Rstr_max.step2_lcp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.common_sub_strings.Rstr_max.step2_lcp" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.common_sub_strings.Rstr_max.step3_rstr">
<code class="sig-name descname">step3_rstr</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/common_sub_strings.html#Rstr_max.step3_rstr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.common_sub_strings.Rstr_max.step3_rstr" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="function">
<dt id="pydna.common_sub_strings.common_sub_strings">
<code class="sig-prename descclassname">pydna.common_sub_strings.</code><code class="sig-name descname">common_sub_strings</code><span class="sig-paren">(</span><em class="sig-param">stringx: str</em>, <em class="sig-param">stringy: str</em>, <em class="sig-param">limit=25</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/common_sub_strings.html#common_sub_strings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.common_sub_strings.common_sub_strings" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds all common substrings between stringx and stringy
longer than limit. This function is case sensitive.
The substrings may overlap.</p>
<p>returns a list of tuples describing the substrings
The list is sorted longest -&gt; shortest.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>stringx</strong><span class="classifier">str</span></dt><dd></dd>
<dt><strong>stringy</strong><span class="classifier">str</span></dt><dd></dd>
<dt><strong>limit</strong><span class="classifier">int, optional</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl>
<dt>list of tuple</dt><dd><p>[(startx1,starty1,length1),(startx2,starty2,length2), …]</p>
<p>startx1 = startposition in x, where substring 1 starts
starty1 = position in y where substring 1 starts
length1 = lenght of substring</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.common_sub_strings</span> <span class="k">import</span> <span class="n">common_sub_strings</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">common_sub_strings</span><span class="p">(</span><span class="s2">&quot;gatgatttcggtagtta&quot;</span><span class="p">,</span> <span class="s2">&quot;gtcagtatgtctatctatcgcg&quot;</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="go">[(1, 6, 3), (7, 17, 3), (10, 4, 3), (12, 3, 3)]</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Overlaps</span>   <span class="n">Symbols</span>
<span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span>  <span class="mi">3</span><span class="p">)</span>   <span class="o">---</span>
<span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">17</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>   <span class="o">+++</span>
<span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>   <span class="o">...</span>
<span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>   <span class="o">===</span>    

            <span class="o">===</span>       
<span class="n">gatgatttcggtagtta</span>           <span class="n">stringx</span>
 <span class="o">---</span>   <span class="o">+++...</span>  

    <span class="o">...</span>           
<span class="n">gtcagtatgtctatctatcgcg</span>      <span class="n">stringy</span>
   <span class="o">===---</span>        <span class="o">+++</span>  
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.common_sub_strings.direct_kark_sort">
<code class="sig-prename descclassname">pydna.common_sub_strings.</code><code class="sig-name descname">direct_kark_sort</code><span class="sig-paren">(</span><em class="sig-param">s</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/common_sub_strings.html#direct_kark_sort"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.common_sub_strings.direct_kark_sort" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.common_sub_strings.kark_sort">
<code class="sig-prename descclassname">pydna.common_sub_strings.</code><code class="sig-name descname">kark_sort</code><span class="sig-paren">(</span><em class="sig-param">s</em>, <em class="sig-param">SA</em>, <em class="sig-param">n</em>, <em class="sig-param">K</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/common_sub_strings.html#kark_sort"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.common_sub_strings.kark_sort" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.common_sub_strings.radixpass">
<code class="sig-prename descclassname">pydna.common_sub_strings.</code><code class="sig-name descname">radixpass</code><span class="sig-paren">(</span><em class="sig-param">a</em>, <em class="sig-param">b</em>, <em class="sig-param">r</em>, <em class="sig-param">s</em>, <em class="sig-param">n</em>, <em class="sig-param">k</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/common_sub_strings.html#radixpass"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.common_sub_strings.radixpass" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.common_sub_strings.terminal_overlap">
<code class="sig-prename descclassname">pydna.common_sub_strings.</code><code class="sig-name descname">terminal_overlap</code><span class="sig-paren">(</span><em class="sig-param">stringx: str</em>, <em class="sig-param">stringy: str</em>, <em class="sig-param">limit=15</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/common_sub_strings.html#terminal_overlap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.common_sub_strings.terminal_overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds the the flanking common substrings between stringx and stringy
longer than limit. This means that the results only contains substrings
that starts or ends at the the ends of stringx and stringy.</p>
<p>This function is case sensitive.</p>
<p>returns a list of tuples describing the substrings
The list is sorted longest -&gt; shortest.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>stringx</strong><span class="classifier">str</span></dt><dd></dd>
<dt><strong>stringy</strong><span class="classifier">str</span></dt><dd></dd>
<dt><strong>limit</strong><span class="classifier">int, optional</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl>
<dt>list of tuple</dt><dd><p>[(startx1,starty1,length1),(startx2,starty2,length2), …]</p>
<p>startx1 = startposition in x, where substring 1 starts
starty1 = position in y where substring 1 starts
length1 = lenght of substring</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.common_sub_strings</span> <span class="k">import</span> <span class="n">terminal_overlap</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">terminal_overlap</span><span class="p">(</span><span class="s2">&quot;agctatgtatcttgcatcgta&quot;</span><span class="p">,</span> <span class="s2">&quot;gcatcgtagtctatttgcttac&quot;</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="go">[(13, 0, 8)]</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>             <span class="o">&lt;--</span> <span class="mi">8</span> <span class="o">-&gt;</span>
<span class="o">&lt;----</span> <span class="mi">13</span> <span class="o">---&gt;</span>               
<span class="n">agctatgtatcttgcatcgta</span>                    <span class="n">stringx</span>
             <span class="n">gcatcgtagtctatttgcttac</span>      <span class="n">stringy</span>
             <span class="mi">0</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="module-pydna.contig">
<span id="pydna-contig-module"></span><h2>pydna.contig module<a class="headerlink" href="#module-pydna.contig" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="pydna.contig.Contig">
<em class="property">class </em><code class="sig-prename descclassname">pydna.contig.</code><code class="sig-name descname">Contig</code><span class="sig-paren">(</span><em class="sig-param">record</em>, <em class="sig-param">*args</em>, <em class="sig-param">graph=None</em>, <em class="sig-param">nodemap=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/contig.html#Contig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.contig.Contig" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord</span></code></a></p>
<p>This class holds information about a DNA assembly. This class is instantiated by
the <code class="xref py py-class docutils literal notranslate"><span class="pre">Assembly</span></code> class and is not meant to be used directly.</p>
<dl class="method">
<dt id="pydna.contig.Contig.detailed_figure">
<code class="sig-name descname">detailed_figure</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/contig.html#Contig.detailed_figure"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.contig.Contig.detailed_figure" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a text representation of the assembled fragments.</p>
<p>Linear:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">acgatgctatactgCCCCCtgtgctgtgctcta</span>
                   <span class="n">TGTGCTGTGCTCTA</span>
                   <span class="n">tgtgctgtgctctaTTTTTtattctggctgtatc</span>
</pre></div>
</div>
<p>Circular:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">||||||||||||||</span>
<span class="n">acgatgctatactgCCCCCtgtgctgtgctcta</span>
                   <span class="n">TGTGCTGTGCTCTA</span>
                   <span class="n">tgtgctgtgctctaTTTTTtattctggctgtatc</span>
                                      <span class="n">TATTCTGGCTGTATC</span>
                                      <span class="n">tattctggctgtatcGGGGGtacgatgctatactg</span>
                                                           <span class="n">ACGATGCTATACTG</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.contig.Contig.figure">
<code class="sig-name descname">figure</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/contig.html#Contig.figure"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.contig.Contig.figure" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a compact ascii representation of the assembled fragments. Each fragment is
represented by:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Size</span> <span class="n">of</span> <span class="n">common</span> <span class="mi">5</span><span class="s1">&#39; substring|Name and size of DNA fragment| Size of common 5&#39;</span> <span class="n">substring</span>
</pre></div>
</div>
<p>Linear:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">frag20</span><span class="o">|</span> <span class="mi">6</span>
       \<span class="o">/</span>
       <span class="o">/</span>\
        <span class="mi">6</span><span class="o">|</span><span class="n">frag23</span><span class="o">|</span> <span class="mi">6</span>
                 \<span class="o">/</span>
                 <span class="o">/</span>\
                  <span class="mi">6</span><span class="o">|</span><span class="n">frag14</span>
</pre></div>
</div>
<p>Circular:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="o">-|</span><span class="mi">2577</span><span class="o">|</span><span class="mi">61</span>
<span class="o">|</span>       \<span class="o">/</span>
<span class="o">|</span>       <span class="o">/</span>\
<span class="o">|</span>       <span class="mi">61</span><span class="o">|</span><span class="mi">5681</span><span class="o">|</span><span class="mi">98</span>
<span class="o">|</span>               \<span class="o">/</span>
<span class="o">|</span>               <span class="o">/</span>\
<span class="o">|</span>               <span class="mi">98</span><span class="o">|</span><span class="mi">2389</span><span class="o">|</span><span class="mi">557</span>
<span class="o">|</span>                       \<span class="o">/</span>
<span class="o">|</span>                       <span class="o">/</span>\
<span class="o">|</span>                       <span class="mi">557</span><span class="o">-</span>
<span class="o">|</span>                          <span class="o">|</span>
 <span class="o">--------------------------</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.contig.Contig.from_SeqRecord">
<em class="property">classmethod </em><code class="sig-name descname">from_SeqRecord</code><span class="sig-paren">(</span><em class="sig-param">record</em>, <em class="sig-param">*args</em>, <em class="sig-param">graph=None</em>, <em class="sig-param">nodemap=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/contig.html#Contig.from_SeqRecord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.contig.Contig.from_SeqRecord" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.contig.Contig.from_string">
<em class="property">classmethod </em><code class="sig-name descname">from_string</code><span class="sig-paren">(</span><em class="sig-param">record: str = ''</em>, <em class="sig-param">*args</em>, <em class="sig-param">graph=None</em>, <em class="sig-param">nodemap=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/contig.html#Contig.from_string"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.contig.Contig.from_string" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.contig.Contig.rc">
<code class="sig-name descname">rc</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.contig.Contig.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reverse complement.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.contig.Contig.reverse_complement">
<code class="sig-name descname">reverse_complement</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/contig.html#Contig.reverse_complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.contig.Contig.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reverse complement.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-pydna.design">
<span id="pydna-design-module"></span><h2>pydna.design module<a class="headerlink" href="#module-pydna.design" title="Permalink to this headline">¶</a></h2>
<p>This module contain functions for primer design for various purposes.</p>
<ul class="simple">
<li><p>:func:primer_design for designing primers for a sequence or a matching primer for an existing primer. Returns an <code class="xref py py-class docutils literal notranslate"><span class="pre">Amplicon</span></code> object (same as the <code class="xref py py-mod docutils literal notranslate"><span class="pre">amplify</span></code> module returns).</p></li>
<li><p>:func:assembly_fragments Adds tails to primers for a linear assembly through homologous recombination or Gibson assembly.</p></li>
<li><p>:func:circular_assembly_fragments Adds tails to primers for a circular assembly through homologous recombination or Gibson assembly.</p></li>
</ul>
<dl class="function">
<dt id="pydna.design.assembly_fragments">
<code class="sig-prename descclassname">pydna.design.</code><code class="sig-name descname">assembly_fragments</code><span class="sig-paren">(</span><em class="sig-param">f</em>, <em class="sig-param">overlap=35</em>, <em class="sig-param">maxlink=40</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/design.html#assembly_fragments"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.design.assembly_fragments" title="Permalink to this definition">¶</a></dt>
<dd><p>This function return a list of <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> objects where 
primers have been modified with tails so that the fragments can be fused in 
the order they appear in the list by for example Gibson assembly or homologous 
recombination.</p>
<p>Given that we have two linear <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> objects a and b</p>
<p>we can modify the reverse primer of a and forward primer of b with tails to allow 
fusion by fusion PCR, Gibson assembly or in-vivo homologous recombination.
The basic requirements for the primers for the three techniques are the same.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>                          <span class="o">&lt;--&gt;</span>

 <span class="n">_________</span> <span class="n">a</span> <span class="n">_________</span>           <span class="n">__________</span> <span class="n">b</span> <span class="n">________</span>
<span class="o">/</span>                     \         <span class="o">/</span>                     \
<span class="n">agcctatcatcttggtctctgca</span>         <span class="n">TTTATATCGCATGACTCTTCTTT</span>
<span class="o">|||||||||||||||||||||||</span>         <span class="o">|||||||||||||||||||||||</span>
                 <span class="o">&lt;</span><span class="n">gacgt</span>                          <span class="o">&lt;</span><span class="n">AGAAA</span>
<span class="n">agcct</span><span class="o">&gt;</span>                          <span class="n">TTTAT</span><span class="o">&gt;</span>
<span class="o">|||||||||||||||||||||||</span>         <span class="o">|||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgt</span>         <span class="n">AAATATAGCGTACTGAGAAGAAA</span>

     <span class="n">agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
     <span class="o">||||||||||||||||||||||||||||||||||||||||||||||</span>
     <span class="n">tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
     \<span class="n">___________________</span> <span class="n">c</span> <span class="n">______________________</span><span class="o">/</span>
</pre></div>
</div>
<p>Design tailed primers incorporating a part of the next or previous fragment to be assembled.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgca</span>
<span class="o">|||||||||||||||||||||||</span>
                <span class="n">gagacgtAAATATA</span>

<span class="o">|||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgt</span>

                       <span class="n">TTTATATCGCATGACTCTTCTTT</span>
                       <span class="o">|||||||||||||||||||||||</span>

                <span class="n">ctctgcaTTTATAT</span>
                       <span class="o">|||||||||||||||||||||||</span>
                       <span class="n">AAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>PCR products with flanking sequences are formed in the PCR process.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATAT</span>
<span class="o">||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATA</span>
                \<span class="n">____________</span><span class="o">/</span>

                   <span class="n">identical</span>
                   <span class="n">sequences</span>
                 <span class="n">____________</span>
                <span class="o">/</span>            \
                <span class="n">ctctgcaTTTATATCGCATGACTCTTCTTT</span>
                <span class="o">||||||||||||||||||||||||||||||</span>
                <span class="n">gagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>The fragments can be fused by any of the techniques mentioned earlier to form c:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="o">||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>The first argument of this function is a list of sequence objects containing 
Amplicons and other similar objects.</p>
<p><strong>At least every second sequence object needs to be an Amplicon</strong></p>
<p>This rule exists because if a sequence object is that is not a PCR product
is to be fused with another fragment, that other fragment needs to be an Amplicon
so that the primer of the other object can be modified to include the whole stretch
of sequence homology needed for the fusion. See the example below where a is a 
non-amplicon (a linear plasmid  vector for instance)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="n">_________</span> <span class="n">a</span> <span class="n">_________</span>           <span class="n">__________</span> <span class="n">b</span> <span class="n">________</span>
<span class="o">/</span>                     \         <span class="o">/</span>                     \
<span class="n">agcctatcatcttggtctctgca</span>   <span class="o">&lt;--&gt;</span>  <span class="n">TTTATATCGCATGACTCTTCTTT</span>
<span class="o">|||||||||||||||||||||||</span>         <span class="o">|||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgt</span>                          <span class="o">&lt;</span><span class="n">AGAAA</span>
                                <span class="n">TTTAT</span><span class="o">&gt;</span>
                                <span class="o">|||||||||||||||||||||||</span>
                          <span class="o">&lt;--&gt;</span>  <span class="n">AAATATAGCGTACTGAGAAGAAA</span>

     <span class="n">agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
     <span class="o">||||||||||||||||||||||||||||||||||||||||||||||</span>
     <span class="n">tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
     \<span class="n">___________________</span> <span class="n">c</span> <span class="n">______________________</span><span class="o">/</span>
</pre></div>
</div>
<p>In this case only the forward primer of b is fitted with a tail with a part a:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgca</span>
<span class="o">|||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgt</span>

                       <span class="n">TTTATATCGCATGACTCTTCTTT</span>
                       <span class="o">|||||||||||||||||||||||</span>
                                        <span class="o">&lt;</span><span class="n">AGAAA</span>
         <span class="n">tcttggtctctgcaTTTATAT</span>
                       <span class="o">|||||||||||||||||||||||</span>
                       <span class="n">AAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>PCR products with flanking sequences are formed in the PCR process.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATAT</span>
<span class="o">||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATA</span>
                \<span class="n">____________</span><span class="o">/</span>

                   <span class="n">identical</span>
                   <span class="n">sequences</span>
                 <span class="n">____________</span>
                <span class="o">/</span>            \
                <span class="n">ctctgcaTTTATATCGCATGACTCTTCTTT</span>
                <span class="o">||||||||||||||||||||||||||||||</span>
                <span class="n">gagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>The fragments can be fused by for example Gibson assembly:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATAT</span>
<span class="o">||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaacca</span>

                             <span class="n">TCGCATGACTCTTCTTT</span>
                <span class="o">||||||||||||||||||||||||||||||</span>
                <span class="n">gagacgtAAATATAGCGTACTGAGAAGAAA</span> 
</pre></div>
</div>
<p>to form c:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT</span>
<span class="o">||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="n">tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA</span>
</pre></div>
</div>
<p>The first argument of this function is a list of sequence objects containing 
Amplicons and other similar objects.</p>
<p>The overlap argument controls how many base pairs of overlap required between 
adjacent sequence fragments. In the junction between Amplicons, tails with the 
length of about half of this value is added to the two primers
closest to the junction.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt;       &lt;
Amplicon1
         Amplicon2
         &gt;       &lt;

         ⇣

&gt;       &lt;-
Amplicon1
         Amplicon2
        -&gt;       &lt;                     
</pre></div>
</div>
<p>In the case of an Amplicon adjacent to a Dseqrecord object, the tail will 
be twice as long (1*overlap) since the 
recombining sequence is present entirely on this primer:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Dseqrecd1
         Amplicon1
         &gt;       &lt;

         ⇣

Dseqrecd1
         Amplicon1
       --&gt;       &lt;
</pre></div>
</div>
<p>Note that if the sequence of DNA fragments starts or stops with an Amplicon, 
the very first and very last prinmer will not be modified i.e. assembles are 
always assumed to be linear. There are simple tricks around that for circular
assemblies depicted in the last two examples below.</p>
<p>The maxlink arguments controls the cut off length for sequences that will be
synhtesized by adding them to primers for the adjacent fragment(s). The 
argument list may contain short spacers (such as spacers between fusion proteins).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Example 1: Linear assembly of PCR products (pydna.amplicon.Amplicon class objects) ------

&gt;       &lt;         &gt;       &lt;
Amplicon1         Amplicon3
         Amplicon2         Amplicon4
         &gt;       &lt;         &gt;       &lt;

                     ⇣
                     pydna.design.assembly_fragments
                     ⇣ 

&gt;       &lt;-       -&gt;       &lt;-                      pydna.assembly.Assembly
Amplicon1         Amplicon3                         
         Amplicon2         Amplicon4     ➤  Amplicon1Amplicon2Amplicon3Amplicon4
        -&gt;       &lt;-       -&gt;       &lt;

Example 2: Linear assembly of alternating Amplicons and other fragments

&gt;       &lt;         &gt;       &lt;
Amplicon1         Amplicon2
         Dseqrecd1         Dseqrecd2

                     ⇣
                     pydna.design.assembly_fragments
                     ⇣ 

&gt;       &lt;--     --&gt;       &lt;--                     pydna.assembly.Assembly
Amplicon1         Amplicon2
         Dseqrecd1         Dseqrecd2     ➤  Amplicon1Dseqrecd1Amplicon2Dseqrecd2

Example 3: Linear assembly of alternating Amplicons and other fragments

Dseqrecd1         Dseqrecd2
         Amplicon1         Amplicon2
         &gt;       &lt;       --&gt;       &lt;

                     ⇣
             pydna.design.assembly_fragments
                     ⇣
                                                  pydna.assembly.Assembly
Dseqrecd1         Dseqrecd2
         Amplicon1         Amplicon2     ➤  Dseqrecd1Amplicon1Dseqrecd2Amplicon2
       --&gt;       &lt;--     --&gt;       &lt;

Example 4: Circular assembly of alternating Amplicons and other fragments

                 -&gt;       &lt;==
Dseqrecd1         Amplicon2
         Amplicon1         Dseqrecd1
       --&gt;       &lt;-
                     ⇣
                     pydna.design.assembly_fragments
                     ⇣ 
                                                   pydna.assembly.Assembly
                 -&gt;       &lt;==
Dseqrecd1         Amplicon2                    -Dseqrecd1Amplicon1Amplicon2-  
         Amplicon1                       ➤    |                             |
       --&gt;       &lt;-                            -----------------------------

------ Example 5: Circular assembly of Amplicons

&gt;       &lt;         &gt;       &lt;
Amplicon1         Amplicon3
         Amplicon2         Amplicon1
         &gt;       &lt;         &gt;       &lt;

                     ⇣
                     pydna.design.assembly_fragments
                     ⇣ 

&gt;       &lt;=       -&gt;       &lt;-        
Amplicon1         Amplicon3                  
         Amplicon2         Amplicon1
        -&gt;       &lt;-       +&gt;       &lt;

                     ⇣
             make new Amplicon using the Amplicon1.template and 
             the last fwd primer and the first rev primer.
                     ⇣
                                                   pydna.assembly.Assembly
+&gt;       &lt;=       -&gt;       &lt;-        
 Amplicon1         Amplicon3                  -Amplicon1Amplicon2Amplicon3-
          Amplicon2                      ➤   |                             |
         -&gt;       &lt;-                          -----------------------------
</pre></div>
</div>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>f</strong><span class="classifier">list of <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> and other Dseqrecord like objects</span></dt><dd><p>list Amplicon and Dseqrecord object for which fusion primers should be constructed.</p>
</dd>
<dt><strong>overlap</strong><span class="classifier">int, optional</span></dt><dd><p>Length of required overlap between fragments.</p>
</dd>
<dt><strong>maxlink</strong><span class="classifier">int, optional</span></dt><dd><p>Maximum length of spacer sequences that may be present in f. These will be included in tails for designed primers.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl>
<dt><strong>seqs</strong><span class="classifier">list of <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> and other Dseqrecord like objects <a class="reference internal" href="#pydna.amplicon.Amplicon" title="pydna.amplicon.Amplicon"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplicon.Amplicon</span></code></a> objects</span></dt><dd><div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">Amplicon1</span><span class="p">,</span>
 <span class="n">Amplicon2</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>    
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.design</span> <span class="k">import</span> <span class="n">primer_design</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">primer_design</span><span class="p">(</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatg&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">primer_design</span><span class="p">(</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ccaaacccaccaggtaccttatgtaagtacttcaagtcgccagaagacttcttggtcaagttgcc&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">primer_design</span><span class="p">(</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;tgtactggtgctgaaccttgtatcaagttgggtgttgacgccattgccccaggtggtcgtttcgtt&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.design</span> <span class="k">import</span> <span class="n">assembly_fragments</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># We would like a circular recombination, so the first sequence has to be repeated</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fa1</span><span class="p">,</span><span class="n">fb</span><span class="p">,</span><span class="n">fc</span><span class="p">,</span><span class="n">fa2</span> <span class="o">=</span> <span class="n">assembly_fragments</span><span class="p">([</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span><span class="p">,</span><span class="n">a</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Since all fragments are Amplicons, we need to extract the rp of the 1st and fp of the last fragments.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.amplify</span> <span class="k">import</span> <span class="n">pcr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fa</span> <span class="o">=</span> <span class="n">pcr</span><span class="p">(</span><span class="n">fa2</span><span class="o">.</span><span class="n">forward_primer</span><span class="p">,</span> <span class="n">fa1</span><span class="o">.</span><span class="n">reverse_primer</span><span class="p">,</span> <span class="n">a</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">fa</span><span class="p">,</span><span class="n">fb</span><span class="p">,</span><span class="n">fc</span><span class="p">]</span>
<span class="go">[Amplicon(100), Amplicon(101), Amplicon(102)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.assembly</span> <span class="k">import</span> <span class="n">Assembly</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span> <span class="o">=</span> <span class="n">Assembly</span><span class="p">([</span><span class="n">fa</span><span class="p">,</span><span class="n">fb</span><span class="p">,</span><span class="n">fc</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span>
<span class="go">Assembly</span>
<span class="go">fragments....: 100bp 101bp 102bp</span>
<span class="go">limit(bp)....: 25</span>
<span class="go">G.nodes......: 6</span>
<span class="go">algorithm....: common_sub_strings</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span><span class="o">.</span><span class="n">assemble_linear</span><span class="p">()</span>
<span class="go">[Contig(-231), Contig(-166), Contig(-36)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">cseguid</span><span class="p">()</span>
<span class="go">&#39;V3Mi8zilejgyoH833UbjJOtDMbc&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="o">+</span><span class="n">c</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span><span class="o">.</span><span class="n">cseguid</span><span class="p">()</span>
<span class="go">&#39;V3Mi8zilejgyoH833UbjJOtDMbc&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">assemblyobj</span><span class="o">.</span><span class="n">assemble_circular</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">figure</span><span class="p">())</span>
<span class="go"> -|100bp_PCR_prod|36</span>
<span class="go">|                 \/</span>
<span class="go">|                 /\</span>
<span class="go">|                 36|101bp_PCR_prod|36</span>
<span class="go">|                                   \/</span>
<span class="go">|                                   /\</span>
<span class="go">|                                   36|102bp_PCR_prod|36</span>
<span class="go">|                                                     \/</span>
<span class="go">|                                                     /\</span>
<span class="go">|                                                     36-</span>
<span class="go">|                                                        |</span>
<span class="go"> --------------------------------------------------------</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.design.circular_assembly_fragments">
<code class="sig-prename descclassname">pydna.design.</code><code class="sig-name descname">circular_assembly_fragments</code><span class="sig-paren">(</span><em class="sig-param">f</em>, <em class="sig-param">overlap=35</em>, <em class="sig-param">maxlink=40</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/design.html#circular_assembly_fragments"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.design.circular_assembly_fragments" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.design.primer_design">
<code class="sig-prename descclassname">pydna.design.</code><code class="sig-name descname">primer_design</code><span class="sig-paren">(</span><em class="sig-param">template</em>, <em class="sig-param">fp=None</em>, <em class="sig-param">rp=None</em>, <em class="sig-param">target_tm=55.0</em>, <em class="sig-param">fprimerc=1000.0</em>, <em class="sig-param">rprimerc=1000.0</em>, <em class="sig-param">saltc=50.0</em>, <em class="sig-param">limit=13</em>, <em class="sig-param">formula=&lt;function tmbresluc at 0x7f342598c730&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/design.html#primer_design"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.design.primer_design" title="Permalink to this definition">¶</a></dt>
<dd><p>This function designs a forward primer and a reverse primer for PCR amplification 
of a given template sequence.</p>
<p>The template argument is a Dseqrecord object or equivalent containing the template sequence.</p>
<p>The optional fp and rp arguments can contain an existing primer for the sequence (either the forward or reverse primer).
One or the other primers can be specified, not both (since then there is nothing to design!, use the pydna.amplify.pcr function instead).</p>
<p>If one of the primers is given, the other primer is designed to match in terms of Tm.
If both primers are designed, they will be designed to target_tm</p>
<p>fprimerc, rprimerc and saltc are formward and reverse primer concentration (nM). Saltc is the salt concentration. 
These arguments might affect how Tm is calculated.</p>
<p>formula is a function that can take at least three arguments f( str, primerc=float, saltc=float).
There are several of these in the pydna.tm module.</p>
<p>The function returns a pydna.amplicon.Amplicon class instance. This object has 
the object.forward_primer and object.reverse_primer properties which contain the designed primers.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>template</strong><span class="classifier">pydna.dseqrecord.Dseqrecord</span></dt><dd><p>a Dseqrecord object. The only required argument.</p>
</dd>
<dt><strong>fp, rp</strong><span class="classifier">pydna.primer.Primer, optional</span></dt><dd><p>optional pydna.primer.Primer objects containing one primer each.</p>
</dd>
<dt><strong>target_tm</strong><span class="classifier">float, optional</span></dt><dd><p>target tm for the primers, set to 55°C by default.</p>
</dd>
<dt><strong>fprimerc</strong><span class="classifier">float, optional</span></dt><dd><p>Concentration of forward primer in nM, set to 1000.0 nM by default.</p>
</dd>
<dt><strong>rprimerc</strong><span class="classifier">float, optional</span></dt><dd><p>Concentration of reverse primer in nM, set to 1000.0 nM by default.</p>
</dd>
<dt><strong>saltc</strong><span class="classifier">float, optional</span></dt><dd><p>Salt concentration (monovalet cations) <code class="xref py py-mod docutils literal notranslate"><span class="pre">tmbresluc</span></code> set to 50.0 mM by default</p>
</dd>
<dt><strong>formula</strong><span class="classifier">function</span></dt><dd><p>formula used for tm calculation
this is the name of a function.
built in options are:</p>
<ol class="arabic simple">
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.amplify.tmbresluc()</span></code> (default)</p></li>
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.amplify.basictm()</span></code></p></li>
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.amplify.tmstaluc98()</span></code></p></li>
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.amplify.tmbreslauer86()</span></code></p></li>
</ol>
<p>These functions are imported from the <a class="reference internal" href="#module-pydna.amplify" title="pydna.amplify"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pydna.amplify</span></code></a> module, but can be
substituted for some other custom made function.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>result</strong><span class="classifier">Amplicon</span></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;atgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span>
<span class="go">Dseqrecord(-64)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.design</span> <span class="k">import</span> <span class="n">primer_design</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span> <span class="o">=</span> <span class="n">primer_design</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span>
<span class="go">Amplicon(64)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span><span class="o">.</span><span class="n">forward_primer</span>
<span class="go">f64 18-mer:5&#39;-atgactgctaacccttcc-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span><span class="o">.</span><span class="n">reverse_primer</span>
<span class="go">r64 19-mer:5&#39;-catcgtaagtttcgaacga-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">ampl</span><span class="o">.</span><span class="n">figure</span><span class="p">())</span>
<span class="go">5atgactgctaacccttcc...tcgttcgaaacttacgatg3</span>
<span class="go">                      ||||||||||||||||||| tm 53.8 (dbd) 60.6</span>
<span class="go">                     3agcaagctttgaatgctac5</span>
<span class="go">5atgactgctaacccttcc3</span>
<span class="go"> |||||||||||||||||| tm 54.4 (dbd) 58.4</span>
<span class="go">3tactgacgattgggaagg...agcaagctttgaatgctac5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span> <span class="o">=</span> <span class="s2">&quot;GGATCC&quot;</span> <span class="o">+</span> <span class="n">ampl</span><span class="o">.</span><span class="n">forward_primer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span> <span class="o">=</span> <span class="s2">&quot;GGATCC&quot;</span> <span class="o">+</span> <span class="n">ampl</span><span class="o">.</span><span class="n">reverse_primer</span>  
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span>
<span class="go">f64 24-mer:5&#39;-GGATCCatgactgct..tcc-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span>
<span class="go">r64 25-mer:5&#39;-GGATCCcatcgtaag..cga-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.amplify</span> <span class="k">import</span> <span class="n">pcr</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr_prod</span> <span class="o">=</span> <span class="n">pcr</span><span class="p">(</span><span class="n">pf</span><span class="p">,</span> <span class="n">pr</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pcr_prod</span><span class="o">.</span><span class="n">figure</span><span class="p">())</span>
<span class="go">      5atgactgctaacccttcc...tcgttcgaaacttacgatg3</span>
<span class="go">                            ||||||||||||||||||| tm 53.8 (dbd) 60.6</span>
<span class="go">                           3agcaagctttgaatgctacCCTAGG5</span>
<span class="go">5GGATCCatgactgctaacccttcc3</span>
<span class="go">       |||||||||||||||||| tm 54.4 (dbd) 58.4</span>
<span class="go">      3tactgacgattgggaagg...agcaagctttgaatgctac5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pcr_prod</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="go">GGATCCatgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatgGGATCC</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.primer</span> <span class="k">import</span> <span class="n">Primer</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span> <span class="o">=</span> <span class="n">Primer</span><span class="p">(</span><span class="s2">&quot;atgactgctaacccttccttggtgttg&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;myprimer&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span> <span class="o">=</span> <span class="n">primer_design</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">fp</span> <span class="o">=</span> <span class="n">pf</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span><span class="o">.</span><span class="n">forward_primer</span>
<span class="go">myprimer 27-mer:5&#39;-atgactgctaaccct..ttg-3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span><span class="o">.</span><span class="n">reverse_primer</span>
<span class="go">r64 28-mer:5&#39;-catcgtaagtttcga..gtc-3&#39;</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="module-pydna.download">
<span id="pydna-download-module"></span><h2>pydna.download module<a class="headerlink" href="#module-pydna.download" title="Permalink to this headline">¶</a></h2>
<p>Provides a function for downloading online text files</p>
</div>
<div class="section" id="module-pydna.editor">
<span id="pydna-editor-module"></span><h2>pydna.editor module<a class="headerlink" href="#module-pydna.editor" title="Permalink to this headline">¶</a></h2>
<p>This module provides a class for opening a sequence using an editor
that accepts a file as a command line argument.</p>
<p>ApE - A plasmid Editor <a class="footnote-reference brackets" href="#id18" id="id17">9</a> is and excellent editor for this purpose.</p>
<div class="section" id="references">
<h3>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h3>
<dl class="footnote brackets">
<dt class="label" id="id18"><span class="brackets"><a class="fn-backref" href="#id17">9</a></span></dt>
<dd><p><a class="reference external" href="http://biologylabs.utah.edu/jorgensen/wayned/ape/">http://biologylabs.utah.edu/jorgensen/wayned/ape/</a></p>
</dd>
</dl>
<dl class="class">
<dt id="pydna.editor.Editor">
<em class="property">class </em><code class="sig-prename descclassname">pydna.editor.</code><code class="sig-name descname">Editor</code><span class="sig-paren">(</span><em class="sig-param">shell_command_for_editor</em>, <em class="sig-param">tmpdir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/editor.html#Editor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.editor.Editor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.6/library/functions.html#object" title="(in Python v3.6)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>The Editor class needs to be instantiated before use.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>shell_command_for_editor</strong><span class="classifier">str</span></dt><dd><p>String containing the path to the editor</p>
</dd>
<dt><strong>tmpdir</strong><span class="classifier">str, optional</span></dt><dd><p>String containing path to the temprary directory where sequence
files are stored before opening.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#ape = pydna.Editor(&quot;tclsh8.6 /home/bjorn/.ApE/apeextractor/ApE.vfs/lib/app-AppMain/AppMain.tcl&quot;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#ape.open(&quot;aaa&quot;) # This command opens the sequence in the ApE editor</span>
</pre></div>
</div>
<dl class="method">
<dt id="pydna.editor.Editor.open">
<code class="sig-name descname">open</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">seq_to_open</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/editor.html#Editor.open"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.editor.Editor.open" title="Permalink to this definition">¶</a></dt>
<dd><p>Open a sequence for editing in an external (DNA) editor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>args</strong><span class="classifier">SeqRecord or Dseqrecord object</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="pydna.editor.ape">
<code class="sig-prename descclassname">pydna.editor.</code><code class="sig-name descname">ape</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/editor.html#ape"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.editor.ape" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
</div>
<div class="section" id="module-pydna.gel">
<span id="pydna-gel-module"></span><h2>pydna.gel module<a class="headerlink" href="#module-pydna.gel" title="Permalink to this headline">¶</a></h2>
<p>Provides the class <cite>Gel</cite> for the simulation of agarose slab-gel
electrophoresis of DNA at constant electric field.</p>
<div class="section" id="note">
<h3>Note<a class="headerlink" href="#note" title="Permalink to this headline">¶</a></h3>
<p>This code is at an early stage of development and documentation.</p>
<dl class="function">
<dt id="pydna.gel.DKuhn">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">DKuhn</code><span class="sig-paren">(</span><em class="sig-param">kB</em>, <em class="sig-param">T</em>, <em class="sig-param">eta</em>, <em class="sig-param">l</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.DKuhn" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.Dblob">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">Dblob</code><span class="sig-paren">(</span><em class="sig-param">kB</em>, <em class="sig-param">T</em>, <em class="sig-param">eta</em>, <em class="sig-param">a</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.Dblob" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pydna.gel.Fakeseq">
<em class="property">class </em><code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">Fakeseq</code><span class="sig-paren">(</span><em class="sig-param">l</em>, <em class="sig-param">n=1e-11</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#Fakeseq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.Fakeseq" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.6/library/functions.html#object" title="(in Python v3.6)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
</dd></dl>

<dl class="function">
<dt id="pydna.gel.Gauss_FWHM">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">Gauss_FWHM</code><span class="sig-paren">(</span><em class="sig-param">FWTM</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.Gauss_FWHM" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.Gauss_dev">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">Gauss_dev</code><span class="sig-paren">(</span><em class="sig-param">FWHM</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.Gauss_dev" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.Gauss_hgt">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">Gauss_hgt</code><span class="sig-paren">(</span><em class="sig-param">auc</em>, <em class="sig-param">dev</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.Gauss_hgt" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.Gaussian">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">Gaussian</code><span class="sig-paren">(</span><em class="sig-param">x</em>, <em class="sig-param">hgt</em>, <em class="sig-param">ctr</em>, <em class="sig-param">dev</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.Gaussian" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="pydna.gel.Gel">
<em class="property">class </em><code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">Gel</code><span class="sig-paren">(</span><em class="sig-param">samples, names=None, percentgel=&lt;Quantity(1.0, 'gram / milliliter')&gt;, electrfield=&lt;Quantity(5.0, 'volt / centimeter')&gt;, temperature=&lt;Quantity(295.15, 'kelvin')&gt;, gel_len=&lt;Quantity(8, 'centimeter')&gt;, wellxy=&lt;Quantity([7 2], 'millimeter')&gt;, wellsep=&lt;Quantity(2, 'millimeter')&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#Gel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.Gel" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.6/library/functions.html#object" title="(in Python v3.6)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Gel object for DNA slab-gel electrophoresis simulation.</p>
<p>Class designed to store information regarding an electrophoresis
experimental setup. Includes gel’s and well’s dimensions,
DNA samples loaded, agarose concentration, electric field
intensity and temperature.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>samples</strong><span class="classifier">list of lists of pydna.Dseqrecord objects</span></dt><dd><p>List of samples with the DNA fragments to separate.
Each sample corresponds to a different well.
Accepts lists of pydna.Dseq objects and converts them to
Dseqrecords with the default quantity (<cite>n</cite> parameter).</p>
</dd>
<dt><strong>names</strong><span class="classifier">list of str, optional</span></dt><dd><p>List with the identifiers of the samples.
Defaults to None, in which case the identifiers are of the form
“lane i” where i is the index.</p>
</dd>
<dt><strong>percentgel</strong><span class="classifier">pint.unit.Quantity object, float or int, optional</span></dt><dd><p>Agarose concentration in the gel.
Defaults to Q_(1.0, ‘(g/(100 mL))*100’).
If a float or int is given assumes grams of agarose per milliliter
of buffer.</p>
</dd>
<dt><strong>electrfield</strong><span class="classifier">pint.unit.Quantity object, float or int, optional</span></dt><dd><p>Electric field intensity.
Defaults to Q_(5.0, ‘V/cm’).
If a float or int is given assumes Volts per centimeter.</p>
</dd>
<dt><strong>temperature</strong><span class="classifier">pint.unit.Quantity object, float or int, optional</span></dt><dd><p>Absolute temperature.
Defaults to Q_(295.15, ‘K’).
If a float or int is given assumes Kelvin.
The temperature is only considered in the calculation of the
diffusional component of the band width.</p>
</dd>
<dt><strong>gel_len</strong><span class="classifier">pint.unit.Quantity object, float or int, optional</span></dt><dd><p>Gel length (as measured from the bottom of the well in the
direction of the DNA migration).
Defaults to Q_(8, ‘cm’).
If a float or int is given assumes centimeters.</p>
</dd>
<dt><strong>wellxy</strong><span class="classifier">iterable of pint.unit.Quantity object, float or int, optional</span></dt><dd><p>Well’s dimensions.
Well’s width (perpendicular to the migration direction) and
well’s height (migration direction).
Defaults to Q_([7, 2], ‘mm’).
If a float or int is given assumes millimeters.
The well’s width is merely used for aesthetic purposes.
The well’s height is used to compute the initial band width.</p>
</dd>
<dt><strong>wellsep</strong><span class="classifier">pint.unit.Quantity object, float or int, optional</span></dt><dd><p>Separation between wells.
Defaults to Q_(2, ‘mm’).
Merelly for aestetic purposes.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>The DNA is diluted in the full volume of the well.</p></li>
<li><p>The previous implies that the injection size, which relates to</p></li>
</ul>
<p>the initial bandwidth, is assumed to be well height (welly).
* The DNA electrophoretic mobility in the well is given by the free
solution mobility and is equal for every fragment.</p>
<dl class="method">
<dt id="pydna.gel.Gel.run">
<code class="sig-name descname">run</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">till_len=0.75</em>, <em class="sig-param">till_time=None</em>, <em class="sig-param">exposure=0.5</em>, <em class="sig-param">plot=True</em>, <em class="sig-param">res=&lt;Quantity(500</em>, <em class="sig-param">'pixel / inch')&gt;</em>, <em class="sig-param">cursor_ovr={'hover': False}</em>, <em class="sig-param">back_col=0.3</em>, <em class="sig-param">band_col=1</em>, <em class="sig-param">well_col=0.05</em>, <em class="sig-param">noise=0.015</em>, <em class="sig-param">detectlim=0.04</em>, <em class="sig-param">interpol='linear'</em>, <em class="sig-param">dset_name='vertical'</em>, <em class="sig-param">replNANs=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#Gel.run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.Gel.run" title="Permalink to this definition">¶</a></dt>
<dd><p>Run electrophoresis.</p>
<p>Run the electrophoresis procedure until a specified time or until
the quickest band reaches a specified gel length.
If both conditions are given the most stringent will be respected.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>till_len</strong><span class="classifier">float (0 to 1), optional</span></dt><dd><p>Fraction of the gel length (measured from the well bottom) to
serve as finish line.
Defaults to 0.75.</p>
</dd>
<dt><strong>till_time</strong><span class="classifier">pint.unit.Quantity object, float or int, optional</span></dt><dd><p>Time at which to stop the electrophoresis.
Defaults to None.
If float or int is given assumes hours.</p>
</dd>
<dt><strong>exposure</strong><span class="classifier">float (0 to 1)</span></dt><dd><p>Fraction of signal saturation.
If exposure is set to 0 every band’s light intensity will be
given by a perfect Gaussian curve. Wider bands with lower
amounts of DNA might be hard to see.
The closer to 1 the exposure the higher the saturation, which
favors the weaker bands visualization.
If exposure is set to 1 only the weakest band will be a
Gaussian, all others will be saturated.
Defaults to 0.5.</p>
</dd>
<dt><strong>plot</strong><span class="classifier">{True, False}, optional</span></dt><dd><p>Whether to draw and return the gel picture or not.
The electrophoresis data (such as migrated distances) will be
stored in the Gel object either way.</p>
</dd>
<dt><strong>res</strong><span class="classifier">pint.unit.Quantity object, float or int, optional</span></dt><dd><p>Resolution used to construct an array of light intensities
corresponding to the gel picture.
The dimensions of the intensities array are given by the gel’s
dimensions multiplied by the resolution.
A higher resolution implies more calculations and might cause
the method to become slow.
A lower resolution might result in a very pixelated image.
Defaults to Q_(500, ‘px/in’).
If float or int is given assumes pixels per inch.
This resolution won’t necessarily convey to the final gel
picture since the intensity array is processed by
matplotlib.pyplot.imshow.</p>
</dd>
<dt><strong>cursor_ovr</strong><span class="classifier">dict, optional</span></dt><dd><p>Key arguments to be passed to mpldatacursor.datacursor which
provides interactive cursor functions to the resulting plot.
Defaults to dict(hover=False).</p>
</dd>
<dt><strong>back_col</strong><span class="classifier">float or int, optional</span></dt><dd><p>Background color (light intensity).
Defaults to 0.3.
Solid black (0) allows the maximum contrast.</p>
</dd>
<dt><strong>band_col</strong><span class="classifier">float or int, optional</span></dt><dd><p>Band color (maximum light intensity).
Defaults to 1.
White (1) allows for maximum contrast.</p>
</dd>
<dt><strong>well_col</strong><span class="classifier">float or int, optional</span></dt><dd><p>Well color (light intensity).
Defaults to 0.05.</p>
</dd>
<dt><strong>noise</strong><span class="classifier">float or int, optional</span></dt><dd><p>Standard deviation used to generate a Normal distribution of
noise centered in the background color. This effect is purely
aesthetic.
Defaults to 0.015.</p>
</dd>
<dt><strong>detectlim</strong><span class="classifier">float, optional</span></dt><dd><p>Minimal light intensity difference between the center of the band
and the background color (back_col) below which the band is
considered to be indistinguishable and a white doted outline is
drawn around it.
Defaults to 0.04.</p>
</dd>
<dt><strong>interpol</strong><span class="classifier">{‘linear’, ‘cubic’, ‘nearest’}, optional</span></dt><dd><p>Interpolation method. This is passed to
scipy.interpolate.griddata for the interpolation of the
vWBR’s parameters from the experimental datasets
(mu_S, mu_L, gamma = f(field, agarose concentration) <span id="id19">[1-3]</span>.
Defaults to ‘linear’ which is more conservative.
‘nearest’ is not expected to provide good results.</p>
</dd>
<dt><strong>dset_name</strong><span class="classifier">{‘vertical’, ‘horizontal’}, optional</span></dt><dd><p>Dataset identifier (str). Identifies the dataset to use.
Currently two datasets are available (‘vertical’ and
‘horizontal’) <span id="id20">[1-3]</span>.
The names alude to the geometry of the experimental setup.
The geometry however is not expected to cause significant
differences on the results.</p>
</dd>
<dt><strong>replNANs</strong><span class="classifier">{True, False}, optional</span></dt><dd><p>Whether to replace NANs (not a number) by ‘nearest’
interpolation. NANs will only be produced if the conditions
of electric field intensity and agarose concentration go
beyond the concave space provided by the dataset.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="citation">
<dt class="label" id="re9fd70fb779e-1"><span class="brackets">1</span></dt>
<dd><p>Van Winkle, D.H., Beheshti, A., Rill, R.L.: DNA</p>
</dd>
</dl>
<p>electrophoresis in agarose gels: A simple relation describing the
length dependence of mobility. ELECTROPHORESIS 23(1), 15–19 (2002)</p>
<dl class="citation">
<dt class="label" id="re9fd70fb779e-2"><span class="brackets">2</span></dt>
<dd><p>Rill, R.L., Beheshti, A., Van Winkle, D.H.: DNA</p>
</dd>
</dl>
<p>electrophoresis in agarose gels: Effects of field and gel
concentration on the exponential dependence of reciprocal mobility
on DNA length. ELECTROPHORESIS 23(16), 2710–2719 (2002)</p>
<dl class="citation">
<dt class="label" id="re9fd70fb779e-3"><span class="brackets">3</span></dt>
<dd><p>Beheshti, A.: DNA Electrophoresis in th Agarose Gels: A New</p>
</dd>
</dl>
<p>Mobility vs. DNA Length Dependence. Electronic Theses, Treatises and
Dissertations (Paper 1207) (2002)</p>
</dd></dl>

<dl class="method">
<dt id="pydna.gel.Gel.set_DNAspace_for_mu0">
<code class="sig-name descname">set_DNAspace_for_mu0</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">DNA_space</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#Gel.set_DNAspace_for_mu0"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.Gel.set_DNAspace_for_mu0" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.gel.Gel.set_DNAspace_for_vWBRfit">
<code class="sig-name descname">set_DNAspace_for_vWBRfit</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">DNA_space</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#Gel.set_DNAspace_for_vWBRfit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.Gel.set_DNAspace_for_vWBRfit" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.gel.Gel.set_Tvals_for_mu0">
<code class="sig-name descname">set_Tvals_for_mu0</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">Tvals</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#Gel.set_Tvals_for_mu0"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.Gel.set_Tvals_for_mu0" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.gel.Gel.set_field">
<code class="sig-name descname">set_field</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">electrfield</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#Gel.set_field"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.Gel.set_field" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.gel.Gel.set_gelength">
<code class="sig-name descname">set_gelength</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">gel_len</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#Gel.set_gelength"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.Gel.set_gelength" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.gel.Gel.set_percentgel">
<code class="sig-name descname">set_percentgel</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">percentgel</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#Gel.set_percentgel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.Gel.set_percentgel" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.gel.Gel.set_temperature">
<code class="sig-name descname">set_temperature</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">temperature</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#Gel.set_temperature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.Gel.set_temperature" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.gel.Gel.set_wellsep">
<code class="sig-name descname">set_wellsep</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">wellsep</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#Gel.set_wellsep"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.Gel.set_wellsep" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.gel.Gel.set_wellx">
<code class="sig-name descname">set_wellx</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">wellx</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#Gel.set_wellx"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.Gel.set_wellx" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.gel.Gel.set_welly">
<code class="sig-name descname">set_welly</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">welly</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#Gel.set_welly"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.Gel.set_welly" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="function">
<dt id="pydna.gel.H2Oviscosity">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">H2Oviscosity</code><span class="sig-paren">(</span><em class="sig-param">T</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.H2Oviscosity" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.NKuhn_to_N">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">NKuhn_to_N</code><span class="sig-paren">(</span><em class="sig-param">NKuhn</em>, <em class="sig-param">l</em>, <em class="sig-param">a</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.NKuhn_to_N" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.NKuhn_to_Nbp">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">NKuhn_to_Nbp</code><span class="sig-paren">(</span><em class="sig-param">NKuhn</em>, <em class="sig-param">b</em>, <em class="sig-param">l</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.NKuhn_to_Nbp" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.N_to_NKuhn">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">N_to_NKuhn</code><span class="sig-paren">(</span><em class="sig-param">N</em>, <em class="sig-param">a</em>, <em class="sig-param">l</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.N_to_NKuhn" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.N_to_Nbp">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">N_to_Nbp</code><span class="sig-paren">(</span><em class="sig-param">N</em>, <em class="sig-param">a</em>, <em class="sig-param">b</em>, <em class="sig-param">l</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.N_to_Nbp" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.Nbp_to_N">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">Nbp_to_N</code><span class="sig-paren">(</span><em class="sig-param">Nbp</em>, <em class="sig-param">a</em>, <em class="sig-param">b</em>, <em class="sig-param">l</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.Nbp_to_N" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.Nbp_to_NKuhn">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">Nbp_to_NKuhn</code><span class="sig-paren">(</span><em class="sig-param">Nbp</em>, <em class="sig-param">b</em>, <em class="sig-param">l</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.Nbp_to_NKuhn" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.Ogston_Rouse">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">Ogston_Rouse</code><span class="sig-paren">(</span><em class="sig-param">Nbp</em>, <em class="sig-param">kB</em>, <em class="sig-param">T</em>, <em class="sig-param">a</em>, <em class="sig-param">eta</em>, <em class="sig-param">b</em>, <em class="sig-param">l</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.Ogston_Rouse" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.Ogston_Zimm">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">Ogston_Zimm</code><span class="sig-paren">(</span><em class="sig-param">D0</em>, <em class="sig-param">g</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.Ogston_Zimm" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.Zimm_Rouse">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">Zimm_Rouse</code><span class="sig-paren">(</span><em class="sig-param">x0</em>, <em class="sig-param">args</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.Zimm_Rouse" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.Zimm_g">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">Zimm_g</code><span class="sig-paren">(</span><em class="sig-param">Nbp</em>, <em class="sig-param">DRouse</em>, <em class="sig-param">qeff</em>, <em class="sig-param">mu0</em>, <em class="sig-param">kB</em>, <em class="sig-param">T</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.Zimm_g" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.accel_plateau">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">accel_plateau</code><span class="sig-paren">(</span><em class="sig-param">epsilon</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.accel_plateau" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.assign_quantitiesB">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">assign_quantitiesB</code><span class="sig-paren">(</span><em class="sig-param">samples</em>, <em class="sig-param">maxdef=&lt;Quantity(150</em>, <em class="sig-param">'nanogram')&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#assign_quantitiesB"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.assign_quantitiesB" title="Permalink to this definition">¶</a></dt>
<dd><p>Assigns quantities (masses in nanograms) to the DNA fragments without
corresponding quantity assuming a linear relationship between the DNA
length (in basepairs) and its mass. As if the fragments originated in
a restriction procedure.
For each sample takes the maximum quantity (either from the other samples
or from the default) and assigns it to the fragment with greater length.</p>
</dd></dl>

<dl class="function">
<dt id="pydna.gel.bandbroadening">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">bandbroadening</code><span class="sig-paren">(</span><em class="sig-param">D</em>, <em class="sig-param">time</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.bandbroadening" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.contour_length">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">contour_length</code><span class="sig-paren">(</span><em class="sig-param">Nbp</em>, <em class="sig-param">b</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.contour_length" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.diff_Zimm_Rouse">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">diff_Zimm_Rouse</code><span class="sig-paren">(</span><em class="sig-param">Nbp</em>, <em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#diff_Zimm_Rouse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.diff_Zimm_Rouse" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.diffusion_coefficient">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">diffusion_coefficient</code><span class="sig-paren">(</span><em class="sig-param">Nbp</em>, <em class="sig-param">N_lim1</em>, <em class="sig-param">N_lim2</em>, <em class="sig-param">N_lim3</em>, <em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#diffusion_coefficient"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.diffusion_coefficient" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.dim_or_units">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">dim_or_units</code><span class="sig-paren">(</span><em class="sig-param">quantity</em>, <em class="sig-param">reference</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#dim_or_units"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.dim_or_units" title="Permalink to this definition">¶</a></dt>
<dd><p>If &lt;quantity&gt; is not an instance of &lt;Quantity&gt;, instantiate a &lt;Quantity&gt;
object with &lt;quantity&gt; as magnitude and as &lt;reference.units&gt; as units.
If &lt;quantity&gt; is an instance of &lt;Quantity&gt;, check whether it has the same
dimensionality as &lt;reference&gt;.</p>
</dd></dl>

<dl class="function">
<dt id="pydna.gel.equil_accel">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">equil_accel</code><span class="sig-paren">(</span><em class="sig-param">epsilon</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.equil_accel" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.ferguson_to_mu0">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">ferguson_to_mu0</code><span class="sig-paren">(</span><em class="sig-param">field</em>, <em class="sig-param">Tvals</em>, <em class="sig-param">DNAvals</em>, <em class="sig-param">dataset</em>, <em class="sig-param">mu_func</em>, <em class="sig-param">adjmethod='linear'</em>, <em class="sig-param">replNANs=True</em>, <em class="sig-param">plot=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#ferguson_to_mu0"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.ferguson_to_mu0" title="Permalink to this definition">¶</a></dt>
<dd><p>This function extrapolates the free solution mobility (mu0)
for a specified electric field intensity (field), via Ferguson plot
(ln(mobility) vs. %agarose).</p>
<p>Mobiliy calculation method:
[E,T,muS,muL,gamma] -&gt; [E,T,mu(L)] -(L*)-&gt; [E,T,mu] -(E*,T*,interp.)-&gt; mu*</p>
</dd></dl>

<dl class="function">
<dt id="pydna.gel.flatten">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">flatten</code><span class="sig-paren">(</span><em class="sig-param">List</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#flatten"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.flatten" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.free_solution">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">free_solution</code><span class="sig-paren">(</span><em class="sig-param">kB</em>, <em class="sig-param">T</em>, <em class="sig-param">eta</em>, <em class="sig-param">Rh</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.free_solution" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.gelplot_imshow">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">gelplot_imshow</code><span class="sig-paren">(</span><em class="sig-param">distances</em>, <em class="sig-param">bandwidths</em>, <em class="sig-param">intensities</em>, <em class="sig-param">lanes</em>, <em class="sig-param">names</em>, <em class="sig-param">gel_len</em>, <em class="sig-param">wellx</em>, <em class="sig-param">welly</em>, <em class="sig-param">wellsep</em>, <em class="sig-param">res</em>, <em class="sig-param">cursor_ovr</em>, <em class="sig-param">back_col</em>, <em class="sig-param">band_col</em>, <em class="sig-param">well_col</em>, <em class="sig-param">noise</em>, <em class="sig-param">Itol</em>, <em class="sig-param">detectlim</em>, <em class="sig-param">title</em>, <em class="sig-param">FWTM</em>, <em class="sig-param">show=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#gelplot_imshow"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.gelplot_imshow" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.gen_sample">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">gen_sample</code><span class="sig-paren">(</span><em class="sig-param">sizes</em>, <em class="sig-param">quantities</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#gen_sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.gen_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Return list of pydna Dseqrecords of given size and quantity.</p>
<p>If a single quantity is given it is divided by the DNA fragments
in proportion to their length.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sizes</strong><span class="classifier">iterable of pint.unit.Quantity objects or ints</span></dt><dd><p>List of DNA sizes in base pairs (bp).</p>
</dd>
<dt><strong>quantities</strong><span class="classifier">iterable of pint.unit.Quantity objects, floats or ints</span></dt><dd><p>List of DNA weights in nanograms (ng).
If a single quantity is given (pint.unit.Quantity object, float
or int) it is divided linearly by the DNA fragments length.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Direct quantity assignment without declared units.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sizes</span> <span class="o">=</span> <span class="p">[</span><span class="mi">3000</span><span class="p">,</span> <span class="mi">500</span><span class="p">,</span> <span class="mi">1500</span><span class="p">]</span>  <span class="c1"># bp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">qts</span> <span class="o">=</span> <span class="p">[</span><span class="mf">70.0</span><span class="p">,</span> <span class="mf">11.7</span><span class="p">,</span> <span class="mf">35.0</span><span class="p">]</span>  <span class="c1"># ng</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sample</span> <span class="o">=</span> <span class="n">gen_sample</span><span class="p">(</span><span class="n">sizes</span><span class="p">,</span> <span class="n">qts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sample</span>
<span class="go">[Dseqrecord(-3000), Dseqrecord(-500), Dseqrecord(-1500)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">float</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0:.2f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">sample</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">m</span><span class="p">()</span> <span class="o">*</span> <span class="mf">1E6</span><span class="p">))</span>  <span class="c1"># µg</span>
<span class="go">0.07</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Q_</span><span class="p">([</span><span class="n">dna</span><span class="o">.</span><span class="n">m</span><span class="p">()</span> <span class="k">for</span> <span class="n">dna</span> <span class="ow">in</span> <span class="n">sample</span><span class="p">],</span> <span class="s1">&#39;g&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="s1">&#39;ng&#39;</span><span class="p">)</span> <span class="c1"># doctest: +SKIP</span>
<span class="go">&lt;Quantity([70.  11.7 35. ], &#39;nanogram&#39;)&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Q_</span><span class="p">([</span><span class="n">dna</span><span class="o">.</span><span class="n">n</span> <span class="k">for</span> <span class="n">dna</span> <span class="ow">in</span> <span class="n">sample</span><span class="p">],</span> <span class="s1">&#39;mol&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="s1">&#39;pmol&#39;</span><span class="p">)</span> <span class="c1"># doctest: +SKIP</span>
<span class="go">&lt;Quantity([0.03776682 0.03786665 0.03776521], &#39;picomole&#39;)&gt;</span>
</pre></div>
</div>
<p>Direct quantity assignment with declared units.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sizes</span> <span class="o">=</span> <span class="n">Q_</span><span class="p">([</span><span class="mi">3000</span><span class="p">,</span> <span class="mi">500</span><span class="p">,</span> <span class="mi">1500</span><span class="p">],</span> <span class="s1">&#39;bp&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">qts</span> <span class="o">=</span> <span class="n">Q_</span><span class="p">([</span><span class="mf">70.0</span><span class="p">,</span> <span class="mf">11.7</span><span class="p">,</span> <span class="mf">35.0</span><span class="p">],</span>  <span class="s1">&#39;ng&#39;</span><span class="p">)</span> 
<span class="gp">&gt;&gt;&gt; </span><span class="n">sample</span> <span class="o">=</span> <span class="n">gen_sample</span><span class="p">(</span><span class="n">sizes</span><span class="p">,</span> <span class="n">qts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sample</span>
<span class="go">[Dseqrecord(-3000), Dseqrecord(-500), Dseqrecord(-1500)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Q_</span><span class="p">([</span><span class="n">dna</span><span class="o">.</span><span class="n">m</span><span class="p">()</span> <span class="k">for</span> <span class="n">dna</span> <span class="ow">in</span> <span class="n">sample</span><span class="p">],</span> <span class="s1">&#39;g&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="s1">&#39;ng&#39;</span><span class="p">)</span> <span class="c1"># doctest: +SKIP</span>
<span class="go">&lt;Quantity([70.  11.7 35. ], &#39;nanogram&#39;)&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Q_</span><span class="p">([</span><span class="n">dna</span><span class="o">.</span><span class="n">n</span> <span class="k">for</span> <span class="n">dna</span> <span class="ow">in</span> <span class="n">sample</span><span class="p">],</span> <span class="s1">&#39;mol&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="s1">&#39;pmol&#39;</span><span class="p">)</span> <span class="c1"># doctest: +SKIP</span>
<span class="go">&lt;Quantity([0.03776682 0.03786665 0.03776521], &#39;picomole&#39;)&gt;</span>
</pre></div>
</div>
<p>Assignment of total quantity (with declared units).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sample</span> <span class="o">=</span> <span class="n">gen_sample</span><span class="p">(</span><span class="n">Q_</span><span class="p">([</span><span class="mi">3000</span><span class="p">,</span> <span class="mi">500</span><span class="p">,</span> <span class="mi">1500</span><span class="p">],</span> <span class="s1">&#39;bp&#39;</span><span class="p">),</span> <span class="n">Q_</span><span class="p">(</span><span class="mi">200</span><span class="p">,</span> <span class="s1">&#39;ng&#39;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sample</span>
<span class="go">[Dseqrecord(-3000), Dseqrecord(-500), Dseqrecord(-1500)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Q_</span><span class="p">([</span><span class="n">dna</span><span class="o">.</span><span class="n">m</span><span class="p">()</span> <span class="k">for</span> <span class="n">dna</span> <span class="ow">in</span> <span class="n">sample</span><span class="p">],</span> <span class="s1">&#39;g&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="s1">&#39;ng&#39;</span><span class="p">)</span> <span class="c1"># doctest: +SKIP</span>
<span class="go">&lt;Quantity([120.  20.  60.], &#39;nanogram&#39;)&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Q_</span><span class="p">([</span><span class="n">dna</span><span class="o">.</span><span class="n">n</span> <span class="k">for</span> <span class="n">dna</span> <span class="ow">in</span> <span class="n">sample</span><span class="p">],</span> <span class="s1">&#39;mol&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="s1">&#39;pmol&#39;</span><span class="p">)</span> <span class="c1"># doctest: +SKIP</span>
<span class="go">&lt;Quantity([0.06474311 0.06472932 0.06474035], &#39;picomole&#39;)&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.gel.lin_div_Qty">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">lin_div_Qty</code><span class="sig-paren">(</span><em class="sig-param">sample</em>, <em class="sig-param">quantity</em>, <em class="sig-param">criteria=&lt;built-in function len&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#lin_div_Qty"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.lin_div_Qty" title="Permalink to this definition">¶</a></dt>
<dd><p>Linearly divides a quantity by the elements of sample considering a
criteria. Criteria must by a function that returns a number upon being
applied to each element of sample.</p>
</dd></dl>

<dl class="function">
<dt id="pydna.gel.logspace_int">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">logspace_int</code><span class="sig-paren">(</span><em class="sig-param">minimum</em>, <em class="sig-param">maximum</em>, <em class="sig-param">divs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#logspace_int"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.logspace_int" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.pore_size">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">pore_size</code><span class="sig-paren">(</span><em class="sig-param">gamma</em>, <em class="sig-param">muL</em>, <em class="sig-param">mu0</em>, <em class="sig-param">lp</em>, <em class="sig-param">b</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.pore_size" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.pore_size_fit">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">pore_size_fit</code><span class="sig-paren">(</span><em class="sig-param">C</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.pore_size_fit" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.radius_gyration">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">radius_gyration</code><span class="sig-paren">(</span><em class="sig-param">L</em>, <em class="sig-param">lp</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.radius_gyration" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.random_Dseqs">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">random_Dseqs</code><span class="sig-paren">(</span><em class="sig-param">sizes</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#random_Dseqs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.random_Dseqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of pydna Dseqs of given sizes and random sequences.</p>
</dd></dl>

<dl class="function">
<dt id="pydna.gel.reduced_field">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">reduced_field</code><span class="sig-paren">(</span><em class="sig-param">eta</em>, <em class="sig-param">a</em>, <em class="sig-param">mu0</em>, <em class="sig-param">E</em>, <em class="sig-param">kB</em>, <em class="sig-param">T</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.reduced_field" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.reduced_field_Kuhn">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">reduced_field_Kuhn</code><span class="sig-paren">(</span><em class="sig-param">eta</em>, <em class="sig-param">l</em>, <em class="sig-param">mu0</em>, <em class="sig-param">E</em>, <em class="sig-param">kB</em>, <em class="sig-param">T</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.reduced_field_Kuhn" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.reptation_accelerated">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">reptation_accelerated</code><span class="sig-paren">(</span><em class="sig-param">Dblob</em>, <em class="sig-param">epsilon</em>, <em class="sig-param">N</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.reptation_accelerated" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.reptation_equilibrium">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">reptation_equilibrium</code><span class="sig-paren">(</span><em class="sig-param">Dblob</em>, <em class="sig-param">N</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.reptation_equilibrium" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.reptation_plateau">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">reptation_plateau</code><span class="sig-paren">(</span><em class="sig-param">Dblob</em>, <em class="sig-param">epsilon</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.reptation_plateau" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.rundistance">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">rundistance</code><span class="sig-paren">(</span><em class="sig-param">time</em>, <em class="sig-param">mobility</em>, <em class="sig-param">field</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.rundistance" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.runtime">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">runtime</code><span class="sig-paren">(</span><em class="sig-param">distance</em>, <em class="sig-param">mobility</em>, <em class="sig-param">field</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.gel.runtime" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt>
<code class="sig-name descname">size_to_mobility(dna_len, field, percentage, mu_func=&lt;function vWBR.&lt;locals&gt;.&lt;lambda&gt; at 0x7f3414047400&gt;, dataset={'E': &lt;Quantity([0.62 0.93 1.24 1.55 1.86 2.17 2.48 3.1  4.97 6.21 0.62 0.93 1.24 1.55</code></dt>
<dt>
<code class="sig-name descname">1.86 2.17 2.48 3.1  4.97 6.21 0.62 0.93 1.24 1.55 1.86 2.17 2.48 3.1</code></dt>
<dt>
<code class="sig-name descname">4.97 6.21 0.62 0.93 1.24 1.55 1.86 2.17 2.48 3.1  4.97 6.21], 'volt / centimeter')&gt;, 'T': &lt;Quantity([0.4 0.4 0.4 0.4 0.4 0.4 0.4 0.4 0.4 0.4 0.7 0.7 0.7 0.7 0.7 0.7 0.7 0.7</code></dt>
<dt>
<code class="sig-name descname">0.7 0.7 1.  1.  1.  1.  1.  1.  1.  1.  1.  1.  1.3 1.3 1.3 1.3 1.3 1.3</code></dt>
<dt>
<code class="sig-name descname">1.3 1.3 1.3 1.3], 'gram / milliliter')&gt;, 'muS': &lt;Quantity([2.67e-08 2.54e-08 3.67e-08 3.18e-08 3.12e-08 3.07e-08 3.13e-08 3.14e-08</code></dt>
<dt>
<code class="sig-name descname">3.34e-08 3.11e-08 2.86e-08 2.38e-08 3.71e-08 3.61e-08 3.61e-08 3.07e-08</code></dt>
<dt>
<code class="sig-name descname">3.42e-08 3.14e-08 3.08e-08 3.24e-08 2.48e-08 2.41e-08 3.34e-08 3.25e-08</code></dt>
<dt>
<code class="sig-name descname">3.69e-08 2.87e-08 3.39e-08 3.36e-08 3.04e-08 3.13e-08 2.12e-08 1.96e-08</code></dt>
<dt>
<code class="sig-name descname">3.32e-08 3.47e-08 3.97e-08 2.76e-08 3.26e-08 3.37e-08 2.68e-08 2.70e-08], 'meter ** 2 / second / volt')&gt;, 'muL': &lt;Quantity([2.10e-09 3.40e-09 8.10e-09 7.70e-09 7.80e-09 8.90e-09 9.00e-09 9.80e-09</code></dt>
<dt>
<code class="sig-name descname">1.22e-08 1.20e-08 8.36e-14 9.00e-10 4.90e-09 5.30e-09 6.20e-09 5.50e-09</code></dt>
<dt>
<code class="sig-name descname">7.20e-09 6.90e-09 9.30e-09 1.03e-08 1.82e-22 1.37e-21 1.70e-09 3.30e-09</code></dt>
<dt>
<code class="sig-name descname">4.70e-09 3.80e-09 5.20e-09 6.10e-09 7.70e-09 8.80e-09 1.87e-18 3.80e-14</code></dt>
<dt>
<code class="sig-name descname">2.00e-10 2.70e-09 3.60e-09 2.10e-09 4.50e-09 5.00e-09 4.90e-09 5.90e-09], 'meter ** 2 / second / volt')&gt;, 'gamma': &lt;Quantity([5.27990e+01 2.53040e+01 1.49220e+01 1.44880e+01 1.26030e+01 1.19110e+01</code></dt>
<dt>
<code class="sig-name descname">1.26480e+01 9.42300e+00 9.63000e+00 9.21400e+00 7.32000e+05 5.42640e+01</code></dt>
<dt>
<code class="sig-name descname">1.52550e+01 1.31250e+01 9.83400e+00 1.01510e+01 9.03600e+00 7.71300e+00</code></dt>
<dt>
<code class="sig-name descname">6.18800e+00 6.37100e+00 1.83000e+14 2.33000e+13 2.58550e+01 1.27730e+01</code></dt>
<dt>
<code class="sig-name descname">8.50200e+00 9.37200e+00 7.20900e+00 5.85200e+00 4.74900e+00 4.41700e+00</code></dt>
<dt>
<code class="sig-name descname">1.11000e+10 6.10000e+05 1.79189e+02 1.19200e+01 8.59400e+00 1.24120e+01</code></dt>
<dt>
<code class="sig-name descname">7.46400e+00 5.15900e+00 4.04800e+00 3.55500e+00], 'kilobase_pair')&gt;}, method='linear', replNANs=True)</code></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.to_units">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">to_units</code><span class="sig-paren">(</span><em class="sig-param">quantity</em>, <em class="sig-param">units</em>, <em class="sig-param">var_name=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#to_units"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.to_units" title="Permalink to this definition">¶</a></dt>
<dd><p>Asserts that the quantity has the proper dimensions
(inferred from the default units) if the quantity is an instance of
pint.unit.Quantity or assigns the default units if it’s not.</p>
</dd></dl>

<dl class="function">
<dt id="pydna.gel.vWBR">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">vWBR</code><span class="sig-paren">(</span><em class="sig-param">muS</em>, <em class="sig-param">muL</em>, <em class="sig-param">gamma</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#vWBR"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.vWBR" title="Permalink to this definition">¶</a></dt>
<dd><p>vWBR equation</p>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-name descname">vWBRfit(field, percentage, DNAvals=array([  100.        ,   604.04040404,  1108.08080808,  1612.12121212,</code></dt>
<dt>
<code class="sig-name descname">2116.16161616,  2620.2020202 ,  3124.24242424,  3628.28282828,</code></dt>
<dt>
<code class="sig-name descname">4132.32323232,  4636.36363636,  5140.4040404 ,  5644.44444444,</code></dt>
<dt>
<code class="sig-name descname">6148.48484848,  6652.52525253,  7156.56565657,  7660.60606061,</code></dt>
<dt>
<code class="sig-name descname">8164.64646465,  8668.68686869,  9172.72727273,  9676.76767677,</code></dt>
<dt>
<code class="sig-name descname">10180.80808081, 10684.84848485, 11188.88888889, 11692.92929293,</code></dt>
<dt>
<code class="sig-name descname">12196.96969697, 12701.01010101, 13205.05050505, 13709.09090909,</code></dt>
<dt>
<code class="sig-name descname">14213.13131313, 14717.17171717, 15221.21212121, 15725.25252525,</code></dt>
<dt>
<code class="sig-name descname">16229.29292929, 16733.33333333, 17237.37373737, 17741.41414141,</code></dt>
<dt>
<code class="sig-name descname">18245.45454545, 18749.49494949, 19253.53535354, 19757.57575758,</code></dt>
<dt>
<code class="sig-name descname">20261.61616162, 20765.65656566, 21269.6969697 , 21773.73737374,</code></dt>
<dt>
<code class="sig-name descname">22277.77777778, 22781.81818182, 23285.85858586, 23789.8989899 ,</code></dt>
<dt>
<code class="sig-name descname">24293.93939394, 24797.97979798, 25302.02020202, 25806.06060606,</code></dt>
<dt>
<code class="sig-name descname">26310.1010101 , 26814.14141414, 27318.18181818, 27822.22222222,</code></dt>
<dt>
<code class="sig-name descname">28326.26262626, 28830.3030303 , 29334.34343434, 29838.38383838,</code></dt>
<dt>
<code class="sig-name descname">30342.42424242, 30846.46464646, 31350.50505051, 31854.54545455,</code></dt>
<dt>
<code class="sig-name descname">32358.58585859, 32862.62626263, 33366.66666667, 33870.70707071,</code></dt>
<dt>
<code class="sig-name descname">34374.74747475, 34878.78787879, 35382.82828283, 35886.86868687,</code></dt>
<dt>
<code class="sig-name descname">36390.90909091, 36894.94949495, 37398.98989899, 37903.03030303,</code></dt>
<dt>
<code class="sig-name descname">38407.07070707, 38911.11111111, 39415.15151515, 39919.19191919,</code></dt>
<dt>
<code class="sig-name descname">40423.23232323, 40927.27272727, 41431.31313131, 41935.35353535,</code></dt>
<dt>
<code class="sig-name descname">42439.39393939, 42943.43434343, 43447.47474747, 43951.51515152,</code></dt>
<dt>
<code class="sig-name descname">44455.55555556, 44959.5959596 , 45463.63636364, 45967.67676768,</code></dt>
<dt>
<code class="sig-name descname">46471.71717172, 46975.75757576, 47479.7979798 , 47983.83838384,</code></dt>
<dt>
<code class="sig-name descname">48487.87878788, 48991.91919192, 49495.95959596, 50000.        ]), dataset={'E': &lt;Quantity([0.62 0.93 1.24 1.55 1.86 2.17 2.48 3.1  4.97 6.21 0.62 0.93 1.24 1.55</code></dt>
<dt>
<code class="sig-name descname">1.86 2.17 2.48 3.1  4.97 6.21 0.62 0.93 1.24 1.55 1.86 2.17 2.48 3.1</code></dt>
<dt>
<code class="sig-name descname">4.97 6.21 0.62 0.93 1.24 1.55 1.86 2.17 2.48 3.1  4.97 6.21], 'volt / centimeter')&gt;, 'T': &lt;Quantity([0.4 0.4 0.4 0.4 0.4 0.4 0.4 0.4 0.4 0.4 0.7 0.7 0.7 0.7 0.7 0.7 0.7 0.7</code></dt>
<dt>
<code class="sig-name descname">0.7 0.7 1.  1.  1.  1.  1.  1.  1.  1.  1.  1.  1.3 1.3 1.3 1.3 1.3 1.3</code></dt>
<dt>
<code class="sig-name descname">1.3 1.3 1.3 1.3], 'gram / milliliter')&gt;, 'muS': &lt;Quantity([2.67e-08 2.54e-08 3.67e-08 3.18e-08 3.12e-08 3.07e-08 3.13e-08 3.14e-08</code></dt>
<dt>
<code class="sig-name descname">3.34e-08 3.11e-08 2.86e-08 2.38e-08 3.71e-08 3.61e-08 3.61e-08 3.07e-08</code></dt>
<dt>
<code class="sig-name descname">3.42e-08 3.14e-08 3.08e-08 3.24e-08 2.48e-08 2.41e-08 3.34e-08 3.25e-08</code></dt>
<dt>
<code class="sig-name descname">3.69e-08 2.87e-08 3.39e-08 3.36e-08 3.04e-08 3.13e-08 2.12e-08 1.96e-08</code></dt>
<dt>
<code class="sig-name descname">3.32e-08 3.47e-08 3.97e-08 2.76e-08 3.26e-08 3.37e-08 2.68e-08 2.70e-08], 'meter ** 2 / second / volt')&gt;, 'muL': &lt;Quantity([2.10e-09 3.40e-09 8.10e-09 7.70e-09 7.80e-09 8.90e-09 9.00e-09 9.80e-09</code></dt>
<dt>
<code class="sig-name descname">1.22e-08 1.20e-08 8.36e-14 9.00e-10 4.90e-09 5.30e-09 6.20e-09 5.50e-09</code></dt>
<dt>
<code class="sig-name descname">7.20e-09 6.90e-09 9.30e-09 1.03e-08 1.82e-22 1.37e-21 1.70e-09 3.30e-09</code></dt>
<dt>
<code class="sig-name descname">4.70e-09 3.80e-09 5.20e-09 6.10e-09 7.70e-09 8.80e-09 1.87e-18 3.80e-14</code></dt>
<dt>
<code class="sig-name descname">2.00e-10 2.70e-09 3.60e-09 2.10e-09 4.50e-09 5.00e-09 4.90e-09 5.90e-09], 'meter ** 2 / second / volt')&gt;, 'gamma': &lt;Quantity([5.27990e+01 2.53040e+01 1.49220e+01 1.44880e+01 1.26030e+01 1.19110e+01</code></dt>
<dt>
<code class="sig-name descname">1.26480e+01 9.42300e+00 9.63000e+00 9.21400e+00 7.32000e+05 5.42640e+01</code></dt>
<dt>
<code class="sig-name descname">1.52550e+01 1.31250e+01 9.83400e+00 1.01510e+01 9.03600e+00 7.71300e+00</code></dt>
<dt>
<code class="sig-name descname">6.18800e+00 6.37100e+00 1.83000e+14 2.33000e+13 2.58550e+01 1.27730e+01</code></dt>
<dt>
<code class="sig-name descname">8.50200e+00 9.37200e+00 7.20900e+00 5.85200e+00 4.74900e+00 4.41700e+00</code></dt>
<dt>
<code class="sig-name descname">1.11000e+10 6.10000e+05 1.79189e+02 1.19200e+01 8.59400e+00 1.24120e+01</code></dt>
<dt>
<code class="sig-name descname">7.46400e+00 5.15900e+00 4.04800e+00 3.55500e+00], 'kilobase_pair')&gt;}, mu_func=&lt;function vWBR.&lt;locals&gt;.&lt;lambda&gt; at 0x7f3414047400&gt;, method='linear', replNANs=True, plot=True)</code></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.gel.weight_standard_sample">
<code class="sig-prename descclassname">pydna.gel.</code><code class="sig-name descname">weight_standard_sample</code><span class="sig-paren">(</span><em class="sig-param">key</em>, <em class="sig-param">qty=&lt;Quantity(500</em>, <em class="sig-param">'nanogram')&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/gel.html#weight_standard_sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.gel.weight_standard_sample" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
</div>
<div class="section" id="module-pydna.genbank">
<span id="pydna-genbank-module"></span><h2>pydna.genbank module<a class="headerlink" href="#module-pydna.genbank" title="Permalink to this headline">¶</a></h2>
<p>This module provides a class for downloading sequences from genbank
called Genbank and an function that does the same thing called genbank.</p>
<p>The function can be used if the environmental variable <strong>pydna_email</strong> has 
been set to a valid email address. The easiest way to do this permanantly is to edit the 
<cite>pydna.ini</cite> file. See the documentation of <a class="reference internal" href="#pydna.open_config_folder" title="pydna.open_config_folder"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.open_config_folder()</span></code></a></p>
<dl class="class">
<dt id="pydna.genbank.Genbank">
<em class="property">class </em><code class="sig-prename descclassname">pydna.genbank.</code><code class="sig-name descname">Genbank</code><span class="sig-paren">(</span><em class="sig-param">users_email: str</em>, <em class="sig-param">*args</em>, <em class="sig-param">tool='pydna'</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbank.html#Genbank"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbank.Genbank" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.6/library/functions.html#object" title="(in Python v3.6)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Class to facilitate download from genbank. It is easier and 
quicker to use the <a class="reference internal" href="#pydna.genbank.genbank" title="pydna.genbank.genbank"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.genbank.genbank()</span></code></a> function directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>users_email</strong><span class="classifier">string</span></dt><dd><p>Has to be a valid email address. You should always tell
Genbanks who you are, so that they can contact you.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.genbank</span> <span class="k">import</span> <span class="n">Genbank</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gb</span><span class="o">=</span><span class="n">Genbank</span><span class="p">(</span><span class="s2">&quot;bjornjobb@gmail.com&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rec</span> <span class="o">=</span> <span class="n">gb</span><span class="o">.</span><span class="n">nucleotide</span><span class="p">(</span><span class="s2">&quot;AJ515731&quot;</span><span class="p">)</span>   <span class="c1"># &lt;- entry from genbank</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rec</span><span class="p">))</span>
<span class="go">19</span>
</pre></div>
</div>
<dl class="method">
<dt id="pydna.genbank.Genbank.nucleotide">
<code class="sig-name descname">nucleotide</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.genbank.Genbank.nucleotide" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="function">
<dt id="pydna.genbank.genbank">
<code class="sig-prename descclassname">pydna.genbank.</code><code class="sig-name descname">genbank</code><span class="sig-paren">(</span><em class="sig-param">accession: str = 'CS570233.1'</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbank.html#genbank"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbank.genbank" title="Permalink to this definition">¶</a></dt>
<dd><p>Download a genbank nuclotide record. This function takes the same paramenters as the 
:func:pydna.genbank.Genbank.nucleotide method. The email address stored in the <cite>pydna_email</cite> 
environment variable is used. The easiest way set this permanantly is to edit the 
<cite>pydna.ini</cite> file. See the documentation of <a class="reference internal" href="#pydna.open_config_folder" title="pydna.open_config_folder"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.open_config_folder()</span></code></a></p>
<p>if no accession is given, a <a class="reference external" href="https://www.ncbi.nlm.nih.gov/nuccore/CS570233.1">very short Genbank entry</a>
is used as an example. This can be usefuk for testing the connection to Genbank.
Please note that this result is also cached by default by settings in the pydna.ini file.
See the documentation of <a class="reference internal" href="#pydna.open_config_folder" title="pydna.open_config_folder"><code class="xref py py-func docutils literal notranslate"><span class="pre">pydna.open_config_folder()</span></code></a></p>
</dd></dl>

</div>
<div class="section" id="module-pydna.genbankfile">
<span id="pydna-genbankfile-module"></span><h2>pydna.genbankfile module<a class="headerlink" href="#module-pydna.genbankfile" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="pydna.genbankfile.GenbankFile">
<em class="property">class </em><code class="sig-prename descclassname">pydna.genbankfile.</code><code class="sig-name descname">GenbankFile</code><span class="sig-paren">(</span><em class="sig-param">record</em>, <em class="sig-param">*args</em>, <em class="sig-param">path=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfile.html#GenbankFile"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankfile.GenbankFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord</span></code></a></p>
<dl class="method">
<dt id="pydna.genbankfile.GenbankFile.from_SeqRecord">
<em class="property">classmethod </em><code class="sig-name descname">from_SeqRecord</code><span class="sig-paren">(</span><em class="sig-param">record</em>, <em class="sig-param">*args</em>, <em class="sig-param">path=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfile.html#GenbankFile.from_SeqRecord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankfile.GenbankFile.from_SeqRecord" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.genbankfile.GenbankFile.rc">
<code class="sig-name descname">rc</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.genbankfile.GenbankFile.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reverse complement.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.genbankfile.GenbankFile.reverse_complement">
<code class="sig-name descname">reverse_complement</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfile.html#GenbankFile.reverse_complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankfile.GenbankFile.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reverse complement.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-pydna.genbankfixer">
<span id="pydna-genbankfixer-module"></span><h2>pydna.genbankfixer module<a class="headerlink" href="#module-pydna.genbankfixer" title="Permalink to this headline">¶</a></h2>
<p>This module provides the <a class="reference internal" href="#pydna.genbankfixer.gbtext_clean" title="pydna.genbankfixer.gbtext_clean"><code class="xref py py-func docutils literal notranslate"><span class="pre">gbtext_clean()</span></code></a> function which can clean up broken Genbank files enough to 
pass the BioPython Genbank parser</p>
<p>Almost all of this code was lifted from BioJSON (<a class="reference external" href="https://github.com/levskaya/BioJSON">https://github.com/levskaya/BioJSON</a>) by Anselm Levskaya.
The original code was not accompanied by any software licence. This parser is based on pyparsing.</p>
<p>There are some modifications to deal with fringe cases.</p>
<p>The parser first produces JSON as an intermediate format which is then formatted back into a 
string in Genbank format.</p>
<p>The parser is not complete, so some fields do not survive the roundtrip (see below). 
This should not be a difficult fix. The returned result has two properties,  
.jseq which is the intermediate JSON produced by the parser and .gbtext 
which is the formatted genbank string.</p>
<dl class="function">
<dt id="pydna.genbankfixer.concat_dict">
<code class="sig-prename descclassname">pydna.genbankfixer.</code><code class="sig-name descname">concat_dict</code><span class="sig-paren">(</span><em class="sig-param">dlist</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#concat_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankfixer.concat_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>more or less dict(list of string pairs) but merges
vals with the same keys so no duplicates occur</p>
</dd></dl>

<dl class="function">
<dt id="pydna.genbankfixer.gbtext_clean">
<code class="sig-prename descclassname">pydna.genbankfixer.</code><code class="sig-name descname">gbtext_clean</code><span class="sig-paren">(</span><em class="sig-param">gbtext</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#gbtext_clean"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankfixer.gbtext_clean" title="Permalink to this definition">¶</a></dt>
<dd><p>This function takes a string containing <strong>one</strong> genbank sequence
in Genbank format and returns a named tuple containing two fields,
the gbtext containing a string with the corrected genbank sequence and
jseq which contains the JSON intermediate.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="s1">&#39;&#39;&#39;LOCUS       New_DNA      3 bp    DNA   CIRCULAR SYN        19-JUN-2013</span>
<span class="gp">... </span><span class="s1">DEFINITION  .</span>
<span class="gp">... </span><span class="s1">ACCESSION   </span>
<span class="gp">... </span><span class="s1">VERSION     </span>
<span class="gp">... </span><span class="s1">SOURCE      .</span>
<span class="gp">... </span><span class="s1">  ORGANISM  .</span>
<span class="gp">... </span><span class="s1">COMMENT     </span>
<span class="gp">... </span><span class="s1">COMMENT     ApEinfo:methylated:1</span>
<span class="gp">... </span><span class="s1">ORIGIN</span>
<span class="gp">... </span><span class="s1">        1 aaa     </span>
<span class="gp">... </span><span class="s1">//&#39;&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.readers</span> <span class="k">import</span> <span class="n">read</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">read</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>  <span class="c1"># doctest: +SKIP</span>
<span class="go">/home/bjorn/anaconda3/envs/bjorn36/lib/python3.6/site-packages/Bio/GenBank/Scanner.py:1388: BiopythonParserWarning: Malformed LOCUS line found - is this correct?</span>
<span class="go">:&#39;LOCUS       New_DNA      3 bp    DNA   CIRCULAR SYN        19-JUN-2013\n&#39;</span>
<span class="go">  &quot;correct?\n:%r&quot; % line, BiopythonParserWarning)</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;/home/bjorn/python_packages/pydna/pydna/readers.py&quot;</span>, line <span class="m">48</span>, in <span class="n">read</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
<span class="gr">IndexError</span>: <span class="n">pop from empty list</span>
<span class="go">&lt;BLANKLINE&gt;</span>
<span class="go">During handling of the above exception, another exception occurred:</span>
<span class="go">&lt;BLANKLINE&gt;</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/home/bjorn/python_packages/pydna/pydna/readers.py&quot;</span>, line <span class="m">50</span>, in <span class="n">read</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No sequences found in data:</span><span class="se">\n</span><span class="s2">(</span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">data</span><span class="p">[:</span><span class="mi">79</span><span class="p">]))</span>
<span class="gr">ValueError</span>: <span class="n">No sequences found in data:</span>
<span class="go">(LOCUS       New_DNA      3 bp    DNA   CIRCULAR SYN        19-JUN-2013</span>
<span class="go">DEFINITI)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.genbankfixer</span> <span class="k">import</span> <span class="n">gbtext_clean</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span><span class="p">,</span> <span class="n">j2</span> <span class="o">=</span> <span class="n">gbtext_clean</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">s2</span><span class="p">)</span>
<span class="go">LOCUS       New_DNA                    3 bp ds-DNA     circular SYN 19-JUN-2013</span>
<span class="go">DEFINITION  .</span>
<span class="go">ACCESSION   </span>
<span class="go">VERSION     </span>
<span class="go">SOURCE      .</span>
<span class="go">ORGANISM  .</span>
<span class="go">COMMENT     </span>
<span class="go">COMMENT     ApEinfo:methylated:1</span>
<span class="go">FEATURES             Location/Qualifiers</span>
<span class="go">ORIGIN</span>
<span class="go">        1 aaa     </span>
<span class="go">//</span>
<span class="go">&lt;BLANKLINE&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s3</span> <span class="o">=</span> <span class="n">read</span><span class="p">(</span><span class="n">s2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s3</span>
<span class="go">Dseqrecord(o3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">s3</span><span class="o">.</span><span class="n">format</span><span class="p">())</span>
<span class="go">LOCUS       New_DNA                    3 bp    ds-DNA  circular SYN 19-JUN-2013</span>
<span class="go">DEFINITION  .</span>
<span class="go">ACCESSION   New_DNA</span>
<span class="go">VERSION     New_DNA</span>
<span class="go">KEYWORDS    .</span>
<span class="go">SOURCE      </span>
<span class="go">  ORGANISM  .</span>
<span class="go">            .</span>
<span class="go">COMMENT     </span>
<span class="go">            ApEinfo:methylated:1</span>
<span class="go">FEATURES             Location/Qualifiers</span>
<span class="go">ORIGIN</span>
<span class="go">        1 aaa</span>
<span class="go">//</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.genbankfixer.locstr">
<code class="sig-prename descclassname">pydna.genbankfixer.</code><code class="sig-name descname">locstr</code><span class="sig-paren">(</span><em class="sig-param">locs</em>, <em class="sig-param">strand</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#locstr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankfixer.locstr" title="Permalink to this definition">¶</a></dt>
<dd><p>genbank formatted location string, assumes no join’d combo of rev and fwd seqs</p>
</dd></dl>

<dl class="function">
<dt id="pydna.genbankfixer.originstr">
<code class="sig-prename descclassname">pydna.genbankfixer.</code><code class="sig-name descname">originstr</code><span class="sig-paren">(</span><em class="sig-param">sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#originstr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankfixer.originstr" title="Permalink to this definition">¶</a></dt>
<dd><p>formats dna sequence as broken, numbered lines ala genbank</p>
</dd></dl>

<dl class="function">
<dt id="pydna.genbankfixer.parseGBLoc">
<code class="sig-prename descclassname">pydna.genbankfixer.</code><code class="sig-name descname">parseGBLoc</code><span class="sig-paren">(</span><em class="sig-param">s</em>, <em class="sig-param">l</em>, <em class="sig-param">t</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#parseGBLoc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankfixer.parseGBLoc" title="Permalink to this definition">¶</a></dt>
<dd><p>retwingles parsed genbank location strings, assumes no joins of RC and FWD sequences</p>
</dd></dl>

<dl class="function">
<dt id="pydna.genbankfixer.strip_indent">
<code class="sig-prename descclassname">pydna.genbankfixer.</code><code class="sig-name descname">strip_indent</code><span class="sig-paren">(</span><em class="sig-param">str</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#strip_indent"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankfixer.strip_indent" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.genbankfixer.strip_multiline">
<code class="sig-prename descclassname">pydna.genbankfixer.</code><code class="sig-name descname">strip_multiline</code><span class="sig-paren">(</span><em class="sig-param">s</em>, <em class="sig-param">l</em>, <em class="sig-param">t</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#strip_multiline"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankfixer.strip_multiline" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.genbankfixer.toGB">
<code class="sig-prename descclassname">pydna.genbankfixer.</code><code class="sig-name descname">toGB</code><span class="sig-paren">(</span><em class="sig-param">jseq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#toGB"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankfixer.toGB" title="Permalink to this definition">¶</a></dt>
<dd><p>parses json jseq data and prints out ApE compatible genbank</p>
</dd></dl>

<dl class="function">
<dt id="pydna.genbankfixer.toInt">
<code class="sig-prename descclassname">pydna.genbankfixer.</code><code class="sig-name descname">toInt</code><span class="sig-paren">(</span><em class="sig-param">s</em>, <em class="sig-param">l</em>, <em class="sig-param">t</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#toInt"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankfixer.toInt" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.genbankfixer.toJSON">
<code class="sig-prename descclassname">pydna.genbankfixer.</code><code class="sig-name descname">toJSON</code><span class="sig-paren">(</span><em class="sig-param">gbkstring</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#toJSON"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankfixer.toJSON" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.genbankfixer.wrapstring">
<code class="sig-prename descclassname">pydna.genbankfixer.</code><code class="sig-name descname">wrapstring</code><span class="sig-paren">(</span><em class="sig-param">str_</em>, <em class="sig-param">rowstart</em>, <em class="sig-param">rowend</em>, <em class="sig-param">padfirst=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankfixer.html#wrapstring"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankfixer.wrapstring" title="Permalink to this definition">¶</a></dt>
<dd><p>wraps the provided string in lines of length rowend-rowstart
and padded on the left by rowstart.
-&gt; if padfirst is false the first line is not padded</p>
</dd></dl>

</div>
<div class="section" id="module-pydna.genbankrecord">
<span id="pydna-genbankrecord-module"></span><h2>pydna.genbankrecord module<a class="headerlink" href="#module-pydna.genbankrecord" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="pydna.genbankrecord.GenbankRecord">
<em class="property">class </em><code class="sig-prename descclassname">pydna.genbankrecord.</code><code class="sig-name descname">GenbankRecord</code><span class="sig-paren">(</span><em class="sig-param">record</em>, <em class="sig-param">*args</em>, <em class="sig-param">item='accession'</em>, <em class="sig-param">start=None</em>, <em class="sig-param">stop=None</em>, <em class="sig-param">strand=1</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankrecord.html#GenbankRecord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.dseqrecord.Dseqrecord" title="pydna.dseqrecord.Dseqrecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.dseqrecord.Dseqrecord</span></code></a></p>
<dl class="method">
<dt id="pydna.genbankrecord.GenbankRecord.from_SeqRecord">
<em class="property">classmethod </em><code class="sig-name descname">from_SeqRecord</code><span class="sig-paren">(</span><em class="sig-param">record</em>, <em class="sig-param">*args</em>, <em class="sig-param">item='accession'</em>, <em class="sig-param">start=None</em>, <em class="sig-param">stop=None</em>, <em class="sig-param">strand=1</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankrecord.html#GenbankRecord.from_SeqRecord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord.from_SeqRecord" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.genbankrecord.GenbankRecord.rc">
<code class="sig-name descname">rc</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reverse complement.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.genbankrecord.GenbankRecord.reverse_complement">
<code class="sig-name descname">reverse_complement</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/genbankrecord.html#GenbankRecord.reverse_complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.genbankrecord.GenbankRecord.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reverse complement.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pydna.dseq.Dseq.reverse_complement" title="pydna.dseq.Dseq.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pydna.dseq.Dseq.reverse_complement</span></code></a></dt><dd></dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-pydna.myprimers">
<span id="pydna-myprimers-module"></span><h2>pydna.myprimers module<a class="headerlink" href="#module-pydna.myprimers" title="Permalink to this headline">¶</a></h2>
<p>This module provides three ways to access a primer list specified 
in the primers entry in the pydna.ini file or in the pydna_primers
environment variable.</p>
<p>The list has to have sequences in FASTA, Genbank or EMBL formats.
The primer list can have the format below for example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">first_primer</span>
<span class="n">tgatcgtcatgctgactatactat</span>
<span class="o">&gt;</span><span class="n">second_primer</span>
<span class="n">ctaggatcgtagatctagctg</span>
<span class="o">...</span>
</pre></div>
</div>
<p>list_primers is a list <a class="reference internal" href="#pydna.primer.Primer" title="pydna.primer.Primer"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.primer.Primer</span></code></a> objects
dict_primers is a dict where the key is the id of the object</p>
<p>each primer is instantiated as p001, p002, … for all primers</p>
<dl class="function">
<dt id="pydna.myprimers.append_primer_list">
<code class="sig-prename descclassname">pydna.myprimers.</code><code class="sig-name descname">append_primer_list</code><span class="sig-paren">(</span><em class="sig-param">primers: list</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/myprimers.html#append_primer_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.myprimers.append_primer_list" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-pydna.parsers">
<span id="pydna-parsers-module"></span><h2>pydna.parsers module<a class="headerlink" href="#module-pydna.parsers" title="Permalink to this headline">¶</a></h2>
<p>Provides two functions, parse and parse_primers</p>
<dl class="function">
<dt id="pydna.parsers.parse">
<code class="sig-prename descclassname">pydna.parsers.</code><code class="sig-name descname">parse</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">ds=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/parsers.html#parse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.parsers.parse" title="Permalink to this definition">¶</a></dt>
<dd><p>This function returns <em>all</em> DNA sequences found in data. If no
sequences are found, an empty list is returned. This is a greedy
function, use carefully.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>data</strong><span class="classifier">string or iterable</span></dt><dd><p>The data parameter is a string containing:</p>
<ol class="arabic simple">
<li><p>an absolute path to a local file.
The file will be read in text
mode and parsed for EMBL, FASTA
and Genbank sequences.</p></li>
<li><p>a string containing one or more
sequences in EMBL, GENBANK,
or FASTA format. Mixed formats
are allowed.</p></li>
<li><p>data can be a list or other iterable where the elements are 1 or 2</p></li>
</ol>
</dd>
<dt><strong>ds</strong><span class="classifier">bool</span></dt><dd><p>If True double stranded <code class="xref py py-class docutils literal notranslate"><span class="pre">Dseqrecord</span></code> objects are returned.
If False single stranded <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord</span></code> <a class="footnote-reference brackets" href="#id25" id="id24">10</a> objects are returned.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>list</dt><dd><p>contains Dseqrecord or SeqRecord objects</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">read</span></code></dt><dd></dd>
</dl>
</div>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id25"><span class="brackets"><a class="fn-backref" href="#id24">10</a></span></dt>
<dd><p><a class="reference external" href="http://biopython.org/wiki/SeqRecord">http://biopython.org/wiki/SeqRecord</a></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pydna.parsers.parse_primers">
<code class="sig-prename descclassname">pydna.parsers.</code><code class="sig-name descname">parse_primers</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/parsers.html#parse_primers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.parsers.parse_primers" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-pydna.primer">
<span id="pydna-primer-module"></span><h2>pydna.primer module<a class="headerlink" href="#module-pydna.primer" title="Permalink to this headline">¶</a></h2>
<p>This module contain provide the Primer class that is a subclass of the biopython SeqRecord.</p>
<dl class="class">
<dt id="pydna.primer.Primer">
<em class="property">class </em><code class="sig-prename descclassname">pydna.primer.</code><code class="sig-name descname">Primer</code><span class="sig-paren">(</span><em class="sig-param">record</em>, <em class="sig-param">*args</em>, <em class="sig-param">template=None</em>, <em class="sig-param">position=None</em>, <em class="sig-param">footprint=0</em>, <em class="sig-param">concentration=1000.0</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/primer.html#Primer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.primer.Primer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.seqrecord.SeqRecord" title="pydna.seqrecord.SeqRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">pydna.seqrecord.SeqRecord</span></code></a></p>
<p>This class can hold information about a primer and its position on a template 
footprint and tail.</p>
<dl class="method">
<dt id="pydna.primer.Primer.footprint">
<em class="property">property </em><code class="sig-name descname">footprint</code><a class="headerlink" href="#pydna.primer.Primer.footprint" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.primer.Primer.tail">
<em class="property">property </em><code class="sig-name descname">tail</code><a class="headerlink" href="#pydna.primer.Primer.tail" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.primer.Primer.tm">
<code class="sig-name descname">tm</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">saltc=50.0</em>, <em class="sig-param">formula=&lt;function tmbresluc at 0x7f342598c730&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/primer.html#Primer.tm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.primer.Primer.tm" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="module-pydna.readers">
<span id="pydna-readers-module"></span><h2>pydna.readers module<a class="headerlink" href="#module-pydna.readers" title="Permalink to this headline">¶</a></h2>
<p>Provides two functions, read and read_primer.</p>
<dl class="function">
<dt id="pydna.readers.read">
<code class="sig-prename descclassname">pydna.readers.</code><code class="sig-name descname">read</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">ds=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/readers.html#read"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.readers.read" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is similar the <code class="xref py py-func docutils literal notranslate"><span class="pre">parse()</span></code> function but expects one and only
one sequence or and exception is thrown.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>data</strong><span class="classifier">string</span></dt><dd><p>see below</p>
</dd>
<dt><strong>ds</strong><span class="classifier">bool</span></dt><dd><p>Double stranded or single stranded DNA, if True return
Dseqrecord objects, else Bio.SeqRecord objects.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>Dseqrecord</dt><dd><p>contains the first Dseqrecord or SeqRecord object parsed.</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse</span></code></dt><dd></dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>The data parameter is similar to the data parameter for <code class="xref py py-func docutils literal notranslate"><span class="pre">parse()</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="pydna.readers.read_primer">
<code class="sig-prename descclassname">pydna.readers.</code><code class="sig-name descname">read_primer</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/readers.html#read_primer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.readers.read_primer" title="Permalink to this definition">¶</a></dt>
<dd><p>Use this function to read a primer sequence from a string or a local file. 
The usage is similar to the <code class="xref py py-func docutils literal notranslate"><span class="pre">parse_primer()</span></code> function.</p>
</dd></dl>

</div>
<div class="section" id="module-pydna.seqrecord">
<span id="pydna-seqrecord-module"></span><h2>pydna.seqrecord module<a class="headerlink" href="#module-pydna.seqrecord" title="Permalink to this headline">¶</a></h2>
<p>This module provide a subclass of the Biopython SeqRecord class.
It has a number of extra methods and uses 
the <code class="xref py py-class docutils literal notranslate"><span class="pre">pydna._pretty_str.pretty_str</span></code> class instread of str for a 
nicer output in the IPython shell.</p>
<dl class="class">
<dt id="pydna.seqrecord.SeqRecord">
<em class="property">class </em><code class="sig-prename descclassname">pydna.seqrecord.</code><code class="sig-name descname">SeqRecord</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqRecord.SeqRecord</span></code></p>
<dl class="method">
<dt id="pydna.seqrecord.SeqRecord.accession">
<em class="property">property </em><code class="sig-name descname">accession</code><a class="headerlink" href="#pydna.seqrecord.SeqRecord.accession" title="Permalink to this definition">¶</a></dt>
<dd><p>alias for id property</p>
</dd></dl>

<dl class="method">
<dt id="pydna.seqrecord.SeqRecord.add_colors_to_features_for_ape">
<code class="sig-name descname">add_colors_to_features_for_ape</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.add_colors_to_features_for_ape"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.add_colors_to_features_for_ape" title="Permalink to this definition">¶</a></dt>
<dd><p>This method assigns colors to features compatible with the 
<a class="reference external" href="http://jorgensen.biology.utah.edu/wayned/ape/">ApE editor</a></p>
</dd></dl>

<dl class="method">
<dt id="pydna.seqrecord.SeqRecord.add_feature">
<code class="sig-name descname">add_feature</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">x=None</em>, <em class="sig-param">y=None</em>, <em class="sig-param">seq=None</em>, <em class="sig-param">type='misc'</em>, <em class="sig-param">strand=1</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.add_feature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.add_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a feature of type misc to the feature list of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>x</strong><span class="classifier">int</span></dt><dd><p>Indicates start of the feature</p>
</dd>
<dt><strong>y</strong><span class="classifier">int</span></dt><dd><p>Indicates end of the feature</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="k">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">features</span>
<span class="go">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">features</span>
<span class="go">[SeqFeature(FeatureLocation(ExactPosition(2), ExactPosition(4), strand=1), type=&#39;misc&#39;)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.seqrecord.SeqRecord.definition">
<em class="property">property </em><code class="sig-name descname">definition</code><a class="headerlink" href="#pydna.seqrecord.SeqRecord.definition" title="Permalink to this definition">¶</a></dt>
<dd><p>alias for description property</p>
</dd></dl>

<dl class="method">
<dt id="pydna.seqrecord.SeqRecord.extract_feature">
<code class="sig-name descname">extract_feature</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">n</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.extract_feature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.extract_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts a feature and creates a new SeqRecord object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>n</strong><span class="classifier">int</span></dt><dd><p>Indicates the feature to extract</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="k">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">extract_feature</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">SeqRecord(seq=Seq(&#39;gt&#39;, IUPACAmbiguousDNA()), id=&#39;ft2&#39;, name=&#39;ft2&#39;, description=&#39;description&#39;, dbxrefs=[])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.seqrecord.SeqRecord.gc">
<code class="sig-name descname">gc</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.gc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.gc" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns GC content</p>
</dd></dl>

<dl class="method">
<dt id="pydna.seqrecord.SeqRecord.isorf">
<code class="sig-name descname">isorf</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">table=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.isorf"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.isorf" title="Permalink to this definition">¶</a></dt>
<dd><p>Detects if sequence is an open reading frame (orf) in the 5’-3’ direction.
Translation tables are numbers according to the NCBI numbering <a class="footnote-reference brackets" href="#id27" id="id26">11</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>table</strong><span class="classifier">int</span></dt><dd><p>Sets the translation table, default is 1 (standard code)</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>bool</dt><dd><p>True if sequence is an orf, False otherwise.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id27"><span class="brackets"><a class="fn-backref" href="#id26">11</a></span></dt>
<dd><p><a class="reference external" href="http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c">http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="k">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atgaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;atttaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.seqrecord.SeqRecord.list_features">
<code class="sig-name descname">list_features</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.list_features"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.list_features" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints an ASCII table with all features.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Alphabet.IUPAC</span> <span class="k">import</span> <span class="n">IUPACAmbiguousDNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="k">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="k">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;atgtaa&quot;</span><span class="p">,</span><span class="n">IUPACAmbiguousDNA</span><span class="p">()))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">list_features</span><span class="p">())</span>
<span class="go">+-----+---------------+-----+-----+-----+-----+------+------+</span>
<span class="go">| Ft# | Label or Note | Dir | Sta | End | Len | type | orf? |</span>
<span class="go">+-----+---------------+-----+-----+-----+-----+------+------+</span>
<span class="go">|   0 | L:ft2         | --&gt; | 2   | 4   |   2 | misc |  no  |</span>
<span class="go">+-----+---------------+-----+-----+-----+-----+------+------+</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.seqrecord.SeqRecord.locus">
<em class="property">property </em><code class="sig-name descname">locus</code><a class="headerlink" href="#pydna.seqrecord.SeqRecord.locus" title="Permalink to this definition">¶</a></dt>
<dd><p>alias for name property</p>
</dd></dl>

<dl class="method">
<dt id="pydna.seqrecord.SeqRecord.olaps">
<code class="sig-name descname">olaps</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">other</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.olaps"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.olaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the overlaps between the sequence and another sequence,
The other sequence can be a string, Seq, SeqRecord, Dseq or DseqRecord</p>
</dd></dl>

<dl class="method">
<dt id="pydna.seqrecord.SeqRecord.reverse_complement">
<code class="sig-name descname">reverse_complement</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.reverse_complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Return new SeqRecord with reverse complement sequence.</p>
<p>By default the new record does NOT preserve the sequence identifier,
name, description, general annotation or database cross-references -
these are unlikely to apply to the reversed sequence.</p>
<p>You can specify the returned record’s id, name and description as
strings, or True to keep that of the parent, or False for a default.</p>
<p>You can specify the returned record’s features with a list of
SeqFeature objects, or True to keep that of the parent, or False to
omit them. The default is to keep the original features (with the
strand and locations adjusted).</p>
<p>You can also specify both the returned record’s annotations and
letter_annotations as dictionaries, True to keep that of the parent,
or False to omit them. The default is to keep the original
annotations (with the letter annotations reversed).</p>
<p>To show what happens to the pre-letter annotations, consider an
example Solexa variant FASTQ file with a single entry, which we’ll
read in as a SeqRecord:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;Quality/solexa_faked.fastq&quot;</span><span class="p">,</span> <span class="s2">&quot;fastq-solexa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
<span class="go">slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">letter_annotations</span><span class="p">))</span>
<span class="go">[&#39;solexa_quality&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">letter_annotations</span><span class="p">[</span><span class="s2">&quot;solexa_quality&quot;</span><span class="p">])</span>
<span class="go">[40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2, -3, -4, -5]</span>
</pre></div>
</div>
<p>Now take the reverse complement, here we explicitly give a new
identifier (the old identifier with a suffix):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">rc_record</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">(</span><span class="nb">id</span><span class="o">=</span><span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="o">+</span> <span class="s2">&quot;_rc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">rc_record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">rc_record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
<span class="go">slxa_0001_1_0001_01_rc NNNNNNACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT</span>
</pre></div>
</div>
<p>Notice that the per-letter-annotations have also been reversed,
although this may not be appropriate for all cases.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">rc_record</span><span class="o">.</span><span class="n">letter_annotations</span><span class="p">[</span><span class="s2">&quot;solexa_quality&quot;</span><span class="p">])</span>
<span class="go">[-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40]</span>
</pre></div>
</div>
<p>Now for the features, we need a different example. Parsing a GenBank
file is probably the easiest way to get an nice example with features
in it…</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;GenBank/pBAD30.gb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">plasmid</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s2">&quot;gb&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">plasmid</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">plasmid</span><span class="p">)))</span>
<span class="go">pBAD30 4923</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plasmid</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Seq(&#39;GCTAGCGGAGTGTATACTGGCTTACTATGTTGGCACTGATGAGGGTGTCAGTGA...ATG&#39;, IUPACAmbiguousDNA())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">plasmid</span><span class="o">.</span><span class="n">features</span><span class="p">)</span>
<span class="go">13</span>
</pre></div>
</div>
<p>Now, let’s take the reverse complement of this whole plasmid:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">rc_plasmid</span> <span class="o">=</span> <span class="n">plasmid</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">(</span><span class="nb">id</span><span class="o">=</span><span class="n">plasmid</span><span class="o">.</span><span class="n">id</span><span class="o">+</span><span class="s2">&quot;_rc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">rc_plasmid</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">rc_plasmid</span><span class="p">)))</span>
<span class="go">pBAD30_rc 4923</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rc_plasmid</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Seq(&#39;CATGGGCAAATATTATACGCAAGGCGACAAGGTGCTGATGCCGCTGGCGATTCA...AGC&#39;, IUPACAmbiguousDNA())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">rc_plasmid</span><span class="o">.</span><span class="n">features</span><span class="p">)</span>
<span class="go">13</span>
</pre></div>
</div>
<p>Let’s compare the first CDS feature - it has gone from being the
second feature (index 1) to the second last feature (index -2), its
strand has changed, and the location switched round.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">plasmid</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="go">type: CDS</span>
<span class="go">location: [1081:1960](-)</span>
<span class="go">qualifiers:</span>
<span class="go">    Key: label, Value: [&#39;araC&#39;]</span>
<span class="go">    Key: note, Value: [&#39;araC regulator of the arabinose BAD promoter&#39;]</span>
<span class="go">    Key: vntifkey, Value: [&#39;4&#39;]</span>
<span class="go">&lt;BLANKLINE&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">rc_plasmid</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">])</span>
<span class="go">type: CDS</span>
<span class="go">location: [2963:3842](+)</span>
<span class="go">qualifiers:</span>
<span class="go">    Key: label, Value: [&#39;araC&#39;]</span>
<span class="go">    Key: note, Value: [&#39;araC regulator of the arabinose BAD promoter&#39;]</span>
<span class="go">    Key: vntifkey, Value: [&#39;4&#39;]</span>
<span class="go">&lt;BLANKLINE&gt;</span>
</pre></div>
</div>
<p>You can check this new location, based on the length of the plasmid:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">plasmid</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1081</span>
<span class="go">3842</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">plasmid</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1960</span>
<span class="go">2963</span>
</pre></div>
</div>
<p>Note that if the SeqFeature annotation includes any strand specific
information (e.g. base changes for a SNP), this information is not
amended, and would need correction after the reverse complement.</p>
<p>Note trying to reverse complement a protein SeqRecord raises an
exception:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="k">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="k">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="k">import</span> <span class="n">IUPAC</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protein_rec</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MAIVMGR&quot;</span><span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">protein</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Test&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protein_rec</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">Proteins do not have complements!</span>
</pre></div>
</div>
<p>Also note you can reverse complement a SeqRecord using a MutableSeq:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="k">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="k">import</span> <span class="n">MutableSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="k">import</span> <span class="n">generic_dna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rec</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">,</span> <span class="n">generic_dna</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Test&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;T&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
<span class="go">Test TCGT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rc</span> <span class="o">=</span> <span class="n">rec</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">(</span><span class="nb">id</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">rc</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">rc</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
<span class="go">Test ACGA</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.seqrecord.SeqRecord.seguid">
<code class="sig-name descname">seguid</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.seguid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.seguid" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the url safe SEGUID <a class="footnote-reference brackets" href="#id30" id="id28">12</a> for the sequence.
This checksum is the same as seguid but with base64.urlsafe
encoding <a href="#id35"><span class="problematic" id="id36"><span id="id29"></span>[#]_</span></a> instead of the normal base 64. This means that
the characters + and / are replaced with - and _ so that
the checksum can be a part of and URL or a filename.</p>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id30"><span class="brackets"><a class="fn-backref" href="#id28">12</a></span></dt>
<dd><p><a class="reference external" href="http://wiki.christophchamp.com/index.php/SEGUID">http://wiki.christophchamp.com/index.php/SEGUID</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="k">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;aaaaaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span> <span class="c1"># original seguid is +bKGnebMkia5kNg/gF7IORXMnIU</span>
<span class="go">&#39;-bKGnebMkia5kNg_gF7IORXMnIU&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.seqrecord.SeqRecord.stamp">
<code class="sig-name descname">stamp</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/seqrecord.html#SeqRecord.stamp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.seqrecord.SeqRecord.stamp" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a SEGUID or cSEGUID checksum and a datestring to the description property.
This will show in the genbank format.</p>
<p>For linear sequences:</p>
<p><code class="docutils literal notranslate"><span class="pre">SEGUID_&lt;seguid&gt;_&lt;datestring&gt;</span></code></p>
<p>For circular sequences:</p>
<p><code class="docutils literal notranslate"><span class="pre">cSEGUID_&lt;seguid&gt;_&lt;datestring&gt;</span></code></p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.seqrecord</span> <span class="k">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">SeqRecord</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">stamp</span><span class="p">()</span>
<span class="go">&#39;SEGUID_YG7G6b2Kj_KtFOX63j8mRHHoIlE...&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">description</span>
<span class="go">&#39;SEGUID_YG7G6b2Kj_KtFOX63j8mRHHoIlE...&#39;</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-pydna.tm">
<span id="pydna-tm-module"></span><h2>pydna.tm module<a class="headerlink" href="#module-pydna.tm" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pydna.tm.Q5">
<code class="sig-prename descclassname">pydna.tm.</code><code class="sig-name descname">Q5</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/tm.html#Q5"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.tm.Q5" title="Permalink to this definition">¶</a></dt>
<dd><p>For Q5 Ta they take the lower of the two Tms and add 1C 
(up to 72C). For Phusion they take the lower of the two 
and add 3C (up to 72C).</p>
</dd></dl>

<dl class="function">
<dt id="pydna.tm.basictm">
<code class="sig-prename descclassname">pydna.tm.</code><code class="sig-name descname">basictm</code><span class="sig-paren">(</span><em class="sig-param">primer</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/tm.html#basictm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.tm.basictm" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the melting temperature (Tm) of the primer using
the basic formula. This function returns the same value as
the Biopython Bio.SeqUtils.MeltingTemp.Tm_Wallace</p>
<div class="line-block">
<div class="line">Tm = (wA+xT)*2 + (yG+zC)*4 assumed 50mM monovalent cations</div>
<div class="line"><br /></div>
<div class="line">w = number of A in primer</div>
<div class="line">x = number of T in primer</div>
<div class="line">y = number of G in primer</div>
<div class="line">z = number of C in primer</div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>primer</strong><span class="classifier">string</span></dt><dd><p>Primer sequence 5’-3’</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>tm</strong><span class="classifier">int</span></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.tm</span> <span class="k">import</span> <span class="n">basictm</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">basictm</span><span class="p">(</span><span class="s2">&quot;ggatcc&quot;</span><span class="p">)</span>
<span class="go">20</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.tm.tmbreslauer86">
<code class="sig-prename descclassname">pydna.tm.</code><code class="sig-name descname">tmbreslauer86</code><span class="sig-paren">(</span><em class="sig-param">primer</em>, <em class="sig-param">*args</em>, <em class="sig-param">dnac=500.0</em>, <em class="sig-param">saltc=50</em>, <em class="sig-param">thermodynamics=False</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/tm.html#tmbreslauer86"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.tm.tmbreslauer86" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the melting temperature (Tm) of the primer using
the nearest neighbour algorithm. Formula and thermodynamic data
is taken from Breslauer 1986.</p>
<p>These data are no longer widely used.</p>
<p>Breslauer 1986, table 2 <a class="reference internal" href="#rce0e7b96d429-2" id="id31">[2]</a></p>
<table class="docutils align-default">
<colgroup>
<col style="width: 29%" />
<col style="width: 29%" />
<col style="width: 24%" />
<col style="width: 18%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>pair</p></th>
<th class="head"><p>dH</p></th>
<th class="head"><p>dS</p></th>
<th class="head"><p>dG</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>AA/TT</p></td>
<td><p>9.1</p></td>
<td><p>24.0</p></td>
<td><p>1.9</p></td>
</tr>
<tr class="row-odd"><td><p>AT/TA</p></td>
<td><p>8.6</p></td>
<td><p>23.9</p></td>
<td><p>1.5</p></td>
</tr>
<tr class="row-even"><td><p>TA/AT</p></td>
<td><p>6.0</p></td>
<td><p>16.9</p></td>
<td><p>0.9</p></td>
</tr>
<tr class="row-odd"><td><p>CA/GT</p></td>
<td><p>5.8</p></td>
<td><p>12.9</p></td>
<td><p>1.9</p></td>
</tr>
<tr class="row-even"><td><p>GT/CA</p></td>
<td><p>6.5</p></td>
<td><p>17.3</p></td>
<td><p>1.3</p></td>
</tr>
<tr class="row-odd"><td><p>CT/GA</p></td>
<td><p>7.8</p></td>
<td><p>20.8</p></td>
<td><p>1.6</p></td>
</tr>
<tr class="row-even"><td><p>GA/CT</p></td>
<td><p>5.6</p></td>
<td><p>13.5</p></td>
<td><p>1.6</p></td>
</tr>
<tr class="row-odd"><td><p>CG/GC</p></td>
<td><p>11.9</p></td>
<td><p>27.8</p></td>
<td><p>3.6</p></td>
</tr>
<tr class="row-even"><td><p>GC/CG</p></td>
<td><p>11.1</p></td>
<td><p>26.7</p></td>
<td><p>3.1</p></td>
</tr>
<tr class="row-odd"><td><p>GG/CC</p></td>
<td><p>11.0</p></td>
<td><p>26.6</p></td>
<td><p>3.1</p></td>
</tr>
</tbody>
</table>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>primer</strong><span class="classifier">string</span></dt><dd><p>Primer sequence 5’-3’</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>tm</strong><span class="classifier">float</span></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="citation">
<dt class="label" id="rce0e7b96d429-2"><span class="brackets"><a class="fn-backref" href="#id31">2</a></span></dt>
<dd><p>K.J. Breslauer et al., “Predicting DNA Duplex Stability from the Base Sequence,” Proceedings of the National Academy of Sciences 83, no. 11 (1986): 3746.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.tm</span> <span class="k">import</span> <span class="n">tmbreslauer86</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tmbreslauer86</span><span class="p">(</span><span class="s2">&quot;ACGTCATCGACACTATCATCGAC&quot;</span><span class="p">)</span>
<span class="go">64.28863985851899</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.tm.tmbresluc">
<code class="sig-prename descclassname">pydna.tm.</code><code class="sig-name descname">tmbresluc</code><span class="sig-paren">(</span><em class="sig-param">primer</em>, <em class="sig-param">*args</em>, <em class="sig-param">primerc=500.0</em>, <em class="sig-param">saltc=50</em>, <em class="sig-param">thermodynamics=False</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/tm.html#tmbresluc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.tm.tmbresluc" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the tm for a primer using a formula adapted to polymerases
with a DNA binding domain, such as the Phusion polymerase.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>primer</strong><span class="classifier">string</span></dt><dd><p>primer sequence 5’-3’</p>
</dd>
<dt><strong>primerc</strong><span class="classifier">float</span></dt><dd><p>primer concentration in nM), set to 500.0 nm by default.</p>
</dd>
<dt><strong>saltc</strong><span class="classifier">float, optional</span></dt><dd><p>Monovalent cation concentration in mM, set to 50.0 mM by default.</p>
</dd>
<dt><strong>thermodynamics</strong><span class="classifier">bool, optional</span></dt><dd><p>prints details of the thermodynamic data to stdout. For
debugging only.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>tm</strong><span class="classifier">float</span></dt><dd><p>the tm of the primer</p>
</dd>
<dt><strong>tm,dH,dS</strong><span class="classifier">tuple</span></dt><dd><p>tm and dH and dS used for the calculation</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pydna.tm.tmstaluc98">
<code class="sig-prename descclassname">pydna.tm.</code><code class="sig-name descname">tmstaluc98</code><span class="sig-paren">(</span><em class="sig-param">primer</em>, <em class="sig-param">*args</em>, <em class="sig-param">dnac=50</em>, <em class="sig-param">saltc=50</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/tm.html#tmstaluc98"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.tm.tmstaluc98" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the melting temperature (Tm) of the primer using
the nearest neighbour algorithm. Formula and thermodynamic data
is taken from SantaLucia 1998 <a class="reference internal" href="#r620dfe066a50-1" id="id33">[1]</a>. This implementation gives the same
answer as the one provided by Biopython (See Examples).</p>
<p>Thermodynamic data used:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 38%" />
<col style="width: 31%" />
<col style="width: 31%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>pair</p></th>
<th class="head"><p>dH</p></th>
<th class="head"><p>dS</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>AA/TT</p></td>
<td><p>7.9</p></td>
<td><p>22.2</p></td>
</tr>
<tr class="row-odd"><td><p>AT/TA</p></td>
<td><p>7.2</p></td>
<td><p>20.4</p></td>
</tr>
<tr class="row-even"><td><p>TA/AT</p></td>
<td><p>7.2</p></td>
<td><p>21.3</p></td>
</tr>
<tr class="row-odd"><td><p>CA/GT</p></td>
<td><p>8.5</p></td>
<td><p>22.7</p></td>
</tr>
<tr class="row-even"><td><p>GT/CA</p></td>
<td><p>8.4</p></td>
<td><p>22.4</p></td>
</tr>
<tr class="row-odd"><td><p>CT/GA</p></td>
<td><p>7.8</p></td>
<td><p>21.0</p></td>
</tr>
<tr class="row-even"><td><p>GA/CT</p></td>
<td><p>8.2</p></td>
<td><p>22.2</p></td>
</tr>
<tr class="row-odd"><td><p>CG/GC</p></td>
<td><p>10.6</p></td>
<td><p>27.2</p></td>
</tr>
<tr class="row-even"><td><p>GC/CG</p></td>
<td><p>9.8</p></td>
<td><p>24.4</p></td>
</tr>
<tr class="row-odd"><td><p>GG/CC</p></td>
<td><p>8.0</p></td>
<td><p>19.9</p></td>
</tr>
</tbody>
</table>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>primer</strong><span class="classifier">string</span></dt><dd><p>Primer sequence 5’-3’</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>tm</strong><span class="classifier">float</span></dt><dd><p>tm of the primer</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="citation">
<dt class="label" id="r620dfe066a50-1"><span class="brackets"><a class="fn-backref" href="#id33">1</a></span></dt>
<dd><p>SantaLucia J Jr. A unified view of polymer, dumbbell, and oligonucleotide DNA nearest-neighbor thermodynamics. Proc Natl Acad Sci U S A 1998;95:1460–5.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.tm</span> <span class="k">import</span> <span class="n">tmstaluc98</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqUtils.MeltingTemp</span> <span class="k">import</span> <span class="n">Tm_NN</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tmstaluc98</span><span class="p">(</span><span class="s2">&quot;ACGTCATCGACACTATCATCGAC&quot;</span><span class="p">)</span>
<span class="go">54.55597724052518</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Tm_NN</span><span class="p">(</span><span class="s2">&quot;ACGTCATCGACACTATCATCGAC&quot;</span><span class="p">)</span>
<span class="go">54.55597724052518</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="module-pydna.utils">
<span id="pydna-utils-module"></span><h2>pydna.utils module<a class="headerlink" href="#module-pydna.utils" title="Permalink to this headline">¶</a></h2>
<p>This module provides miscellaneous functions.</p>
<dl class="function">
<dt id="pydna.utils.SmallestRotation">
<code class="sig-prename descclassname">pydna.utils.</code><code class="sig-name descname">SmallestRotation</code><span class="sig-paren">(</span><em class="sig-param">s</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#SmallestRotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.SmallestRotation" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.utils.complement">
<code class="sig-prename descclassname">pydna.utils.</code><code class="sig-name descname">complement</code><span class="sig-paren">(</span><em class="sig-param">sequence: str</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.complement" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the complement of sequence (string)
accepts mixed DNA/RNA</p>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.cseguid">
<code class="sig-prename descclassname">pydna.utils.</code><code class="sig-name descname">cseguid</code><span class="sig-paren">(</span><em class="sig-param">seq: str</em><span class="sig-paren">)</span> &#x2192; pydna._pretty.pretty_str<a class="reference internal" href="_modules/pydna/utils.html#cseguid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.cseguid" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the url safe cSEGUID for the sequence. The cSEGUID is the SEGUID checksum
calculated for the lexicographically minimal string rotation of a DNA sequence.
Only defined for circular sequences.</p>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.eq">
<code class="sig-prename descclassname">pydna.utils.</code><code class="sig-name descname">eq</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#eq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.eq" title="Permalink to this definition">¶</a></dt>
<dd><p>Compares two or more DNA sequences for equality i.e. they
represent the same DNA molecule. Comparisons are case insensitive.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>args</strong><span class="classifier">iterable</span></dt><dd><p>iterable containing sequences
args can be strings, Biopython Seq or SeqRecord, Dseqrecord
or dsDNA objects.</p>
</dd>
<dt><strong>circular</strong><span class="classifier">bool, optional</span></dt><dd><p>Consider all molecules circular or linear</p>
</dd>
<dt><strong>linear</strong><span class="classifier">bool, optional</span></dt><dd><p>Consider all molecules circular or linear</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>eq</strong><span class="classifier">bool</span></dt><dd><p>Returns True or False</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Compares two or more DNA sequences for equality i.e. if they
represent the same DNA molecule.</p>
<p>Two linear sequences are considiered equal if either:</p>
<ul class="simple">
<li><p>They have the same sequence (case insensitive)</p></li>
<li><p>One sequence is the reverse complement of the other (case insensitive)</p></li>
</ul>
<p>Two circular sequences are considered equal if they are circular permutations:</p>
<ol class="arabic simple">
<li><p>They have the same lengt, AND</p></li>
<li><p>One sequence or can be found in the concatenation of the other sequence with it    , OR</p></li>
<li><p>The reverse complement can be found in the concatenation of the other sequence with itself.</p></li>
</ol>
<p>The topology for the comparison can be set using one of the keywords
linear or circular to True or False.</p>
<p>If circular or linear is not set, it will be deduced from the topology of
each sequence for sequences that have a linear or circular attribute
(like Dseq and Dseqrecord).</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.dseqrecord</span> <span class="k">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.utils</span> <span class="k">import</span> <span class="n">eq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;AAA&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;AAA&quot;</span><span class="p">,</span><span class="s2">&quot;TTT&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;AAA&quot;</span><span class="p">,</span><span class="s2">&quot;TTT&quot;</span><span class="p">,</span><span class="s2">&quot;tTt&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;aaa&quot;</span><span class="p">,</span><span class="s2">&quot;AAA&quot;</span><span class="p">,</span><span class="s2">&quot;TTT&quot;</span><span class="p">,</span><span class="s2">&quot;tTt&quot;</span><span class="p">,</span> <span class="n">linear</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;Taaa&quot;</span><span class="p">,</span><span class="s2">&quot;aTaa&quot;</span><span class="p">,</span> <span class="n">linear</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;Taaa&quot;</span><span class="p">,</span><span class="s2">&quot;aTaa&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;Taaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s2">&quot;aTaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">b</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">linear</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">linear</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;ggatcc&quot;</span><span class="p">,</span><span class="s2">&quot;GGATCC&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;ggatcca&quot;</span><span class="p">,</span><span class="s2">&quot;GGATCCa&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s2">&quot;ggatcca&quot;</span><span class="p">,</span><span class="s2">&quot;tGGATCC&quot;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.flatten">
<code class="sig-prename descclassname">pydna.utils.</code><code class="sig-name descname">flatten</code><span class="sig-paren">(</span><em class="sig-param">*args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#flatten"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.flatten" title="Permalink to this definition">¶</a></dt>
<dd><p>Flattens an iterable of iterables down to str, bytes, bytearray or any of the pydna or Biopython seq objects</p>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.identifier_from_string">
<code class="sig-prename descclassname">pydna.utils.</code><code class="sig-name descname">identifier_from_string</code><span class="sig-paren">(</span><em class="sig-param">s: str</em><span class="sig-paren">)</span> &#x2192; str<a class="reference internal" href="_modules/pydna/utils.html#identifier_from_string"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.identifier_from_string" title="Permalink to this definition">¶</a></dt>
<dd><p>This function returns a string that is a valid python identifier based on the argument s or an empty string</p>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.lseguid">
<code class="sig-prename descclassname">pydna.utils.</code><code class="sig-name descname">lseguid</code><span class="sig-paren">(</span><em class="sig-param">seq: str</em><span class="sig-paren">)</span> &#x2192; pydna._pretty.pretty_str<a class="reference internal" href="_modules/pydna/utils.html#lseguid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.lseguid" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the url safe lSEGUID checksum for the sequence (seq). This is the SEGUID
checksum with the ‘+’ and ‘/’ characters of standard Base64 encoding are respectively
replaced by ‘-‘ and ‘_’.</p>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.memorize">
<code class="sig-prename descclassname">pydna.utils.</code><code class="sig-name descname">memorize</code><span class="sig-paren">(</span><em class="sig-param">filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#memorize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.memorize" title="Permalink to this definition">¶</a></dt>
<dd><p>Decorator for caching fucntions and classes, see pydna.download and pydna.Assembly for use</p>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.rc">
<code class="sig-prename descclassname">pydna.utils.</code><code class="sig-name descname">rc</code><span class="sig-paren">(</span><em class="sig-param">sequence: str</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pydna/utils.html#rc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the reverse complement of sequence (string)
accepts mixed DNA/RNA</p>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.seguid">
<code class="sig-prename descclassname">pydna.utils.</code><code class="sig-name descname">seguid</code><span class="sig-paren">(</span><em class="sig-param">seq: str</em><span class="sig-paren">)</span> &#x2192; pydna._pretty.pretty_str<a class="reference internal" href="_modules/pydna/utils.html#seguid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.seguid" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the url safe SEGUID checksum for the sequence. This is the SEGUID
checksum with the ‘+’ and ‘/’ characters of standard Base64 encoding are respectively
replaced by ‘-‘ and ‘_’.</p>
</dd></dl>

</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Welcome to Pydna’s documentation!</a><ul>
<li><a class="reference internal" href="#module-pydna">Module contents</a><ul>
<li><a class="reference internal" href="#pydna">pydna</a><ul>
<li><a class="reference internal" href="#pydna-package-layout">pydna package layout</a></li>
<li><a class="reference internal" href="#how-to-use-the-documentation">How to use the documentation</a></li>
<li><a class="reference internal" href="#pydna-source-code">pydna source code</a></li>
<li><a class="reference internal" href="#how-to-get-more-help">How to get more help</a></li>
<li><a class="reference internal" href="#examples-of-pydna-in-use">Examples of pydna in use</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.dseq">pydna.dseq module</a></li>
<li><a class="reference internal" href="#module-pydna.dseqrecord">pydna.dseqrecord module</a></li>
<li><a class="reference internal" href="#module-pydna.amplicon">pydna.amplicon module</a></li>
<li><a class="reference internal" href="#module-pydna.amplify">pydna.amplify module</a></li>
<li><a class="reference internal" href="#module-pydna.assembly">pydna.assembly module</a></li>
<li><a class="reference internal" href="#module-pydna.common_sub_strings">pydna.common_sub_strings module</a></li>
<li><a class="reference internal" href="#module-pydna.contig">pydna.contig module</a></li>
<li><a class="reference internal" href="#module-pydna.design">pydna.design module</a></li>
<li><a class="reference internal" href="#module-pydna.download">pydna.download module</a></li>
<li><a class="reference internal" href="#module-pydna.editor">pydna.editor module</a><ul>
<li><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.gel">pydna.gel module</a><ul>
<li><a class="reference internal" href="#note">Note</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.genbank">pydna.genbank module</a></li>
<li><a class="reference internal" href="#module-pydna.genbankfile">pydna.genbankfile module</a></li>
<li><a class="reference internal" href="#module-pydna.genbankfixer">pydna.genbankfixer module</a></li>
<li><a class="reference internal" href="#module-pydna.genbankrecord">pydna.genbankrecord module</a></li>
<li><a class="reference internal" href="#module-pydna.myprimers">pydna.myprimers module</a></li>
<li><a class="reference internal" href="#module-pydna.parsers">pydna.parsers module</a></li>
<li><a class="reference internal" href="#module-pydna.primer">pydna.primer module</a></li>
<li><a class="reference internal" href="#module-pydna.readers">pydna.readers module</a></li>
<li><a class="reference internal" href="#module-pydna.seqrecord">pydna.seqrecord module</a></li>
<li><a class="reference internal" href="#module-pydna.tm">pydna.tm module</a></li>
<li><a class="reference internal" href="#module-pydna.utils">pydna.utils module</a></li>
</ul>
</li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">Pydna 3.0.1+2.g47ae1f65a3.dirty documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2013-2019, Björn Johansson.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.1.2.
    </div>
  </body>
</html>