

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pydna package &mdash; Pydna 0.6.2 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.6.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="Pydna 0.6.2 documentation" href="index.html" />
    <link rel="next" title="pydna.py_rstr_max package" href="pydna.py_rstr_max.html" />
    <link rel="prev" title="Welcome to Pydna’s documentation!" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="np-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pydna.py_rstr_max.html" title="pydna.py_rstr_max package"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to Pydna’s documentation!"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">Pydna 0.6.2 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="pydna-package">
<h1>pydna package<a class="headerlink" href="#pydna-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="pydna.py_rstr_max.html">pydna.py_rstr_max package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pydna.py_rstr_max.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna.py_rstr_max.html#module-pydna.py_rstr_max.rstr_max">pydna.py_rstr_max.rstr_max module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna.py_rstr_max.html#module-pydna.py_rstr_max.tools_karkkainen_sanders">pydna.py_rstr_max.tools_karkkainen_sanders module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pydna.py_rstr_max.html#module-pydna.py_rstr_max">Module contents</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-pydna.amplify">
<span id="pydna-amplify-module"></span><h2>pydna.amplify module<a class="headerlink" href="#module-pydna.amplify" title="Permalink to this headline">¶</a></h2>
<p>This module provides functions for PCR. Primers with 5&#8217; tails as well as inverse PCR on
circular templates are handled correctly.</p>
<dl class="class">
<dt id="pydna.amplify.Amplicon">
<em class="property">class </em><tt class="descclassname">pydna.amplify.</tt><tt class="descname">Amplicon</tt><big>(</big><em>record</em>, <em>template=None</em>, <em>forward_primer=None</em>, <em>reverse_primer=None</em>, <em>saltc=None</em>, <em>forward_primer_concentration=None</em>, <em>reverse_primer_concentration=None</em>, <em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Amplicon"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Amplicon" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.dsdna.Dseqrecord" title="pydna.dsdna.Dseqrecord"><tt class="xref py py-class docutils literal"><span class="pre">pydna.dsdna.Dseqrecord</span></tt></a></p>
<p>The Amplicon class holds information about a PCR reaction involving two
primers and one template. This class is used by the Anneal class and is not
meant to be instantiated directly.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>forward_primer</strong> : SeqRecord(Biopython)</p>
<blockquote>
<div><p>SeqRecord object holding the forward (sense) primer</p>
</div></blockquote>
<p><strong>reverse_primer</strong> : SeqRecord(Biopython)</p>
<blockquote>
<div><p>SeqRecord object holding the reverse (antisense) primer</p>
</div></blockquote>
<p><strong>template</strong> : Dseqrecord</p>
<blockquote>
<div><p>Dseqrecord object holding the template (circular or linear)</p>
</div></blockquote>
<p><strong>saltc</strong> : float, optional</p>
<blockquote>
<div><p>saltc = monovalent cations (mM) (Na,K..)
default value is 50mM
This is used for Tm calculations.</p>
</div></blockquote>
<p><strong>forward_primer_concentration</strong> : float, optional</p>
<blockquote>
<div><p>primer concentration (nM)
default set to 1000nM = 1µM
This is used for Tm calculations.</p>
</div></blockquote>
<p><strong>rc</strong> : float, optional</p>
<blockquote class="last">
<div><p>primer concentration (nM)
default set to 1000nM = 1µM
This is used for Tm calculations.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="pydna.amplify.Amplicon.figure">
<tt class="descname">figure</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Amplicon.figure"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Amplicon.figure" title="Permalink to this definition">¶</a></dt>
<dd><p>This method returns a simple figure of the two primers binding to a part
of the template.</p>
<div class="highlight-python"><pre>5gctactacacacgtactgactg3
 |||||||||||||||||||||| tm 52.6 (dbd) 58.3
5gctactacacacgtactgactg...caagatagagtcagtaaccaca3
3cgatgatgtgtgcatgactgac...gttctatctcagtcattggtgt5
                          |||||||||||||||||||||| tm 49.1 (dbd) 57.7
                         3gttctatctcagtcattggtgt5</pre>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>figure:string</strong> :</p>
<blockquote class="last">
<div><p>A string containing a text representation of the primers
annealing on the template (see example above).</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Notes</p>
<p>tm in the figure above is the melting temperature (tm) for each primer calculated according to
SantaLucia 1998 <a class="footnote-reference" href="#id4" id="id1">[1]</a>.</p>
<p>dbd is the tm calculation for enzymes with dsDNA binding domains like Pfu-Sso7d <a class="footnote-reference" href="#id5" id="id2">[2]</a>. See <a class="footnote-reference" href="#id6" id="id3">[3]</a>
for more information.</p>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id4" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[1]</a></td><td><ol class="first last upperalpha simple" start="10">
<li>SantaLucia, “A Unified View of Polymer, Dumbbell, and Oligonucleotide DNA Nearest-neighbor Thermodynamics,” Proceedings of the National Academy of Sciences 95, no. 4 (1998): 1460.</li>
</ol>
</td></tr>
</tbody>
</table>
<table class="docutils footnote" frame="void" id="id5" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id2">[2]</a></td><td><ol class="first last upperalpha simple" start="13">
<li>Nørholm, “A Mutant Pfu DNA Polymerprimerase Designed for Advanced Uracil-excision DNA Engineering,” BMC Biotechnology 10, no. 1 (2010): 21, doi:10.1186/1472-6750-10-21.</li>
</ol>
</td></tr>
</tbody>
</table>
<table class="docutils footnote" frame="void" id="id6" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id3">[3]</a></td><td><a class="reference external" href="http://www.thermoscientificbio.com/webtools/tmc/">http://www.thermoscientificbio.com/webtools/tmc/</a></td></tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pydna.amplify.Amplicon.flankdn">
<tt class="descname">flankdn</tt><big>(</big><em>flankdnlength=50</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Amplicon.flankdn"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Amplicon.flankdn" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a Dseqrecord object containing flankdnlength bases downstream of the reverse primer footprint.
Truncated if the template is not long enough.</p>
<div class="highlight-python"><pre>                               &lt;---- flankdn ------&gt;

                3actactgactatctTAATAA5
                 ||||||||||||||
acgcattcagctactgtactactgactatctatcgtacatgtactatcgtat</pre>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.amplify.Amplicon.flankup">
<tt class="descname">flankup</tt><big>(</big><em>flankuplength=50</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Amplicon.flankup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Amplicon.flankup" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a Dseqrecord object containing flankuplength bases upstream of the forward primer footprint,
Truncated if the template is not long enough.</p>
<div class="highlight-python"><pre>&lt;--- flankup ---&gt;

          5TAATAAactactgactatct3
                 ||||||||||||||
acgcattcagctactgtactactgactatctatcg</pre>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.amplify.Amplicon.program">
<tt class="descname">program</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Amplicon.program"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Amplicon.program" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a string containing a text representation of two proposed
PCR programs. The first program is adapted for Taq poymerase, while
the second is adapted for Pfu-Sso7d or other polymerases with DNA binding domain.</p>
<div class="highlight-python"><pre>Taq (rate 30 nt/s)
Three-step|         30 cycles     |      |SantaLucia 1998
94.0°C    |94.0°C                 |      |SaltC 50mM
__________|_____          72.0°C  |72.0°C|
04min00s  |30s  \         ________|______|
          |      \ 46.0°C/ 0min 1s|10min |
          |       \_____/         |      |
          |         30s           |      |4-8°C

Pfu-Sso7d (rate 15s/kb)
Three-step|          30 cycles   |      |Breslauer1986,SantaLucia1998
98.0°C    |98.0°C                |      |SaltC 50mM
__________|_____          72.0°C |72.0°C|Primer1C   1µM
00min30s  |10s  \ 61.0°C ________|______|Primer2C   1µM
          |      \______/ 0min 0s|10min |
          |        10s           |      |4-8°C</pre>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pydna.amplify.Anneal">
<em class="property">class </em><tt class="descclassname">pydna.amplify.</tt><tt class="descname">Anneal</tt><big>(</big><em>primers</em>, <em>template</em>, <em>limit=13</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Anneal"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Anneal" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">object</span></tt></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>primers</strong> : iterable containing SeqRecord objects</p>
<blockquote>
<div><p>Primer sequences 5&#8217;-3&#8217;.</p>
</div></blockquote>
<p><strong>template</strong> : Dseqrecord object</p>
<blockquote>
<div><p>The template sequence 5&#8217;-3&#8217;.</p>
</div></blockquote>
<p><strong>limit</strong> : int, optional</p>
<blockquote class="last">
<div><p>limit length of the annealing part of the primers.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">template</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;tacactcaccgtctatcattatctactatcgactgtatcatctgatagcac&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p1</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&quot;&gt;p1</span><span class="se">\n</span><span class="s">tacactcaccgtctatcattatc&quot;</span><span class="p">,</span> <span class="n">ds</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&quot;&gt;p2</span><span class="se">\n</span><span class="s">gtgctatcagatgatacagtcg&quot;</span><span class="p">,</span> <span class="n">ds</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ann</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">Anneal</span><span class="p">((</span><span class="n">p1</span><span class="p">,</span><span class="n">p2</span><span class="p">),</span> <span class="n">template</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">ann</span><span class="o">.</span><span class="n">report</span><span class="p">()</span>
<span class="go">Template &lt;unknown name&gt; 51 nt linear:</span>
<span class="go">Primer p1 anneals forward at position 23</span>
<span class="go">Primer p2 anneals reverse at position 29</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ann</span><span class="o">.</span><span class="n">products</span>
<span class="go">[Amplicon(51)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">amplicon_list</span> <span class="o">=</span> <span class="n">ann</span><span class="o">.</span><span class="n">products</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">amplicon</span> <span class="o">=</span> <span class="n">amplicon_list</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">amplicon</span>
<span class="go">Amplicon(51)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">amplicon</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="go">5tacactcaccgtctatcattatc...cgactgtatcatctgatagcac3</span>
<span class="go">                           |||||||||||||||||||||| tm 50.6 (dbd) 60.5</span>
<span class="go">                          3gctgacatagtagactatcgtg5</span>
<span class="go">5tacactcaccgtctatcattatc3</span>
<span class="go"> ||||||||||||||||||||||| tm 49.4 (dbd) 58.8</span>
<span class="go">3atgtgagtggcagatagtaatag...gctgacatagtagactatcgtg5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">amplicon</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;02-FEB-2013&#39;</span>   <span class="c"># Set the date for this example to pass the doctest</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">amplicon</span>
<span class="go">Dseqrecord</span>
<span class="go">circular: False</span>
<span class="go">size: 51</span>
<span class="go">ID: 51bp U96+TO06Y6pFs74SQx8M1IVTBiY</span>
<span class="go">Name: 51bp_PCR_prod</span>
<span class="go">Description: Product_p1_p2</span>
<span class="go">Number of features: 2</span>
<span class="go">/date=02-FEB-2013</span>
<span class="go">Dseq(-51)</span>
<span class="go">taca..gcac</span>
<span class="go">atgt..cgtg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">amplicon</span><span class="o">.</span><span class="n">program</span><span class="p">()</span>

<span class="go">Taq (rate 30 nt/s)</span>
<span class="go">Three-step|         30 cycles     |      |SantaLucia 1998</span>
<span class="go">94.0°C    |94.0°C                 |      |SaltC 50mM</span>
<span class="go">__________|_____          72.0°C  |72.0°C|</span>
<span class="go">04min00s  |30s  \         ________|______|</span>
<span class="go">          |      \ 44.0°C/ 0min 1s|10min |</span>
<span class="go">          |       \_____/         |      |</span>
<span class="go">          |         30s           |      |4-8°C</span>

<span class="go">Pfu-Sso7d (rate 15s/kb)</span>
<span class="go">Three-step|          30 cycles   |      |Breslauer1986,SantaLucia1998</span>
<span class="go">98.0°C    |98.0°C                |      |SaltC 50mM</span>
<span class="go">__________|_____          72.0°C |72.0°C|Primer1C   1µM</span>
<span class="go">00min30s  |10s  \ 62.0°C ________|______|Primer2C   1µM</span>
<span class="go">          |      \______/ 0min 0s|10min |</span>
<span class="go">          |        10s           |      |4-8°C</span>

<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
<p class="rubric">Attributes</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="85%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>products: list</td>
<td>A list of Amplicon objects, one for each primer pair that may form a PCR product.</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="pydna.amplify.Anneal.products">
<tt class="descname">products</tt><a class="reference internal" href="_modules/pydna/amplify.html#Anneal.products"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Anneal.products" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.amplify.Anneal.report">
<tt class="descname">report</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Anneal.report"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Anneal.report" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is an alias of str(Annealobj).
Returns a short string representation.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pydna.amplify.Primer">
<em class="property">class </em><tt class="descclassname">pydna.amplify.</tt><tt class="descname">Primer</tt><big>(</big><em>seq_obj</em>, <em>position</em>, <em>footprint</em>, <em>tail</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#Primer"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.Primer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">Bio.SeqRecord.SeqRecord</span></tt></p>
<p>This class holds information about a primer on a template</p>
</dd></dl>

<dl class="function">
<dt id="pydna.amplify.basictm">
<tt class="descclassname">pydna.amplify.</tt><tt class="descname">basictm</tt><big>(</big><em>primer</em>, <em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#basictm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.basictm" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the melting temperature (Tm) of the primer using
the basic formula.</p>
<div class="line-block">
<div class="line">Tm = (wA+xT)*2 + (yG+zC)*4 assumed 50mM monovalent cations</div>
<div class="line"><br /></div>
<div class="line">w = number of A in primer</div>
<div class="line">x = number of T in primer</div>
<div class="line">y = number of G in primer</div>
<div class="line">z = number of C in primer</div>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>primer</strong> : string</p>
<blockquote>
<div><p>Primer sequence 5&#8217;-3&#8217;</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first last"><strong>tm</strong> : int</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.amplify</span> <span class="kn">import</span> <span class="n">basictm</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">basictm</span><span class="p">(</span><span class="s">&quot;ggatcc&quot;</span><span class="p">)</span>
<span class="go">20</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.amplify.pcr">
<tt class="descclassname">pydna.amplify.</tt><tt class="descname">pcr</tt><big>(</big><em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#pcr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.pcr" title="Permalink to this definition">¶</a></dt>
<dd><p>pcr is a convenience function for Anneal to simplify its usage,
especially from the command line. If more than one PCR product is
formed, an exception is raised.</p>
<p>args is any iterable of sequences or an iterable of iterables of sequences.
args will be greedily flattened.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>args</strong> : iterable containing sequence objects</p>
<blockquote>
<div><p>Several arguments are also accepted.</p>
</div></blockquote>
<p><strong>limit</strong> : int = 13, optional</p>
<blockquote>
<div><p>limit length of the annealing part of the primers.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>product</strong> : Dseqrecord</p>
<blockquote class="last">
<div><p>a Dseqrecord object representing the PCR product.
The direction of the PCR product will be the same as
for the template sequence.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Notes</p>
<p>sequences in args could be of type:</p>
<ul class="simple">
<li>string</li>
<li>Seq</li>
<li>SeqRecord</li>
<li>Dseqrecord</li>
</ul>
<p>The last sequence will be interpreted as the template
all preceeding sequences as primers.</p>
<p>This is a powerful function, use with care!</p>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">template</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;tacactcaccgtctatcattatctactatcgactgtatcatctgatagcac&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p1</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&quot;&gt;p1</span><span class="se">\n</span><span class="s">tacactcaccgtctatcattatc&quot;</span><span class="p">,</span> <span class="n">ds</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&quot;&gt;p2</span><span class="se">\n</span><span class="s">cgactgtatcatctgatagcac&quot;</span><span class="p">,</span> <span class="n">ds</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">pcr</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">,</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Amplicon(51)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">pcr</span><span class="p">([</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">],</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Amplicon(51)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">pcr</span><span class="p">((</span><span class="n">p1</span><span class="p">,</span><span class="n">p2</span><span class="p">,),</span> <span class="n">template</span><span class="p">)</span>
<span class="go">Amplicon(51)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.amplify.tmbreslauer86">
<tt class="descclassname">pydna.amplify.</tt><tt class="descname">tmbreslauer86</tt><big>(</big><em>primer</em>, <em>dnac=500.0</em>, <em>saltc=50</em>, <em>thermodynamics=False</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#tmbreslauer86"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.tmbreslauer86" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the melting temperature (Tm) of the primer using
the nearest neighbour algorithm. Formula and thermodynamic data
is taken from Breslauer 1986. These data are no longer widely used.</p>
<p>Breslauer 1986, table 2 <a class="footnote-reference" href="#id8" id="id7">[4]</a></p>
<table border="1" class="docutils">
<colgroup>
<col width="29%" />
<col width="29%" />
<col width="24%" />
<col width="18%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">pair</th>
<th class="head">dH</th>
<th class="head">dS</th>
<th class="head">dG</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>AA/TT</td>
<td>9.1</td>
<td>24.0</td>
<td>1.9</td>
</tr>
<tr class="row-odd"><td>AT/TA</td>
<td>8.6</td>
<td>23.9</td>
<td>1.5</td>
</tr>
<tr class="row-even"><td>TA/AT</td>
<td>6.0</td>
<td>16.9</td>
<td>0.9</td>
</tr>
<tr class="row-odd"><td>CA/GT</td>
<td>5.8</td>
<td>12.9</td>
<td>1.9</td>
</tr>
<tr class="row-even"><td>GT/CA</td>
<td>6.5</td>
<td>17.3</td>
<td>1.3</td>
</tr>
<tr class="row-odd"><td>CT/GA</td>
<td>7.8</td>
<td>20.8</td>
<td>1.6</td>
</tr>
<tr class="row-even"><td>GA/CT</td>
<td>5.6</td>
<td>13.5</td>
<td>1.6</td>
</tr>
<tr class="row-odd"><td>CG/GC</td>
<td>11.9</td>
<td>27.8</td>
<td>3.6</td>
</tr>
<tr class="row-even"><td>GC/CG</td>
<td>11.1</td>
<td>26.7</td>
<td>3.1</td>
</tr>
<tr class="row-odd"><td>GG/CC</td>
<td>11.0</td>
<td>26.6</td>
<td>3.1</td>
</tr>
</tbody>
</table>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>primer</strong> : string</p>
<blockquote>
<div><p>Primer sequence 5&#8217;-3&#8217; in UPPERCASE</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first last"><strong>tm</strong> : float</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id8" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id7">[4]</a></td><td>K.J. Breslauer et al., “Predicting DNA Duplex Stability from the Base Sequence,” Proceedings of the National Academy of Sciences 83, no. 11 (1986): 3746.</td></tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.amplify</span> <span class="kn">import</span> <span class="n">tmbreslauer86</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tmbreslauer86</span><span class="p">(</span><span class="s">&quot;ACGTCATCGACACTATCATCGAC&quot;</span><span class="p">)</span>
<span class="go">64.28863985851899</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.amplify.tmbresluc">
<tt class="descclassname">pydna.amplify.</tt><tt class="descname">tmbresluc</tt><big>(</big><em>primer</em>, <em>primerc=500.0</em>, <em>saltc=50</em>, <em>thermodynamics=False</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#tmbresluc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.tmbresluc" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the tm for a primer using a formula adapted to polymerases
with a DNA binding domain.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>primer</strong> : string</p>
<blockquote>
<div><p>primer sequence 5&#8217;-3&#8217;</p>
</div></blockquote>
<p><strong>primerc</strong> : float</p>
<blockquote>
<div><p>concentration (nM)</p>
</div></blockquote>
<p><strong>saltc</strong> : float, optional</p>
<blockquote>
<div><p>Monovalent cation concentration (mM)</p>
</div></blockquote>
<p><strong>thermodynamics</strong> : bool, optional</p>
<blockquote>
<div><p>prints details of the thermodynamic data to stdout. For
debugging only.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>tm</strong> : float</p>
<blockquote>
<div><p>the tm of the primer</p>
</div></blockquote>
<p><strong>tm,dH,dS</strong> : tuple</p>
<blockquote class="last">
<div><p>tm and dH and dS used for the calculation</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pydna.amplify.tmstaluc98">
<tt class="descclassname">pydna.amplify.</tt><tt class="descname">tmstaluc98</tt><big>(</big><em>primer</em>, <em>dnac=50</em>, <em>saltc=50</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/amplify.html#tmstaluc98"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.amplify.tmstaluc98" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the melting temperature (Tm) of the primer using
the nearest neighbour algorithm. Formula and thermodynamic data
is taken from SantaLucia 1998 <a class="footnote-reference" href="#id4" id="id9">[1]</a>. This implementation gives the same
answer as the one provided by Biopython (See Examples).</p>
<p>Thermodynamic data used:</p>
<table border="1" class="docutils">
<colgroup>
<col width="38%" />
<col width="31%" />
<col width="31%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">pair</th>
<th class="head">dH</th>
<th class="head">dS</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>AA/TT</td>
<td>7.9</td>
<td>22.2</td>
</tr>
<tr class="row-odd"><td>AT/TA</td>
<td>7.2</td>
<td>20.4</td>
</tr>
<tr class="row-even"><td>TA/AT</td>
<td>7.2</td>
<td>21.3</td>
</tr>
<tr class="row-odd"><td>CA/GT</td>
<td>8.5</td>
<td>22.7</td>
</tr>
<tr class="row-even"><td>GT/CA</td>
<td>8.4</td>
<td>22.4</td>
</tr>
<tr class="row-odd"><td>CT/GA</td>
<td>7.8</td>
<td>21.0</td>
</tr>
<tr class="row-even"><td>GA/CT</td>
<td>8.2</td>
<td>22.2</td>
</tr>
<tr class="row-odd"><td>CG/GC</td>
<td>10.6</td>
<td>27.2</td>
</tr>
<tr class="row-even"><td>GC/CG</td>
<td>9.8</td>
<td>24.4</td>
</tr>
<tr class="row-odd"><td>GG/CC</td>
<td>8.0</td>
<td>19.9</td>
</tr>
</tbody>
</table>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>primer</strong> : string</p>
<blockquote>
<div><p>Primer sequence 5&#8217;-3&#8217; in UPPERCASE</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>tm</strong> : float</p>
<blockquote class="last">
<div><p>tm of the primer</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna.amplify</span> <span class="kn">import</span> <span class="n">tmstaluc98</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqUtils.MeltingTemp</span> <span class="kn">import</span> <span class="n">Tm_staluc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tmstaluc98</span><span class="p">(</span><span class="s">&quot;ACGTCATCGACACTATCATCGAC&quot;</span><span class="p">)</span>
<span class="go">54.55597724052518</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Tm_staluc</span><span class="p">(</span><span class="s">&quot;ACGTCATCGACACTATCATCGAC&quot;</span><span class="p">)</span>
<span class="go">54.555977240525124</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="module-pydna.assembly">
<span id="pydna-assembly-module"></span><h2>pydna.assembly module<a class="headerlink" href="#module-pydna.assembly" title="Permalink to this headline">¶</a></h2>
<p>This module provides functions for assembly of sequences by homologous recombination and other
related techniques. Given a list of sequences (Dseqrecords), all sequences will be analyzed for
overlapping regions of DNA (common substrings).</p>
<p>The assembly algorithm is based on graph theory where each overlapping region forms a node and 
sequences separating the overlapping regions form edges.</p>
<dl class="class">
<dt id="pydna.assembly.Assembly">
<em class="property">class </em><tt class="descclassname">pydna.assembly.</tt><tt class="descname">Assembly</tt><big>(</big><em>dsrecs</em>, <em>limit=25</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Assembly"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Assembly" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">object</span></tt></p>
<p>Assembly of a list of linear DNA fragments into linear or circular constructs.   
Accepts a list of Dseqrecords (source fragments) to initiate an Assembly object. 
Several methods are available for analysis of overlapping sequences, graph construction
and assembly.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>dsrecs</strong> : list</p>
<blockquote class="last">
<div><p>a list of Dseqrecord objects.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna</span> <span class="kn">import</span> <span class="n">Assembly</span><span class="p">,</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;acgatgctatactgCCCCCtgtgctgtgctcta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;tgtgctgtgctctaTTTTTtattctggctgtatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;tattctggctgtatcGGGGGtacgatgctatactg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">Assembly</span><span class="p">((</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span><span class="p">),</span> <span class="n">limit</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">analyze_overlaps</span><span class="p">()</span>
<span class="go">&#39;6 sequences analyzed of which 3 have shared homologies with totally 3 overlaps&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">create_graph</span><span class="p">()</span>
<span class="go">&quot;A graph with 5&#39;, 3&#39; and 3 internal nodes was created&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">assemble_circular_from_graph</span><span class="p">()</span>
<span class="go">&#39;1 circular products were formed&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">Assembly object:</span>
<span class="go">Sequences........................: 33, 34, 35</span>
<span class="go">Sequences with shared homologies.: 33, 34, 35</span>
<span class="go">Homology limit (bp)..............: 14</span>
<span class="go">Number of overlaps...............: 3</span>
<span class="go">Nodes in graph...................: 5</span>
<span class="go">Assembly protocol................: overlaps</span>
<span class="go">Circular products................: 59</span>
<span class="go">Linear products..................: No linear products</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">circular_products</span>
<span class="go">[Dseqrecord(o59)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">circular_products</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">watson</span>
<span class="go">&#39;CCCCCtgtgctgtgctctaTTTTTtattctggctgtatcGGGGGtacgatgctatactg&#39;</span>
</pre></div>
</div>
<dl class="method">
<dt id="pydna.assembly.Assembly.analyze_overlaps">
<tt class="descname">analyze_overlaps</tt><big>(</big><em>limit=None</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Assembly.analyze_overlaps"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Assembly.analyze_overlaps" title="Permalink to this definition">¶</a></dt>
<dd><p>This method carries out a pairwise analysis for common sub strings (shared homologous sequences) 
among DNA fragments stored in the dsrecs property.</p>
<ul class="simple">
<li>Common sub strings need to be equal to or longer than limit.</li>
<li>Common sub strings can be present anywhere within the sequences stored in the dsrecs property.</li>
</ul>
<p>The analyzed_dsrec property will contain copies of the sequences in the dsrecs property that have 
common substrings. Common sub strings are added as features to the sequences in the analyzed_dsrec 
property.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>limit</strong> : int</p>
<blockquote class="last">
<div><p>The shortest shared sub sequence to be considered. This will also set the limit property.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition-see-also admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#pydna.assembly.Assembly.analyze_terminal_overlaps" title="pydna.assembly.Assembly.analyze_terminal_overlaps"><tt class="xref py py-obj docutils literal"><span class="pre">pydna.assembly.Assembly.analyze_terminal_overlaps</span></tt></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.assembly.Assembly.analyze_terminal_overlaps">
<tt class="descname">analyze_terminal_overlaps</tt><big>(</big><em>limit=None</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Assembly.analyze_terminal_overlaps"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Assembly.analyze_terminal_overlaps" title="Permalink to this definition">¶</a></dt>
<dd><p>This method carries out a pairwise analysis for shared sub sequences equal to or longer than
limit. The difference between this method and <a class="reference internal" href="#pydna.assembly.Assembly.analyze_overlaps" title="pydna.assembly.Assembly.analyze_overlaps"><tt class="xref py py-func docutils literal"><span class="pre">analyze_overlaps()</span></tt></a> is that shared sequences 
must start or end at the extremities of the source fragments.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>limit</strong> : int</p>
<blockquote class="last">
<div><p>The shortest shared sub sequence to be considered. This will also set the limit property.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition-see-also admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#pydna.assembly.Assembly.analyze_overlaps" title="pydna.assembly.Assembly.analyze_overlaps"><tt class="xref py py-obj docutils literal"><span class="pre">pydna.assembly.Assembly.analyze_overlaps</span></tt></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.assembly.Assembly.assemble_circular_from_graph">
<tt class="descname">assemble_circular_from_graph</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Assembly.assemble_circular_from_graph"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Assembly.assemble_circular_from_graph" title="Permalink to this definition">¶</a></dt>
<dd><p>Assembles all circular products from the graph created by <a class="reference internal" href="#pydna.assembly.Assembly.create_graph" title="pydna.assembly.Assembly.create_graph"><tt class="xref py py-func docutils literal"><span class="pre">create_graph()</span></tt></a></p>
</dd></dl>

<dl class="method">
<dt id="pydna.assembly.Assembly.assemble_fusion_pcr_linear">
<tt class="descname">assemble_fusion_pcr_linear</tt><big>(</big><em>limit=None</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Assembly.assemble_fusion_pcr_linear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Assembly.assemble_fusion_pcr_linear" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience method to assembles all possible linear fusion PCR products.
This method is a synonym of <a class="reference internal" href="#pydna.assembly.Assembly.assemble_gibson_linear" title="pydna.assembly.Assembly.assemble_gibson_linear"><tt class="xref py py-func docutils literal"><span class="pre">assemble_gibson_linear()</span></tt></a></p>
</dd></dl>

<dl class="method">
<dt id="pydna.assembly.Assembly.assemble_gibson_circular">
<tt class="descname">assemble_gibson_circular</tt><big>(</big><em>limit=None</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Assembly.assemble_gibson_circular"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Assembly.assemble_gibson_circular" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience method to assembles all possible circular Gibson assembly products
expected from in-vivo homologous recombination by the following steps:</p>
<ol class="arabic simple">
<li>Analysis of source fragments using the <a class="reference internal" href="#pydna.assembly.Assembly.analyze_terminal_overlaps" title="pydna.assembly.Assembly.analyze_terminal_overlaps"><tt class="xref py py-func docutils literal"><span class="pre">analyze_terminal_overlaps()</span></tt></a> method</li>
<li>Construction of a graph using the <a class="reference internal" href="#pydna.assembly.Assembly.create_graph" title="pydna.assembly.Assembly.create_graph"><tt class="xref py py-func docutils literal"><span class="pre">create_graph()</span></tt></a> method</li>
<li>Assembly of possible circular products using the <a class="reference internal" href="#pydna.assembly.Assembly.assemble_circular_from_graph" title="pydna.assembly.Assembly.assemble_circular_from_graph"><tt class="xref py py-func docutils literal"><span class="pre">assemble_circular_from_graph()</span></tt></a> method</li>
</ol>
<p>The property &#8220;circular_products&#8221; is initiated with a list of <a class="reference internal" href="#pydna.assembly.Fragment" title="pydna.assembly.Fragment"><tt class="xref py py-class docutils literal"><span class="pre">Fragment</span></tt></a> objects.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.assembly.Assembly.assemble_gibson_linear">
<tt class="descname">assemble_gibson_linear</tt><big>(</big><em>limit=None</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Assembly.assemble_gibson_linear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Assembly.assemble_gibson_linear" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience method to assembles all possible linear Gibson assembly products
expected from in-vivo homologous recombination by the following steps:</p>
<ol class="arabic simple">
<li>Analysis of source fragments using the <a class="reference internal" href="#pydna.assembly.Assembly.analyze_terminal_overlaps" title="pydna.assembly.Assembly.analyze_terminal_overlaps"><tt class="xref py py-func docutils literal"><span class="pre">analyze_terminal_overlaps()</span></tt></a> method</li>
<li>Construction of a graph using the <a class="reference internal" href="#pydna.assembly.Assembly.create_graph" title="pydna.assembly.Assembly.create_graph"><tt class="xref py py-func docutils literal"><span class="pre">create_graph()</span></tt></a> method</li>
<li>Assembly of possible circular products using the <a class="reference internal" href="#pydna.assembly.Assembly.assemble_linear_from_graph" title="pydna.assembly.Assembly.assemble_linear_from_graph"><tt class="xref py py-func docutils literal"><span class="pre">assemble_linear_from_graph()</span></tt></a> method</li>
</ol>
<p>The property &#8220;linear_products&#8221; is initiated with a list of <a class="reference internal" href="#pydna.assembly.Fragment" title="pydna.assembly.Fragment"><tt class="xref py py-class docutils literal"><span class="pre">Fragment</span></tt></a> objects.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.assembly.Assembly.assemble_golden_gate">
<tt class="descname">assemble_golden_gate</tt><big>(</big><em>limit=None</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Assembly.assemble_golden_gate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Assembly.assemble_golden_gate" title="Permalink to this definition">¶</a></dt>
<dd><p>Not implemented</p>
</dd></dl>

<dl class="method">
<dt id="pydna.assembly.Assembly.assemble_hr_circular">
<tt class="descname">assemble_hr_circular</tt><big>(</big><em>limit=None</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Assembly.assemble_hr_circular"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Assembly.assemble_hr_circular" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience method to assembles all possible circular homologous recombination (hr) products
expected from in-vivo homologous recombination by the following steps:</p>
<ol class="arabic simple">
<li>Analysis of source fragments using the <a class="reference internal" href="#pydna.assembly.Assembly.analyze_overlaps" title="pydna.assembly.Assembly.analyze_overlaps"><tt class="xref py py-func docutils literal"><span class="pre">analyze_overlaps()</span></tt></a> method</li>
<li>Construction of a graph using the <a class="reference internal" href="#pydna.assembly.Assembly.create_graph" title="pydna.assembly.Assembly.create_graph"><tt class="xref py py-func docutils literal"><span class="pre">create_graph()</span></tt></a> method</li>
<li>Assembly of possible circular products using the <a class="reference internal" href="#pydna.assembly.Assembly.assemble_circular_from_graph" title="pydna.assembly.Assembly.assemble_circular_from_graph"><tt class="xref py py-func docutils literal"><span class="pre">assemble_circular_from_graph()</span></tt></a> method</li>
</ol>
<p>The property &#8220;circular_products&#8221; is initiated with a list of <a class="reference internal" href="#pydna.assembly.Fragment" title="pydna.assembly.Fragment"><tt class="xref py py-class docutils literal"><span class="pre">Fragment</span></tt></a> objects.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.assembly.Assembly.assemble_hr_linear">
<tt class="descname">assemble_hr_linear</tt><big>(</big><em>limit=None</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Assembly.assemble_hr_linear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Assembly.assemble_hr_linear" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience method to assembles all possible linear homologous recombination (hr) products
expected from in-vivo homologous recombination by the following steps:</p>
<ol class="arabic simple">
<li>Analysis of source fragments using the <a class="reference internal" href="#pydna.assembly.Assembly.analyze_overlaps" title="pydna.assembly.Assembly.analyze_overlaps"><tt class="xref py py-func docutils literal"><span class="pre">analyze_overlaps()</span></tt></a> method</li>
<li>Construction of a graph using the <a class="reference internal" href="#pydna.assembly.Assembly.create_graph" title="pydna.assembly.Assembly.create_graph"><tt class="xref py py-func docutils literal"><span class="pre">create_graph()</span></tt></a> method</li>
<li>Assembly of possible linear products using the <a class="reference internal" href="#pydna.assembly.Assembly.assemble_linear_from_graph" title="pydna.assembly.Assembly.assemble_linear_from_graph"><tt class="xref py py-func docutils literal"><span class="pre">assemble_linear_from_graph()</span></tt></a> method</li>
</ol>
<p>The property &#8220;linear_products&#8221; is initiated with a list of <a class="reference internal" href="#pydna.assembly.Fragment" title="pydna.assembly.Fragment"><tt class="xref py py-class docutils literal"><span class="pre">Fragment</span></tt></a> objects.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.assembly.Assembly.assemble_linear_from_graph">
<tt class="descname">assemble_linear_from_graph</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Assembly.assemble_linear_from_graph"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Assembly.assemble_linear_from_graph" title="Permalink to this definition">¶</a></dt>
<dd><p>Assembles all linear products from the graph created by <a class="reference internal" href="#pydna.assembly.Assembly.create_graph" title="pydna.assembly.Assembly.create_graph"><tt class="xref py py-func docutils literal"><span class="pre">create_graph()</span></tt></a></p>
</dd></dl>

<dl class="method">
<dt id="pydna.assembly.Assembly.assemble_slic">
<tt class="descname">assemble_slic</tt><big>(</big><em>limit=None</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Assembly.assemble_slic"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Assembly.assemble_slic" title="Permalink to this definition">¶</a></dt>
<dd><p>Not implemented</p>
</dd></dl>

<dl class="method">
<dt id="pydna.assembly.Assembly.create_graph">
<tt class="descname">create_graph</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Assembly.create_graph"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Assembly.create_graph" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a graph from analyzed source fragments. Nodes in the graph are shared sub sequences
found as features in the source sequences. Edges are the sequences that connect the shared 
sub sequences in each source fragment. Two additional nodes called 5&#8217; and 3&#8217; are also added to 
represent DNA ends.</p>
</dd></dl>

<dl class="attribute">
<dt id="pydna.assembly.Assembly.dsrecs">
<tt class="descname">dsrecs</tt><em class="property"> = None</em><a class="headerlink" href="#pydna.assembly.Assembly.dsrecs" title="Permalink to this definition">¶</a></dt>
<dd><p>Sequences fed to this class is stored in this property</p>
</dd></dl>

<dl class="attribute">
<dt id="pydna.assembly.Assembly.limit">
<tt class="descname">limit</tt><em class="property"> = None</em><a class="headerlink" href="#pydna.assembly.Assembly.limit" title="Permalink to this definition">¶</a></dt>
<dd><p>The shortest common sub strings to be considered</p>
</dd></dl>

<dl class="attribute">
<dt id="pydna.assembly.Assembly.max_nodes">
<tt class="descname">max_nodes</tt><em class="property"> = None</em><a class="headerlink" href="#pydna.assembly.Assembly.max_nodes" title="Permalink to this definition">¶</a></dt>
<dd><p>The max number of nodes allowed. This can be reset to some other value</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pydna.assembly.Contig">
<em class="property">class </em><tt class="descclassname">pydna.assembly.</tt><tt class="descname">Contig</tt><big>(</big><em>record</em>, <em>source_fragments=</em><span class="optional">[</span><span class="optional">]</span>, <em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Contig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Contig" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.dsdna.Dseqrecord" title="pydna.dsdna.Dseqrecord"><tt class="xref py py-class docutils literal"><span class="pre">pydna.dsdna.Dseqrecord</span></tt></a></p>
<p>This class holds information about a DNA assembly. This class is instantiated by 
the <a class="reference internal" href="#pydna.assembly.Assembly" title="pydna.assembly.Assembly"><tt class="xref py py-class docutils literal"><span class="pre">Assembly</span></tt></a> class and is not meant to be instantiated directly.</p>
<dl class="method">
<dt id="pydna.assembly.Contig.detailed_fig">
<tt class="descname">detailed_fig</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Contig.detailed_fig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Contig.detailed_fig" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pydna.assembly.Contig.detailed_figure">
<tt class="descname">detailed_figure</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Contig.detailed_figure"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Contig.detailed_figure" title="Permalink to this definition">¶</a></dt>
<dd><p>Synonym of <a class="reference internal" href="#pydna.assembly.Contig.detailed_fig" title="pydna.assembly.Contig.detailed_fig"><tt class="xref py py-func docutils literal"><span class="pre">detailed_fig()</span></tt></a></p>
</dd></dl>

<dl class="method">
<dt id="pydna.assembly.Contig.small_fig">
<tt class="descname">small_fig</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Contig.small_fig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Contig.small_fig" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a small ascii representation of the assembled fragments. Each fragment is 
represented by:</p>
<div class="highlight-python"><pre>Size of common 5' substring|Name and size of DNA fragment| Size of common 5' substring</pre>
</div>
<p>Linear:</p>
<div class="highlight-python"><pre>frag20| 6
       \/
       /\
        6|frag23| 6
                 \/
                 /\
                  6|frag14</pre>
</div>
<p>Circular:</p>
<div class="highlight-python"><pre> -|2577|61
|       \/
|       /\
|       61|5681|98
|               \/
|               /\
|               98|2389|557
|                       \/
|                       /\
|                       557-
|                          |
 --------------------------</pre>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.assembly.Contig.small_figure">
<tt class="descname">small_figure</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Contig.small_figure"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Contig.small_figure" title="Permalink to this definition">¶</a></dt>
<dd><p>Synonym of <a class="reference internal" href="#pydna.assembly.Contig.small_fig" title="pydna.assembly.Contig.small_fig"><tt class="xref py py-func docutils literal"><span class="pre">small_fig()</span></tt></a></p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pydna.assembly.Fragment">
<em class="property">class </em><tt class="descclassname">pydna.assembly.</tt><tt class="descname">Fragment</tt><big>(</big><em>record</em>, <em>start1=0</em>, <em>end1=0</em>, <em>start2=0</em>, <em>end2=0</em>, <em>alignment=0</em>, <em>i=0</em>, <em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/assembly.html#Fragment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.assembly.Fragment" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#pydna.dsdna.Dseqrecord" title="pydna.dsdna.Dseqrecord"><tt class="xref py py-class docutils literal"><span class="pre">pydna.dsdna.Dseqrecord</span></tt></a></p>
<p>This class holds information about a DNA fragment in an assembly.
This class is instantiated by the <a class="reference internal" href="#pydna.assembly.Assembly" title="pydna.assembly.Assembly"><tt class="xref py py-class docutils literal"><span class="pre">Assembly</span></tt></a> class and is not 
meant to be instantiated directly.</p>
</dd></dl>

</div>
<div class="section" id="module-pydna.download">
<span id="pydna-download-module"></span><h2>pydna.download module<a class="headerlink" href="#module-pydna.download" title="Permalink to this headline">¶</a></h2>
<p>Provides a class for downloading sequences from genbank.</p>
<dl class="class">
<dt id="pydna.download.Genbank">
<em class="property">class </em><tt class="descclassname">pydna.download.</tt><tt class="descname">Genbank</tt><big>(</big><em>users_email</em>, <em>proxy=None</em>, <em>tool='pydna'</em><big>)</big><a class="reference internal" href="_modules/pydna/download.html#Genbank"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.download.Genbank" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to facilitate download from genbank.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>users_email</strong> : string</p>
<blockquote>
<div><p>Has to be a valid email address. You should always tell
Genbanks who you are, so that they can contact you.</p>
</div></blockquote>
<p><strong>proxy</strong> : string, optional</p>
<blockquote>
<div><p>String containing a proxy url:
&#8220;proxy = &#8220;<a class="reference external" href="http://umiho.proxy.com:3128">http://umiho.proxy.com:3128</a>&#8220;</p>
</div></blockquote>
<p><strong>tool</strong> : string, optional</p>
<blockquote class="last">
<div><p>Default is &#8220;pydna&#8221;. This is to tell Genbank which tool you are
using.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span> 
<span class="gp">&gt;&gt;&gt; </span><span class="n">gb</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Genbank</span><span class="p">(</span><span class="s">&quot;me@mail.se&quot;</span><span class="p">,</span> <span class="n">proxy</span> <span class="o">=</span> <span class="s">&quot;http://proxy.com:3128&quot;</span><span class="p">)</span>                  
<span class="gp">&gt;&gt;&gt; </span><span class="n">rec</span> <span class="o">=</span> <span class="n">gb</span><span class="o">.</span><span class="n">nucleotide</span><span class="p">(</span><span class="s">&quot;L09137&quot;</span><span class="p">)</span> <span class="c"># &lt;- pUC19 from genbank                            </span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">rec</span><span class="p">)</span>                                                                   
<span class="go">2686                                                                                 </span>
</pre></div>
</div>
<dl class="method">
<dt id="pydna.download.Genbank.nucleotide">
<tt class="descname">nucleotide</tt><big>(</big><em>item</em>, <em>start=None</em>, <em>stop=None</em>, <em>strand='watson'</em><big>)</big><a class="reference internal" href="_modules/pydna/download.html#Genbank.nucleotide"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.download.Genbank.nucleotide" title="Permalink to this definition">¶</a></dt>
<dd><p>Download a genbank nuclotide record.</p>
<p>Item is a string containing one genbank acession number <a class="footnote-reference" href="#id12" id="id10">[5]</a>
for a nucleotide file. Start and stop are intervals to be
downloaded. This is useful as some genbank records are large.
If strand is &#8220;c&#8221;, &#8220;C&#8221;, &#8220;crick&#8221;, &#8220;Crick&#8221;, &#8220;antisense&#8221;,&#8221;Antisense&#8221;, 
&#8220;2&#8221; or 2, the watson(antisense) strand is returned, otherwise
the sense strand is returned.</p>
<p>Alternatively, item can be a string containing an url that returns a 
sequence in genbank or FASTA format.</p>
<p>The genbank E-utilities can provide such urls <a class="footnote-reference" href="#id13" id="id11">[6]</a>.</p>
<p>Result is returned as a Dseqrecord object.</p>
<p>Genbank nucleotide accession numbers have this format:</p>
<div class="line-block">
<div class="line">A12345   = 1 letter  + 5 numerals</div>
<div class="line">AB123456 = 2 letters + 6 numerals</div>
</div>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id12" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id10">[5]</a></td><td><a class="reference external" href="http://www.dsimb.inserm.fr/~fuchs/M2BI/AnalSeq/Annexes/Sequences/Accession_Numbers.htm">http://www.dsimb.inserm.fr/~fuchs/M2BI/AnalSeq/Annexes/Sequences/Accession_Numbers.htm</a></td></tr>
</tbody>
</table>
<table class="docutils footnote" frame="void" id="id13" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id11">[6]</a></td><td><a class="reference external" href="http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch">http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch</a></td></tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pydna.download.Genbank.test">
<tt class="descname">test</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/download.html#Genbank.test"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.download.Genbank.test" title="Permalink to this definition">¶</a></dt>
<dd><p>Test downloading the pUC19 plasmid sequence from genbank
returns True if successful. Can be used to test proxy
and other network settings.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-pydna.dsdna">
<span id="pydna-dsdna-module"></span><h2>pydna.dsdna module<a class="headerlink" href="#module-pydna.dsdna" title="Permalink to this headline">¶</a></h2>
<p>Provides two classes, Dseq and Dseqrecord, for handling double stranded
DNA sequences. Dseq and Dseqrecord are subclasses of Biopythons
Seq and SeqRecord classes, respectively. These classes support the
notion of circular and linear DNA.</p>
<dl class="class">
<dt id="pydna.dsdna.Dseq">
<em class="property">class </em><tt class="descclassname">pydna.dsdna.</tt><tt class="descname">Dseq</tt><big>(</big><em>watson</em>, <em>crick=None</em>, <em>ovhg=None</em>, <em>linear=None</em>, <em>circular=None</em>, <em>alphabet=IUPACAmbiguousDNA()</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">Bio.Seq.Seq</span></tt></p>
<p>Dseq is a class designed to hold information for a double stranded
DNA fragment. Dseq also holds information describing the topology of
the DNA fragment (linear or circular).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>watson</strong> : str</p>
<blockquote>
<div><p>a string representing the watson (sense) DNA strand.</p>
</div></blockquote>
<p><strong>crick</strong> : str, optional</p>
<blockquote>
<div><p>a string representing the crick (antisense) DNA strand.</p>
</div></blockquote>
<p><strong>ovhg</strong> : int, optional</p>
<blockquote>
<div><p>A positive or negative number to describe the stagger between the watson and crick strands.
see below for a detailed explanation.</p>
</div></blockquote>
<p><strong>linear</strong> : bool, optional</p>
<blockquote>
<div><p>True indicates that sequence is linear, False that it is circular.</p>
</div></blockquote>
<p><strong>circular</strong> : bool, optional</p>
<blockquote>
<div><p>True indicates that sequence is circular, False that it is linear.</p>
</div></blockquote>
<p><strong>alphabet</strong> : Bio.Alphabet, optional</p>
<blockquote class="last">
<div><p>Bio.Alphabet.IUPAC.IUPACAmbiguousDNA from the Biopython package is set as default.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition-see-also admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#pydna.dsdna.Dseqrecord" title="pydna.dsdna.Dseqrecord"><tt class="xref py py-obj docutils literal"><span class="pre">pydna.dsdna.Dseqrecord</span></tt></a></p>
</div>
<p class="rubric">Examples</p>
<p>Dseq is a subclass of the Biopython Seq object. It stores two
strings representing the watson (sense) and crick(antisense) strands.
two properties called linear and circular, and a numeric value ovhg
(overhang) describing the stagger for the watson and crick strand
in the 5&#8217; end of the fragment.</p>
<p>The most common usage is probably to create a Dseq object as a
part of a Dseqrecord object (see <a class="reference internal" href="#pydna.dsdna.Dseqrecord" title="pydna.dsdna.Dseqrecord"><tt class="xref py py-class docutils literal"><span class="pre">pydna.dsdna.Dseqrecord</span></tt></a>).</p>
<p>There are three ways of creating a Dseq object directly:</p>
<p>Only one argument (string):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<p>The given string will be interpreted as the watson strand of a
blunt, linear double stranded sequence object. The crick strand
is created automatically from the watson strand.</p>
<p>Two arguments (string, string):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;gggaaat&quot;</span><span class="p">,</span><span class="s">&quot;ttt&quot;</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">gggaaat</span>
<span class="go">   ttt</span>
</pre></div>
</div>
<p>If both watson and crick are given, but not ovhg an attempt
will be made to find the best annealing between the strands.
There are limitations to this! For long fragments it is quite
slow. The length of the annealing sequences have to be at least
half the length of the shortest of the strands.</p>
<p>Three arguments (string, string, ovhg=int):</p>
<p>The ovhg parameter controls the stagger at the five prime end:</p>
<div class="highlight-python"><pre>ovhg=-2

XXXXX
  XXXXX

ovhg=-1

XXXXX
 XXXXX

ovhg=0

XXXXX
XXXXX

ovhg=1

 XXXXX
XXXXX

ovhg=2

  XXXXX
XXXXX</pre>
</div>
<p>Example of creating Dseq objects with different amounts of stagger:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-7)</span>
<span class="go">agt</span>
<span class="go">  attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-6)</span>
<span class="go">agt</span>
<span class="go"> attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">agt</span>
<span class="go">attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go"> agt</span>
<span class="go">attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">  agt</span>
<span class="go">attca</span>
</pre></div>
</div>
<p>If the ovhg parameter is psecified a crick strand also needs to be supplied,
otherwise an exception is raised.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s">&quot;agt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/dist-packages/pydna_/dsdna.py&quot;</span>, line <span class="m">169</span>, in <span class="n">__init__</span>
    <span class="k">else</span><span class="p">:</span>
<span class="gr">Exception</span>: <span class="n">ovhg defined without crick strand!</span>
</pre></div>
</div>
<p>The default alphabet is set to Biopython IUPACAmbiguousDNA</p>
<p>The shape of the fragment is set by either:</p>
<ol class="arabic simple">
<li>linear   = False, True</li>
<li>circular = True, False</li>
</ol>
<p>Note that both ends of the DNA fragment has to be compatible to set
circular = True (or linear = False).</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span><span class="s">&quot;ttt&quot;</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span><span class="s">&quot;ttt&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">linear</span> <span class="o">=</span> <span class="bp">False</span> <span class="p">,</span><span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(o3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="bp">True</span> <span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">Dseq(o3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
</pre></div>
</div>
<p>Coercing to string</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;tttcccc&quot;</span><span class="p">,</span><span class="s">&quot;aaacccc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-11)</span>
<span class="go">    tttcccc</span>
<span class="go">ccccaaa</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<span class="go">&#39;ggggtttcccc&#39;</span>
</pre></div>
</div>
<p>The double stranded part is accessible through the dsdata property:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">dsdata</span>
<span class="go">&#39;ttt&#39;</span>
</pre></div>
</div>
<p>A Dseq object can be longer that either the watson or crick strands.</p>
<div class="highlight-python"><pre>&lt;-- length --&gt;
GATCCTTT
     AAAGCCTAG</pre>
</div>
<p>The slicing of a linear Dseq object works mostly as it does for a string.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s">&quot;ggatcc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">&#39;a&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
<span class="go">&#39;at&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[::</span><span class="mi">2</span><span class="p">]</span>
<span class="go">&#39;gac&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">linear</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-1)</span>
<span class="go">a</span>
<span class="go">t</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
<span class="go">Dseq(-2)</span>
<span class="go">at</span>
<span class="go">ta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">4</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">Dseq(-0)</span>


<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[::</span><span class="mi">2</span><span class="p">]</span>
<span class="go">Dseq(-3)</span>
<span class="go">gac</span>
<span class="go">ctg</span>
</pre></div>
</div>
<p>The slicing of a circular Dseq object has a slightly different meaning.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s">&quot;ggAtCc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span>
<span class="go">Dseq(o6)</span>
<span class="go">ggAtCc</span>
<span class="go">ccTaGg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">4</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-5)</span>
<span class="go">CcggA</span>
<span class="go">GgccT</span>
</pre></div>
</div>
<p>The slice [X:X] produces an empty slice for a string, while this will return
the linearized sequence starting at X:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">=</span><span class="s">&quot;ggatcc&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span>
<span class="go">Dseq(o6)</span>
<span class="go">ggatcc</span>
<span class="go">cctagg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">[</span><span class="mi">3</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="go">Dseq(-6)</span>
<span class="go">tccgga</span>
<span class="go">aggcct</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
<dl class="method">
<dt id="pydna.dsdna.Dseq.T4">
<tt class="descname">T4</tt><big>(</big><em>nucleotides=None</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.T4"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.T4" title="Permalink to this definition">¶</a></dt>
<dd><p>Fill in of five prime protruding ends and chewing back of
three prime protruding ends by a DNA polymerase providing both
5&#8217;-3&#8217; DNA polymerase activity and 3&#8217;-5&#8217; nuclease acitivty
(such as T4 DNA polymerase). This can be done in presence of any
combination of the four A, G, C or T. Default are all four
nucleotides together.</p>
<p>Alias for the <a class="reference internal" href="#pydna.dsdna.Dseq.t4" title="pydna.dsdna.Dseq.t4"><tt class="xref py py-func docutils literal"><span class="pre">t4()</span></tt></a> method</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><strong>nucleotides</strong> : str</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;gatcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">()</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatc</span>
<span class="go">ctagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s">&quot;t&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgat</span>
<span class="go"> tagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s">&quot;a&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcga</span>
<span class="go">  agctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s">&quot;g&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcg</span>
<span class="go">   gctag</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pydna.dsdna.Dseq.circular">
<tt class="descname">circular</tt><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.circular"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.circular" title="Permalink to this definition">¶</a></dt>
<dd><p>The circular property</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.cut">
<tt class="descname">cut</tt><big>(</big><em>*enzymes</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.cut"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.cut" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of linear Dseq fragments produced in the digestion.
If there are no cuts, an empty list is returned.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>enzymes</strong> : enzyme object or iterable of such objects</p>
<blockquote>
<div><p>A Bio.Restriction.XXX restriction objects or iterable.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>frags</strong> : list</p>
<blockquote class="last">
<div><p>list of Dseq objects formed by the digestion</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna</span> <span class="kn">import</span> <span class="n">Dseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">=</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;ggatccnnngaattc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Dseq(-15)</span>
<span class="go">ggatccnnngaattc</span>
<span class="go">cctaggnnncttaag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Restriction</span> <span class="kn">import</span> <span class="n">BamHI</span><span class="p">,</span><span class="n">EcoRI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">))</span>
<span class="go">&lt;type &#39;list&#39;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">frag</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">frag</span><span class="o">.</span><span class="n">fig</span><span class="p">()</span>
<span class="go">Dseq(-5)</span>
<span class="go">g</span>
<span class="go">cctag</span>
<span class="go">Dseq(-14)</span>
<span class="go">gatccnnngaattc</span>
<span class="go">    gnnncttaag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">,</span> <span class="n">BamHI</span><span class="p">)</span> <span class="o">==</span>  <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">,</span> <span class="n">EcoRI</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">EcoRI</span><span class="p">,</span> <span class="n">BamHI</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="o">+</span><span class="n">c</span>
<span class="go">Dseq(-15)</span>
<span class="go">ggatccnnngaattc</span>
<span class="go">cctaggnnncttaag</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.fig">
<tt class="descname">fig</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.fig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.fig" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a representation of the sequence, truncated if
longer than 30 bp:</p>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;atcgcttactagcgtactgatcatctgact&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-30)</span>
<span class="go">atcgcttactagcgtactgatcatctgact</span>
<span class="go">tagcgaatgatcgcatgactagtagactga</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">+=</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;A&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-31)</span>
<span class="go">atcg..actA</span>
<span class="go">tagc..tgaT</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.fill_in">
<tt class="descname">fill_in</tt><big>(</big><em>nucleotides=None</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.fill_in"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.fill_in" title="Permalink to this definition">¶</a></dt>
<dd><p>Fill in of five prime protruding end with a DNA polymerase
that has only DNA polymerase activity (such as exo-klenow <a class="footnote-reference" href="#id15" id="id14">[7]</a>)
and any combination of A, G, C or T. Default are all four
nucleotides together.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><strong>nucleotides</strong> : str</td>
</tr>
</tbody>
</table>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id15" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id14">[7]</a></td><td><a class="reference external" href="http://en.wikipedia.org/wiki/Klenow_fragment#The_exo-_Klenow_fragment">http://en.wikipedia.org/wiki/Klenow_fragment#The_exo-_Klenow_fragment</a></td></tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;caaa&quot;</span><span class="p">,</span> <span class="s">&quot;cttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaa</span>
<span class="go"> tttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaag</span>
<span class="go">gtttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">(</span><span class="s">&quot;g&quot;</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaag</span>
<span class="go">gtttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">fill_in</span><span class="p">(</span><span class="s">&quot;tac&quot;</span><span class="p">)</span>
<span class="go">Dseq(-5)</span>
<span class="go">caaa</span>
<span class="go"> tttc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaac&quot;</span><span class="p">,</span> <span class="s">&quot;tttg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span>
<span class="go">Dseq(-5)</span>
<span class="go"> aaac</span>
<span class="go">gttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">fill_in</span><span class="p">()</span>
<span class="go">Dseq(-5)</span>
<span class="go"> aaac</span>
<span class="go">gttt</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.find">
<tt class="descname">find</tt><big>(</big><em>sub</em>, <em>start=0</em>, <em>end=9223372036854775807</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.find"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.find" title="Permalink to this definition">¶</a></dt>
<dd><p>Find method, like that of a python string.</p>
<p>This behaves like the python string method of the same name.</p>
<p>Returns an integer, the index of the first occurrence of substring
argument sub in the (sub)sequence given by [start:end].</p>
<p>Returns -1 if the subsequence is NOT found.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>sub</strong> : string or Seq object</p>
<blockquote>
<div><p>a string or another Seq object to look for.</p>
</div></blockquote>
<p><strong>start</strong> : int, optional</p>
<blockquote>
<div><p>slice start.</p>
</div></blockquote>
<p><strong>end</strong> : int, optional</p>
<blockquote class="last">
<div><p>slice end.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;atcgactgacgtgtt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Dseq(-15)</span>
<span class="go">atcgactgacgtgtt</span>
<span class="go">tagctgactgcacaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&quot;gac&quot;</span><span class="p">)</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="n">watson</span><span class="o">=</span><span class="s">&quot;agt&quot;</span><span class="p">,</span><span class="n">crick</span><span class="o">=</span><span class="s">&quot;actta&quot;</span><span class="p">,</span><span class="n">ovhg</span><span class="o">=-</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Dseq(-7)</span>
<span class="go">agt</span>
<span class="go">  attca</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s">&quot;taa&quot;</span><span class="p">)</span>
<span class="go">2</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.five_prime_end">
<tt class="descname">five_prime_end</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.five_prime_end"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.five_prime_end" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a tuple describing the structure of the 5&#8217; end of
the DNA fragment</p>
<div class="admonition-see-also admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#pydna.dsdna.Dseq.three_prime_end" title="pydna.dsdna.Dseq.three_prime_end"><tt class="xref py py-obj docutils literal"><span class="pre">pydna.dsdna.Dseq.three_prime_end</span></tt></a></p>
</div>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&#39;blunt&#39;, &#39;&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go"> aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&quot;3&#39;&quot;, &#39;t&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go">aaa</span>
<span class="go"> ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">five_prime_end</span><span class="p">()</span>
<span class="go">(&quot;5&#39;&quot;, &#39;a&#39;)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pydna.dsdna.Dseq.linear">
<tt class="descname">linear</tt><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.linear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.linear" title="Permalink to this definition">¶</a></dt>
<dd><p>The linear property</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.looped">
<tt class="descname">looped</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.looped"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.looped" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a circularized Dseq object. This can only be done if the
two ends are compatible, otherwise a TypeError is raised.</p>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;catcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="go">Dseq(o8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s">&quot;t&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgat</span>
<span class="go"> tagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s">&quot;t&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="go">Dseq(o7)</span>
<span class="go">catcgat</span>
<span class="go">gtagcta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s">&quot;a&quot;</span><span class="p">)</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcga</span>
<span class="go">  agctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">T4</span><span class="p">(</span><span class="s">&quot;a&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/dist-packages/pydna/dsdna.py&quot;</span>, line <span class="m">357</span>, in <span class="n">looped</span>
    <span class="k">if</span> <span class="n">type5</span> <span class="o">==</span> <span class="n">type3</span> <span class="ow">and</span> <span class="nb">str</span><span class="p">(</span><span class="n">sticky5</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">rc</span><span class="p">(</span><span class="n">sticky3</span><span class="p">)):</span>
<span class="gr">TypeError</span>: <span class="n">DNA cannot be circularized.</span>
<span class="go">5&#39; and 3&#39; sticky ends not compatible!</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.mung">
<tt class="descname">mung</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.mung"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.mung" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulates treatment a nuclease with 5&#8217;-3&#8217; and 3&#8217;-5&#8217; single
strand specific exonuclease activity (such as mung bean nuclease <a class="footnote-reference" href="#id17" id="id16">[8]</a>)</p>
<div class="highlight-python"><pre>    ggatcc    -&gt;     gatcc
     ctaggg          ctagg

     ggatcc   -&gt;      ggatc
    tcctag            cctag

&gt;&gt;&gt; import pydna
&gt;&gt;&gt; b=pydna.Dseq("caaa", "cttt")
&gt;&gt;&gt; b
Dseq(-5)
caaa
 tttc
&gt;&gt;&gt; b.mung()
Dseq(-3)
aaa
ttt
&gt;&gt;&gt; c=pydna.Dseq("aaac", "tttg")
&gt;&gt;&gt; c
Dseq(-5)
 aaac
gttt
&gt;&gt;&gt; c.mung()
Dseq(-3)
aaa
ttt</pre>
</div>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id17" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id16">[8]</a></td><td><a class="reference external" href="http://en.wikipedia.org/wiki/Mung_bean_nuclease">http://en.wikipedia.org/wiki/Mung_bean_nuclease</a></td></tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="pydna.dsdna.Dseq.ovhg">
<tt class="descname">ovhg</tt><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.ovhg"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.ovhg" title="Permalink to this definition">¶</a></dt>
<dd><p>The ovhg property</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.rc">
<tt class="descname">rc</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.rc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias of the reverse_complement method</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.reverse_complement">
<tt class="descname">reverse_complement</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.reverse_complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a Dseq object where watson and crick have switched
places.</p>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;catcgatc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseq(-8)</span>
<span class="go">gatcgatg</span>
<span class="go">ctagctac</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.t4">
<tt class="descname">t4</tt><big>(</big><em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.t4"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.t4" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for the <a class="reference internal" href="#pydna.dsdna.Dseq.T4" title="pydna.dsdna.Dseq.T4"><tt class="xref py py-func docutils literal"><span class="pre">T4()</span></tt></a> method</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.three_prime_end">
<tt class="descname">three_prime_end</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.three_prime_end"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.three_prime_end" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a tuple describing the structure of the 5&#8217; end of
the DNA fragment</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&#39;blunt&#39;, &#39;&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go"> aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&quot;3&#39;&quot;, &#39;a&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="s">&quot;ttt&quot;</span><span class="p">,</span> <span class="n">ovhg</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(-4)</span>
<span class="go">aaa</span>
<span class="go"> ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">three_prime_end</span><span class="p">()</span>
<span class="go">(&quot;5&#39;&quot;, &#39;t&#39;)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition-see-also admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#pydna.dsdna.Dseq.five_prime_end" title="pydna.dsdna.Dseq.five_prime_end"><tt class="xref py py-obj docutils literal"><span class="pre">pydna.dsdna.Dseq.five_prime_end</span></tt></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseq.tolinear">
<tt class="descname">tolinear</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseq.tolinear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseq.tolinear" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a blunt, linear copy of a circular Dseq object. This can
only be done if the Dseq object is circular, otherwise a
TypeError is raised.</p>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseq</span><span class="p">(</span><span class="s">&quot;catcgatc&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseq(o8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">tolinear</span><span class="p">()</span>
<span class="go">Dseq(-8)</span>
<span class="go">catcgatc</span>
<span class="go">gtagctag</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="pydna.dsdna.Dseqrecord">
<em class="property">class </em><tt class="descclassname">pydna.dsdna.</tt><tt class="descname">Dseqrecord</tt><big>(</big><em>record</em>, <em>circular=None</em>, <em>linear=None</em>, <em>n=1e-11</em>, <em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">Bio.SeqRecord.SeqRecord</span></tt></p>
<p>Dseqrecord is a double stranded version of the Biopython SeqRecord <a class="footnote-reference" href="#id19" id="id18">[9]</a> class.
The Dseqrecord object holds a Dseq object describing the sequence.
Additionally, Dseqrecord hold meta information about the sequence in the
from of a list of SeqFeatures, in the same way as the SeqRecord does.
The Dseqrecord can be initialized with a string, Seq, Dseq, SeqRecord
or another Dseqrecord. The sequence information will be stored in a
Dseq object in all cases. Dseqrecord objects can be read or parsed
from sequences in Fasta, Embl or Genbank format.</p>
<p>There is a short representation associated with the Dseqrecord.
<tt class="docutils literal"><span class="pre">Dseqrecord(-3)</span></tt> represents a linear sequence of length 2
while <tt class="docutils literal"><span class="pre">Dseqrecord(o7)</span></tt>
represents a circular sequence of length 7.</p>
<p>Dseqrecord and Dseq share the same concept of length</p>
<div class="highlight-python"><pre>&lt;-- length --&gt;
GATCCTTT
     AAAGCCTAG</pre>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>record</strong> : string, Seq, SeqRecord, Dseq or other Dseqrecord object</p>
<blockquote>
<div><p>This data will be used to form the seq property</p>
</div></blockquote>
<p><strong>circular</strong> : bool, optional</p>
<blockquote>
<div><p>True or False reflecting the shape of the DNA molecule</p>
</div></blockquote>
<p><strong>linear</strong> : bool, optional</p>
<blockquote class="last">
<div><p>True or False reflecting the shape of the DNA molecule</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id19" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id18">[9]</a></td><td><a class="reference external" href="http://biopython.org/wiki/SeqRecord">http://biopython.org/wiki/SeqRecord</a></td></tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-3)</span>
<span class="go">aaa</span>
<span class="go">ttt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">alphabet</span>
<span class="go">IUPACAmbiguousDNA()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">alphabet</span>
<span class="go">IUPACAmbiguousDNA()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">alphabet</span>
<span class="go">IUPACAmbiguousDNA()</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.add_feature">
<tt class="descname">add_feature</tt><big>(</big><em>x</em>, <em>y</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.add_feature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.add_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a feature of type misc to the feature list of the sequence.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>x</strong> : int</p>
<blockquote>
<div><p>Indicates start of the feature</p>
</div></blockquote>
<p><strong>y</strong> : int</p>
<blockquote class="last">
<div><p>Indicates end of the feature</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">features</span>
<span class="go">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">features</span>
<span class="go">[SeqFeature(FeatureLocation(ExactPosition(2), ExactPosition(4)), type=&#39;misc&#39;)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pydna.dsdna.Dseqrecord.circular">
<tt class="descname">circular</tt><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.circular"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.circular" title="Permalink to this definition">¶</a></dt>
<dd><p>The circular property</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.cseguid">
<tt class="descname">cseguid</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.cseguid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.cseguid" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the cSEGUID for the sequence. The cSEGUID is the SEGUID
checksum calculated for the lexicographically minimal string rotation
of the sequence or its reverse complement.</p>
<p>Only defined for circular sequences.</p>
<p>The cSEGUID checksum uniqely identifies a circular sequence.</p>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaat&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">cseguid</span><span class="p">()</span>
<span class="go">&#39;oopV+6158nHJqedi8lsshIfcqYA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;ataa&quot;</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">cseguid</span><span class="p">()</span>
<span class="go">&#39;oopV+6158nHJqedi8lsshIfcqYA&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.cut">
<tt class="descname">cut</tt><big>(</big><em>*enzymes</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.cut"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.cut" title="Permalink to this definition">¶</a></dt>
<dd><p>Digest the Dseqrecord object with one or more restriction enzymes.
returns a list of linear Dseqrecords. If there are no cuts, an empty
list is returned.</p>
<p>See also <a class="reference internal" href="#pydna.dsdna.Dseq.cut" title="pydna.dsdna.Dseq.cut"><tt class="xref py py-func docutils literal"><span class="pre">Dseq.cut()</span></tt></a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>enzymes</strong> : enzyme object or iterable of such objects</p>
<blockquote>
<div><p>A Bio.Restriction.XXX restriction object or iterable of such.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>Dseqrecord_frags</strong> : list</p>
<blockquote class="last">
<div><p>list of Dseqrecord objects formed by the digestion</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;ggatcc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Restriction</span> <span class="kn">import</span> <span class="n">BamHI</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">)</span>
<span class="go">[Dseqrecord(-5), Dseqrecord(-5)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag1</span><span class="p">,</span> <span class="n">frag2</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">BamHI</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag1</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-5)</span>
<span class="go">g</span>
<span class="go">cctag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frag2</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-5)</span>
<span class="go">gatcc</span>
<span class="go">    g</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.extract_feature">
<tt class="descname">extract_feature</tt><big>(</big><em>n</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.extract_feature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.extract_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts a feature and creates a new Dseqrecord object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>n</strong> : int</p>
<blockquote class="last">
<div><p>Indicates the feature to extract</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">extract_feature</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-2)</span>
<span class="go">gt</span>
<span class="go">ca</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.format">
<tt class="descname">format</tt><big>(</big><em>f='gb'</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.format"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.format" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the sequence as a string using a format supported by Biopython
SeqIO <a class="footnote-reference" href="#id21" id="id20">[10]</a>. Default is &#8220;gb&#8221; which is short for Genbank.</p>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id21" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id20">[10]</a></td><td><a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a></td></tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;02-FEB-2013&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">a</span><span class="o">.</span><span class="n">format</span><span class="p">()</span>
<span class="go">LOCUS       .                          3 bp    DNA     linear   UNK 02-FEB-2013</span>
<span class="go">DEFINITION  .</span>
<span class="go">ACCESSION   &lt;unknown id&gt;</span>
<span class="go">VERSION     &lt;unknown id&gt;</span>
<span class="go">KEYWORDS    .</span>
<span class="go">SOURCE      .</span>
<span class="go">  ORGANISM  .</span>
<span class="go">            .</span>
<span class="go">FEATURES             Location/Qualifiers</span>
<span class="go">ORIGIN</span>
<span class="go">        1 aaa</span>
<span class="go">//</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.isorf">
<tt class="descname">isorf</tt><big>(</big><em>table=1</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.isorf"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.isorf" title="Permalink to this definition">¶</a></dt>
<dd><p>Detects if sequence is an open reading frame (orf) in the 5&#8217;-3&#8217; direction.
Translation tables are numbers according to the NCBI numbering <a class="footnote-reference" href="#id23" id="id22">[11]</a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>table</strong> : int</p>
<blockquote>
<div><p>Sets the translation table, default is 1 (standard code)</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>bool</strong> :</p>
<blockquote class="last">
<div><p>True if sequence is an orf, False otherwise.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id23" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id22">[11]</a></td><td><a class="reference external" href="http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c">http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c</a></td></tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;atgaaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;atttaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">isorf</span><span class="p">()</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pydna.dsdna.Dseqrecord.linear">
<tt class="descname">linear</tt><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.linear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.linear" title="Permalink to this definition">¶</a></dt>
<dd><p>The linear property</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.linearize">
<tt class="descname">linearize</tt><big>(</big><em>*enzymes</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.linearize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.linearize" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is similar to <a class="reference internal" href="#pydna.dsdna.Dseqrecord.cut" title="pydna.dsdna.Dseqrecord.cut"><tt class="xref py py-func docutils literal"><span class="pre">cut()</span></tt></a> but throws an exception if there
is not excactly on cut i.e. none or more than one digestion products.</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.list_features">
<tt class="descname">list_features</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.list_features"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.list_features" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints an ASCII table with all features.</p>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna</span> <span class="kn">import</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;atgtaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">list_features</span><span class="p">()</span>
<span class="go">+----------+-----------+-------+-----+--------+--------------+------+------+</span>
<span class="go">| Feature# | Direction | Start | End | Length | id           | type | orf? |</span>
<span class="go">+----------+-----------+-------+-----+--------+--------------+------+------+</span>
<span class="go">| 0        |    None   |   2   |  4  |      2 | &lt;unknown id&gt; | misc |  no  |</span>
<span class="go">+----------+-----------+-------+-----+--------+--------------+------+------+</span>
<span class="gp">&gt;&gt;&gt; </span>               
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.looped">
<tt class="descname">looped</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.looped"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.looped" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a circular version of the Dseqrecord object. The
underlying Dseq object has to have compatible ends.</p>
<div class="admonition-see-also admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#pydna.dsdna.Dseq.looped" title="pydna.dsdna.Dseq.looped"><tt class="xref py py-obj docutils literal"><span class="pre">pydna.dsdna.Dseq.looped</span></tt></a></p>
</div>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(o3)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.rc">
<tt class="descname">rc</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.rc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of the reverse_complement method</p>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.reverse_complement">
<tt class="descname">reverse_complement</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.reverse_complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reverse complement.</p>
<div class="admonition-see-also admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#pydna.dsdna.Dseq.reverse_complement" title="pydna.dsdna.Dseq.reverse_complement"><tt class="xref py py-obj docutils literal"><span class="pre">pydna.dsdna.Dseq.reverse_complement</span></tt></a></p>
</div>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;ggaatt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">ggaatt</span>
<span class="go">ccttaa</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-6)</span>
<span class="go">aattcc</span>
<span class="go">ttaagg</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.seguid">
<tt class="descname">seguid</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.seguid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.seguid" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the SEGUID <a class="footnote-reference" href="#id25" id="id24">[12]</a> for the sequence.</p>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id25" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id24">[12]</a></td><td><a class="reference external" href="http://wiki.christophchamp.com/index.php/SEGUID">http://wiki.christophchamp.com/index.php/SEGUID</a></td></tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;YG7G6b2Kj/KtFOX63j8mRHHoIlE&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.shifted">
<tt class="descname">shifted</tt><big>(</big><em>shift</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.shifted"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.shifted" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a circular Dseqrecord with a new origin &lt;shift&gt;.
This only works on circular Dseqrecords. If we consider the following
circular sequence:</p>
<div class="line-block">
<div class="line"><tt class="docutils literal"><span class="pre">GAAAT</span>&nbsp;&nbsp; <span class="pre">&lt;--</span> <span class="pre">watson</span> <span class="pre">strand</span></tt></div>
<div class="line"><tt class="docutils literal"><span class="pre">CTTTA</span>&nbsp;&nbsp; <span class="pre">&lt;--</span> <span class="pre">crick</span> <span class="pre">strand</span></tt></div>
</div>
<p>The T and the G on the watson strand are linked together as well
as the A and the C of the of the crick strand.</p>
<p>if <tt class="docutils literal"><span class="pre">shift</span></tt> is 1, this indicates a new origin at position 1:</p>
<div class="line-block">
<div class="line-block">
<div class="line">new origin at the | symbol:</div>
<div class="line"><br /></div>
</div>
<div class="line"><tt class="docutils literal"><span class="pre">G|AAAT</span></tt></div>
<div class="line"><tt class="docutils literal"><span class="pre">C|TTTA</span></tt></div>
</div>
<p>new sequence:</p>
<div class="line-block">
<div class="line"><tt class="docutils literal"><span class="pre">AAATG</span></tt></div>
<div class="line"><tt class="docutils literal"><span class="pre">TTTAC</span></tt></div>
</div>
<p>Shift is always positive and 0&lt;shift&lt;length, so in the example
below, permissible values of shift are 1,2 and 3</p>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaat&quot;</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(o4)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">aaat</span>
<span class="go">ttta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">shifted</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(o4)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">aata</span>
<span class="go">ttat</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.stamp">
<tt class="descname">stamp</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.stamp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.stamp" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a seguid stamp to the description property. This will
show in the genbank format. The following string:</p>
<p><tt class="docutils literal"><span class="pre">SEGUID</span> <span class="pre">&lt;seguid&gt;</span></tt></p>
<p>will be appended to the description property of the Dseqrecord
object (string).</p>
<p>The stamp can be verified with <a class="reference internal" href="#pydna.dsdna.Dseqrecord.verify_stamp" title="pydna.dsdna.Dseqrecord.verify_stamp"><tt class="xref py py-func docutils literal"><span class="pre">verify_stamp()</span></tt></a></p>
<div class="admonition-see-also admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#pydna.dsdna.Dseqrecord.verify_stamp" title="pydna.dsdna.Dseqrecord.verify_stamp"><tt class="xref py py-obj docutils literal"><span class="pre">pydna.dsdna.Dseqrecord.verify_stamp</span></tt></a></p>
</div>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">stamp</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">description</span>
<span class="go">&#39;&lt;unknown description&gt; SEGUID YG7G6b2Kj/KtFOX63j8mRHHoIlE&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">verify_stamp</span><span class="p">()</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.synced">
<tt class="descname">synced</tt><big>(</big><em>ref</em>, <em>limit=25</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.synced"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.synced" title="Permalink to this definition">¶</a></dt>
<dd><p>This function returns a new circular sequence (Dseqrecord object), which has ben rotated
in such a way that there is maximum overlap between the sequence and
ref, which may be a string, Biopython Seq, SeqRecord object or
another Dseqrecord object.</p>
<p>The reason for using this could be to rotate a recombinant plasmid so
that it starts at the same position after cloning. See the example below:</p>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;gaat&quot;</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o4)</span>
<span class="go">gaat</span>
<span class="go">ctta</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">a</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span> <span class="o">+</span> <span class="n">a</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(-4)</span>
<span class="go">atga</span>
<span class="go">tact</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">insert</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;CCC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recombinant</span> <span class="o">=</span> <span class="p">(</span><span class="n">d</span><span class="o">+</span><span class="n">insert</span><span class="p">)</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recombinant</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o7)</span>
<span class="go">atgaCCC</span>
<span class="go">tactGGG</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recombinant</span><span class="o">.</span><span class="n">synced</span><span class="p">(</span><span class="n">a</span><span class="p">)</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Dseq(o7)</span>
<span class="go">gaCCCat</span>
<span class="go">ctGGGta</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.tolinear">
<tt class="descname">tolinear</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.tolinear"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.tolinear" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a linear, blunt copy of a circular Dseqrecord object. The
underlying Dseq object has to be circular.</p>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(o3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">tolinear</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span>
<span class="go">Dseqrecord(-3)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.verify_stamp">
<tt class="descname">verify_stamp</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.verify_stamp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.verify_stamp" title="Permalink to this definition">¶</a></dt>
<dd><p>Verifies the SEGUID stamp in the description property is
valid. True if stamp match the sequid calculated from the sequence.
Exception raised if no stamp can be found.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s">&#39;date&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;02-FEB-2013&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;YG7G6b2Kj/KtFOX63j8mRHHoIlE&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">a</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s">&quot;gb&quot;</span><span class="p">)</span>
<span class="go">LOCUS       .                          3 bp    DNA     linear   UNK 02-FEB-2013</span>
<span class="go">DEFINITION  .</span>
<span class="go">ACCESSION   &lt;unknown id&gt;</span>
<span class="go">VERSION     &lt;unknown id&gt;</span>
<span class="go">KEYWORDS    .</span>
<span class="go">SOURCE      .</span>
<span class="go">  ORGANISM  .</span>
<span class="go">            .</span>
<span class="go">FEATURES             Location/Qualifiers</span>
<span class="go">ORIGIN</span>
<span class="go">        1 aaa</span>
<span class="go">//</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">stamp</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">Dseqrecord(-3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">a</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s">&quot;gb&quot;</span><span class="p">)</span>
<span class="go">LOCUS       .                          3 bp    DNA     linear   UNK 02-FEB-2013</span>
<span class="go">DEFINITION  &lt;unknown description&gt; SEGUID YG7G6b2Kj/KtFOX63j8mRHHoIlE</span>
<span class="go">ACCESSION   &lt;unknown id&gt;</span>
<span class="go">VERSION     &lt;unknown id&gt;</span>
<span class="go">KEYWORDS    .</span>
<span class="go">SOURCE      .</span>
<span class="go">  ORGANISM  .</span>
<span class="go">            .</span>
<span class="go">FEATURES             Location/Qualifiers</span>
<span class="go">ORIGIN</span>
<span class="go">        1 aaa</span>
<span class="go">//</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">verify_stamp</span><span class="p">()</span>
<span class="go">True</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
<div class="admonition-see-also admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#pydna.dsdna.Dseqrecord.stamp" title="pydna.dsdna.Dseqrecord.stamp"><tt class="xref py py-obj docutils literal"><span class="pre">pydna.dsdna.Dseqrecord.stamp</span></tt></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="pydna.dsdna.Dseqrecord.write">
<tt class="descname">write</tt><big>(</big><em>filename=None</em>, <em>f='gb'</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#Dseqrecord.write"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.Dseqrecord.write" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes the Dseqrecord to a file using the format f, which must
be a format supported by Biopython SeqIO for writing <a class="footnote-reference" href="#id27" id="id26">[13]</a>. Default
is &#8220;gb&#8221; which is short for Genbank. Note that Biopython SeqIO reads
more formats than it writes.</p>
<p>Filename is the path to the file where the sequece is to be
written. The filename is optional, if it is not given, the
description property (string) is used together with the format.</p>
<p>If obj is the Dseqrecord object, the default file name will be:</p>
<p><tt class="docutils literal"><span class="pre">&lt;obj.description&gt;.&lt;f&gt;</span></tt></p>
<p>Where &lt;f&gt; is &#8220;gb&#8221; by default. If the filename already exists and
AND the sequence it contains is different, a new file name will be
used so that the old file is not lost:</p>
<p><tt class="docutils literal"><span class="pre">&lt;obj.description&gt;_NEW.&lt;f&gt;</span></tt></p>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id27" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id26">[13]</a></td><td><a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a></td></tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="pydna.dsdna.parse">
<tt class="descclassname">pydna.dsdna.</tt><tt class="descname">parse</tt><big>(</big><em>data</em>, <em>ds=True</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#parse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.parse" title="Permalink to this definition">¶</a></dt>
<dd><p>This function returns <em>all</em> DNA sequences found in data. If no
sequences are found, an empty list is returned. This is a greedy
function, use carefully.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>data</strong> : string or iterable</p>
<blockquote>
<div><p>The data parameter is a string containing:</p>
<ol class="arabic simple">
<li>an absolute path to a local file.
The file will be read in text
mode and parsed for EMBL, FASTA
and Genbank sequences.</li>
<li>an absolute path to a local directory.
All files in the directory will be
read and parsed as in 1.</li>
<li>a string containing one or more
sequences in EMBL, GENBANK,
or FASTA format. Mixed formats
are allowed.</li>
<li>data can be a list or other iterable of 1 - 3</li>
</ol>
</div></blockquote>
<p><strong>ds</strong> : bool</p>
<blockquote>
<div><p>If True double stranded <a class="reference internal" href="#pydna.dsdna.Dseqrecord" title="pydna.dsdna.Dseqrecord"><tt class="xref py py-class docutils literal"><span class="pre">Dseqrecord</span></tt></a> objects are returned.
If False single stranded <tt class="xref py py-class docutils literal"><span class="pre">Bio.SeqRecord</span></tt> <a class="footnote-reference" href="#id29" id="id28">[14]</a> objects are returned.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>list</strong> :</p>
<blockquote class="last">
<div><p>contains Dseqrecord or SeqRecord objects</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition-see-also admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#pydna.dsdna.read" title="pydna.dsdna.read"><tt class="xref py py-obj docutils literal"><span class="pre">read</span></tt></a></p>
</div>
<p class="rubric">References</p>
<table class="docutils footnote" frame="void" id="id29" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id28">[14]</a></td><td><a class="reference external" href="http://biopython.org/wiki/SeqRecord">http://biopython.org/wiki/SeqRecord</a></td></tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pydna.dsdna.parse_new">
<tt class="descclassname">pydna.dsdna.</tt><tt class="descname">parse_new</tt><big>(</big><em>data</em>, <em>ds=True</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#parse_new"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.parse_new" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pydna.dsdna.rc">
<tt class="descclassname">pydna.dsdna.</tt><tt class="descname">rc</tt><big>(</big><em>sequence</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#rc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the reverse complement of sequence (string)
accepts mixed DNA/RNA</p>
</dd></dl>

<dl class="function">
<dt id="pydna.dsdna.read">
<tt class="descclassname">pydna.dsdna.</tt><tt class="descname">read</tt><big>(</big><em>data</em>, <em>ds=True</em><big>)</big><a class="reference internal" href="_modules/pydna/dsdna.html#read"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.dsdna.read" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is similar the <a class="reference internal" href="#pydna.dsdna.parse" title="pydna.dsdna.parse"><tt class="xref py py-func docutils literal"><span class="pre">parse()</span></tt></a> function but expects one and only
one sequence or and exception is thrown.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>data</strong> : string</p>
<blockquote>
<div><p>see below</p>
</div></blockquote>
<p><strong>ds</strong> : bool</p>
<blockquote>
<div><p>Double stranded or single stranded DNA, Return
&#8220;Dseqrecord&#8221; or &#8220;SeqRecord&#8221; objects.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>Dseqrecord</strong> :</p>
<blockquote class="last">
<div><p>contains the first Dseqrecord or SeqRecord object parsed.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition-see-also admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#pydna.dsdna.parse" title="pydna.dsdna.parse"><tt class="xref py py-obj docutils literal"><span class="pre">parse</span></tt></a></p>
</div>
<p class="rubric">Notes</p>
<p>The data parameter is similar to the data parameter for <a class="reference internal" href="#pydna.dsdna.parse" title="pydna.dsdna.parse"><tt class="xref py py-func docutils literal"><span class="pre">parse()</span></tt></a>.</p>
</dd></dl>

</div>
<div class="section" id="module-pydna.editor">
<span id="pydna-editor-module"></span><h2>pydna.editor module<a class="headerlink" href="#module-pydna.editor" title="Permalink to this headline">¶</a></h2>
<p>This module provides a class for opening a sequence using an editor
that accepts a file as a command line argument.</p>
<p>ApE - A plasmid Editor <a class="footnote-reference" href="#id31" id="id30">[15]</a> is and excellent editor for this purpose.</p>
<div class="section" id="references">
<h3>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h3>
<table class="docutils footnote" frame="void" id="id31" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id30">[15]</a></td><td><a class="reference external" href="http://biologylabs.utah.edu/jorgensen/wayned/ape/">http://biologylabs.utah.edu/jorgensen/wayned/ape/</a></td></tr>
</tbody>
</table>
<dl class="class">
<dt id="pydna.editor.Editor">
<em class="property">class </em><tt class="descclassname">pydna.editor.</tt><tt class="descname">Editor</tt><big>(</big><em>shell_command_for_editor</em>, <em>tmpdir=None</em><big>)</big><a class="reference internal" href="_modules/pydna/editor.html#Editor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.editor.Editor" title="Permalink to this definition">¶</a></dt>
<dd><p>The Editor class needs to be instantiated before use.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>shell_command_for_editor</strong> : str</p>
<blockquote>
<div><p>String containing the path to the editor</p>
</div></blockquote>
<p><strong>tmpdir</strong> : str, optional</p>
<blockquote class="last">
<div><p>String containing path to the temprary directory where sequence 
files are stored before opening.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ape</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">Editor</span><span class="p">(</span><span class="s">&quot;tclsh8.6 /home/bjorn/.ApE/apeextractor/ApE.vfs/lib/app-AppMain/AppMain.tcl&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ape</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">)</span> <span class="c"># This command opens the sequence in the ApE editor</span>
</pre></div>
</div>
<dl class="method">
<dt id="pydna.editor.Editor.open">
<tt class="descname">open</tt><big>(</big><em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/editor.html#Editor.open"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.editor.Editor.open" title="Permalink to this definition">¶</a></dt>
<dd><p>Open a sequence for editing in an external (DNA) editor.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><strong>args</strong> : sequence or iterable of sequences</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
</div>
<div class="section" id="module-pydna.findsubstrings_suffix_arrays_python">
<span id="pydna-findsubstrings-suffix-arrays-python-module"></span><h2>pydna.findsubstrings_suffix_arrays_python module<a class="headerlink" href="#module-pydna.findsubstrings_suffix_arrays_python" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pydna.findsubstrings_suffix_arrays_python.common_sub_strings">
<tt class="descclassname">pydna.findsubstrings_suffix_arrays_python.</tt><tt class="descname">common_sub_strings</tt><big>(</big><em>stringx</em>, <em>stringy</em>, <em>limit=25</em><big>)</big><a class="reference internal" href="_modules/pydna/findsubstrings_suffix_arrays_python.html#common_sub_strings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.findsubstrings_suffix_arrays_python.common_sub_strings" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds all the common substrings between stringx and stringy
longer than limit. This function is case sensitive.</p>
<p>returns a list of tuples describing the substrings
The list is sorted longest -&gt; shortest.</p>
<p class="rubric">Examples</p>
<p>[(startx1,starty1,length1),(startx2,starty2,length2), ...]</p>
<p>startx1 = position in x where substring 1 starts
starty1 = position in y where substring 1 starts
length  = lenght of substring</p>
</dd></dl>

<dl class="function">
<dt id="pydna.findsubstrings_suffix_arrays_python.terminal_overlap">
<tt class="descclassname">pydna.findsubstrings_suffix_arrays_python.</tt><tt class="descname">terminal_overlap</tt><big>(</big><em>stringx</em>, <em>stringy</em>, <em>limit=15</em><big>)</big><a class="reference internal" href="_modules/pydna/findsubstrings_suffix_arrays_python.html#terminal_overlap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.findsubstrings_suffix_arrays_python.terminal_overlap" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="pydna-gel-module">
<h2>pydna.gel module<a class="headerlink" href="#pydna-gel-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-pydna.primer_design">
<span id="pydna-primer-design-module"></span><h2>pydna.primer_design module<a class="headerlink" href="#module-pydna.primer_design" title="Permalink to this headline">¶</a></h2>
<p>This module contain functions for primer design.</p>
<dl class="function">
<dt id="pydna.primer_design.assembly_primers">
<tt class="descclassname">pydna.primer_design.</tt><tt class="descname">assembly_primers</tt><big>(</big><em>templates</em>, <em>minlength=16</em>, <em>maxlength=50</em>, <em>min_olap=35</em>, <em>target_tm=55.0</em>, <em>primerc=1000.0</em>, <em>saltc=50.0</em>, <em>formula=&lt;function tmbresluc at 0x2b5ec5da6b18&gt;</em><big>)</big><a class="reference internal" href="_modules/pydna/primer_design.html#assembly_primers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.primer_design.assembly_primers" title="Permalink to this definition">¶</a></dt>
<dd><p>This function return primer pairs that are useful for fusion of DNA sequences given in template.
Given two sequences that we wish to fuse (a and b) to form fragment c.</p>
<div class="highlight-python"><pre> _________ a _________           __________ b ________
/                     \         /                     \
agcctatcatcttggtctctgca   &lt;--&gt;  TTTATATCGCATGACTCTTCTTT
|||||||||||||||||||||||         |||||||||||||||||||||||
tcggatagtagaaccagagacgt   &lt;--&gt;  AAATATAGCGTACTGAGAAGAAA

     agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT
     ||||||||||||||||||||||||||||||||||||||||||||||
     tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA  
     \___________________ c ______________________/</pre>
</div>
<p>We can design tailed primers to fuse a and b by fusion PCR, Gibson assembly or 
in-vivo homologous recombination. The basic requirements for the primers for 
the three techniques are the same.</p>
<p>Design tailed primers incorporating a part of the next or previous fragment to be assembled.</p>
<div class="highlight-python"><pre>agcctatcatcttggtctctgca     
|||||||||||||||||||||||     
                gagacgtAAATATA

|||||||||||||||||||||||     
tcggatagtagaaccagagacgt

                       TTTATATCGCATGACTCTTCTTT
                       |||||||||||||||||||||||

                ctctgcaTTTATAT
                       |||||||||||||||||||||||
                       AAATATAGCGTACTGAGAAGAAA</pre>
</div>
<p>PCR products with flanking sequences are formed in the PCR process.</p>
<div class="highlight-python"><pre>agcctatcatcttggtctctgcaTTTATAT     
||||||||||||||||||||||||||||||     
tcggatagtagaaccagagacgtAAATATA
                \____________/

                   identical 
                   sequences
                 ____________        
                /            \        
                ctctgcaTTTATATCGCATGACTCTTCTTT
                ||||||||||||||||||||||||||||||
                gagacgtAAATATAGCGTACTGAGAAGAAA</pre>
</div>
<p>The fragments can be fused by any of the techniques mentioned earlier.</p>
<div class="highlight-python"><pre>agcctatcatcttggtctctgcaTTTATATCGCATGACTCTTCTTT
||||||||||||||||||||||||||||||||||||||||||||||
tcggatagtagaaccagagacgtAAATATAGCGTACTGAGAAGAAA  </pre>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>templates</strong> : list of Dseqrecord</p>
<blockquote>
<div><p>list Dseqrecord object for which fusion primers should be constructed.</p>
</div></blockquote>
<p><strong>minlength</strong> : int, optional</p>
<blockquote>
<div><p>Minimum length of the annealing part of the primer.</p>
</div></blockquote>
<p><strong>maxlength</strong> : int, optional</p>
<blockquote>
<div><p>Maximum length (including tail) for designed primers.</p>
</div></blockquote>
<p><strong>tot_length</strong> : int, optional</p>
<blockquote>
<div><p>Maximum total length of a the primers</p>
</div></blockquote>
<p><strong>target_tm</strong> : float, optional</p>
<blockquote>
<div><p>target tm for the primers</p>
</div></blockquote>
<p><strong>primerc</strong> : float, optional</p>
<blockquote>
<div><p>Concentration of each primer in nM, set to 1000.0 nM by default</p>
</div></blockquote>
<p><strong>saltc</strong> : float, optional</p>
<blockquote>
<div><p>Salt concentration (monovalet cations) <tt class="xref py py-mod docutils literal"><span class="pre">tmbresluc</span></tt> set to 50.0 mM by default</p>
</div></blockquote>
<p><strong>formula</strong> : function</p>
<blockquote>
<div><p>formula used for tm calculation
this is the name of a function.
built in options are:</p>
<ol class="arabic simple">
<li><a class="reference internal" href="#pydna.amplify.tmbresluc" title="pydna.amplify.tmbresluc"><tt class="xref py py-func docutils literal"><span class="pre">pydna.amplify.tmbresluc()</span></tt></a> (default)</li>
<li><a class="reference internal" href="#pydna.amplify.basictm" title="pydna.amplify.basictm"><tt class="xref py py-func docutils literal"><span class="pre">pydna.amplify.basictm()</span></tt></a></li>
<li><a class="reference internal" href="#pydna.amplify.tmstaluc98" title="pydna.amplify.tmstaluc98"><tt class="xref py py-func docutils literal"><span class="pre">pydna.amplify.tmstaluc98()</span></tt></a></li>
<li><a class="reference internal" href="#pydna.amplify.tmbreslauer86" title="pydna.amplify.tmbreslauer86"><tt class="xref py py-func docutils literal"><span class="pre">pydna.amplify.tmbreslauer86()</span></tt></a></li>
</ol>
<p>These functions are imported from the <a class="reference internal" href="#module-pydna.amplify" title="pydna.amplify"><tt class="xref py py-mod docutils literal"><span class="pre">pydna.amplify</span></tt></a> module, but can be 
substituted for some other custom made function.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>primer_pairs</strong> : list of tuples of <tt class="xref py py-mod docutils literal"><span class="pre">Bio.Seqrecord</span></tt> objects</p>
<blockquote class="last">
<div><div class="highlight-python"><div class="highlight"><pre><span class="p">[(</span><span class="n">forward_primer_1</span><span class="p">,</span> <span class="n">reverse_primer_1</span><span class="p">),</span>
 <span class="p">(</span><span class="n">forward_primer_2</span><span class="p">,</span> <span class="n">reverse_primer_2</span><span class="p">),</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;atgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;ccaaacccaccaggtaccttatgtaagtacttcaagtcgccagaagacttcttggtcaagttgcc&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;tgtactggtgctgaaccttgtatcaagttgggtgttgacgccattgccccaggtggtcgtttcgtt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">primer_pairs</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">assembly_primers</span><span class="p">([</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span><span class="o">=</span><span class="p">[]</span>                                      
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">t</span><span class="p">,</span> <span class="p">(</span><span class="n">f</span><span class="p">,</span><span class="n">r</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">([</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">c</span><span class="p">],</span> <span class="n">primer_pairs</span><span class="p">):</span> <span class="n">p</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pydna</span><span class="o">.</span><span class="n">pcr</span><span class="p">(</span><span class="n">f</span><span class="p">,</span><span class="n">r</span><span class="p">,</span><span class="n">t</span><span class="p">))</span>  
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span>
<span class="go">[Amplicon(82), Amplicon(101), Amplicon(84)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">Assembly</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>     
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span>
<span class="go">Assembly object:</span>
<span class="go">Sequences........................: 82, 101, 84</span>
<span class="go">Sequences with shared homologies.: No analyzed sequences</span>
<span class="go">Homology limit (bp)..............: 25</span>
<span class="go">Number of overlaps...............: No overlaps</span>
<span class="go">Nodes in graph...................: No graph</span>
<span class="go">Assembly protocol................: No protocol</span>
<span class="go">Circular products................: No circular products</span>
<span class="go">Linear products..................: No linear products</span>
<span class="gp">&gt;&gt;&gt; </span>   
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span><span class="o">.</span><span class="n">assemble_gibson_linear</span><span class="p">()</span>
<span class="go">&#39;5 linear products were formed with the following sizes (bp): 195, 149, 147, 36, 36&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span><span class="o">.</span><span class="n">linear_products</span>
<span class="go">[Dseqrecord(-195), Dseqrecord(-149), Dseqrecord(-147), Dseqrecord(-36), Dseqrecord(-36)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">assemblyobj</span><span class="o">.</span><span class="n">linear_products</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;1eNv3d/1PqDPP8qJZIVoA45Www8&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="o">+</span><span class="n">c</span><span class="p">)</span><span class="o">.</span><span class="n">seguid</span><span class="p">()</span>
<span class="go">&#39;1eNv3d/1PqDPP8qJZIVoA45Www8&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="o">+</span><span class="n">c</span><span class="p">,</span> <span class="n">assemblyobj</span><span class="o">.</span><span class="n">linear_products</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">True</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">assemblyobj</span><span class="o">.</span><span class="n">linear_products</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">small_fig</span><span class="p">()</span>
<span class="go">82bp_PCR_prod|36</span>
<span class="go">              \/</span>
<span class="go">              /\</span>
<span class="go">              36|101bp_PCR_prod|36</span>
<span class="go">                                \/</span>
<span class="go">                                /\</span>
<span class="go">                                36|84bp_PCR_prod</span>
<span class="gp">&gt;&gt;&gt; </span>                               
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.primer_design.cloning_primers">
<tt class="descclassname">pydna.primer_design.</tt><tt class="descname">cloning_primers</tt><big>(</big><em>template</em>, <em>minlength=16</em>, <em>maxlength=29</em>, <em>fp=None</em>, <em>rp=None</em>, <em>fp_tail=''</em>, <em>rp_tail=''</em>, <em>target_tm=55.0</em>, <em>primerc=1000.0</em>, <em>saltc=50.0</em>, <em>formula=&lt;function tmbresluc at 0x2b5ec5da6b18&gt;</em><big>)</big><a class="reference internal" href="_modules/pydna/primer_design.html#cloning_primers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.primer_design.cloning_primers" title="Permalink to this definition">¶</a></dt>
<dd><p>This function can design primers for PCR amplification of a given sequence. 
This function accepts a Dseqrecord object containing the template sequence and 
returns a tuple cntaining two ::mod`Bio.SeqRecord.SeqRecord` objects describing 
the primers.</p>
<p>Primer tails can optionally be given in the form of strings.</p>
<p>An predesigned primer can be given, either the forward or reverse primers. In this 
case this function tries to design a primer with a Tm to match the given primer.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>template</strong> : Dseqrecord</p>
<blockquote>
<div><p>a Dseqrecord object.</p>
</div></blockquote>
<p><strong>minlength</strong> : int, optional</p>
<blockquote>
<div><p>Minimum length of the annealing part of the primer</p>
</div></blockquote>
<p><strong>maxlength</strong> : int, optional</p>
<blockquote>
<div><p>Maximum length (including tail) for designed primers.</p>
</div></blockquote>
<p><strong>fp, rp</strong> : SeqRecord, optional</p>
<blockquote>
<div><p>optional Biopython SeqRecord objects containing one primer each.</p>
</div></blockquote>
<p><strong>fp_tail, rp_tail</strong> : string, optional</p>
<blockquote>
<div><p>optional tails to be added to the forwars or reverse primers</p>
</div></blockquote>
<p><strong>target_tm</strong> : float, optional</p>
<blockquote>
<div><p>target tm for the primers</p>
</div></blockquote>
<p><strong>primerc</strong> : float, optional</p>
<blockquote>
<div><p>Concentration of each primer in nM, set to 1000.0 nM by default</p>
</div></blockquote>
<p><strong>saltc</strong> : float, optional</p>
<blockquote>
<div><p>Salt concentration (monovalet cations) <tt class="xref py py-mod docutils literal"><span class="pre">tmbresluc</span></tt> set to 50.0 mM by default</p>
</div></blockquote>
<p><strong>formula</strong> : function</p>
<blockquote>
<div><p>formula used for tm calculation
this is the name of a function.
built in options are:</p>
<ol class="arabic simple">
<li><a class="reference internal" href="#pydna.amplify.tmbresluc" title="pydna.amplify.tmbresluc"><tt class="xref py py-func docutils literal"><span class="pre">pydna.amplify.tmbresluc()</span></tt></a> (default)</li>
<li><a class="reference internal" href="#pydna.amplify.basictm" title="pydna.amplify.basictm"><tt class="xref py py-func docutils literal"><span class="pre">pydna.amplify.basictm()</span></tt></a></li>
<li><a class="reference internal" href="#pydna.amplify.tmstaluc98" title="pydna.amplify.tmstaluc98"><tt class="xref py py-func docutils literal"><span class="pre">pydna.amplify.tmstaluc98()</span></tt></a></li>
<li><a class="reference internal" href="#pydna.amplify.tmbreslauer86" title="pydna.amplify.tmbreslauer86"><tt class="xref py py-func docutils literal"><span class="pre">pydna.amplify.tmbreslauer86()</span></tt></a></li>
</ol>
<p>These functions are imported from the <a class="reference internal" href="#module-pydna.amplify" title="pydna.amplify"><tt class="xref py py-mod docutils literal"><span class="pre">pydna.amplify</span></tt></a> module, but can be 
substituted for some other custom made function.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>fp, rp</strong> : tuple</p>
<blockquote class="last">
<div><p>fp is a :mod:Bio.SeqRecord object describing the forward primer
rp is a :mod:Bio.SeqRecord object describing the reverse primer</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="o">=</span><span class="n">pydna</span><span class="o">.</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;atgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatg&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span>
<span class="go">Dseqrecord(-64)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span><span class="p">,</span><span class="n">pr</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">cloning_primers</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span>
<span class="go">SeqRecord(seq=Seq(&#39;atgactgctaacccttc&#39;, Alphabet()), id=&#39;pfw64&#39;, name=&#39;pfw64&#39;, description=&#39;pfw64&#39;, dbxrefs=[])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span>
<span class="go">SeqRecord(seq=Seq(&#39;catcgtaagtttcgaac&#39;, Alphabet()), id=&#39;prv64&#39;, name=&#39;prv64&#39;, description=&#39;prv64&#39;, dbxrefs=[])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr_prod</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">pcr</span><span class="p">(</span><span class="n">pf</span><span class="p">,</span> <span class="n">pr</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr_prod</span>
<span class="go">Amplicon(64)</span>
<span class="gp">&gt;&gt;&gt; </span>              
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">pcr_prod</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="go">5atgactgctaacccttc...gttcgaaacttacgatg3</span>
<span class="go">                     ||||||||||||||||| tm 42.4 (dbd) 52.9</span>
<span class="go">                    3caagctttgaatgctac5</span>
<span class="go">5atgactgctaacccttc3</span>
<span class="go"> ||||||||||||||||| tm 44.5 (dbd) 54.0</span>
<span class="go">3tactgacgattgggaag...caagctttgaatgctac5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span><span class="p">,</span><span class="n">pr</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">cloning_primers</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">fp_tail</span><span class="o">=</span><span class="s">&quot;GGATCC&quot;</span><span class="p">,</span> <span class="n">rp_tail</span><span class="o">=</span><span class="s">&quot;GAATTC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span>
<span class="go">SeqRecord(seq=Seq(&#39;GGATCCatgactgctaacccttc&#39;, Alphabet()), id=&#39;pfw64&#39;, name=&#39;pfw64&#39;, description=&#39;pfw64&#39;, dbxrefs=[])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span>
<span class="go">SeqRecord(seq=Seq(&#39;GAATTCcatcgtaagtttcgaac&#39;, Alphabet()), id=&#39;prv64&#39;, name=&#39;prv64&#39;, description=&#39;prv64&#39;, dbxrefs=[])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pcr_prod</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">pcr</span><span class="p">(</span><span class="n">pf</span><span class="p">,</span> <span class="n">pr</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">pcr_prod</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="go">      5atgactgctaacccttc...gttcgaaacttacgatg3</span>
<span class="go">                           ||||||||||||||||| tm 42.4 (dbd) 52.9</span>
<span class="go">                          3caagctttgaatgctacCTTAAG5</span>
<span class="go">5GGATCCatgactgctaacccttc3</span>
<span class="go">       ||||||||||||||||| tm 44.5 (dbd) 54.0</span>
<span class="go">      3tactgacgattgggaag...caagctttgaatgctac5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">pcr_prod</span><span class="o">.</span><span class="n">seq</span>
<span class="go">GGATCCatgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatgGAATTC</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s">&quot;atgactgctaacccttccttggtgttg&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span><span class="p">,</span><span class="n">pr</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">cloning_primers</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">fp</span> <span class="o">=</span> <span class="n">pf</span><span class="p">,</span> <span class="n">fp_tail</span><span class="o">=</span><span class="s">&quot;GGATCC&quot;</span><span class="p">,</span> <span class="n">rp_tail</span><span class="o">=</span><span class="s">&quot;GAATTC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pf</span>
<span class="go">SeqRecord(seq=Seq(&#39;GGATCCatgactgctaacccttccttggtgttg&#39;, Alphabet()), id=&#39;pfw64&#39;, name=&#39;pfw64&#39;, description=&#39;pfw64&#39;, dbxrefs=[])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span>
<span class="go">SeqRecord(seq=Seq(&#39;GAATTCcatcgtaagtttcgaacgaaatgtcgtc&#39;, Alphabet()), id=&#39;prv64&#39;, name=&#39;prv64&#39;, description=&#39;prv64&#39;, dbxrefs=[])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ampl</span> <span class="o">=</span> <span class="n">pydna</span><span class="o">.</span><span class="n">pcr</span><span class="p">(</span><span class="n">pf</span><span class="p">,</span><span class="n">pr</span><span class="p">,</span><span class="n">t</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">ampl</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="go">      5atgactgctaacccttccttggtgttg...gacgacatttcgttcgaaacttacgatg3</span>
<span class="go">                                     |||||||||||||||||||||||||||| tm 57.5 (dbd) 72.2</span>
<span class="go">                                    3ctgctgtaaagcaagctttgaatgctacCTTAAG5</span>
<span class="go">5GGATCCatgactgctaacccttccttggtgttg3</span>
<span class="go">       ||||||||||||||||||||||||||| tm 59.0 (dbd) 72.3</span>
<span class="go">      3tactgacgattgggaaggaaccacaac...ctgctgtaaagcaagctttgaatgctac5</span>
<span class="gp">&gt;&gt;&gt; </span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.primer_design.primer_pair">
<tt class="descclassname">pydna.primer_design.</tt><tt class="descname">primer_pair</tt><big>(</big><em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/primer_design.html#primer_pair"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.primer_design.primer_pair" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-pydna.utils">
<span id="pydna-utils-module"></span><h2>pydna.utils module<a class="headerlink" href="#module-pydna.utils" title="Permalink to this headline">¶</a></h2>
<p>This module provides miscellaneous functions.</p>
<dl class="function">
<dt id="pydna.utils.ChenFoxLyndon">
<tt class="descclassname">pydna.utils.</tt><tt class="descname">ChenFoxLyndon</tt><big>(</big><em>s</em><big>)</big><a class="reference internal" href="_modules/pydna/utils.html#ChenFoxLyndon"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.ChenFoxLyndon" title="Permalink to this definition">¶</a></dt>
<dd><p>Decompose s into Lyndon words according to the Chen-Fox-Lyndon theorem.
The arguments are the same as for ChenFoxLyndonBreakpoints but the
return values are subsequences of s rather than indices of breakpoints.</p>
<p>Algorithms on strings and sequences based on Lyndon words.
David Eppstein, October 2011.</p>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.ChenFoxLyndonBreakpoints">
<tt class="descclassname">pydna.utils.</tt><tt class="descname">ChenFoxLyndonBreakpoints</tt><big>(</big><em>s</em><big>)</big><a class="reference internal" href="_modules/pydna/utils.html#ChenFoxLyndonBreakpoints"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.ChenFoxLyndonBreakpoints" title="Permalink to this definition">¶</a></dt>
<dd><p>Find starting positions of Chen-Fox-Lyndon decomposition of s.
The decomposition is a set of Lyndon words that start at 0 and
continue until the next position. 0 itself is not output, but
the final breakpoint at the end of s is. The argument s must be
of a type that can be indexed (e.g. a list, tuple, or string).
The algorithm follows Duval, J. Algorithms 1983, but uses 0-based
indexing rather than Duval&#8217;s choice of 1-based indexing.</p>
<p>Algorithms on strings and sequences based on Lyndon words.
David Eppstein, October 2011.</p>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.SmallestRotation">
<tt class="descclassname">pydna.utils.</tt><tt class="descname">SmallestRotation</tt><big>(</big><em>s</em><big>)</big><a class="reference internal" href="_modules/pydna/utils.html#SmallestRotation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.SmallestRotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Find the rotation of s that is smallest in lexicographic order.
Duval 1983 describes how to modify his algorithm to do so but I think
it&#8217;s cleaner and more general to work from the ChenFoxLyndon output.</p>
<p>Algorithms on strings and sequences based on Lyndon words.
David Eppstein, October 2011.</p>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.copy_features">
<tt class="descclassname">pydna.utils.</tt><tt class="descname">copy_features</tt><big>(</big><em>source_sr</em>, <em>target_sr</em>, <em>limit=10</em><big>)</big><a class="reference internal" href="_modules/pydna/utils.html#copy_features"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.copy_features" title="Permalink to this definition">¶</a></dt>
<dd><p>This function tries to copy all features in source_seq and copy
them to target_seq. Source_sr and target_sr are objects with
a features property, such as Dseqrecord or Biopython SeqRecord.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>source_seq</strong> : SeqRecord or Dseqrecord</p>
<blockquote>
<div><p>The sequence to copy features from</p>
</div></blockquote>
<p><strong>target_seq</strong> : SeqRecord or Dseqrecord</p>
<blockquote>
<div><p>The sequence to copy features to</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>bool</strong> : True</p>
<blockquote class="last">
<div><p>This function acts on target_seq in place.
No data is returned.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.cseguid">
<tt class="descclassname">pydna.utils.</tt><tt class="descname">cseguid</tt><big>(</big><em>seq</em><big>)</big><a class="reference internal" href="_modules/pydna/utils.html#cseguid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.cseguid" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the cSEGUID for the sequence. The cSEGUID is the SEGUID checksum 
calculated for the lexicographically minimal string rotation of a DNA sequence.
Only defined for circular sequences.</p>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.eq">
<tt class="descclassname">pydna.utils.</tt><tt class="descname">eq</tt><big>(</big><em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/pydna/utils.html#eq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.eq" title="Permalink to this definition">¶</a></dt>
<dd><p>Compares two or more DNA sequences for equality i.e. they
represent the same DNA molecule. Comparisons are case insensitive.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>args</strong> : iterable</p>
<blockquote>
<div><p>iterable containing sequences
args can be strings, Biopython Seq or SeqRecord, Dseqrecord
or dsDNA objects.</p>
</div></blockquote>
<p><strong>circular</strong> : bool, optional</p>
<blockquote>
<div><p>Consider all molecules circular or linear</p>
</div></blockquote>
<p><strong>linear</strong> : bool, optional</p>
<blockquote>
<div><p>Consider all molecules circular or linear</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>eq</strong> : bool</p>
<blockquote class="last">
<div><p>Returns True or False</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Notes</p>
<p>Compares two or more DNA sequences for equality i.e. if they
represent the same DNA molecule.</p>
<p>Two linear sequences are considiered equal if either:</p>
<ul class="simple">
<li>They have the same sequence (case insensitive)</li>
<li>One sequence is the reverse complement of the other (case insensitive)</li>
</ul>
<p>Two circular sequences are considered equal if they are circular permutations:</p>
<ol class="arabic simple">
<li>They have the same lengt, AND</li>
<li>One sequence or can be found in the concatenation of the other sequence with itself, OR</li>
<li>The reverse complement can be found in the concatenation of the other sequence with itself.</li>
</ol>
<p>The topology for the comparison can be set using one of the keywords
linear or circular to True or False.</p>
<p>If circular or linear is not set, it will be deduced from the topology of
each sequence for sequences that have a linear or circular attribute
(like Dseq and Dseqrecord).</p>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pydna</span> <span class="kn">import</span> <span class="n">eq</span><span class="p">,</span> <span class="n">Dseqrecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span><span class="s">&quot;AAA&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span><span class="s">&quot;AAA&quot;</span><span class="p">,</span><span class="s">&quot;TTT&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span><span class="s">&quot;AAA&quot;</span><span class="p">,</span><span class="s">&quot;TTT&quot;</span><span class="p">,</span><span class="s">&quot;tTt&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;aaa&quot;</span><span class="p">,</span><span class="s">&quot;AAA&quot;</span><span class="p">,</span><span class="s">&quot;TTT&quot;</span><span class="p">,</span><span class="s">&quot;tTt&quot;</span><span class="p">,</span> <span class="n">linear</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;Taaa&quot;</span><span class="p">,</span><span class="s">&quot;aTaa&quot;</span><span class="p">,</span> <span class="n">linear</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;Taaa&quot;</span><span class="p">,</span><span class="s">&quot;aTaa&quot;</span><span class="p">,</span> <span class="n">circular</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;Taaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">Dseqrecord</span><span class="p">(</span><span class="s">&quot;aTaa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">=</span><span class="n">a</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span><span class="o">=</span><span class="n">b</span><span class="o">.</span><span class="n">looped</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">circular</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">linear</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">,</span><span class="n">linear</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;ggatcc&quot;</span><span class="p">,</span><span class="s">&quot;GGATCC&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;ggatcca&quot;</span><span class="p">,</span><span class="s">&quot;GGATCCa&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">eq</span><span class="p">(</span><span class="s">&quot;ggatcca&quot;</span><span class="p">,</span><span class="s">&quot;tGGATCC&quot;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.pairwise">
<tt class="descclassname">pydna.utils.</tt><tt class="descname">pairwise</tt><big>(</big><em>iterable</em><big>)</big><a class="reference internal" href="_modules/pydna/utils.html#pairwise"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.pairwise" title="Permalink to this definition">¶</a></dt>
<dd><p>s -&gt; (s0,s1), (s1,s2), (s2, s3), ...</p>
</dd></dl>

<dl class="function">
<dt id="pydna.utils.shift_origin">
<tt class="descclassname">pydna.utils.</tt><tt class="descname">shift_origin</tt><big>(</big><em>seq</em>, <em>shift</em><big>)</big><a class="reference internal" href="_modules/pydna/utils.html#shift_origin"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pydna.utils.shift_origin" title="Permalink to this definition">¶</a></dt>
<dd><p>Shift the origin of seq which is assumed to be a circular
sequence.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>seq</strong> : string, Biopython Seq, Biopython SeqRecord, Dseq or Dseqrecord</p>
<blockquote>
<div><p>sequence to be shifted.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>new_seq</strong> : string, Biopython Seq, Biopython SeqRecord, Dseq or Dseqrecord</p>
<blockquote class="last">
<div><p>sequence with a new origin.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition-see-also admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#pydna.dsdna.Dseqrecord.shifted" title="pydna.dsdna.Dseqrecord.shifted"><tt class="xref py py-obj docutils literal"><span class="pre">pydna.dsdna.Dseqrecord.shifted</span></tt></a></p>
</div>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pydna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">shift_origin</span><span class="p">(</span><span class="s">&quot;taaa&quot;</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
<span class="go">&#39;aaat&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">shift_origin</span><span class="p">(</span><span class="s">&quot;taaa&quot;</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>
<span class="go">&#39;taaa&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">shift_origin</span><span class="p">(</span><span class="s">&quot;taaa&quot;</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
<span class="go">&#39;aata&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pydna</span><span class="o">.</span><span class="n">shift_origin</span><span class="p">(</span><span class="s">&quot;gatc&quot;</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
<span class="go">&#39;tcga&#39;</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="module-pydna">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-pydna" title="Permalink to this headline">¶</a></h2>
<div class="section" id="python-dna">
<h3>Python-dna<a class="headerlink" href="#python-dna" title="Permalink to this headline">¶</a></h3>
<p>The Python-dna package.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">copyright:</th><td class="field-body">Copyright 2013,2014 by Björn Johansson. All rights reserved.</td>
</tr>
<tr class="field-even field"><th class="field-name">license:</th><td class="field-body">This code is part of the Python-dna distribution and governed by its
license.  Please see the LICENSE.txt file that should have been included
as part of this package.</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">pydna package</a><ul>
<li><a class="reference internal" href="#subpackages">Subpackages</a><ul>
</ul>
</li>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-pydna.amplify">pydna.amplify module</a></li>
<li><a class="reference internal" href="#module-pydna.assembly">pydna.assembly module</a></li>
<li><a class="reference internal" href="#module-pydna.download">pydna.download module</a></li>
<li><a class="reference internal" href="#module-pydna.dsdna">pydna.dsdna module</a></li>
<li><a class="reference internal" href="#module-pydna.editor">pydna.editor module</a><ul>
<li><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-pydna.findsubstrings_suffix_arrays_python">pydna.findsubstrings_suffix_arrays_python module</a></li>
<li><a class="reference internal" href="#pydna-gel-module">pydna.gel module</a></li>
<li><a class="reference internal" href="#module-pydna.primer_design">pydna.primer_design module</a></li>
<li><a class="reference internal" href="#module-pydna.utils">pydna.utils module</a></li>
<li><a class="reference internal" href="#module-pydna">Module contents</a><ul>
<li><a class="reference internal" href="#python-dna">Python-dna</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Welcome to Pydna&#8217;s documentation!</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="pydna.py_rstr_max.html"
                        title="next chapter">pydna.py_rstr_max package</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/pydna.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="np-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="pydna.py_rstr_max.html" title="pydna.py_rstr_max package"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to Pydna’s documentation!"
             >previous</a> |</li>
        <li><a href="index.html">Pydna 0.6.2 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Björn Johansson.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>